{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"toc_visible":true,"authorship_tag":"ABX9TyNxZXh9MdG5UKNx8svU1cNW"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# **Google Drive**"],"metadata":{"id":"caylNmBGD06r"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"tJ2cl8dhD0dr","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714706922588,"user_tz":-570,"elapsed":22517,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"c84a2dd1-4ee9-46e4-ed30-74a011897408"},"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}],"source":["# Mount Google Drive:\n","from google.colab import drive\n","drive.mount('/content/drive')"]},{"cell_type":"markdown","source":["# **R Magic**"],"metadata":{"id":"oZQwzrEPD-Es"}},{"cell_type":"code","source":["# Downgrade rpy:\n","!pip install rpy2==3.5.1"],"metadata":{"id":"YS3yxDxPD5xt","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714706948062,"user_tz":-570,"elapsed":25477,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"9fa7211d-11d3-4a0f-f5d2-581cf98abd80"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting rpy2==3.5.1\n","  Downloading rpy2-3.5.1.tar.gz (201 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m201.7/201.7 kB\u001b[0m \u001b[31m2.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n","Requirement already satisfied: cffi>=1.10.0 in /usr/local/lib/python3.10/dist-packages (from rpy2==3.5.1) (1.16.0)\n","Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from rpy2==3.5.1) (3.1.3)\n","Requirement already satisfied: pytz in /usr/local/lib/python3.10/dist-packages (from rpy2==3.5.1) (2023.4)\n","Requirement already satisfied: tzlocal in /usr/local/lib/python3.10/dist-packages (from rpy2==3.5.1) (5.2)\n","Requirement already satisfied: pycparser in /usr/local/lib/python3.10/dist-packages (from cffi>=1.10.0->rpy2==3.5.1) (2.22)\n","Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->rpy2==3.5.1) (2.1.5)\n","Building wheels for collected packages: rpy2\n","  Building wheel for rpy2 (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for rpy2: filename=rpy2-3.5.1-cp310-cp310-linux_x86_64.whl size=314932 sha256=f30a057a3a6d586d11a358340705cf2ae7913bc73896e815a77c293bad920d64\n","  Stored in directory: /root/.cache/pip/wheels/73/a6/ff/4e75dd1ce1cfa2b9a670cbccf6a1e41c553199e9b25f05d953\n","Successfully built rpy2\n","Installing collected packages: rpy2\n","  Attempting uninstall: rpy2\n","    Found existing installation: rpy2 3.4.2\n","    Uninstalling rpy2-3.4.2:\n","      Successfully uninstalled rpy2-3.4.2\n","Successfully installed rpy2-3.5.1\n"]}]},{"cell_type":"code","source":["# Activate R magic:\n","%load_ext rpy2.ipython"],"metadata":{"id":"TRMLbJTcD7js"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# **GWAS**"],"metadata":{"id":"EXsZsqCKNclY"}},{"cell_type":"markdown","source":["**Paper:**\n","\n","[\"Genome-wide Association Study of Long COVID\" Lammi et al., 2023](https://www.medrxiv.org/content/10.1101/2023.06.29.23292056v1)\n","\n","**Code:**\n","\n","[GitHub: Meta Analysis](https://github.com/long-covid-hg/META_ANALYSIS/tree/main)\n","\n","[GitHub: PCA](https://github.com/long-covid-hg/pca_projection)\n","\n","[GitHub: MR](https://github.com/marcoralab/MRcovid)\n","\n","[GitHub: Tools](https://github.com/long-covid-hg/LongCovidTools/tree/main)\n","\n","**GWAS:**\n","\n","[LocusZoom: GWAS1](https://my.locuszoom.org/gwas/192226/?token=09a18cf9138243db9cdf79ff6930fdf8)\n","\n","[LocusZoom: GWAS2](https://my.locuszoom.org/gwas/826733/?token=c7274597af504bf3811de6d742921bc8)\n","\n","[LocusZoom: GWAS3](https://my.locuszoom.org/gwas/793752/?token=0dc986619af14b6e8a564c580d3220b4)\n","\n","[LocusZoom: GWAS4](https://my.locuszoom.org/gwas/91854/?token=723e672edf13478e817ca44b56c0c068)\n","\n","**More Information:**\n","- Release: 7\n","- Ensembl: 109\n","- Human Genome Build: GRCh38\n","\n","**Cases vs. Controls:**\n","- **GWAS 1:** Long COVID cases after test-verified SARS-CoV-2 infection (strict case definition) (N=3,018) vs. population control (broad control definition) (N=1,093,995).\n","- **GWAS 2:** Long COVID cases after any SARS-CoV-2 Infection (tested and no tested  clinical confirmation) (broad case definition) (N=6,450) vs. population control (broad control definition) (N= 1,093,995).\n","- **GWAS 3:** Long COVID cases after test-verified SARS-CoV-2 infection (strict case definition) (N=3,018) vs. SARS-CoV-2 cases that are not long-COVID (strict control definition) (N=46,208).\n","- **GWAS 4:** Long COVID cases after any SARS-CoV-2 Infection (broad case definition) (N=6,450) vs. SARS-CoV-2 cases that are not long-COVID (strict control definition) (N=46,208).\n","\n","- <font color='green'>**Broad Controls:**</font> General population control.\n","- <font color='pink'>**Strict Controls:**</font> SARS-CoV-2 cases no long-COVID.\n","- <font color='blue'>**Broad Cases:**</font> Long-COVID cases tested and no tested.\n","- <font color='yellow'>**Strict Cases:**</font> Long-COVID cases tested.\n","\n","<img src=\"https://drive.google.com/uc?export=view&id=17vvFmvFv8mt-WKIjp7JMCwEI_fF0g_50\" alt=\"drawing\" width=\"500\"/>"],"metadata":{"id":"KT4eTYWOVhpN"}},{"cell_type":"code","source":["%%R\n","# Step 1: Install dplyr:\n","install.packages(\"dplyr\")\n","library(dplyr)"],"metadata":{"id":"9JEjCqu_Ob4N"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Step 2: Install BiocManager:\n","install.packages(\"BiocManager\")\n","library(BiocManager)"],"metadata":{"id":"nnRVSeMsFVxC"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Step 3: Install biomaRt:\n","BiocManager::install(\"biomaRt\")\n","library(biomaRt)"],"metadata":{"id":"6aYKsYZKFa2M"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Read the first dataset:\n","# Long COVID cases after test-verified SARS-CoV-2 infection (strict case definition) (N=3,018) vs. population control (broad control definition) (N=1,093,995).\n","gwas_1 <- base::gzfile(\"/content/drive/MyDrive/Colab/Long_COVID/GWAS/long_covid_gwas_1.gz\") %>%\n","  utils::read.table(header = FALSE,\n","                    sep = \"\\t\",\n","                    stringsAsFactors = FALSE)\n","\n","print(head(gwas_1))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Luef5wZkNewV","executionInfo":{"status":"ok","timestamp":1691563335004,"user_tz":-570,"elapsed":225423,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"8aee9440-d4a9-4bd0-ba37-b728f404820b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["  V1     V2          V3 V4 V5          V6          V7        V8        V9\n","1  1 727242  rs61769339  G  A 0.659672903 -0.08914350 0.0725127 0.1419410\n","2  1 729886 rs539032812  T  C 0.203850456 -0.08556010 0.1752460 0.0278364\n","3  1 758351  rs12238997  A  G 0.642450093 -0.08377110 0.0694591 0.1499760\n","4  1 758443  rs61769351  G  C 0.462277845 -0.06721850 0.0711697 0.1478110\n","5  1 770988  rs12029736  A  G 0.331720340  0.04515580 0.0619260 0.4910420\n","6  1 771398 rs113462541  G  A 0.009011522  0.00159483 0.0619572 0.7187970\n"]}]},{"cell_type":"code","source":["%%R\n","# Rename the columns of gwas_1:\n","names(gwas_1) <- c(\"chr\", \"position\", \"variant_id\", \"ref_all\", \"alt_all\", \"logOR\", \"beta\", \"SE\", \"alt_all_freq\")\n","\n","print(dim(gwas_1))\n","print(length(unique(gwas_1$variant_id)))\n","print(head(gwas_1))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"HPxXZqsOP8i5","executionInfo":{"status":"ok","timestamp":1691563339176,"user_tz":-570,"elapsed":4183,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"2f9693b7-884e-4cac-b4c6-15d345cf6a74"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 9510587       9\n","[1] 9023988\n","  chr position  variant_id ref_all alt_all       logOR        beta        SE\n","1   1   727242  rs61769339       G       A 0.659672903 -0.08914350 0.0725127\n","2   1   729886 rs539032812       T       C 0.203850456 -0.08556010 0.1752460\n","3   1   758351  rs12238997       A       G 0.642450093 -0.08377110 0.0694591\n","4   1   758443  rs61769351       G       C 0.462277845 -0.06721850 0.0711697\n","5   1   770988  rs12029736       A       G 0.331720340  0.04515580 0.0619260\n","6   1   771398 rs113462541       G       A 0.009011522  0.00159483 0.0619572\n","  alt_all_freq\n","1    0.1419410\n","2    0.0278364\n","3    0.1499760\n","4    0.1478110\n","5    0.4910420\n","6    0.7187970\n"]}]},{"cell_type":"code","source":["%%R\n","# Subset the previous dataset to have the most important columns for MtRobin:\n","gwas1_subset <- subset(gwas_1, select = c(\"variant_id\", \"beta\", \"SE\"))\n","\n","print(dim(gwas1_subset))\n","print(length(unique(gwas1_subset$variant_id)))\n","print(head(gwas1_subset))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"aHcUDmmt-ssi","executionInfo":{"status":"ok","timestamp":1691563346907,"user_tz":-570,"elapsed":7741,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"a0d5db33-7b84-4522-cb19-a51ed25e1e9d"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 9510587       3\n","[1] 9023988\n","   variant_id        beta        SE\n","1  rs61769339 -0.08914350 0.0725127\n","2 rs539032812 -0.08556010 0.1752460\n","3  rs12238997 -0.08377110 0.0694591\n","4  rs61769351 -0.06721850 0.0711697\n","5  rs12029736  0.04515580 0.0619260\n","6 rs113462541  0.00159483 0.0619572\n"]}]},{"cell_type":"code","source":["%%R\n","# Save the output data in a file:\n","write.table(gwas1_subset,\n","            file = \"/content/drive/MyDrive/Colab/Long_COVID/GWAS/long_covid_gwas_1_subset.txt\",\n","            sep = \"\\t\",\n","            row.names = FALSE,\n","            col.names = TRUE,\n","            quote = FALSE)"],"metadata":{"id":"Dz_6Mq1qC5VA"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Read the second dataset:\n","# Long COVID cases after any SARS-CoV-2 Infection (broad case definition) (N=6,450) vs. population control (broad control definition) (N= 1,093,995).\n","gwas_2 <- base::gzfile(\"/content/drive/MyDrive/Colab/Long_COVID/GWAS/long_covid_gwas_2.gz\") %>%\n","  utils::read.table(header = FALSE,\n","                    sep = \"\\t\",\n","                    stringsAsFactors = FALSE)\n","\n","print(head(gwas_2))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SZ9rj9N0Ot1L","executionInfo":{"status":"ok","timestamp":1691563499966,"user_tz":-570,"elapsed":153068,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"cbeae8d4-8259-4dae-ee37-49659383c027"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["  V1     V2          V3 V4 V5         V6          V7        V8        V9\n","1  1 758351  rs12238997  A  G 1.42869305 -0.07519460 0.0361019 0.1500900\n","2  1 794707 rs148120343  T  C 0.01924556  0.00276278 0.0508297 0.0597876\n","3  1 796338  rs58276399  T  C 1.25511506 -0.06778640 0.0354096 0.1560230\n","4  1 796652  rs61770163  A  C 2.13629358 -0.15318400 0.0571040 0.1545600\n","5  1 797429  rs12131618  T  C 1.25887284  0.13449700 0.0701197 0.0792994\n","6  1 798969 rs141242758  T  C 1.78633086 -0.08296060 0.0345542 0.1549950\n"]}]},{"cell_type":"code","source":["%%R\n","# Rename the columns of gwas_2:\n","names(gwas_2) <- c(\"chr\", \"position\", \"variant_id\", \"ref_all\", \"alt_all\", \"logOR\", \"beta\", \"SE\", \"alt_all_freq\")\n","\n","print(dim(gwas_2))\n","print(length(unique(gwas_2$variant_id)))\n","print(head(gwas_2))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"oycsjoQkQSx0","outputId":"cd7ea6c5-e8f3-4c1e-cc9b-ce37a6bc8264","executionInfo":{"status":"ok","timestamp":1691563505603,"user_tz":-570,"elapsed":5652,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}}},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 9722678       9\n","[1] 9135125\n","  chr position  variant_id ref_all alt_all      logOR        beta        SE\n","1   1   758351  rs12238997       A       G 1.42869305 -0.07519460 0.0361019\n","2   1   794707 rs148120343       T       C 0.01924556  0.00276278 0.0508297\n","3   1   796338  rs58276399       T       C 1.25511506 -0.06778640 0.0354096\n","4   1   796652  rs61770163       A       C 2.13629358 -0.15318400 0.0571040\n","5   1   797429  rs12131618       T       C 1.25887284  0.13449700 0.0701197\n","6   1   798969 rs141242758       T       C 1.78633086 -0.08296060 0.0345542\n","  alt_all_freq\n","1    0.1500900\n","2    0.0597876\n","3    0.1560230\n","4    0.1545600\n","5    0.0792994\n","6    0.1549950\n"]}]},{"cell_type":"code","source":["%%R\n","# Subset the previous dataset to have the most important columns for MtRobin:\n","gwas2_subset <- subset(gwas_2, select = c(\"variant_id\", \"beta\", \"SE\"))\n","\n","print(dim(gwas2_subset))\n","print(length(unique(gwas2_subset$variant_id)))\n","print(head(gwas2_subset))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"q1i0796nBZiE","executionInfo":{"status":"ok","timestamp":1691563514223,"user_tz":-570,"elapsed":8129,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"2eabf053-631d-42ae-b2a6-30116e3cc529"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 9722678       3\n","[1] 9135125\n","   variant_id        beta        SE\n","1  rs12238997 -0.07519460 0.0361019\n","2 rs148120343  0.00276278 0.0508297\n","3  rs58276399 -0.06778640 0.0354096\n","4  rs61770163 -0.15318400 0.0571040\n","5  rs12131618  0.13449700 0.0701197\n","6 rs141242758 -0.08296060 0.0345542\n"]}]},{"cell_type":"code","source":["%%R\n","# Save the output data in a file:\n","write.table(gwas2_subset,\n","            file = \"/content/drive/MyDrive/Colab/Long_COVID/GWAS/long_covid_gwas_2_subset.txt\",\n","            sep = \"\\t\",\n","            row.names = FALSE,\n","            col.names = TRUE,\n","            quote = FALSE)"],"metadata":{"id":"HwxKJmatCRkZ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Read the third dataset:\n","# Long COVID cases after test-verified SARS-CoV-2 infection (strict case definition) (N=3,018) vs. SARS-CoV-2 cases that are not long-COVID (strict control definition) (N=46,208).\n","gwas_3 <- base::gzfile(\"/content/drive/MyDrive/Colab/Long_COVID/GWAS/long_covid_gwas_3.gz\") %>%\n","  utils::read.table(header = FALSE,\n","                    sep = \"\\t\",\n","                    stringsAsFactors = FALSE)\n","\n","print(head(gwas_3))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"7DdSuRWeOu7K","executionInfo":{"status":"ok","timestamp":1691563656005,"user_tz":-570,"elapsed":141792,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"2bb68e1b-18cd-42f6-d8f3-197aa63cac08"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["  V1     V2          V3 V4 V5         V6         V7        V8        V9\n","1  1 727242  rs61769339  G  A 0.30395306 -0.0524928 0.0772207 0.1261010\n","2  1 729886 rs539032812  T  C 0.23106754 -0.0898624 0.1656110 0.0239551\n","3  1 758351  rs12238997  A  G 0.32343624 -0.0508974 0.0712253 0.1338670\n","4  1 758443  rs61769351  G  C 0.20615027 -0.0373298 0.0757353 0.1304840\n","5  1 771398 rs113462541  G  A 0.01595811  0.0028148 0.0622293 0.7284580\n","6  1 782207 rs144155419  G  A 0.39331559 -0.1775650 0.2129080 0.0128376\n"]}]},{"cell_type":"code","source":["%%R\n","# Rename the columns of gwas_3:\n","names(gwas_3) <- c(\"chr\", \"position\", \"variant_id\", \"ref_all\", \"alt_all\", \"logOR\", \"beta\", \"SE\", \"alt_all_freq\")\n","\n","print(dim(gwas_3))\n","print(length(unique(gwas_3$variant_id)))\n","print(head(gwas_3))"],"metadata":{"id":"Gz4qsMr7QlsC","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1691563657730,"user_tz":-570,"elapsed":1741,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"1741e312-1db7-4768-d5ff-8fb019bf981e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 9738584       9\n","[1] 9120808\n","  chr position  variant_id ref_all alt_all      logOR       beta        SE\n","1   1   727242  rs61769339       G       A 0.30395306 -0.0524928 0.0772207\n","2   1   729886 rs539032812       T       C 0.23106754 -0.0898624 0.1656110\n","3   1   758351  rs12238997       A       G 0.32343624 -0.0508974 0.0712253\n","4   1   758443  rs61769351       G       C 0.20615027 -0.0373298 0.0757353\n","5   1   771398 rs113462541       G       A 0.01595811  0.0028148 0.0622293\n","6   1   782207 rs144155419       G       A 0.39331559 -0.1775650 0.2129080\n","  alt_all_freq\n","1    0.1261010\n","2    0.0239551\n","3    0.1338670\n","4    0.1304840\n","5    0.7284580\n","6    0.0128376\n"]}]},{"cell_type":"code","source":["%%R\n","# Subset the previous dataset to have the most important columns for MtRobin:\n","gwas3_subset <- subset(gwas_3, select = c(\"variant_id\", \"beta\", \"SE\"))\n","\n","print(dim(gwas3_subset))\n","print(length(unique(gwas3_subset$variant_id)))\n","print(head(gwas3_subset))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"wOZbuXYhBrrb","executionInfo":{"status":"ok","timestamp":1691563666247,"user_tz":-570,"elapsed":8524,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"4707d375-387c-4db3-eae9-cc967a473847"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 9738584       3\n","[1] 9120808\n","   variant_id       beta        SE\n","1  rs61769339 -0.0524928 0.0772207\n","2 rs539032812 -0.0898624 0.1656110\n","3  rs12238997 -0.0508974 0.0712253\n","4  rs61769351 -0.0373298 0.0757353\n","5 rs113462541  0.0028148 0.0622293\n","6 rs144155419 -0.1775650 0.2129080\n"]}]},{"cell_type":"code","source":["%%R\n","# Save the output data in a file:\n","write.table(gwas3_subset,\n","            file = \"/content/drive/MyDrive/Colab/Long_COVID/GWAS/long_covid_gwas_3_subset.txt\",\n","            sep = \"\\t\",\n","            row.names = FALSE,\n","            col.names = TRUE,\n","            quote = FALSE)"],"metadata":{"id":"gIuupEzWCw13"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Read the fourth dataset:\n","# Long COVID cases after any SARS-CoV-2 Infection (broad case definition) (N=6,450) vs. SARS-CoV-2 cases that are not long-COVID (strict control definition) (N=46,208).\n","gwas_4 <- base::gzfile(\"/content/drive/MyDrive/Colab/Long_COVID/GWAS/long_covid_gwas_4.gz\") %>%\n","  utils::read.table(header = FALSE,\n","                    sep = \"\\t\",\n","                    stringsAsFactors = FALSE)\n","\n","print(head(gwas_4))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5iMmsB5hOwr7","executionInfo":{"status":"ok","timestamp":1691563802058,"user_tz":-570,"elapsed":135831,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"1d4c63a2-95ef-4d54-84cd-c6de79638225"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["  V1     V2          V3 V4 V5         V6          V7        V8        V9\n","1  1 758351  rs12238997  A  G 1.18554130 -0.07748130 0.0420257 0.1384160\n","2  1 794707 rs148120343  T  C 0.06082033 -0.00965152 0.0586638 0.0602122\n","3  1 796338  rs58276399  T  C 1.31144582 -0.08097980 0.0411023 0.1438010\n","4  1 796652  rs61770163  A  C 1.15127158 -0.11419100 0.0631543 0.1372740\n","5  1 797429  rs12131618  T  C 0.87743683  0.11802000 0.0784760 0.0777810\n","6  1 798969 rs141242758  T  C 1.78767653 -0.09648930 0.0401701 0.1431910\n"]}]},{"cell_type":"code","source":["%%R\n","# Rename the columns of gwas_4:\n","names(gwas_4) <- c(\"chr\", \"position\", \"variant_id\", \"ref_all\", \"alt_all\", \"logOR\", \"beta\", \"SE\", \"alt_all_freq\")\n","\n","print(dim(gwas_4))\n","print(length(unique(gwas_4$variant_id)))\n","print(head(gwas_4))"],"metadata":{"id":"ebfM-cS4Qo6r","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1691563807449,"user_tz":-570,"elapsed":5421,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"3a66173b-2267-4e39-f687-da1b7d648038"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 9753825       9\n","[1] 9173788\n","  chr position  variant_id ref_all alt_all      logOR        beta        SE\n","1   1   758351  rs12238997       A       G 1.18554130 -0.07748130 0.0420257\n","2   1   794707 rs148120343       T       C 0.06082033 -0.00965152 0.0586638\n","3   1   796338  rs58276399       T       C 1.31144582 -0.08097980 0.0411023\n","4   1   796652  rs61770163       A       C 1.15127158 -0.11419100 0.0631543\n","5   1   797429  rs12131618       T       C 0.87743683  0.11802000 0.0784760\n","6   1   798969 rs141242758       T       C 1.78767653 -0.09648930 0.0401701\n","  alt_all_freq\n","1    0.1384160\n","2    0.0602122\n","3    0.1438010\n","4    0.1372740\n","5    0.0777810\n","6    0.1431910\n"]}]},{"cell_type":"code","source":["%%R\n","# Subset the previous dataset to have the most important columns for MtRobin:\n","gwas4_subset <- subset(gwas_4, select = c(\"variant_id\", \"beta\", \"SE\"))\n","\n","print(dim(gwas4_subset))\n","print(length(unique(gwas4_subset$variant_id)))\n","print(head(gwas4_subset))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"gfnvVxnPB8HC","executionInfo":{"status":"ok","timestamp":1691563819275,"user_tz":-570,"elapsed":11838,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"71a89553-795f-40a0-e070-127ce52014c6"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 9753825       3\n","[1] 9173788\n","   variant_id        beta        SE\n","1  rs12238997 -0.07748130 0.0420257\n","2 rs148120343 -0.00965152 0.0586638\n","3  rs58276399 -0.08097980 0.0411023\n","4  rs61770163 -0.11419100 0.0631543\n","5  rs12131618  0.11802000 0.0784760\n","6 rs141242758 -0.09648930 0.0401701\n"]}]},{"cell_type":"code","source":["%%R\n","# Save the output data in a file:\n","write.table(gwas4_subset,\n","            file = \"/content/drive/MyDrive/Colab/Long_COVID/GWAS/long_covid_gwas_4_subset.txt\",\n","            sep = \"\\t\",\n","            row.names = FALSE,\n","            col.names = TRUE,\n","            quote = FALSE)"],"metadata":{"id":"UDuyEdALC_Hf"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# **eQTL**"],"metadata":{"id":"-YZZEi39Zqk4"}},{"cell_type":"markdown","source":["**Database:**\n","\n","[GTEx](https://gtexportal.org/home/datasets)\n","- Version: 8\n","- Ensembl: 99\n","- Human Genome Build: GRCh38\n","\n","\n","**Paper:**\n","\n","[\"The GTEx Consortium atlas of genetic regulatory effects across\n","human tissues\" Aguet et al., 2020](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7737656/pdf/nihms-1651674.pdf)"],"metadata":{"id":"nBby013EZ1w9"}},{"cell_type":"code","source":["%%R\n","# Create a function to pre-process the eQTL data files to merge them and obtain one dataset:\n","# This function pre-processes eQTL datasets for the MR_MtRobin method:\n","process_data <- function(input_folder,\n","                         pattern,\n","                         columns_to_select,\n","                         column_names,\n","                         output_path,\n","                         output_delimiter) {\n","\n","  # Load the required libraries:\n","  required_packages <- c(\"devtools\", \"dplyr\", \"utils\", \"tidyverse\", \"data.table\", \"R.utils\")\n","  installed_packages <- rownames(installed.packages())\n","  for (pkg in required_packages) {\n","    if (!pkg %in% installed_packages) {\n","      install.packages(pkg)\n","    }\n","    library(pkg, character.only = TRUE)\n","  }\n","\n","  # Read the input data:\n","  file_list <- base::list.files(path = input_folder, pattern = pattern) %>%\n","    base::lapply(function(x) read.table(file.path(input_folder, x),\n","                                        fill = TRUE, # In case the rows have unequal length, blank fields are implicitly added.\n","                                        header = TRUE))\n","\n","  # Check if the files were read:\n","  if (length(file_list) == 0) {\n","    stop(\"No files found matching the pattern: \", pattern)\n","  }\n","\n","  # Create a mapping between gene_id and gene_name:\n","  gene_id_to_name <- file_list %>%\n","    lapply(function(df) {\n","      df %>%\n","        select(gene_id, gene_name) %>%\n","        distinct()\n","    }) %>%\n","    bind_rows() %>%\n","    distinct()\n","\n","  # Subset, rename, and merge the data frames:\n","  eqtl <- file_list %>%\n","    lapply(function(df) {\n","      df %>%\n","        subset(select = columns_to_select) %>%\n","        rename(!!column_names$gene_id := gene_id,\n","               !!column_names$variant_id := rs_id_dbSNP151_GRCh38p7,\n","               !!column_names$chr_id := variant_id) %>%\n","        select(-gene_name) # Remove the gene_name column\n","    }) %>%\n","    # Join by gene_id, variant_id and chr_id\n","    # Don't join by gene_name because it is ambiguous\n","    purrr::reduce(full_join, by = c(\"gene_id\", \"variant_id\", \"chr_id\"))\n","\n","  # Add gene_name as the second column:\n","  eqtl <- eqtl %>%\n","    left_join(gene_id_to_name, by = \"gene_id\") %>%\n","    relocate(gene_name, .after = gene_id)\n","\n","  # Find the number of input files:\n","  num_files <- length(file_list)\n","\n","  # Generate column indices for beta, SE, and pvalue columns:\n","  beta_indices <- seq(5, 5 + 3 * (num_files - 1), 3)\n","  SE_indices <- seq(6, 6 + 3 * (num_files - 1), 3)\n","  pvalue_indices <- seq(7, 7 + 3 * (num_files - 1), 3)\n","\n","  # Subset the data frame:\n","  eqtl <- eqtl[, c(1, 2, 3, 4,\n","                   beta_indices,\n","                   SE_indices,\n","                   pvalue_indices)]\n","\n","  # Generate column names for beta, SE, and pvalue columns:\n","  if (is.null(column_names$beta_names)) {\n","    column_names$beta_names <- paste0(\"beta_\", 1:num_files)\n","  }\n","  if (is.null(column_names$SE_names)) {\n","    column_names$SE_names <- paste0(\"SE_\", 1:num_files)\n","  }\n","  if (is.null(column_names$pvalue_names)) {\n","    column_names$pvalue_names <- paste0(\"pvalue_\", 1:num_files)\n","  }\n","\n","  # Rename the columns:\n","  colnames(eqtl) <- c(column_names$gene_id,\n","                      column_names$gene_name,\n","                      column_names$variant_id,\n","                      column_names$chr_id,\n","                      column_names$beta_names,\n","                      column_names$SE_names,\n","                      column_names$pvalue_names)\n","\n","  # Define a function to count the non-NA values for beta_value, SE and p-value columns in each row:\n","  count_non_na <- function(row) {\n","    pval_count <- sum(!is.na(row[column_names$pvalue_names]))\n","    beta_count <- sum(!is.na(row[column_names$beta_names]))\n","    SE_count <- sum(!is.na(row[column_names$SE_names]))\n","    return(pval_count == beta_count && beta_count == SE_count)\n","  }\n","\n","  # Retain the rows that have the values in the beta, SE and pvalue corresponding columns:\n","  eqtl <- eqtl[apply(eqtl, 1, count_non_na),]\n","\n","  # Export the processed data to a .txt file:\n","  write.table(eqtl,\n","              file = file.path(output_path),\n","              sep = output_delimiter,\n","              row.names = FALSE,\n","              col.names = TRUE,\n","              quote = FALSE)\n","}"],"metadata":{"id":"rfMYPQMTZstX"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Call the function:\n","process_data(input_folder = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/Input\",\n","             pattern = \"*.txt.gz\",\n","             columns_to_select = c(\"gene_id\",\n","                                   \"gene_name\",\n","                                   \"rs_id_dbSNP151_GRCh38p7\",\n","                                   \"variant_id\",\n","                                   \"slope\",\n","                                   \"slope_se\",\n","                                   \"pval_nominal\"),\n","             column_names = list(gene_id = \"gene_id\",\n","                                 gene_name = \"gene_name\",\n","                                 variant_id = \"variant_id\",\n","                                 chr_id = \"chr_id\",\n","                                 beta_names = NULL,\n","                                 SE_names = NULL,\n","                                 pvalue_names = NULL),\n","             output_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues.txt\",\n","             output_delimiter = \"\\t\")"],"metadata":{"id":"BHSDSTWLaXUj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the previous file:\n","eqtl_all_tissues <- read.table(file = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues.txt\",\n","                               header = TRUE,\n","                               sep = \"\")\n","\n","print(dim(eqtl_all_tissues)) # Number of rows and columns\n","print(length(unique(eqtl_all_tissues$gene_id))) # Number of unique genes\n","print(length(unique(eqtl_all_tissues$variant_id))) # Number of unique rsIDs\n","print(head(eqtl_all_tissues)) # Check the format of the table"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5tFb_gJBazj-","executionInfo":{"status":"ok","timestamp":1691723475993,"user_tz":-570,"elapsed":56388,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"033d4edf-a9bd-422e-c7e1-fa2a70d5d3fe"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1008494     151\n","[1] 39832\n","[1] 828935\n","            gene_id     gene_name   variant_id              chr_id    beta_1\n","1 ENSG00000227232.5        WASH7P  rs769952832  chr1_64764_C_T_b38  0.586346\n","2 ENSG00000268903.1 RP11-34P13.15  rs866355763 chr1_103147_C_T_b38 -0.612097\n","3 ENSG00000269981.1 RP11-34P13.16   rs62642117 chr1_108826_G_C_b38  0.431229\n","4 ENSG00000241860.6 RP11-34P13.13  rs201327123  chr1_14677_G_A_b38  0.700658\n","5 ENSG00000279457.4 RP11-34P13.18  rs188376087 chr1_599167_G_A_b38 -0.687794\n","6 ENSG00000228463.9    AP006222.2 rs1206875823 chr1_280550_G_A_b38 -1.822030\n","     beta_2 beta_3   beta_4   beta_5    beta_6 beta_7    beta_8   beta_9\n","1        NA     NA       NA       NA        NA     NA        NA       NA\n","2        NA     NA       NA       NA        NA     NA        NA       NA\n","3        NA     NA 0.727510       NA  0.907156     NA        NA       NA\n","4  0.955911     NA 0.844491 0.841275  0.924220     NA  0.832645 0.855631\n","5        NA     NA       NA       NA -0.672016     NA        NA       NA\n","6 -2.041440     NA       NA       NA        NA     NA -1.708880       NA\n","  beta_10   beta_11 beta_12 beta_13  beta_14  beta_15  beta_16  beta_17 beta_18\n","1      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","2      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","3 1.13727  1.237960      NA      NA       NA 0.651921 0.616972       NA      NA\n","4 1.06242  0.919054 1.18909 1.07671 0.996264       NA 1.124170       NA      NA\n","5      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","6      NA -1.647010      NA      NA       NA       NA       NA -1.90221      NA\n","  beta_19   beta_20   beta_21 beta_22   beta_23  beta_24  beta_25 beta_26\n","1      NA        NA        NA      NA        NA 0.680547 0.525474      NA\n","2      NA        NA        NA      NA        NA       NA       NA      NA\n","3      NA        NA        NA      NA  0.647392 0.882308 0.717440      NA\n","4      NA  0.782287  0.870255      NA  0.944220 1.030600 1.001690      NA\n","5      NA -0.733600 -0.462612      NA        NA       NA       NA      NA\n","6      NA -2.062780        NA      NA -1.510250       NA       NA      NA\n","   beta_27  beta_28 beta_29 beta_30   beta_31   beta_32 beta_33   beta_34\n","1       NA       NA      NA      NA        NA        NA      NA        NA\n","2       NA       NA      NA      NA        NA        NA      NA        NA\n","3 0.596756 0.581917      NA      NA  0.700484        NA      NA        NA\n","4 0.984300 1.000880      NA      NA  0.890830  0.747842      NA        NA\n","5       NA       NA      NA      NA -0.696659 -0.588254      NA -0.534898\n","6       NA       NA      NA      NA        NA        NA      NA        NA\n","    beta_35  beta_36 beta_37  beta_38 beta_39  beta_40   beta_41 beta_42\n","1        NA       NA      NA       NA      NA 0.524447        NA      NA\n","2        NA       NA      NA       NA      NA       NA        NA      NA\n","3        NA       NA      NA 1.151320      NA 0.586588  0.732238      NA\n","4  1.050980 0.935812      NA 0.882921 1.00593 0.911987  0.857182 1.00837\n","5 -0.584404       NA      NA       NA      NA       NA -0.527948      NA\n","6        NA       NA      NA       NA      NA       NA        NA      NA\n","   beta_43 beta_44   beta_45   beta_46  beta_47 beta_48   beta_49     SE_1\n","1       NA      NA        NA        NA       NA      NA        NA 0.100677\n","2       NA      NA        NA        NA       NA      NA        NA 0.169958\n","3       NA 0.86881        NA  0.713296       NA      NA        NA 0.111226\n","4 0.626596      NA        NA  0.767395       NA 1.36956        NA 0.128601\n","5       NA      NA -0.714849 -0.532982       NA      NA -0.544131 0.129230\n","6       NA      NA        NA        NA -2.34025      NA        NA 0.189149\n","      SE_2 SE_3     SE_4     SE_5     SE_6 SE_7     SE_8     SE_9    SE_10\n","1       NA   NA       NA       NA       NA   NA       NA       NA       NA\n","2       NA   NA       NA       NA       NA   NA       NA       NA       NA\n","3       NA   NA 0.163159       NA 0.135021   NA       NA       NA 0.257834\n","4 0.148454   NA 0.166414 0.233686 0.134661   NA 0.270479 0.206367 0.255493\n","5       NA   NA       NA       NA 0.129430   NA       NA       NA       NA\n","6 0.210970   NA       NA       NA       NA   NA 0.388371       NA       NA\n","     SE_11    SE_12    SE_13    SE_14    SE_15    SE_16    SE_17 SE_18 SE_19\n","1       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","2       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","3 0.262474       NA       NA       NA 0.176732 0.152881       NA    NA    NA\n","4 0.229093 0.225087 0.216332 0.255306       NA 0.198307       NA    NA    NA\n","5       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","6 0.435729       NA       NA       NA       NA       NA 0.487461    NA    NA\n","     SE_20    SE_21 SE_22    SE_23    SE_24    SE_25 SE_26    SE_27    SE_28\n","1       NA       NA    NA       NA 0.110002 0.124357    NA       NA       NA\n","2       NA       NA    NA       NA       NA       NA    NA       NA       NA\n","3       NA       NA    NA 0.197378 0.172344 0.171137    NA 0.148811 0.155418\n","4 0.164811 0.143738    NA 0.186397 0.202106 0.189964    NA 0.152923 0.173899\n","5 0.143504 0.140358    NA       NA       NA       NA    NA       NA       NA\n","6 0.288671       NA    NA 0.279751       NA       NA    NA       NA       NA\n","  SE_29 SE_30    SE_31    SE_32 SE_33    SE_34    SE_35    SE_36 SE_37   SE_38\n","1    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","2    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","3    NA    NA 0.175460       NA    NA       NA       NA       NA    NA 0.26540\n","4    NA    NA 0.230550 0.139465    NA       NA 0.128636 0.264854    NA 0.21388\n","5    NA    NA 0.188241 0.136197    NA 0.105454 0.126366       NA    NA      NA\n","6    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","     SE_39    SE_40    SE_41    SE_42    SE_43    SE_44    SE_45    SE_46\n","1       NA 0.103390       NA       NA       NA       NA       NA       NA\n","2       NA       NA       NA       NA       NA       NA       NA       NA\n","3       NA 0.154386 0.150034       NA       NA 0.203341       NA 0.147434\n","4 0.226305 0.135983 0.129206 0.259747 0.201859       NA       NA 0.140433\n","5       NA       NA 0.119271       NA       NA       NA 0.176693 0.118812\n","6       NA       NA       NA       NA       NA       NA       NA       NA\n","     SE_47    SE_48    SE_49    pvalue_1    pvalue_2 pvalue_3    pvalue_4\n","1       NA       NA       NA 1.01661e-08          NA       NA          NA\n","2       NA       NA       NA 3.47332e-04          NA       NA          NA\n","3       NA       NA       NA 1.19525e-04          NA       NA 1.14483e-05\n","4       NA 0.341024       NA 7.91948e-08 3.45758e-10       NA 6.62770e-07\n","5       NA       NA 0.110842 1.53694e-07          NA       NA          NA\n","6 0.479929       NA       NA 2.72051e-20 4.93052e-20       NA          NA\n","     pvalue_5    pvalue_6 pvalue_7    pvalue_8   pvalue_9   pvalue_10\n","1          NA          NA       NA          NA         NA          NA\n","2          NA          NA       NA          NA         NA          NA\n","3          NA 4.87933e-11       NA          NA         NA 2.08612e-05\n","4 0.000415358 1.94366e-11       NA 2.57082e-03 5.5663e-05 5.66701e-05\n","5          NA 3.00058e-07       NA          NA         NA          NA\n","6          NA          NA       NA 2.33112e-05         NA          NA\n","    pvalue_11   pvalue_12   pvalue_13   pvalue_14   pvalue_15   pvalue_16\n","1          NA          NA          NA          NA          NA          NA\n","2          NA          NA          NA          NA          NA          NA\n","3 5.00013e-06          NA          NA          NA 0.000330066 8.37712e-05\n","4 9.03769e-05 3.97305e-07 1.93296e-06 0.000154857          NA 6.43853e-08\n","5          NA          NA          NA          NA          NA          NA\n","6 2.17289e-04          NA          NA          NA          NA          NA\n","    pvalue_17 pvalue_18 pvalue_19   pvalue_20   pvalue_21 pvalue_22   pvalue_23\n","1          NA        NA        NA          NA          NA        NA          NA\n","2          NA        NA        NA          NA          NA        NA          NA\n","3          NA        NA        NA          NA          NA        NA 1.17837e-03\n","4          NA        NA        NA 3.10205e-06 3.16761e-09        NA 7.67178e-07\n","5          NA        NA        NA 5.44298e-07 1.06566e-03        NA          NA\n","6 0.000152164        NA        NA 5.89145e-12          NA        NA 1.50137e-07\n","    pvalue_24   pvalue_25 pvalue_26   pvalue_27   pvalue_28 pvalue_29 pvalue_30\n","1 2.02537e-09 3.24586e-05        NA          NA          NA        NA        NA\n","2          NA          NA        NA          NA          NA        NA        NA\n","3 5.50613e-07 3.72830e-05        NA 7.25685e-05 2.16661e-04        NA        NA\n","4 6.07252e-07 2.71848e-07        NA 3.54545e-10 2.12053e-08        NA        NA\n","5          NA          NA        NA          NA          NA        NA        NA\n","6          NA          NA        NA          NA          NA        NA        NA\n","    pvalue_31   pvalue_32 pvalue_33   pvalue_34   pvalue_35   pvalue_36\n","1          NA          NA        NA          NA          NA          NA\n","2          NA          NA        NA          NA          NA          NA\n","3 9.76311e-05          NA        NA          NA          NA          NA\n","4 1.59114e-04 1.32367e-07        NA          NA 2.97619e-15 0.000572981\n","5 2.90924e-04 1.93308e-05        NA 5.16275e-07 4.87731e-06          NA\n","6          NA          NA        NA          NA          NA          NA\n","  pvalue_37   pvalue_38   pvalue_39   pvalue_40   pvalue_41   pvalue_42\n","1        NA          NA          NA 5.76471e-07          NA          NA\n","2        NA          NA          NA          NA          NA          NA\n","3        NA 2.29415e-05          NA 1.65067e-04 1.39875e-06          NA\n","4        NA 5.39767e-05 1.52886e-05 6.03296e-11 8.00361e-11 0.000161822\n","5        NA          NA          NA          NA 1.16151e-05          NA\n","6        NA          NA          NA          NA          NA          NA\n","   pvalue_43   pvalue_44   pvalue_45   pvalue_46   pvalue_47   pvalue_48\n","1         NA          NA          NA          NA          NA          NA\n","2         NA          NA          NA          NA          NA          NA\n","3         NA 2.68288e-05          NA 1.74502e-06          NA          NA\n","4 0.00220496          NA          NA 7.31007e-08          NA 0.000105414\n","5         NA          NA 6.83456e-05 9.00081e-06          NA          NA\n","6         NA          NA          NA          NA 3.90116e-06          NA\n","    pvalue_49\n","1          NA\n","2          NA\n","3          NA\n","4          NA\n","5 1.18089e-06\n","6          NA\n"]}]},{"cell_type":"markdown","source":["# **Ensembl**"],"metadata":{"id":"QoGjiXvjS-CO"}},{"cell_type":"markdown","source":["Each dataset was built in different Ensembl versions. It is a good practice to check and update them to the latest and same Ensembl version.\n","\n","- **Gene Annotations:** Different Ensembl versions might have different annotations for genes, transcripts, and regulatory regions. This could affect the identification of genetic variants that are used as instrumental variables in MR.\n","\n","- **Variant Mapping:** SNP IDs and their locations can change between versions. This may lead to inconsistencies in variant mapping across datasets, affecting the selection of instrumental variables.\n","\n","- **InDels and Structural Variants:** Information about insertions, deletions, and structural variants might vary across versions, which could impact the analysis if these types of variants are included.\n","\n","- **Functional Annotations:** Different versions may have different functional annotations (e.g., missense, synonymous) for variants, affecting the biological interpretation.\n","\n","- **Compatibility with Other Tools:** Tools used in MR analysis may be optimized for specific Ensembl versions, so using different versions might lead to incompatibilities or suboptimal performance.\n","\n","- **Population Stratification and Cohort Differences:** If the eQTL and GWAS datasets are from different populations or cohorts, differences in Ensembl versions might exacerbate underlying heterogeneity, further complicating the analysis."],"metadata":{"id":"JJL5WlqULwe-"}},{"cell_type":"markdown","source":["## **eQTL**"],"metadata":{"id":"ocHFRWlz_1km"}},{"cell_type":"markdown","source":["**Ensembl 88 to 110:**\n","\n","- Built using the GENCODE version 26 = Ensembl release version 88.\n","- It should be updated to the Ensembl release version 110.\n","\n","https://gtexportal.org/home/releaseInfoPage\n","\n","<img src=\"https://drive.google.com/uc?export=view&id=1COIP2zk_tZfJOzMSouefKDFKFoHQ9IGA\" alt=\"drawing\" width=\"300\"/>\n","\n","https://www.gencodegenes.org/human/releases.html\n","\n","<img src=\"https://drive.google.com/uc?export=view&id=1NlUXWwaVdW5cmm_nPF_kEvoHTXIfIZSa\" alt=\"drawing\" width=\"500\"/>"],"metadata":{"id":"RTecwGIl_ztZ"}},{"cell_type":"markdown","source":["**ENSG_IDs**\n","\n","Gene stable IDs (without version numbers) tend to be consistent, but it's still a good practice to verify."],"metadata":{"id":"8AGVyAH1C2ib"}},{"cell_type":"code","source":["%%R\n","# Step 1: Install and load dplyr:\n","install.packages(\"dplyr\")\n","library(dplyr)"],"metadata":{"id":"rq47kADNKu7t"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Step 2: Install and load BiocManager:\n","install.packages(\"BiocManager\")\n","library(BiocManager)"],"metadata":{"id":"roUazsR7Kye7"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Step 3: Install and load biomaRt:\n","BiocManager::install(\"biomaRt\")\n","library(biomaRt)"],"metadata":{"id":"YbaTWa7hK2EL"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Step 4: Open the eQTL file:\n","eqtl_all_tissues <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues.txt\",\n","                               header = TRUE,\n","                               sep = \"\",\n","                               stringsAsFactors = FALSE)\n","print(dim(eqtl_all_tissues))\n","print(head(eqtl_all_tissues))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mKvzZkAqK_XB","executionInfo":{"status":"ok","timestamp":1692594571206,"user_tz":-570,"elapsed":90255,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"940c5554-f291-4f6e-f1ba-9fd15168c13a"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1008494     151\n","            gene_id     gene_name   variant_id              chr_id    beta_1\n","1 ENSG00000227232.5        WASH7P  rs769952832  chr1_64764_C_T_b38  0.586346\n","2 ENSG00000268903.1 RP11-34P13.15  rs866355763 chr1_103147_C_T_b38 -0.612097\n","3 ENSG00000269981.1 RP11-34P13.16   rs62642117 chr1_108826_G_C_b38  0.431229\n","4 ENSG00000241860.6 RP11-34P13.13  rs201327123  chr1_14677_G_A_b38  0.700658\n","5 ENSG00000279457.4 RP11-34P13.18  rs188376087 chr1_599167_G_A_b38 -0.687794\n","6 ENSG00000228463.9    AP006222.2 rs1206875823 chr1_280550_G_A_b38 -1.822030\n","     beta_2 beta_3   beta_4   beta_5    beta_6 beta_7    beta_8   beta_9\n","1        NA     NA       NA       NA        NA     NA        NA       NA\n","2        NA     NA       NA       NA        NA     NA        NA       NA\n","3        NA     NA 0.727510       NA  0.907156     NA        NA       NA\n","4  0.955911     NA 0.844491 0.841275  0.924220     NA  0.832645 0.855631\n","5        NA     NA       NA       NA -0.672016     NA        NA       NA\n","6 -2.041440     NA       NA       NA        NA     NA -1.708880       NA\n","  beta_10   beta_11 beta_12 beta_13  beta_14  beta_15  beta_16  beta_17 beta_18\n","1      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","2      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","3 1.13727  1.237960      NA      NA       NA 0.651921 0.616972       NA      NA\n","4 1.06242  0.919054 1.18909 1.07671 0.996264       NA 1.124170       NA      NA\n","5      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","6      NA -1.647010      NA      NA       NA       NA       NA -1.90221      NA\n","  beta_19   beta_20   beta_21 beta_22   beta_23  beta_24  beta_25 beta_26\n","1      NA        NA        NA      NA        NA 0.680547 0.525474      NA\n","2      NA        NA        NA      NA        NA       NA       NA      NA\n","3      NA        NA        NA      NA  0.647392 0.882308 0.717440      NA\n","4      NA  0.782287  0.870255      NA  0.944220 1.030600 1.001690      NA\n","5      NA -0.733600 -0.462612      NA        NA       NA       NA      NA\n","6      NA -2.062780        NA      NA -1.510250       NA       NA      NA\n","   beta_27  beta_28 beta_29 beta_30   beta_31   beta_32 beta_33   beta_34\n","1       NA       NA      NA      NA        NA        NA      NA        NA\n","2       NA       NA      NA      NA        NA        NA      NA        NA\n","3 0.596756 0.581917      NA      NA  0.700484        NA      NA        NA\n","4 0.984300 1.000880      NA      NA  0.890830  0.747842      NA        NA\n","5       NA       NA      NA      NA -0.696659 -0.588254      NA -0.534898\n","6       NA       NA      NA      NA        NA        NA      NA        NA\n","    beta_35  beta_36 beta_37  beta_38 beta_39  beta_40   beta_41 beta_42\n","1        NA       NA      NA       NA      NA 0.524447        NA      NA\n","2        NA       NA      NA       NA      NA       NA        NA      NA\n","3        NA       NA      NA 1.151320      NA 0.586588  0.732238      NA\n","4  1.050980 0.935812      NA 0.882921 1.00593 0.911987  0.857182 1.00837\n","5 -0.584404       NA      NA       NA      NA       NA -0.527948      NA\n","6        NA       NA      NA       NA      NA       NA        NA      NA\n","   beta_43 beta_44   beta_45   beta_46  beta_47 beta_48   beta_49     SE_1\n","1       NA      NA        NA        NA       NA      NA        NA 0.100677\n","2       NA      NA        NA        NA       NA      NA        NA 0.169958\n","3       NA 0.86881        NA  0.713296       NA      NA        NA 0.111226\n","4 0.626596      NA        NA  0.767395       NA 1.36956        NA 0.128601\n","5       NA      NA -0.714849 -0.532982       NA      NA -0.544131 0.129230\n","6       NA      NA        NA        NA -2.34025      NA        NA 0.189149\n","      SE_2 SE_3     SE_4     SE_5     SE_6 SE_7     SE_8     SE_9    SE_10\n","1       NA   NA       NA       NA       NA   NA       NA       NA       NA\n","2       NA   NA       NA       NA       NA   NA       NA       NA       NA\n","3       NA   NA 0.163159       NA 0.135021   NA       NA       NA 0.257834\n","4 0.148454   NA 0.166414 0.233686 0.134661   NA 0.270479 0.206367 0.255493\n","5       NA   NA       NA       NA 0.129430   NA       NA       NA       NA\n","6 0.210970   NA       NA       NA       NA   NA 0.388371       NA       NA\n","     SE_11    SE_12    SE_13    SE_14    SE_15    SE_16    SE_17 SE_18 SE_19\n","1       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","2       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","3 0.262474       NA       NA       NA 0.176732 0.152881       NA    NA    NA\n","4 0.229093 0.225087 0.216332 0.255306       NA 0.198307       NA    NA    NA\n","5       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","6 0.435729       NA       NA       NA       NA       NA 0.487461    NA    NA\n","     SE_20    SE_21 SE_22    SE_23    SE_24    SE_25 SE_26    SE_27    SE_28\n","1       NA       NA    NA       NA 0.110002 0.124357    NA       NA       NA\n","2       NA       NA    NA       NA       NA       NA    NA       NA       NA\n","3       NA       NA    NA 0.197378 0.172344 0.171137    NA 0.148811 0.155418\n","4 0.164811 0.143738    NA 0.186397 0.202106 0.189964    NA 0.152923 0.173899\n","5 0.143504 0.140358    NA       NA       NA       NA    NA       NA       NA\n","6 0.288671       NA    NA 0.279751       NA       NA    NA       NA       NA\n","  SE_29 SE_30    SE_31    SE_32 SE_33    SE_34    SE_35    SE_36 SE_37   SE_38\n","1    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","2    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","3    NA    NA 0.175460       NA    NA       NA       NA       NA    NA 0.26540\n","4    NA    NA 0.230550 0.139465    NA       NA 0.128636 0.264854    NA 0.21388\n","5    NA    NA 0.188241 0.136197    NA 0.105454 0.126366       NA    NA      NA\n","6    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","     SE_39    SE_40    SE_41    SE_42    SE_43    SE_44    SE_45    SE_46\n","1       NA 0.103390       NA       NA       NA       NA       NA       NA\n","2       NA       NA       NA       NA       NA       NA       NA       NA\n","3       NA 0.154386 0.150034       NA       NA 0.203341       NA 0.147434\n","4 0.226305 0.135983 0.129206 0.259747 0.201859       NA       NA 0.140433\n","5       NA       NA 0.119271       NA       NA       NA 0.176693 0.118812\n","6       NA       NA       NA       NA       NA       NA       NA       NA\n","     SE_47    SE_48    SE_49    pvalue_1    pvalue_2 pvalue_3    pvalue_4\n","1       NA       NA       NA 1.01661e-08          NA       NA          NA\n","2       NA       NA       NA 3.47332e-04          NA       NA          NA\n","3       NA       NA       NA 1.19525e-04          NA       NA 1.14483e-05\n","4       NA 0.341024       NA 7.91948e-08 3.45758e-10       NA 6.62770e-07\n","5       NA       NA 0.110842 1.53694e-07          NA       NA          NA\n","6 0.479929       NA       NA 2.72051e-20 4.93052e-20       NA          NA\n","     pvalue_5    pvalue_6 pvalue_7    pvalue_8   pvalue_9   pvalue_10\n","1          NA          NA       NA          NA         NA          NA\n","2          NA          NA       NA          NA         NA          NA\n","3          NA 4.87933e-11       NA          NA         NA 2.08612e-05\n","4 0.000415358 1.94366e-11       NA 2.57082e-03 5.5663e-05 5.66701e-05\n","5          NA 3.00058e-07       NA          NA         NA          NA\n","6          NA          NA       NA 2.33112e-05         NA          NA\n","    pvalue_11   pvalue_12   pvalue_13   pvalue_14   pvalue_15   pvalue_16\n","1          NA          NA          NA          NA          NA          NA\n","2          NA          NA          NA          NA          NA          NA\n","3 5.00013e-06          NA          NA          NA 0.000330066 8.37712e-05\n","4 9.03769e-05 3.97305e-07 1.93296e-06 0.000154857          NA 6.43853e-08\n","5          NA          NA          NA          NA          NA          NA\n","6 2.17289e-04          NA          NA          NA          NA          NA\n","    pvalue_17 pvalue_18 pvalue_19   pvalue_20   pvalue_21 pvalue_22   pvalue_23\n","1          NA        NA        NA          NA          NA        NA          NA\n","2          NA        NA        NA          NA          NA        NA          NA\n","3          NA        NA        NA          NA          NA        NA 1.17837e-03\n","4          NA        NA        NA 3.10205e-06 3.16761e-09        NA 7.67178e-07\n","5          NA        NA        NA 5.44298e-07 1.06566e-03        NA          NA\n","6 0.000152164        NA        NA 5.89145e-12          NA        NA 1.50137e-07\n","    pvalue_24   pvalue_25 pvalue_26   pvalue_27   pvalue_28 pvalue_29 pvalue_30\n","1 2.02537e-09 3.24586e-05        NA          NA          NA        NA        NA\n","2          NA          NA        NA          NA          NA        NA        NA\n","3 5.50613e-07 3.72830e-05        NA 7.25685e-05 2.16661e-04        NA        NA\n","4 6.07252e-07 2.71848e-07        NA 3.54545e-10 2.12053e-08        NA        NA\n","5          NA          NA        NA          NA          NA        NA        NA\n","6          NA          NA        NA          NA          NA        NA        NA\n","    pvalue_31   pvalue_32 pvalue_33   pvalue_34   pvalue_35   pvalue_36\n","1          NA          NA        NA          NA          NA          NA\n","2          NA          NA        NA          NA          NA          NA\n","3 9.76311e-05          NA        NA          NA          NA          NA\n","4 1.59114e-04 1.32367e-07        NA          NA 2.97619e-15 0.000572981\n","5 2.90924e-04 1.93308e-05        NA 5.16275e-07 4.87731e-06          NA\n","6          NA          NA        NA          NA          NA          NA\n","  pvalue_37   pvalue_38   pvalue_39   pvalue_40   pvalue_41   pvalue_42\n","1        NA          NA          NA 5.76471e-07          NA          NA\n","2        NA          NA          NA          NA          NA          NA\n","3        NA 2.29415e-05          NA 1.65067e-04 1.39875e-06          NA\n","4        NA 5.39767e-05 1.52886e-05 6.03296e-11 8.00361e-11 0.000161822\n","5        NA          NA          NA          NA 1.16151e-05          NA\n","6        NA          NA          NA          NA          NA          NA\n","   pvalue_43   pvalue_44   pvalue_45   pvalue_46   pvalue_47   pvalue_48\n","1         NA          NA          NA          NA          NA          NA\n","2         NA          NA          NA          NA          NA          NA\n","3         NA 2.68288e-05          NA 1.74502e-06          NA          NA\n","4 0.00220496          NA          NA 7.31007e-08          NA 0.000105414\n","5         NA          NA 6.83456e-05 9.00081e-06          NA          NA\n","6         NA          NA          NA          NA 3.90116e-06          NA\n","    pvalue_49\n","1          NA\n","2          NA\n","3          NA\n","4          NA\n","5 1.18089e-06\n","6          NA\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 5: Delete the .# at the end of each gene_id:\n","eqtl_all_tissues$gene_id <- sub(\"\\\\..*\", \"\", eqtl_all_tissues$gene_id)\n","print(dim(eqtl_all_tissues))\n","print(head(eqtl_all_tissues))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"E9COxd_tRBL2","executionInfo":{"status":"ok","timestamp":1692594575081,"user_tz":-570,"elapsed":3482,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"8ec3c805-f5be-4fe9-9fec-43465593b559"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1008494     151\n","          gene_id     gene_name   variant_id              chr_id    beta_1\n","1 ENSG00000227232        WASH7P  rs769952832  chr1_64764_C_T_b38  0.586346\n","2 ENSG00000268903 RP11-34P13.15  rs866355763 chr1_103147_C_T_b38 -0.612097\n","3 ENSG00000269981 RP11-34P13.16   rs62642117 chr1_108826_G_C_b38  0.431229\n","4 ENSG00000241860 RP11-34P13.13  rs201327123  chr1_14677_G_A_b38  0.700658\n","5 ENSG00000279457 RP11-34P13.18  rs188376087 chr1_599167_G_A_b38 -0.687794\n","6 ENSG00000228463    AP006222.2 rs1206875823 chr1_280550_G_A_b38 -1.822030\n","     beta_2 beta_3   beta_4   beta_5    beta_6 beta_7    beta_8   beta_9\n","1        NA     NA       NA       NA        NA     NA        NA       NA\n","2        NA     NA       NA       NA        NA     NA        NA       NA\n","3        NA     NA 0.727510       NA  0.907156     NA        NA       NA\n","4  0.955911     NA 0.844491 0.841275  0.924220     NA  0.832645 0.855631\n","5        NA     NA       NA       NA -0.672016     NA        NA       NA\n","6 -2.041440     NA       NA       NA        NA     NA -1.708880       NA\n","  beta_10   beta_11 beta_12 beta_13  beta_14  beta_15  beta_16  beta_17 beta_18\n","1      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","2      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","3 1.13727  1.237960      NA      NA       NA 0.651921 0.616972       NA      NA\n","4 1.06242  0.919054 1.18909 1.07671 0.996264       NA 1.124170       NA      NA\n","5      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","6      NA -1.647010      NA      NA       NA       NA       NA -1.90221      NA\n","  beta_19   beta_20   beta_21 beta_22   beta_23  beta_24  beta_25 beta_26\n","1      NA        NA        NA      NA        NA 0.680547 0.525474      NA\n","2      NA        NA        NA      NA        NA       NA       NA      NA\n","3      NA        NA        NA      NA  0.647392 0.882308 0.717440      NA\n","4      NA  0.782287  0.870255      NA  0.944220 1.030600 1.001690      NA\n","5      NA -0.733600 -0.462612      NA        NA       NA       NA      NA\n","6      NA -2.062780        NA      NA -1.510250       NA       NA      NA\n","   beta_27  beta_28 beta_29 beta_30   beta_31   beta_32 beta_33   beta_34\n","1       NA       NA      NA      NA        NA        NA      NA        NA\n","2       NA       NA      NA      NA        NA        NA      NA        NA\n","3 0.596756 0.581917      NA      NA  0.700484        NA      NA        NA\n","4 0.984300 1.000880      NA      NA  0.890830  0.747842      NA        NA\n","5       NA       NA      NA      NA -0.696659 -0.588254      NA -0.534898\n","6       NA       NA      NA      NA        NA        NA      NA        NA\n","    beta_35  beta_36 beta_37  beta_38 beta_39  beta_40   beta_41 beta_42\n","1        NA       NA      NA       NA      NA 0.524447        NA      NA\n","2        NA       NA      NA       NA      NA       NA        NA      NA\n","3        NA       NA      NA 1.151320      NA 0.586588  0.732238      NA\n","4  1.050980 0.935812      NA 0.882921 1.00593 0.911987  0.857182 1.00837\n","5 -0.584404       NA      NA       NA      NA       NA -0.527948      NA\n","6        NA       NA      NA       NA      NA       NA        NA      NA\n","   beta_43 beta_44   beta_45   beta_46  beta_47 beta_48   beta_49     SE_1\n","1       NA      NA        NA        NA       NA      NA        NA 0.100677\n","2       NA      NA        NA        NA       NA      NA        NA 0.169958\n","3       NA 0.86881        NA  0.713296       NA      NA        NA 0.111226\n","4 0.626596      NA        NA  0.767395       NA 1.36956        NA 0.128601\n","5       NA      NA -0.714849 -0.532982       NA      NA -0.544131 0.129230\n","6       NA      NA        NA        NA -2.34025      NA        NA 0.189149\n","      SE_2 SE_3     SE_4     SE_5     SE_6 SE_7     SE_8     SE_9    SE_10\n","1       NA   NA       NA       NA       NA   NA       NA       NA       NA\n","2       NA   NA       NA       NA       NA   NA       NA       NA       NA\n","3       NA   NA 0.163159       NA 0.135021   NA       NA       NA 0.257834\n","4 0.148454   NA 0.166414 0.233686 0.134661   NA 0.270479 0.206367 0.255493\n","5       NA   NA       NA       NA 0.129430   NA       NA       NA       NA\n","6 0.210970   NA       NA       NA       NA   NA 0.388371       NA       NA\n","     SE_11    SE_12    SE_13    SE_14    SE_15    SE_16    SE_17 SE_18 SE_19\n","1       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","2       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","3 0.262474       NA       NA       NA 0.176732 0.152881       NA    NA    NA\n","4 0.229093 0.225087 0.216332 0.255306       NA 0.198307       NA    NA    NA\n","5       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","6 0.435729       NA       NA       NA       NA       NA 0.487461    NA    NA\n","     SE_20    SE_21 SE_22    SE_23    SE_24    SE_25 SE_26    SE_27    SE_28\n","1       NA       NA    NA       NA 0.110002 0.124357    NA       NA       NA\n","2       NA       NA    NA       NA       NA       NA    NA       NA       NA\n","3       NA       NA    NA 0.197378 0.172344 0.171137    NA 0.148811 0.155418\n","4 0.164811 0.143738    NA 0.186397 0.202106 0.189964    NA 0.152923 0.173899\n","5 0.143504 0.140358    NA       NA       NA       NA    NA       NA       NA\n","6 0.288671       NA    NA 0.279751       NA       NA    NA       NA       NA\n","  SE_29 SE_30    SE_31    SE_32 SE_33    SE_34    SE_35    SE_36 SE_37   SE_38\n","1    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","2    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","3    NA    NA 0.175460       NA    NA       NA       NA       NA    NA 0.26540\n","4    NA    NA 0.230550 0.139465    NA       NA 0.128636 0.264854    NA 0.21388\n","5    NA    NA 0.188241 0.136197    NA 0.105454 0.126366       NA    NA      NA\n","6    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","     SE_39    SE_40    SE_41    SE_42    SE_43    SE_44    SE_45    SE_46\n","1       NA 0.103390       NA       NA       NA       NA       NA       NA\n","2       NA       NA       NA       NA       NA       NA       NA       NA\n","3       NA 0.154386 0.150034       NA       NA 0.203341       NA 0.147434\n","4 0.226305 0.135983 0.129206 0.259747 0.201859       NA       NA 0.140433\n","5       NA       NA 0.119271       NA       NA       NA 0.176693 0.118812\n","6       NA       NA       NA       NA       NA       NA       NA       NA\n","     SE_47    SE_48    SE_49    pvalue_1    pvalue_2 pvalue_3    pvalue_4\n","1       NA       NA       NA 1.01661e-08          NA       NA          NA\n","2       NA       NA       NA 3.47332e-04          NA       NA          NA\n","3       NA       NA       NA 1.19525e-04          NA       NA 1.14483e-05\n","4       NA 0.341024       NA 7.91948e-08 3.45758e-10       NA 6.62770e-07\n","5       NA       NA 0.110842 1.53694e-07          NA       NA          NA\n","6 0.479929       NA       NA 2.72051e-20 4.93052e-20       NA          NA\n","     pvalue_5    pvalue_6 pvalue_7    pvalue_8   pvalue_9   pvalue_10\n","1          NA          NA       NA          NA         NA          NA\n","2          NA          NA       NA          NA         NA          NA\n","3          NA 4.87933e-11       NA          NA         NA 2.08612e-05\n","4 0.000415358 1.94366e-11       NA 2.57082e-03 5.5663e-05 5.66701e-05\n","5          NA 3.00058e-07       NA          NA         NA          NA\n","6          NA          NA       NA 2.33112e-05         NA          NA\n","    pvalue_11   pvalue_12   pvalue_13   pvalue_14   pvalue_15   pvalue_16\n","1          NA          NA          NA          NA          NA          NA\n","2          NA          NA          NA          NA          NA          NA\n","3 5.00013e-06          NA          NA          NA 0.000330066 8.37712e-05\n","4 9.03769e-05 3.97305e-07 1.93296e-06 0.000154857          NA 6.43853e-08\n","5          NA          NA          NA          NA          NA          NA\n","6 2.17289e-04          NA          NA          NA          NA          NA\n","    pvalue_17 pvalue_18 pvalue_19   pvalue_20   pvalue_21 pvalue_22   pvalue_23\n","1          NA        NA        NA          NA          NA        NA          NA\n","2          NA        NA        NA          NA          NA        NA          NA\n","3          NA        NA        NA          NA          NA        NA 1.17837e-03\n","4          NA        NA        NA 3.10205e-06 3.16761e-09        NA 7.67178e-07\n","5          NA        NA        NA 5.44298e-07 1.06566e-03        NA          NA\n","6 0.000152164        NA        NA 5.89145e-12          NA        NA 1.50137e-07\n","    pvalue_24   pvalue_25 pvalue_26   pvalue_27   pvalue_28 pvalue_29 pvalue_30\n","1 2.02537e-09 3.24586e-05        NA          NA          NA        NA        NA\n","2          NA          NA        NA          NA          NA        NA        NA\n","3 5.50613e-07 3.72830e-05        NA 7.25685e-05 2.16661e-04        NA        NA\n","4 6.07252e-07 2.71848e-07        NA 3.54545e-10 2.12053e-08        NA        NA\n","5          NA          NA        NA          NA          NA        NA        NA\n","6          NA          NA        NA          NA          NA        NA        NA\n","    pvalue_31   pvalue_32 pvalue_33   pvalue_34   pvalue_35   pvalue_36\n","1          NA          NA        NA          NA          NA          NA\n","2          NA          NA        NA          NA          NA          NA\n","3 9.76311e-05          NA        NA          NA          NA          NA\n","4 1.59114e-04 1.32367e-07        NA          NA 2.97619e-15 0.000572981\n","5 2.90924e-04 1.93308e-05        NA 5.16275e-07 4.87731e-06          NA\n","6          NA          NA        NA          NA          NA          NA\n","  pvalue_37   pvalue_38   pvalue_39   pvalue_40   pvalue_41   pvalue_42\n","1        NA          NA          NA 5.76471e-07          NA          NA\n","2        NA          NA          NA          NA          NA          NA\n","3        NA 2.29415e-05          NA 1.65067e-04 1.39875e-06          NA\n","4        NA 5.39767e-05 1.52886e-05 6.03296e-11 8.00361e-11 0.000161822\n","5        NA          NA          NA          NA 1.16151e-05          NA\n","6        NA          NA          NA          NA          NA          NA\n","   pvalue_43   pvalue_44   pvalue_45   pvalue_46   pvalue_47   pvalue_48\n","1         NA          NA          NA          NA          NA          NA\n","2         NA          NA          NA          NA          NA          NA\n","3         NA 2.68288e-05          NA 1.74502e-06          NA          NA\n","4 0.00220496          NA          NA 7.31007e-08          NA 0.000105414\n","5         NA          NA 6.83456e-05 9.00081e-06          NA          NA\n","6         NA          NA          NA          NA 3.90116e-06          NA\n","    pvalue_49\n","1          NA\n","2          NA\n","3          NA\n","4          NA\n","5 1.18089e-06\n","6          NA\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 6: Print the current Mart database:\n","print(listMarts())"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"cWkWiIk5RcrE","executionInfo":{"status":"ok","timestamp":1692154688961,"user_tz":-570,"elapsed":1000,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"9d09b98b-f2bc-4d80-eadd-260ce21c41fc"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["               biomart                version\n","1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 110\n","2   ENSEMBL_MART_MOUSE      Mouse strains 110\n","3     ENSEMBL_MART_SNP  Ensembl Variation 110\n","4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 110\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 7: Print the archived databases links:\n","attributes <- listEnsemblArchives()\n","print(attributes[,1:4])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"T_SDpr6GSABz","executionInfo":{"status":"ok","timestamp":1692156315847,"user_tz":-570,"elapsed":565,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"0721a55c-0652-49e5-d6d2-43cc3b15fb25"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["             name     date                                 url version\n","1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37\n","2     Ensembl 110 Jul 2023 https://jul2023.archive.ensembl.org     110\n","3     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109\n","4     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108\n","5     Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107\n","6     Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106\n","7     Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105\n","8     Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104\n","9     Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103\n","10    Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102\n","11    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101\n","12    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100\n","13     Ensembl 99 Jan 2020 https://jan2020.archive.ensembl.org      99\n","14     Ensembl 98 Sep 2019 https://sep2019.archive.ensembl.org      98\n","15     Ensembl 97 Jul 2019 https://jul2019.archive.ensembl.org      97\n","16     Ensembl 96 Apr 2019 https://apr2019.archive.ensembl.org      96\n","17     Ensembl 95 Jan 2019 https://jan2019.archive.ensembl.org      95\n","18     Ensembl 94 Oct 2018 https://oct2018.archive.ensembl.org      94\n","19     Ensembl 93 Jul 2018 https://jul2018.archive.ensembl.org      93\n","20     Ensembl 80 May 2015 https://may2015.archive.ensembl.org      80\n","21     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77\n","22     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75\n","23     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 8: Print the latest Mart update 110:\n","print(listMarts(host='https://jul2023.archive.ensembl.org'))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"1LkYkKa1ST9j","executionInfo":{"status":"ok","timestamp":1692154842393,"user_tz":-570,"elapsed":8505,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"4f7acf55-84a1-4ced-ea13-c79f7b7ff14f"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["               biomart                version\n","1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 110\n","2   ENSEMBL_MART_MOUSE      Mouse strains 110\n","3     ENSEMBL_MART_SNP  Ensembl Variation 110\n","4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 110\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 9: Create the query for the Mart 110:\n","# Obs.: The current eQTL version 8 was built using Mart 88:\n","ensembl_110 <- useMart(host='https://jul2023.archive.ensembl.org',\n","                       biomart='ENSEMBL_MART_ENSEMBL',\n","                       dataset='hsapiens_gene_ensembl')"],"metadata":{"id":"jplh7SvZSkWh"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Step 10: Check the current Ensembl attributes to select the desired ones:\n","print(head(listAttributes(ensembl_110)))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yY_BIPSTSxbp","executionInfo":{"status":"ok","timestamp":1692155337421,"user_tz":-570,"elapsed":5,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"f90e454b-9fe5-4306-9c60-a7d6b361cebc"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["                           name                  description         page\n","1               ensembl_gene_id               Gene stable ID feature_page\n","2       ensembl_gene_id_version       Gene stable ID version feature_page\n","3         ensembl_transcript_id         Transcript stable ID feature_page\n","4 ensembl_transcript_id_version Transcript stable ID version feature_page\n","5            ensembl_peptide_id            Protein stable ID feature_page\n","6    ensembl_peptide_id_version    Protein stable ID version feature_page\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 11: Check the current Ensembl filters to select the desired ones:\n","print(head(listFilters(ensembl_110)))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"NKaX8nIwUW6O","executionInfo":{"status":"ok","timestamp":1692155387645,"user_tz":-570,"elapsed":342,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"e1848826-6535-4857-fc23-9fff2c8003aa"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["             name              description\n","1 chromosome_name Chromosome/scaffold name\n","2           start                    Start\n","3             end                      End\n","4      band_start               Band Start\n","5        band_end                 Band End\n","6    marker_start             Marker Start\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 12: Get the Gene Information for Ensembl Version 110:\n","genes_110 <- getBM(attributes = c(\"ensembl_gene_id\"),\n","                   filters = \"ensembl_gene_id\",\n","                   values = eqtl_all_tissues$gene_id,\n","                   mart = ensembl_110)\n","\n","print(dim(genes_110))\n","print(head(genes_110))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"62ngfZIgUxq9","executionInfo":{"status":"ok","timestamp":1692595099188,"user_tz":-570,"elapsed":430191,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"e318e414-9b3d-4f26-db21-95ef680fdc9f"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 997207      1\n","  ensembl_gene_id\n","1 ENSG00000000457\n","2 ENSG00000000460\n","3 ENSG00000000938\n","4 ENSG00000000971\n","5 ENSG00000001460\n","6 ENSG00000001461\n"]}]},{"cell_type":"markdown","source":["Other important attributes for SNPs:\n","- `variation_name`\n","- `chromosome_name`\n","- `start_position`\n","- `end_position`\n","- `with_validated_snp`\n","- `allele`"],"metadata":{"id":"oTW1dzMQVwsc"}},{"cell_type":"code","source":["%%R\n","# Step 13: Compare the number of rows:\n","print(dim(eqtl_all_tissues)) # 1,008,494 rows\n","print(dim(genes_110)) # 997,207 rows\n","print(length(unique(eqtl_all_tissues$gene_id))) # 39,832 unique rows\n","print(length(unique(genes_110$ensembl_gene_id))) # 39,278 unique rows\n","# I lost 554 rows and 403 unique gene_id"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jf3FiRFgWgs3","executionInfo":{"status":"ok","timestamp":1692156332615,"user_tz":-570,"elapsed":334,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"3f483a29-7f99-46de-a3ee-8228b4d780ab"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1008494     151\n","[1] 997207      2\n","[1] 39832\n","[1] 39278\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 14: Check and filter to ensure that each ensembl_gene_id appears only once:\n","genes_110 <- genes_110 %>%\n","  group_by(ensembl_gene_id) %>%\n","  slice(1) %>%\n","  ungroup()"],"metadata":{"id":"N8NLQg6wWz7w"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Step 15: Update the eQTL dataset:\n","eqtl_all_tissues_110 <- eqtl_all_tissues %>%\n","  inner_join(genes_110, by = c(\"gene_id\" = \"ensembl_gene_id\"))\n","print(dim(eqtl_all_tissues_110))\n","print(head(eqtl_all_tissues_110))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Y98aA1TwKiLp","executionInfo":{"status":"ok","timestamp":1692595114270,"user_tz":-570,"elapsed":9863,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"885913d2-90ff-4ce0-aab8-519f740a145a"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 997207    151\n","          gene_id     gene_name   variant_id              chr_id    beta_1\n","1 ENSG00000227232        WASH7P  rs769952832  chr1_64764_C_T_b38  0.586346\n","2 ENSG00000268903 RP11-34P13.15  rs866355763 chr1_103147_C_T_b38 -0.612097\n","3 ENSG00000269981 RP11-34P13.16   rs62642117 chr1_108826_G_C_b38  0.431229\n","4 ENSG00000241860 RP11-34P13.13  rs201327123  chr1_14677_G_A_b38  0.700658\n","5 ENSG00000279457 RP11-34P13.18  rs188376087 chr1_599167_G_A_b38 -0.687794\n","6 ENSG00000228463    AP006222.2 rs1206875823 chr1_280550_G_A_b38 -1.822030\n","     beta_2 beta_3   beta_4   beta_5    beta_6 beta_7    beta_8   beta_9\n","1        NA     NA       NA       NA        NA     NA        NA       NA\n","2        NA     NA       NA       NA        NA     NA        NA       NA\n","3        NA     NA 0.727510       NA  0.907156     NA        NA       NA\n","4  0.955911     NA 0.844491 0.841275  0.924220     NA  0.832645 0.855631\n","5        NA     NA       NA       NA -0.672016     NA        NA       NA\n","6 -2.041440     NA       NA       NA        NA     NA -1.708880       NA\n","  beta_10   beta_11 beta_12 beta_13  beta_14  beta_15  beta_16  beta_17 beta_18\n","1      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","2      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","3 1.13727  1.237960      NA      NA       NA 0.651921 0.616972       NA      NA\n","4 1.06242  0.919054 1.18909 1.07671 0.996264       NA 1.124170       NA      NA\n","5      NA        NA      NA      NA       NA       NA       NA       NA      NA\n","6      NA -1.647010      NA      NA       NA       NA       NA -1.90221      NA\n","  beta_19   beta_20   beta_21 beta_22   beta_23  beta_24  beta_25 beta_26\n","1      NA        NA        NA      NA        NA 0.680547 0.525474      NA\n","2      NA        NA        NA      NA        NA       NA       NA      NA\n","3      NA        NA        NA      NA  0.647392 0.882308 0.717440      NA\n","4      NA  0.782287  0.870255      NA  0.944220 1.030600 1.001690      NA\n","5      NA -0.733600 -0.462612      NA        NA       NA       NA      NA\n","6      NA -2.062780        NA      NA -1.510250       NA       NA      NA\n","   beta_27  beta_28 beta_29 beta_30   beta_31   beta_32 beta_33   beta_34\n","1       NA       NA      NA      NA        NA        NA      NA        NA\n","2       NA       NA      NA      NA        NA        NA      NA        NA\n","3 0.596756 0.581917      NA      NA  0.700484        NA      NA        NA\n","4 0.984300 1.000880      NA      NA  0.890830  0.747842      NA        NA\n","5       NA       NA      NA      NA -0.696659 -0.588254      NA -0.534898\n","6       NA       NA      NA      NA        NA        NA      NA        NA\n","    beta_35  beta_36 beta_37  beta_38 beta_39  beta_40   beta_41 beta_42\n","1        NA       NA      NA       NA      NA 0.524447        NA      NA\n","2        NA       NA      NA       NA      NA       NA        NA      NA\n","3        NA       NA      NA 1.151320      NA 0.586588  0.732238      NA\n","4  1.050980 0.935812      NA 0.882921 1.00593 0.911987  0.857182 1.00837\n","5 -0.584404       NA      NA       NA      NA       NA -0.527948      NA\n","6        NA       NA      NA       NA      NA       NA        NA      NA\n","   beta_43 beta_44   beta_45   beta_46  beta_47 beta_48   beta_49     SE_1\n","1       NA      NA        NA        NA       NA      NA        NA 0.100677\n","2       NA      NA        NA        NA       NA      NA        NA 0.169958\n","3       NA 0.86881        NA  0.713296       NA      NA        NA 0.111226\n","4 0.626596      NA        NA  0.767395       NA 1.36956        NA 0.128601\n","5       NA      NA -0.714849 -0.532982       NA      NA -0.544131 0.129230\n","6       NA      NA        NA        NA -2.34025      NA        NA 0.189149\n","      SE_2 SE_3     SE_4     SE_5     SE_6 SE_7     SE_8     SE_9    SE_10\n","1       NA   NA       NA       NA       NA   NA       NA       NA       NA\n","2       NA   NA       NA       NA       NA   NA       NA       NA       NA\n","3       NA   NA 0.163159       NA 0.135021   NA       NA       NA 0.257834\n","4 0.148454   NA 0.166414 0.233686 0.134661   NA 0.270479 0.206367 0.255493\n","5       NA   NA       NA       NA 0.129430   NA       NA       NA       NA\n","6 0.210970   NA       NA       NA       NA   NA 0.388371       NA       NA\n","     SE_11    SE_12    SE_13    SE_14    SE_15    SE_16    SE_17 SE_18 SE_19\n","1       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","2       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","3 0.262474       NA       NA       NA 0.176732 0.152881       NA    NA    NA\n","4 0.229093 0.225087 0.216332 0.255306       NA 0.198307       NA    NA    NA\n","5       NA       NA       NA       NA       NA       NA       NA    NA    NA\n","6 0.435729       NA       NA       NA       NA       NA 0.487461    NA    NA\n","     SE_20    SE_21 SE_22    SE_23    SE_24    SE_25 SE_26    SE_27    SE_28\n","1       NA       NA    NA       NA 0.110002 0.124357    NA       NA       NA\n","2       NA       NA    NA       NA       NA       NA    NA       NA       NA\n","3       NA       NA    NA 0.197378 0.172344 0.171137    NA 0.148811 0.155418\n","4 0.164811 0.143738    NA 0.186397 0.202106 0.189964    NA 0.152923 0.173899\n","5 0.143504 0.140358    NA       NA       NA       NA    NA       NA       NA\n","6 0.288671       NA    NA 0.279751       NA       NA    NA       NA       NA\n","  SE_29 SE_30    SE_31    SE_32 SE_33    SE_34    SE_35    SE_36 SE_37   SE_38\n","1    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","2    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","3    NA    NA 0.175460       NA    NA       NA       NA       NA    NA 0.26540\n","4    NA    NA 0.230550 0.139465    NA       NA 0.128636 0.264854    NA 0.21388\n","5    NA    NA 0.188241 0.136197    NA 0.105454 0.126366       NA    NA      NA\n","6    NA    NA       NA       NA    NA       NA       NA       NA    NA      NA\n","     SE_39    SE_40    SE_41    SE_42    SE_43    SE_44    SE_45    SE_46\n","1       NA 0.103390       NA       NA       NA       NA       NA       NA\n","2       NA       NA       NA       NA       NA       NA       NA       NA\n","3       NA 0.154386 0.150034       NA       NA 0.203341       NA 0.147434\n","4 0.226305 0.135983 0.129206 0.259747 0.201859       NA       NA 0.140433\n","5       NA       NA 0.119271       NA       NA       NA 0.176693 0.118812\n","6       NA       NA       NA       NA       NA       NA       NA       NA\n","     SE_47    SE_48    SE_49    pvalue_1    pvalue_2 pvalue_3    pvalue_4\n","1       NA       NA       NA 1.01661e-08          NA       NA          NA\n","2       NA       NA       NA 3.47332e-04          NA       NA          NA\n","3       NA       NA       NA 1.19525e-04          NA       NA 1.14483e-05\n","4       NA 0.341024       NA 7.91948e-08 3.45758e-10       NA 6.62770e-07\n","5       NA       NA 0.110842 1.53694e-07          NA       NA          NA\n","6 0.479929       NA       NA 2.72051e-20 4.93052e-20       NA          NA\n","     pvalue_5    pvalue_6 pvalue_7    pvalue_8   pvalue_9   pvalue_10\n","1          NA          NA       NA          NA         NA          NA\n","2          NA          NA       NA          NA         NA          NA\n","3          NA 4.87933e-11       NA          NA         NA 2.08612e-05\n","4 0.000415358 1.94366e-11       NA 2.57082e-03 5.5663e-05 5.66701e-05\n","5          NA 3.00058e-07       NA          NA         NA          NA\n","6          NA          NA       NA 2.33112e-05         NA          NA\n","    pvalue_11   pvalue_12   pvalue_13   pvalue_14   pvalue_15   pvalue_16\n","1          NA          NA          NA          NA          NA          NA\n","2          NA          NA          NA          NA          NA          NA\n","3 5.00013e-06          NA          NA          NA 0.000330066 8.37712e-05\n","4 9.03769e-05 3.97305e-07 1.93296e-06 0.000154857          NA 6.43853e-08\n","5          NA          NA          NA          NA          NA          NA\n","6 2.17289e-04          NA          NA          NA          NA          NA\n","    pvalue_17 pvalue_18 pvalue_19   pvalue_20   pvalue_21 pvalue_22   pvalue_23\n","1          NA        NA        NA          NA          NA        NA          NA\n","2          NA        NA        NA          NA          NA        NA          NA\n","3          NA        NA        NA          NA          NA        NA 1.17837e-03\n","4          NA        NA        NA 3.10205e-06 3.16761e-09        NA 7.67178e-07\n","5          NA        NA        NA 5.44298e-07 1.06566e-03        NA          NA\n","6 0.000152164        NA        NA 5.89145e-12          NA        NA 1.50137e-07\n","    pvalue_24   pvalue_25 pvalue_26   pvalue_27   pvalue_28 pvalue_29 pvalue_30\n","1 2.02537e-09 3.24586e-05        NA          NA          NA        NA        NA\n","2          NA          NA        NA          NA          NA        NA        NA\n","3 5.50613e-07 3.72830e-05        NA 7.25685e-05 2.16661e-04        NA        NA\n","4 6.07252e-07 2.71848e-07        NA 3.54545e-10 2.12053e-08        NA        NA\n","5          NA          NA        NA          NA          NA        NA        NA\n","6          NA          NA        NA          NA          NA        NA        NA\n","    pvalue_31   pvalue_32 pvalue_33   pvalue_34   pvalue_35   pvalue_36\n","1          NA          NA        NA          NA          NA          NA\n","2          NA          NA        NA          NA          NA          NA\n","3 9.76311e-05          NA        NA          NA          NA          NA\n","4 1.59114e-04 1.32367e-07        NA          NA 2.97619e-15 0.000572981\n","5 2.90924e-04 1.93308e-05        NA 5.16275e-07 4.87731e-06          NA\n","6          NA          NA        NA          NA          NA          NA\n","  pvalue_37   pvalue_38   pvalue_39   pvalue_40   pvalue_41   pvalue_42\n","1        NA          NA          NA 5.76471e-07          NA          NA\n","2        NA          NA          NA          NA          NA          NA\n","3        NA 2.29415e-05          NA 1.65067e-04 1.39875e-06          NA\n","4        NA 5.39767e-05 1.52886e-05 6.03296e-11 8.00361e-11 0.000161822\n","5        NA          NA          NA          NA 1.16151e-05          NA\n","6        NA          NA          NA          NA          NA          NA\n","   pvalue_43   pvalue_44   pvalue_45   pvalue_46   pvalue_47   pvalue_48\n","1         NA          NA          NA          NA          NA          NA\n","2         NA          NA          NA          NA          NA          NA\n","3         NA 2.68288e-05          NA 1.74502e-06          NA          NA\n","4 0.00220496          NA          NA 7.31007e-08          NA 0.000105414\n","5         NA          NA 6.83456e-05 9.00081e-06          NA          NA\n","6         NA          NA          NA          NA 3.90116e-06          NA\n","    pvalue_49\n","1          NA\n","2          NA\n","3          NA\n","4          NA\n","5 1.18089e-06\n","6          NA\n"]}]},{"cell_type":"code","source":["%%R\n","# Step 16: Export the dataset:\n","write.table(eqtl_all_tissues_110,\n","            \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","            sep = \"\\t\",\n","            row.names = FALSE,\n","            quote = FALSE,\n","            col.names = TRUE)"],"metadata":{"id":"L1QXJz8QTNqL"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**rsIDs**\n","\n","Check if there are differences in the rsIDs of the eQTL based on Ensembl 88 and the rsIDs from Ensembl 110:"],"metadata":{"id":"7lGP9nCMCtkr"}},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q normal\n","\n","# Request resources:\n","#PBS -l ncpus=1\n","#PBS -l mem=100GB\n","#PBS -l jobfs=10GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Specify paths\n","EQTL_FILE=\"/scratch/sq95/sp6154/eQTL/chrID_to_rsID_eQTL/eqtl_all_tissues.txt\"\n","TEMP_EQTL_RSID=\"/scratch/sq95/sp6154/eQTL/chrID_to_rsID_eQTL/eqtl_rsid.tmp\"\n","TEMP_VCF_RSID=\"/scratch/sq95/sp6154/eQTL/chrID_to_rsID_eQTL/vcf_rsid.tmp\"\n","RESULT_FILE=\"/scratch/sq95/sp6154/eQTL/chrID_to_rsID_eQTL/missing_rsids.txt\"\n","\n","# Extract rsIDs from EQTL dataset\n","awk 'NR>1 {print $3}' $EQTL_FILE > $TEMP_EQTL_RSID\n","\n","# Extract rsIDs from all VCF files, concatenate, sort, and deduplicate\n","> $TEMP_VCF_RSID # Initialize the temp VCF rsID file\n","for i in {1..22} X Y; do\n","    zcat \"/scratch/sq95/sp6154/eQTL/chrID_to_rsID_eQTL/version_110/homo_sapiens-chr${i}.vcf.gz\" | awk '!/^#/ {print $3}' >> $TEMP_VCF_RSID\n","done\n","sort $TEMP_VCF_RSID | uniq > ${TEMP_VCF_RSID}.sorted\n","mv ${TEMP_VCF_RSID}.sorted $TEMP_VCF_RSID\n","\n","# Find rsIDs from EQTL dataset not in VCF files and save to RESULT_FILE\n","grep -v -f $TEMP_VCF_RSID $TEMP_EQTL_RSID > $RESULT_FILE\n","\n","# Cleanup\n","rm $TEMP_EQTL_RSID $TEMP_VCF_RSID\n","\n","echo \"Missing rsIDs saved in $RESULT_FILE\"\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"9lwfsy-AEKo5"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**Result:**\n","\n","I got `0-row` in the `missing_rsids.txt file`. It means, there are not missing rsIDs in the `eqtl_all_tissues.txt` file\n","\n","All the SNPs in the `eqtl_all_tissues.txt` based on Ensembl 88 are present in the Ensembl 110."],"metadata":{"id":"qxYQG4C2OseO"}},{"cell_type":"markdown","source":["## **GWAS**"],"metadata":{"id":"MROSdAFCBYUA"}},{"cell_type":"markdown","source":["**Ensembl 109 to 110**:\n","\n","`rsID`:\n","- Check for any deprecated or merged rsIDs by comparing with Ensembl 110.\n","- The rsIDs will remain consistent across different Ensembl releases and other databases like dbSNP. An rsID, once assigned to a specific variant, does not change its meaning across versions.\n","- **Content**: While the rsID's meaning remains consistent, the content or the number of SNPs in different releases may vary. New SNPs might be added, and some might be removed or merged.\n","- **Annotations**: The annotations or additional information related to each SNP (like its effect, associated traits, etc.) might be updated or refined in newer releases.\n","- **Genomic Coordinates**: As reference genomes get updated, the genomic coordinates of SNPs might change. This doesn't change the rsID, but the position on a chromosome might be different between two versions of the reference genome.\n","- **File Formats and Structure**: The structure or format of the files provided by Ensembl might differ between releases. While the core information remains consistent, there might be additional columns or fields in newer releases.\n","\n","Check if there are differences in the rsIDs of the GWAS datasets based on Ensembl 109 and the rsIDs from Ensembl 110:"],"metadata":{"id":"ZSKoINLVXqhm"}},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q normal\n","\n","# Request resources:\n","#PBS -l ncpus=1\n","#PBS -l mem=100GB\n","#PBS -l jobfs=10GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","\n","# Specify the work to be done:\n","\n","# Temporary file for the rsIDs from the GWAS dataset\n","TEMP_GWAS_RSID=\"/scratch/sq95/sp6154/GWAS/gwas_rsid.tmp\"\n","\n","# Results file prefix\n","RESULT_PREFIX=\"/scratch/sq95/sp6154/GWAS/missing_rsids\"\n","\n","# Loop through the GWAS datasets\n","for GWAS_FILE in long_covid_gwas_1_subset.txt long_covid_gwas_2_subset.txt long_covid_gwas_3_subset.txt long_covid_gwas_4_subset.txt\n","do\n","    # Extract the rsIDs from the current GWAS dataset\n","    awk 'NR>1 {print $1}' $GWAS_FILE > $TEMP_GWAS_RSID\n","\n","    # Generate a distinct result file name based on the current GWAS dataset\n","    RESULT_FILE=\"missing_rsids_$(basename $GWAS_FILE .txt).result\"\n","\n","    # Empty the results file (initialize)\n","    > $RESULT_FILE\n","\n","    # Loop through the VCF files\n","    for i in {1..22} X Y\n","    do\n","        # Path to the VCF file\n","        VCF_FILE=\"/scratch/sq95/sp6154/eQTL/chrID_to_rsID_eQTL/version_110/homo_sapiens-chr${i}.vcf.gz\"\n","\n","        # Use zcat to handle gzipped files and awk to extract the 3rd column (ID)\n","        # Then use grep to find which rsIDs from the GWAS dataset are present in the VCF file\n","        # And use grep again to remove them from the TEMP_GWAS_RSID file\n","        zcat $VCF_FILE | awk '!/^#/ {print $3}' | grep -f - $TEMP_GWAS_RSID | grep -v -f - $TEMP_GWAS_RSID > ${TEMP_GWAS_RSID}.temp && mv ${TEMP_GWAS_RSID}.temp $TEMP_GWAS_RSID\n","    done\n","\n","    # Save the remaining rsIDs to the result file\n","    cat $TEMP_GWAS_RSID >> $RESULT_FILE\n","\n","    echo \"Missing rsIDs for $GWAS_FILE saved in $RESULT_FILE\"\n","done\n","\n","# Remove temporary file\n","rm $TEMP_GWAS_RSID\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"grQRaJhMP7mS"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**Result:**\n","\n","I got `0-row` in the `missing_rsids.txt file`. It means, there are not missing rsIDs in any GWAS datasets.\n","\n","All the SNPs in the GWAS datasets based on Ensembl 109 are present in the Ensembl 110."],"metadata":{"id":"-_LIGmPVQZHn"}},{"cell_type":"markdown","source":["## **GWS**"],"metadata":{"id":"PTJldhdDGKyB"}},{"cell_type":"markdown","source":["**Ensembl 88**:\n","\n","- Subset the WGS dataset with the rsIDs from the eQTL dataset.\n","- Because the rsIDs in the eQTL dataset based on the Ensembl 88 are the same than those based on the Enseml 110, it is not necessary to convert the rsIDs from Ensembl 88 to Ensembl 110.\n","\n","**VCF file to be used if it is necessary to update the rsIDs from one Ensembl version to another one:**\n","\n","Both `homo_sapiens_somatic.vcf.gz` and `homo_sapiens.vcf.gz` are VCF (Variant Call Format) files, which contain information about variants found in the genome. However, they represent different types of variants:\n","\n","1. **homo_sapiens.vcf.gz**:\n","   - This file contains germline variants for Homo sapiens.\n","   - Germline variants are present in every cell of an individual and are inherited from parents. They are found in both somatic (body) cells and germ cells (sperm and egg).\n","   - These are the typical variants one might think of when considering genetic differences between individuals, such as the variants responsible for eye color, height, susceptibility to certain inherited diseases, etc.\n","\n","2. **homo_sapiens_somatic.vcf.gz**:\n","   - This file contains somatic variants for Homo sapiens.\n","   - Somatic variants arise during an individual's lifetime in a particular cell and are passed on only to the descendants of that cell. They are not present in germ cells and, therefore, are not inherited by offspring.\n","   - Somatic mutations are often associated with cancers, as they can accumulate in cells over time and may lead to uncontrolled cell growth if they occur in certain genes."],"metadata":{"id":"E0b3qbBAGNMX"}},{"cell_type":"code","source":["# Run the following script in GADI:\n","\n","# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q normal\n","\n","# Request resources:\n","#PBS -l ncpus=1\n","#PBS -l mem=100GB\n","#PBS -l jobfs=10GB\n","#PBS -l walltime=01:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Step 1: Extract the third (variant_id) and fourth (chr_id) columns from the eQTL file:\n","awk -F, 'FNR>1 && !seen[$3]++ {print $4, $3}' eqtl_all_tissues_110.txt > ids_to_update_unique.txt\n","\n","# Step 2: Create a GTEx.bim file with the rs_ID that matches in the ids_to_update_unique.txt and GTEx.bim files:\n","/scratch/sq95/sp6154/PLINK/plink --bfile WGS_GTEx --update-name ids_to_update_unique.txt --make-bed --out WGS_GTEx_covid_chrID\n","\n","# Step 3: Create a list of rs_IDs from the eQTL file:\n","awk -F, '{print $3}' eqtl_all_tissues_110.txt > rs_IDs.txt\n","\n","# Step 4: Change the chr_ID format to rs_ID in the GTEx file using the list with the rs_IDs obtained from the eQTL file:\n","/scratch/sq95/sp6154/PLINK/plink --bfile WGS_GTEx_covid_chrID --extract rs_IDs.txt --make-bed --out WGS_GTEx_covid_rsID\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"Ygbs57IcRyHj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%bash\n","# Check the results:\n","wc /content/drive/MyDrive/Colab/Long_COVID/GTEx/WGS_GTEx_covid_rsID.bim\n","head /content/drive/MyDrive/Colab/Long_COVID/GTEx/WGS_GTEx_covid_rsID.bim"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mCCeMRfEsXpp","executionInfo":{"status":"ok","timestamp":1692249945159,"user_tz":-570,"elapsed":1391,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"5783e369-8767-4a74-eb59-a009348404a3"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["  820792  4924752 23687328 /content/drive/MyDrive/Colab/MR_MtRobin/Long_COVID/GTEx/WGS_GTEx_covid_rsID.bim\n","1\trs554008981\t0\t13550\tA\tG\n","1\trs201327123\t0\t14677\tA\tG\n","1\trs62636368\t0\t16841\tT\tG\n","1\trs3891260\t0\t16856\tG\tA\n","1\trs372841554\t0\t17407\tA\tG\n","1\trs747093451\t0\t17408\tG\tC\n","1\trs376731495\t0\t17722\tG\tA\n","1\trs1190943651\t0\t17730\tA\tC\n","1\trs1209835338\t0\t20254\tA\tG\n","1\trs141149254\t0\t54490\tA\tG\n"]}]},{"cell_type":"markdown","source":["**Results:**\n","\n","**eQTL:** 828,935 SNPs\n","\n","**GTEx_filtered:** 820,792 SNPs\n","\n","I lost 8,143 SNPs"],"metadata":{"id":"La3I-Wyh9fCX"}},{"cell_type":"markdown","source":["# **MtRobin**"],"metadata":{"id":"wgvq0MITr5p2"}},{"cell_type":"markdown","source":["## **General Setup**"],"metadata":{"id":"KBGTltMRZ22m"}},{"cell_type":"markdown","source":["### **Packages**"],"metadata":{"id":"1Icd77h14fe7"}},{"cell_type":"code","source":["%%R\n","\n","# Install the lme4 package:\n","install.packages(\"lme4\")\n","library(lme4)"],"metadata":{"id":"8y2rf7qtZZXf"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Install and load the dplyr package:\n","install.packages(\"dplyr\")\n","library(dplyr)"],"metadata":{"id":"bEKRTugQZYD0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Install and load the httr package:\n","install.packages(\"httr\")\n","library(httr)"],"metadata":{"id":"1H8rKwlcZVQv"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Install and load the jsonlite package:\n","install.packages(\"jsonlite\")\n","library(jsonlite)"],"metadata":{"id":"Tyc1Fb4hbFdM"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### **GADI**"],"metadata":{"id":"TTmWqZeF4jbl"}},{"cell_type":"code","source":["# Create the .def file for the select_setup environment:\n","# Name: select_setup.def\n","\n","Bootstrap: docker\n","From: rocker/r-base:4.2.2\n","\n","%files\n","    /mnt/c/Users/pinsy007/Downloads/working/libicu60_60.2-3ubuntu3_amd64.deb /mnt/c/Users/pinsy007/Downloads/working/libicu60_60.2-3ubuntu3_amd64.deb\n","\n","%post\n","    apt-get update\n","    apt-get install -y libcurl4-openssl-dev libssl-dev libxml2-dev libicu-dev liblapack-dev libblas-dev gfortran\n","\n","    # Install libicu60 from the provided .deb file\n","    dpkg -i /mnt/c/Users/pinsy007/Downloads/working/libicu60_60.2-3ubuntu3_amd64.deb\n","\n","    echo \"Checking R lib paths...\"\n","    R -e \".libPaths()\"\n","\n","    echo \"Installing the packages separately...\"\n","    # External packages:\n","    R -e \"install.packages('matrixStats', repos='http://cran.rstudio.com/')\"\n","    R -e \"install.packages('reshape2', repos='http://cran.rstudio.com/')\"\n","    R -e \"install.packages('BiocManager', repos='http://cran.rstudio.com/')\"\n","    R -e \"BiocManager::install('snpStats')\"\n","    R -e \"install.packages('stringi', repos='http://cran.rstudio.com/')\"\n","\n","%runscript\n","    exec R \"$@\""],"metadata":{"id":"0YD-9qInRv8y"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Run the following command on my local computer to create the .sif file:\n","$ singularity build --fakeroot select_setup.sif select_setup.def"],"metadata":{"id":"Oc8Cg2OfRxGJ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Create the .def file for the main_function environment:\n","# Name: main_function.def\n","\n","Bootstrap: docker\n","From: rocker/r-base:4.2.2\n","\n","%post\n","    apt-get update && apt-get install -y cmake \\\n","    && apt-get clean \\\n","    && rm -rf /var/lib/apt/lists/*\n","\n","    R -e \"install.packages(c('nloptr', 'lme4', 'data.table'), repos='http://cran.rstudio.com/')\"\n","\n","%runscript\n","    exec R \"$@\""],"metadata":{"id":"mP77GC9JvOhT"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Run the following command on my local computer to create the .sif file:\n","$ singularity build --fakeroot main_function.sif main_function.def"],"metadata":{"id":"eCsR2hu5vn1p"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Create the .def file for the MR_MtRobin_resample environment to calculate the p_values and FDR:\n","# Name: p_values.def\n","\n","Bootstrap: docker\n","From: rocker/r-base:4.2.2\n","\n","%post\n","    apt-get update && apt-get install -y cmake \\\n","    && apt-get clean \\\n","    && rm -rf /var/lib/apt/lists/*\n","\n","    R -e \"install.packages(c('nloptr', 'lme4', 'mvtnorm', 'data.table'), repos='http://cran.rstudio.com/')\"\n","\n","%runscript\n","    exec R \"$@\""],"metadata":{"id":"cjYINoVD6s7V"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Run the following command on my local computer to create the .sif file:\n","$ sudo singularity build p_values.sif p_values.def"],"metadata":{"id":"cpuoIwlm7O6C"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["There are 2 functions for the selection of IVs that are used in statistical modeling, particularly in the context of multiple regression, but they can be applied in various other modeling contexts as well.\n","\n","1. **Forward selection:**\n","- **Starting Point:** Begins with an empty model (i.e., no predictors/independent variables).\n","- **Process**: At each step, the variable that gives the most significant improvement to the model is added.\n","- **Stopping Criterion**: Continues until adding more variables does not produce a statistically significant improvement in the model's fit.\n","- **Advantages**:\n","  - Can be computationally faster when there are many predictors since it doesn't start with a model containing all predictors.\n","  - It might never include some predictors if they don't meet the inclusion criteria.\n","\n","2. **Backward selection:**\n","- **Starting Point**: Begins with a full model that includes all potential predictors/independent variables.\n","- **Process**: At each step, the least significant variable (i.e., the one that contributes the least to the model) is removed.\n","- **Stopping Criterion**: Continues removing variables until removing more would result in a statistically significant loss of fit.\n","- **Advantages**:\n","  - It begins by considering all interactions between predictors, which might be missed in forward selection.\n","  - More likely to find a globally optimal model in certain situations.\n","\n","In summary, forward selection starts with no predictors and sequentially adds them, while backward elimination starts with all predictors and sequentially removes them.\n","\n","<img src=\"https://drive.google.com/uc?export=view&id=1ZxvYoCDcjA6CkqjcQOKcq4br8yoUy2yx\" alt=\"drawing\" width=\"700\"/>\n","\n","**Steps in the functions:**\n","1. **`select_IV_fs` (Forward Selection)**:\n","   - Start with an empty set of SNPs.\n","   - For each SNP (from `snp_set`), compute its p-value (`pval`).\n","   - If this p-value is below a threshold (`pval_thresh`), the SNP is considered significant and added to the SNP set for the gene.\n","   - Check for the LD (Linkage Disequilibrium) threshold and ensure that the newly added SNP is not in high LD with any of the previously added SNPs.\n","   - Filter each SNP based on the number of tissues (`nTiss_thresh`) in which the SNP is significant.\n","\n","2. **`select_IV_bs` (Backward Selection)**:\n","   - Start with a full set of SNPs based on the p-value threshold (`pval_thresh`).\n","   - Iteratively remove SNPs from this set:\n","     - Identify SNPs that are in high LD (greater than `ld_thresh`) with another SNP in the set.\n","     - Among these high LD SNPs, keep the one with the lowest p-value and remove the others.\n","   - Continue until no pairs of SNPs in the set have LD above the threshold.\n","\n","From the above descriptions, the criteria used for the performance (or for SNP inclusion/exclusion) in both functions are:\n","\n","1. **P-value of the SNP**: A threshold (`pval_thresh`) is used to determine if the SNP's association with the gene expression (or another phenotype) is statistically significant.\n","\n","2. **LD (Linkage Disequilibrium) between SNPs**: SNPs that are in high LD can provide redundant information, so there's a threshold (`ld_thresh`) to ensure selected SNPs are not in high LD.\n","\n","3. **Number of Tissues**: In the forward selection method, a SNP needs to be significant in a certain number of tissues (`nTiss_thresh`).\n","\n","To sum it up:\n","- Forward selection adds SNPs based on their p-values, ensuring LD criteria and the number of tissues are met.\n","- Backward selection removes SNPs based on LD criteria. The initial set is based on the p-value threshold."],"metadata":{"id":"v3GMCQ_Zzny1"}},{"cell_type":"markdown","source":["## **Common Functions**"],"metadata":{"id":"3Lt0qT9s61Xn"}},{"cell_type":"code","source":["%%R\n","\n","# Create a function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Load necessary libraries:\n","#install.packages(\"httr\")\n","#library(httr)\n","#install.packages(\"jsonlite\")\n","#library(jsonlite)\n","#install.packages(\"dplyr\")\n","#library(dplyr)\n","#install.packages(\"tidyverse\")\n","#library(tidyverse)\n","\n","analyze_gene_data <- function(result_main_path,\n","                              eqtl_long_covid_exposure_path,\n","                              save_files = list()) {\n","  # Read the input data:\n","  result_main <- readRDS(result_main_path)\n","  eqtl_long_covid_exposure <- read.table(eqtl_long_covid_exposure_path, header = TRUE, sep = \"\\t\", stringsAsFactors = FALSE)\n","  # Filter out errors:\n","  result_main <- result_main[!sapply(result_main, function(x) inherits(x, \"simpleError\"))]\n","  # Create a data frame:\n","  results <- data.frame(gene_id = character(),\n","                        beta_y = numeric(),\n","                        Obs = integer(),\n","                        SNPs = integer(),\n","                        REML = numeric(),\n","                        SD = numeric(),\n","                        Residual_SD = numeric(),\n","                        SNP_IDs = character(),\n","                        stringsAsFactors = FALSE)\n","  # Loop over each gene to extract each value:\n","  for (i in seq_along(result_main)) {\n","    lme_res <- result_main[[i]]$lme_res\n","    gene_id <- names(result_main)[i]  # get the gene id\n","    # Extract each required information:\n","    reml <- -2 * logLik(lme_res)\n","    sd <- attr(VarCorr(lme_res)$snpID, \"stddev\")\n","    res_sd <- sigma(lme_res)\n","    obs <- length(residuals(lme_res))\n","    snps <- length(unique(ranef(lme_res)$snpID[[1]]))\n","    beta_y <- fixef(lme_res)[\"beta_y\"]\n","    # Extract SNP IDs and convert to a single comma-separated string:\n","    snp_ids <- result_main[[i]][[\"lme_res\"]]@pp[[\".->Ut\"]]@Dimnames[[1]]\n","    snp_ids_str <- paste(snp_ids, collapse = \",\")\n","    # Add the results to the previous empty data frame including SNP IDs:\n","    results <- rbind(results, data.frame(gene_id = gene_id,\n","                                         beta_y = beta_y,\n","                                         Obs = obs,\n","                                         SNPs = snps,\n","                                         SNP_IDs = snp_ids_str,\n","                                         REML = reml,\n","                                         SD = sd,\n","                                         Residual_SD = res_sd,\n","                                         row.names = NULL))\n","  }\n","  # Remove duplicates based on 'gene_id', keeping the first occurrence:\n","  eqtl_long_covid_exposure <- eqtl_long_covid_exposure[!duplicated(eqtl_long_covid_exposure$gene_id), ]\n","  # Merge the two tables based on gene_id:\n","  merged_data <- merge(results, eqtl_long_covid_exposure[c(\"gene_id\",\"gene_name\")], by = \"gene_id\")\n","  # Get the column names:\n","  cols <- colnames(merged_data)\n","  # Remove 'gene_name' from the list:\n","  cols <- cols[cols != \"gene_name\"]\n","  # Add 'gene_name' as the second column:\n","  cols <- c(cols[1], \"gene_name\", cols[2:length(cols)])\n","  # Reorder the columns:\n","  merged_data <- merged_data[, cols]\n","  # Order the results by absolute value of the beta_y column from highest to lowest for all genes:\n","  all_genes <- merged_data[order(-abs(merged_data$beta_y)), ]\n","  # Order the results by absolute value of the beta_y column from highest to lowest for TOP 10 genes:\n","  top10_beta <- merged_data[order(-abs(merged_data$beta_y)), ][1:10, ]\n","  # Function to get gene types in bulk from Ensembl ensuring JSON response:\n","  get_gene_types_bulk <- function(gene_names) {\n","    url <- \"https://rest.ensembl.org/lookup/symbol/homo_sapiens\"\n","    body <- list(\n","      symbols = gene_names\n","    )\n","    response <- POST(url, body = toJSON(body), add_headers(\"Content-Type\" = \"application/json\", \"Accept\" = \"application/json\"))\n","    if (http_type(response) == \"application/json\") {\n","      content <- fromJSON(content(response, \"text\", encoding = \"UTF-8\"))\n","      biotypes <- sapply(gene_names, function(gene) {\n","        if (!is.null(content[[gene]])) {\n","          return(content[[gene]]$biotype)\n","        } else {\n","          return(NA)\n","        }\n","      })\n","      return(biotypes)\n","    } else {\n","      return(rep(NA, length(gene_names)))\n","    }\n","  }\n","  # Split genes into smaller chunks (e.g., 100 genes per chunk) to avoid overloading the API:\n","  gene_chunks <- split(merged_data$gene_name, ceiling(seq_along(merged_data$gene_name)/100))\n","  all_biotypes <- unlist(lapply(gene_chunks, get_gene_types_bulk))\n","  # Create a copy of the original dataframe and add gene types to the new dataframe:\n","  all_genes <- merged_data\n","  all_genes$gene_type <- all_biotypes\n","  # Reorder the columns:\n","  all_genes <- merged_data %>%\n","    mutate(gene_type = all_biotypes) %>%\n","    select(1:2, gene_type, everything())\n","  # Filter and sort the new dataframe:\n","  protein_coding_genes <- all_genes %>%\n","    filter(gene_type == \"protein_coding\") %>%\n","    arrange(desc(abs(beta_y)))\n","  # Create a custom name mapping:\n","  name_mapping <- list(\n","    all_genes = \"all_genes\",\n","    protein_coding_genes = \"protein_coding_genes\"\n","  )\n","  # Print dimensions and head of the result data frames:\n","  result_tables <- list(\n","    all_genes = all_genes,\n","    protein_coding_genes = protein_coding_genes\n","  )\n","  # Save the result data frames:\n","  for (name in names(result_tables)) {\n","    custom_name <- name_mapping[[name]]\n","    if (!is.null(custom_name) && custom_name %in% names(save_files)) {\n","      save_path <- save_files[[custom_name]]\n","      if (!dir.exists(dirname(save_path))) {\n","        dir.create(dirname(save_path), recursive = TRUE)\n","      }\n","      # Use tab as the separator for .txt files:\n","      write.table(result_tables[[name]], file = save_path, sep = \"\\t\", row.names = FALSE, quote = FALSE)\n","      cat(paste(\"Saved\", custom_name, \"as\", custom_name, \"to\", save_path, \"\\n\"))  # Use custom_name here for correct referencing\n","    }\n","    cat(\"Dimensions of\", custom_name, \":\", dim(result_tables[[name]]), \"\\n\\n\")\n","    print(head(result_tables[[name]]))\n","    cat(\"\\n---\\n\")\n","  }\n","}"],"metadata":{"id":"eucMTZW_98zP"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Create a function to retrieve more information from the Mt_Robin and filter the significant genes:\n","\n","# Install and Load necessary libraries:\n","#install.packages(\"lme4\")\n","#library(lme4)\n","#install.packages(\"dplyr\")\n","#library(dplyr)\n","\n","# Define the function:\n","process_final_data <- function(input_path_1, input_path_2, input_path_3, output_path) {\n","\n","  # Set the number of decimals:\n","  options(digits=7) # Default by R\n","\n","  # Step 1: Open the input datasets\n","  results_1 <- readRDS(input_path_1)\n","  print(paste(\"Causal genes (noncoding genes and protein coding genes): \"))\n","  print(length(results_1))\n","  print(results_1[[1]])\n","\n","  results_2 <- read.table(input_path_2, header=TRUE, sep=\"\\t\")\n","  print(paste(\"Causal genes (protein coding genes): \"))\n","  print(dim(results_2))\n","  print(head(results_2))\n","\n","  MR_results <- read.table(input_path_3, header=TRUE, sep=\"\\t\")\n","  print(paste(\"Causal genes (noncoding genes and protein coding genes): \"))\n","  print(dim(MR_results))\n","  print(head(MR_results))\n","\n","  # Step 2: Initialize an empty dataframe with additional CI columns\n","  extracted_results <- data.frame(\n","    gene_id = character(),\n","    gene_name = character(),\n","    SNPs = integer(),\n","    SNPs_IDs = character(),\n","    Obs = integer(),\n","    beta_y = numeric(),\n","    Min = numeric(),\n","    First_Q = numeric(),\n","    Median = numeric(),\n","    Third_Q = numeric(),\n","    Max = numeric(),\n","    SD_RE = numeric(),\n","    Res_SD_RE = numeric(),\n","    Var_RE = numeric(),\n","    Res_Var_RE = numeric(),\n","    SE_FE = numeric(),\n","    t_value_FE = numeric(),\n","    REML = numeric(),\n","    p_value = numeric(),\n","    FDR = numeric(),\n","    CI_lower = numeric(),\n","    CI_upper = numeric(),\n","    stringsAsFactors = FALSE\n","  )\n","\n","  # Step 3: Loop over each gene ID in results_2, enrich with MR_results and calculate other statistics:\n","  for (gene_id in results_2$gene_id) {\n","    # Check if the gene ID exists in results_1:\n","    if (gene_id %in% names(results_1)) {\n","      lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","      # If lme_res is NULL, skip to the next iteration of the loop:\n","      if(is.null(lme_res)) next\n","      summary_result <- summary(lme_res)\n","      residuals_summary <- summary_result$residuals\n","      var_corr <- VarCorr(lme_res)\n","      beta_y <- summary_result$coefficients[\"beta_y\", \"Estimate\"]\n","      SE_FE <- summary_result$coefficients[\"beta_y\", \"Std. Error\"]\n","      # Calculate 95% Confidence Intervals:\n","      CI_lower <- beta_y - 1.96 * SE_FE\n","      CI_upper <- beta_y + 1.96 * SE_FE\n","      # Find the corresponding p_value and FDR in MR_results:\n","      MR_row <- MR_results[MR_results$gene_id == gene_id,]\n","      p_value <- ifelse(nrow(MR_row) > 0, MR_row$pvalue, NA)\n","      FDR <- ifelse(nrow(MR_row) > 0, MR_row$fdr, NA)\n","      # Extract and prepare additional metrics from lme_res and results_2:\n","      Min <- min(residuals_summary)\n","      First_Q <- quantile(residuals_summary, 0.25)\n","      Median <- median(residuals_summary)\n","      Third_Q <- quantile(residuals_summary, 0.75)\n","      Max <- max(residuals_summary)\n","      SD_RE <- sqrt(var_corr$snpID[1])\n","      Var_RE <- var_corr$snpID[1]\n","      Res_SD_RE <- attr(var_corr, \"sc\")\n","      Res_Var_RE <- Res_SD_RE^2\n","      t_value_FE <- summary_result$coefficients[\"beta_y\", \"t value\"]\n","      gene_row <- results_2[results_2$gene_id == gene_id,]\n","      gene_name <- gene_row$gene_name\n","      SNPs_IDs <- gene_row$SNP_IDs\n","      SNPs <- length(strsplit(as.character(SNPs_IDs), \",\")[[1]])\n","      Obs <- gene_row$Obs\n","      REML <- gene_row$REML\n","      # Append results to the dataframe:\n","      extracted_results <- rbind(extracted_results, data.frame(\n","        gene_id, gene_name, SNPs, SNPs_IDs, Obs, beta_y,\n","        Min, First_Q, Median, Third_Q, Max,\n","        SD_RE, Res_SD_RE, Var_RE, Res_Var_RE,\n","        SE_FE, t_value_FE, REML, p_value, FDR, CI_lower, CI_upper, stringsAsFactors = FALSE))\n","    }\n","  }\n","\n","  # Step 4: Remove all row names:\n","  rownames(extracted_results) <- NULL\n","\n","  # Step 5: Filter out genes with a p_value greater than 0.05 and an FDR greater than 0.05:\n","  extracted_results_filtered <- extracted_results %>%\n","    filter(p_value <= 0.05 & FDR <= 0.05)\n","\n","  # Step 6: Print dim and head table:\n","  print(paste(\"Final table (protein coding genes): \"))\n","  print(dim(extracted_results_filtered))\n","  print(extracted_results_filtered)\n","\n","  # Step 7: Export the table with just significant values (p_values <= 0.05):\n","  write.table(extracted_results_filtered, file = output_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","}"],"metadata":{"id":"bqWNTGbdF-rL"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Install and load necessary libraries:\n","#install.packages(\"dplyr\")\n","#library(dplyr)\n","#install.packages(\"tidyr\")\n","#library(tidyr)\n","#install.packages(\"matrixStats\")\n","#library(matrixStats)\n","\n","# Function to retrieve the tissues and category tissues from the eQTL to the final table:\n","process_eqtl_mt_robin <- function(input_eqtl_path, input_mt_robin_path, output_path) {\n","\n","  # Load datasets:\n","  eqtl_all_tissues <- read.table(file = input_eqtl_path, header = TRUE, sep = \"\\t\")\n","  mt_robin <- read.table(file = input_mt_robin_path, header = TRUE, sep = \"\\t\")\n","\n","  # Map beta columns to tissue names:\n","  tissue_names <- c(\"Subcutaneous adipose\", \"Visceral omentum adipose\", \"Adrenal gland\",\n","                    \"Aorta artery\", \"Coronary artery\", \"Tibial artery\", \"Amygdala brain\",\n","                    \"Anterior cingulate cortex\", \"Caudate\", \"Cerebellar hemisphere\",\n","                    \"Cerebellum\", \"Cortex\", \"Frontal cortex\", \"Hippocampus\", \"Hypothalamus\",\n","                    \"Nucleus accumbens (basal ganglia)\", \"Putamen (basal ganglia)\", \"Spinal cord\",\n","                    \"Substantia nigra\", \"Breast mammary tissue\", \"Cultured fibroblasts\",\n","                    \"EBV-transformed lymphocytes\", \"Sigmoid colon\", \"Transverse colon\",\n","                    \"Gastroesophageal junction\", \"Esophagus mucosa\", \"Esophagus muscularis\",\n","                    \"Atrial appendage\", \"Left ventricle\", \"Kidney cortex\", \"Liver\", \"Lung\",\n","                    \"Minor salivary gland\", \"Skeletal muscle\", \"Tibial nerve\", \"Ovary\",\n","                    \"Pancreas\", \"Pituitary\", \"Prostate\", \"Not sun-exposed skin (suprapubic)\",\n","                    \"Sun-exposed skin (lower leg)\", \"Small intestine terminal ileum\", \"Spleen\",\n","                    \"Stomach\", \"Testis\", \"Thyroid\", \"Uterus\", \"Vagina\", \"Whole blood\")\n","\n","  # Define a mapping from tissues to categories:\n","  tissue_to_category <- list(\n","    \"Adrenal gland\" = \"Brain and Nervous System\",\n","    \"Amygdala brain\" = \"Brain and Nervous System\",\n","    \"Anterior cingulate cortex\" = \"Brain and Nervous System\",\n","    \"Aorta artery\" = \"Cardiovascular System\",\n","    \"Atrial appendage\" = \"Cardiovascular System\",\n","    \"Breast mammary tissue\" = \"Exocrine System\",\n","    \"Caudate\" = \"Brain and Nervous System\",\n","    \"Cerebellar hemisphere\" = \"Brain and Nervous System\",\n","    \"Cerebellum\" = \"Brain and Nervous System\",\n","    \"Coronary artery\" = \"Cardiovascular System\",\n","    \"Cortex\" = \"Brain and Nervous System\",\n","    \"Cultured fibroblasts\" = \"Connective Tissue\",\n","    \"EBV-transformed lymphocytes\" = \"Immune System\",\n","    \"Esophagus mucosa\" = \"Digestive System\",\n","    \"Esophagus muscularis\" = \"Digestive System\",\n","    \"Frontal cortex\" = \"Brain and Nervous System\",\n","    \"Gastroesophageal junction\" = \"Digestive System\",\n","    \"Hippocampus\" = \"Brain and Nervous System\",\n","    \"Hypothalamus\" = \"Brain and Nervous System\",\n","    \"Kidney cortex\" = \"Excretory System\",\n","    \"Left ventricle\" = \"Cardiovascular System\",\n","    \"Liver\" = \"Digestive System\",\n","    \"Lung\" = \"Respiratory System\",\n","    \"Minor salivary gland\" = \"Exocrine System\",\n","    \"Not sun-exposed skin (suprapubic)\" = \"Skin\",\n","    \"Nucleus accumbens (basal ganglia)\" = \"Brain and Nervous System\",\n","    \"Ovary\" = \"Reproductive System\",\n","    \"Pancreas\" = \"Digestive System\",\n","    \"Pituitary\" = \"Endocrine System\",\n","    \"Prostate\" = \"Reproductive System\",\n","    \"Putamen (basal ganglia)\" = \"Brain and Nervous System\",\n","    \"Sigmoid colon\" = \"Digestive System\",\n","    \"Skeletal muscle\" = \"Muscular System\",\n","    \"Small intestine terminal ileum\" = \"Digestive System\",\n","    \"Spinal cord\" = \"Brain and Nervous System\",\n","    \"Spleen\" = \"Immune System\",\n","    \"Stomach\" = \"Digestive System\",\n","    \"Subcutaneous adipose\" = \"Adipose Tissue\",\n","    \"Substantia nigra\" = \"Brain and Nervous System\",\n","    \"Sun-exposed skin (lower leg)\" = \"Skin\",\n","    \"Testis\" = \"Reproductive System\",\n","    \"Thyroid\" = \"Endocrine System\",\n","    \"Tibial artery\" = \"Cardiovascular System\",\n","    \"Tibial nerve\" = \"Brain and Nervous System\",\n","    \"Transverse colon\" = \"Digestive System\",\n","    \"Uterus\" = \"Reproductive System\",\n","    \"Vagina\" = \"Reproductive System\",\n","    \"Visceral omentum adipose\" = \"Adipose Tissue\",\n","    \"Whole blood\" = \"Blood\"\n","  )\n","\n","  # Create variables for the thresholds:\n","  pval_thresh <- 0.001\n","  nTiss_thresh <- 1\n","\n","  # Create a function to map the tissues for each gene/SNP combination:\n","  process_gene_snp <- function(gene_id, snp_ids) {\n","    # Filter the eQTL data for the current gene and SNPs:\n","    eqtl_gene_snp <- eqtl_all_tissues[eqtl_all_tissues$gene_id == gene_id & eqtl_all_tissues$variant_id %in% snp_ids, ]\n","\n","    # Extract the p-value matrix and beta matrix:\n","    pval_mat <- as.matrix(eqtl_gene_snp[, grep(\"^pvalue_\", colnames(eqtl_gene_snp)), drop = FALSE])\n","    beta_mat <- as.matrix(eqtl_gene_snp[, grep(\"^beta_\", colnames(eqtl_gene_snp)), drop = FALSE])\n","\n","    # Calculate the minimum p-value for each SNP:\n","    min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE)\n","\n","    # Calculate the p-value threshold for the specified number of tissues:\n","    nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x, na.last = TRUE)[nTiss_thresh])\n","\n","    # Subset SNPs that meet the p-value threshold in the specified number of tissues:\n","    eqtl_gene_snp_filtered <- eqtl_gene_snp[min_p < pval_thresh & nTissThresh_p < pval_thresh, ]\n","\n","    # Extract the beta matrix and p-value matrix for the filtered SNPs:\n","    beta_mat_filtered <- as.matrix(eqtl_gene_snp_filtered[, grep(\"^beta_\", colnames(eqtl_gene_snp_filtered)), drop = FALSE])\n","    pval_mat_filtered <- as.matrix(eqtl_gene_snp_filtered[, grep(\"^pvalue_\", colnames(eqtl_gene_snp_filtered)), drop = FALSE])\n","\n","    # Identify tissues with significant eQTL associations:\n","    pvals_lt001 <- pval_mat_filtered < pval_thresh\n","    betas_sig <- beta_mat_filtered * pvals_lt001\n","\n","    # Check if each SNP has both positive and negative significant eQTL associations:\n","    sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15))))\n","    sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15))))\n","\n","    # Filter out SNPs with both positive and negative significant eQTL associations:\n","    eqtl_gene_snp_filtered <- eqtl_gene_snp_filtered[!(sig_pos_eQTL == 1 & sig_neg_eQTL == 1), ]\n","\n","    # Extract the list of beta columns (tissues) for the remaining SNPs:\n","    beta_cols <- colnames(eqtl_gene_snp_filtered)[grepl(\"^beta_\", colnames(eqtl_gene_snp_filtered))]\n","\n","    # Filter out beta columns with all NAs:\n","    beta_cols <- beta_cols[colSums(!is.na(eqtl_gene_snp_filtered[, beta_cols, drop = FALSE])) > 0]\n","\n","    # Count the total number of tissues:\n","    n_tissues <- length(beta_cols)\n","\n","    # Map beta column names to tissue names:\n","    tissue_cols <- tissue_names[as.integer(gsub(\"beta_\", \"\", beta_cols))]\n","\n","    # Map tissue names to categories:\n","    category_cols <- unname(tissue_to_category[tissue_cols])\n","\n","    # Return results:\n","    return(list(tissue_cols = tissue_cols, n_tissues = n_tissues, category_cols = category_cols))\n","  }\n","\n","  # Apply the previous function for each gene/SNP combination:\n","  mt_robin_processed <- apply(mt_robin, 1, function(row) {\n","    gene_id <- row[\"gene_id\"]\n","    snp_ids <- strsplit(row[\"SNPs_IDs\"], \",\")[[1]]\n","    result <- process_gene_snp(gene_id, snp_ids)\n","    return(data.frame(gene_id = gene_id, SNPs_IDs = row[\"SNPs_IDs\"], tissue_cols = paste(result$tissue_cols, collapse = \",\"), n_tissues = result$n_tissues, category_cols = paste(result$category_cols, collapse = \",\")))\n","  })\n","  mt_robin_processed <- do.call(rbind, mt_robin_processed)\n","\n","  # Rename the columns in mt_robin_processed:\n","  colnames(mt_robin_processed) <- c(\"gene_id\", \"SNPs_IDs\", \"Tissues\", \"n_Tissues\", \"Tissue_Categories\")\n","\n","  # Merge mt_robin and mt_robin_processed based on the gene_id column, selecting only the required columns\n","  mt_robin_merged <- merge(mt_robin, mt_robin_processed[, c(\"gene_id\", \"Tissues\", \"Tissue_Categories\")], by = \"gene_id\", all.x = TRUE)\n","\n","  # Print results:\n","  print(mt_robin_merged)\n","\n","  # Export the final table with adjusted delimiter for tissues and categories\n","  write.table(mt_robin_merged, file = output_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","}"],"metadata":{"id":"NRjSNx9z-hFM"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## **forward_selection**"],"metadata":{"id":"VJ-NefaR7Z6p"}},{"cell_type":"markdown","source":["### **Loop GWAS**"],"metadata":{"id":"fwNmgELW75jO"}},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(matrixStats)\n","library(snpStats)\n","library(utils)\n","library(parallel)\n","\n","# Step 2: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 3: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 4: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 5: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 6: Modify the original select_IV_fs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_fs() function --> Starts with no variables and adds one at a time:\n","select_IV_fs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset eQTL dataset to gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  # pval_mat <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  #sig_neg_eQTL <- as.numeric(apply(betas_sig,1,function(x) any(x < -1e-15)))\n","  #sig_pos_eQTL <- as.numeric(apply(betas_sig,1,function(x) any(x > 1e-15)))\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Select the best SNPs remaining in SNP_pool based on minimum p-value:\n","  selected_snps <- NULL\n","  while(length(SNP_pool) > 1){\n","\n","    ## Identify top SNPs in pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    next_snp_idx <- which.min(eqtl_data$min_p)\n","    next_snp <- eqtl_data$variant_id[next_snp_idx]\n","    selected_snps <-c(selected_snps,next_snp)\n","\n","    ## Identify SNPs not in high LD with top SNPs in the pool:\n","    LD_idx <- which(row.names(LD_r2)==next_snp)\n","    LD_keep_idx <- which(LD_r2[LD_idx,] < ld_thresh)\n","    LD_r2 <- LD_r2[LD_keep_idx,LD_keep_idx]\n","    SNP_pool <- names(LD_keep_idx)\n","  }\n","  selected_snps <- c(selected_snps, SNP_pool)\n","  return(selected_snps)\n","}\n","\n","# Step 7: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 8: Loop over the 4 GWAS datasets:\n","gwas_files <- c(\"long_covid_gwas_1_subset.txt\",\n","                \"long_covid_gwas_2_subset.txt\",\n","                \"long_covid_gwas_3_subset.txt\",\n","                \"long_covid_gwas_4_subset.txt\")\n","for (gwas_file in gwas_files) {\n","\n","  # Step 8.1: Open the GWAS input data set:\n","  gwas_covid_outcome <- read.table(paste0(\"/data/\", gwas_file),\n","                                   header = TRUE,\n","                                   sep = \"\",\n","                                   stringsAsFactors = FALSE)\n","\n","  # Step 8.2: Create a unique directory for the LD matrices corresponding to this GWAS dataset:\n","  ld_matrix_dir <- paste0(\"/data/ld_matrix_loop_fs_\",\n","                          gsub(\".txt\", \"\", gwas_file))\n","\n","  # Step 8.3: Create the directory if it doesn't exist:\n","  if (!dir.exists(ld_matrix_dir)) {\n","    dir.create(ld_matrix_dir)\n","  }\n","\n","# Step 9: Apply the function to each gene id:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","\n","  # Step 9.1: Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_fs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","    }, error = function(e) print(e))\n","  }, mc.cores = num_cores)\n","\n","# Step 10: Name the elements of the list with corresponding gene ids:\n","names(result_list) <- all_genes_vector\n","\n","# Step 11: Save the previous result as a .rds file:\n","saveRDS(result_list, file = paste0(\"/data/long_covid_IVs_loop_fs_\",\n","                                   gsub(\".txt\", \"\", gwas_file), \".rds\"))\n","}"],"metadata":{"id":"6rUtM6Qv775h"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q megamem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=2000GB\n","#PBS -l jobfs=1000GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/select_function_fs_loop.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"j2pHmf3p8BkX"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["`Result: Entered into a held queue.`\n","\n","Check the following links:\n","\n","https://opus.nci.org.au/display/Help/Queue+Structure\n","\n","https://opus.nci.org.au/pages/viewpage.action?pageId=90308823\n","\n","https://opus.nci.org.au/display/Help/FAQ+2%3A+What+does+exceeded+memory+allocation+mean\n","\n","**Pay attention:**\n","\n","1. **Queuing**: Jobs typically enter a queue when submitted. The scheduler decides when to start each job based on available resources, job priority, and other parameters. If the cluster doesn't have sufficient resources available to start all the jobs immediately, some of the jobs may be left waiting in the queue.\n","\n","2. **Fair Share and User Priority**: Many clusters implement a \"fair share\" policy, which means that if a single user or group submits a large number of jobs or continually monopolizes resources, their job priority may be temporarily decreased. This prevents any single user from dominating the cluster resources and ensures equitable access for all users.\n","\n","3. **Resource Fragmentation**: Even if the cluster has free resources, they might not be contiguous. For example, if you request 48 cores and there are 50 cores free but spread across multiple nodes, the scheduler might not start the job until 48 cores become available on a single node or in a configuration that meets the job's specifications.\n","\n","4. **Parallel Execution**: If there are enough resources on the cluster, multiple jobs can run simultaneously. If you submit two jobs, each requesting 48 cores, and there are 96 cores available, both jobs could potentially start at the same time.\n","\n","5. **Job Dependencies**: Some schedulers allow you to set dependencies between jobs, meaning one job won't start until another has completed. This can be useful if one job produces data that another job will use.\n","\n","6. **Array Jobs**: If you're running multiple similar jobs with the same resource requirements, some schedulers support array jobs. This allows you to submit a large number of similar jobs as a single array, and the scheduler can then manage and start these jobs efficiently.\n","\n","7. **Backfilling**: Some schedulers use backfilling to optimize the job queue. If a large job is waiting for resources to become available, but there's a gap where smaller jobs can fit without delaying the large job, the scheduler might start the smaller jobs to make efficient use of resources.\n","\n","8. **Job Preemption**: On some clusters, especially those with a partition or Quality of Service (QoS) for high-priority or express jobs, longer-running, lower-priority jobs might be preempted (paused or stopped) to free up resources for a high-priority job. Once the high-priority job completes, the preempted jobs may be restarted or resumed.\n","\n","\n","**Possible solutions:**\n","\n","1. **Optimize the R code**: Review the R scripts to check for potential memory inefficiencies. For instance:\n","    - Use more memory-efficient data structures, like data.table instead of data.frame.\n","    - Clear out large objects that are no longer needed using `rm(objectName)` and then run `gc()` to free up memory.\n","    - Use functions that perform operations in chunks rather than loading entire large datasets into memory.\n","\n","2. **Parallelism**: Consider reducing the number of concurrent processes. While parallelism can speed up computations, it multiplies memory usage. If each R process requires a lot of memory, running many of them at once could quickly deplete available resources.\n","\n","3. **Increase Memory Allocation**: If possible, request a node with more memory or increase the memory allocation for the job. This may involve communicating with the system administrator or adjusting the job submission scripts.\n","\n","4. **Use High Memory Nodes**: If the cluster offers nodes specifically designed for high-memory tasks, request those nodes for the job.\n","\n","5. **Use Disk**: For operations that deal with large datasets, consider using disk-based tools or algorithms that operate on chunks of data at a time, rather than loading everything into memory. For example, the `ff` and `bigmemory` packages in R allow for the analysis of large datasets without loading them fully into memory.\n","\n","6. **Profile Memory**: Tools like `pryr` can help profile memory usage in R. This way, you can identify which parts of the code consume the most memory and target those for optimization.\n","\n","7. **External Help**: Check the link provided in the error message (https://opus.nci.org.au/x/SwGRAQ). It might contain specific guidelines or recommendations for managing memory on the `gadi` system.\n","\n","**Nodes vs. CPUs vs. Cores:**\n","\n","**Node:**\n","\n","- A node generally refers to a single physical or virtual machine in the cluster.\n","- Each node will have its own local memory and possibly local storage.\n","- A node might contain one or multiple CPUs (processors). Each of these CPUs can have multiple cores.\n","- Nodes might also contain other specialized hardware components, such as GPUs (Graphics Processing Units) or FPGAs (Field-Programmable Gate Arrays).\n","- In HPC contexts, there are often two main types of nodes:\n","  - Compute Nodes: Used primarily to execute application code.\n","  - Service or Head Nodes: Used for managing the cluster, submitting jobs, and other administrative tasks.\n","\n","**Core:**\n","\n","- A core is a component of a CPU and represents a single processing unit. Modern CPUs are multi-core, meaning they have multiple cores that can process data independently.\n","- Each core can execute its own thread(s), allowing for parallel processing within a single CPU.\n","- Cores share the resources of the CPU they are part of, such as caches, but often have their own local resources too.\n","- For example, an 8-core CPU can potentially run eight different processes or threads simultaneously.\n","- Cores can be physical or logical. Logical cores (or threads) are seen in technologies like Intel's Hyper-Threading, where each physical core can execute two threads simultaneously, effectively doubling the number of cores from a software perspective.\n","\n","**Example:**\n","\n","ncpus=48 --> 48 cores are requested, which might be distributed across one or multiple nodes, depending on the cluster's configuration and how nodes are set up. If the cluster has nodes with 48 cores each, this might equate to one node. If nodes have 24 cores each, your job would span two nodes."],"metadata":{"id":"WAg18-YP8J9G"}},{"cell_type":"markdown","source":["### **GWAS1**"],"metadata":{"id":"HV7GttP4s7UR"}},{"cell_type":"code","source":["# select_IV function:\n","\n","library(matrixStats) # It needs to be installed separately\n","#library(BiocManager) # It needs to be installed separately\n","library(snpStats) # It needs to be installed separately\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_1_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 4: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 5: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 6: Create a variable for the path to the directory to save the LD matrix results:\n","ld_matrix_dir <- \"/data/ld_matrix_fs_1\"\n","\n","# Step 7: Create the directory if it doesn't exist:\n","if (!dir.exists(ld_matrix_dir)) {\n","  dir.create(ld_matrix_dir)\n","}\n","\n","# Step 8: Modify the original select_IV_fs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_fs() function --> Starts with no variables and adds one at a time:\n","select_IV_fs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset eQTL dataset to gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  # pval_mat <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  #sig_neg_eQTL <- as.numeric(apply(betas_sig,1,function(x) any(x < -1e-15)))\n","  #sig_pos_eQTL <- as.numeric(apply(betas_sig,1,function(x) any(x > 1e-15)))\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Select the best SNPs remaining in SNP_pool based on minimum p-value:\n","  selected_snps <- NULL\n","  while(length(SNP_pool) > 1){\n","\n","    ## Identify top SNPs in pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    next_snp_idx <- which.min(eqtl_data$min_p)\n","    next_snp <- eqtl_data$variant_id[next_snp_idx]\n","    selected_snps <-c(selected_snps,next_snp)\n","\n","    ## Identify SNPs not in high LD with top SNPs in the pool:\n","    LD_idx <- which(row.names(LD_r2)==next_snp)\n","    LD_keep_idx <- which(LD_r2[LD_idx,] < ld_thresh)\n","    LD_r2 <- LD_r2[LD_keep_idx,LD_keep_idx]\n","    SNP_pool <- names(LD_keep_idx)\n","  }\n","  selected_snps <- c(selected_snps, SNP_pool)\n","  return(selected_snps)\n","}\n","\n","# Step 9: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 10: Apply the function to each gene id:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","  # Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_fs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","  }, error = function(e) print(e))\n","}, mc.cores = num_cores)\n","\n","# Step 11: Name the elements of the list with corresponding gene ids:\n","names(result_list) <- all_genes_vector\n","\n","# Step 12: Save the previous result as a .rds file:\n","saveRDS(result_list, file = \"/data/long_covid_IVs_fs_1.rds\")"],"metadata":{"id":"lTDcsTX0zaQv"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/select_function_fs_1.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"3tDrKPLl7k8I"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-23 22:20:17:\n","   Job Id:             93322355.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      804.92\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 257:47:02\n","   Memory Requested:   500.0GB               Memory Used: 391.58GB\n","   Walltime requested: 30:00:00            Walltime Used: 05:35:23\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"Aoj5QM7RUlSB"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Open the output file:\n","result_selectIV_fs_1 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_IVs_fs_1.rds\")\n","\n","# Filter out errors:\n","result_selectIV_fs_1 <- result_selectIV_fs_1[!sapply(result_selectIV_fs_1, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_selectIV_fs_1))\n","print(length(unique(result_selectIV_fs_1)))\n","print(head(result_selectIV_fs_1))"],"metadata":{"id":"ebLipIo3JPbq","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1696971649225,"user_tz":-630,"elapsed":3705,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"dba5ab54-9913-48df-acb7-5636516717f9"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 38574\n","[1] 38480\n","$ENSG00000000003\n"," [1] \"rs8780\"      \"rs1204407\"   \"rs189980168\" \"rs4145090\"   \"rs139645081\"\n"," [6] \"rs112002620\" \"rs113922837\" \"rs17329105\"  \"rs193124579\" \"rs146639579\"\n","[11] \"rs201585484\" \"rs111478058\" \"rs143491698\" \"rs572714672\" \"rs56100956\" \n","[16] \"rs147601806\" \"rs139830751\" \"rs12842251\"  \"rs73630694\" \n","\n","$ENSG00000000005\n"," [1] \"rs6616166\"   \"rs4828038\"   \"rs12006573\"  \"rs73557922\"  \"rs141394856\"\n"," [6] \"rs140468798\" \"rs73555045\"  \"rs5921738\"   \"rs139231094\" \"rs59368006\" \n","[11] \"rs111915382\" \"rs73562536\"  \"rs12832642\"  \"rs201254498\" \"rs186611188\"\n","[16] \"rs113422802\" \"rs796285800\" \"rs180795909\" \"rs1004006\"   \"rs148617015\"\n","[21] \"rs140109110\"\n","\n","$ENSG00000000419\n"," [1] \"rs1292040204\" \"rs79562553\"   \"rs4809840\"    \"rs35201621\"   \"rs35530561\"  \n"," [6] \"rs141892692\"  \"rs944933\"     \"rs11483800\"   \"rs113036491\"  \"rs116892696\" \n","[11] \"rs34483216\"   \"rs60578213\"   \"rs118181672\"  \"rs4811253\"    \"rs11086363\"  \n","[16] \"rs73278571\"   \"rs73131221\"   \"rs78319058\"   \"rs79223715\"   \"rs189417964\" \n","[21] \"rs74422706\"   \"rs112070162\"  \"rs59034704\"   \"rs180710262\"  \"rs577456025\" \n","[26] \"rs117900008\"  \"rs34223230\"   \"rs141885330\"  \"rs78937289\"   \"rs16995896\"  \n","[31] \"rs72020869\"   \"rs142281981\"  \"rs13037515\"   \"rs1883804\"    \"rs17730415\"  \n","[36] \"rs11697209\"   \"rs112566287\"  \"rs149762110\"  \"rs4809839\"    \"rs6021441\"   \n","[41] \"rs75866341\"   \"rs13036755\"   \"rs74177405\"  \n","\n","$ENSG00000000457\n"," [1] \"rs10753796\"  \"rs4656710\"   \"rs115234612\" \"rs3766155\"   \"rs2056980\"  \n"," [6] \"rs4987403\"   \"rs12123695\"  \"rs185963038\" \"rs116684351\" \"rs1076637\"  \n","[11] \"rs7417281\"   \"rs7541615\"   \"rs973608\"    \"rs2982465\"   \"rs113123846\"\n","[16] \"rs4987338\"   \"rs114779274\" \"rs12038549\"  \"rs112361437\"\n","\n","$ENSG00000000460\n"," [1] \"rs35296671\"  \"rs1800016\"   \"rs4656198\"   \"rs3753311\"   \"rs4987391\"  \n"," [6] \"rs180738486\" \"rs4562681\"   \"rs531844\"    \"rs115157324\" \"rs12116424\" \n","[11] \"rs17346300\"  \"rs60843604\"  \"rs12742457\"  \"rs78159465\"  \"rs139679825\"\n","[16] \"rs35662131\" \n","\n","$ENSG00000000938\n"," [1] \"rs3737803\"   \"rs72191132\"  \"rs137984979\" \"rs11273428\"  \"rs75800899\" \n"," [6] \"rs191579330\" \"rs9426420\"   \"rs148432575\" \"rs10430146\"  \"rs139100109\"\n","[11] \"rs6701065\"   \"rs80298610\"  \"rs17361686\"  \"rs484740\"    \"rs114059345\"\n","[16] \"rs72651589\"  \"rs4357550\"   \"rs41284294\"  \"rs138392492\" \"rs148577969\"\n","[21] \"rs151288141\" \"rs71642227\"  \"rs12759594\"  \"rs2076457\"   \"rs141256722\"\n","[26] \"rs76443568\" \n","\n"]}]},{"cell_type":"code","source":["# setup function:\n","\n","library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_1_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the list of genes-SNPs:\n","result_selectedIVs <- readRDS(\"/data/long_covid_IVs_fs_1.rds\")\n","\n","# Step 4: Create a variable for the path to the LD matrix directory:\n","ld_matrix_dir <- \"/data/ld_matrix_fs_1\"\n","\n","# Step 5: Setup the Mt_Robin method using the following setup function:\n","MR_MtRobin_setup <- function(geneID,\n","                             snpID,\n","                             eqtl_data,\n","                             gwas_data,\n","                             ld_matrix_path,\n","                             nTiss){\n","  tryCatch({\n","    ## Check if the gene is present and subset\n","    if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","    eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","    ## Confirm all SNPs are present in the dataset\n","    if(!all(snpID %in% eqtl_data$variant_id)) stop(\"Some SNPs missing in eQTL dataset\")\n","    if(!all(snpID %in% gwas_data$variant_id)) stop(\"Some SNPs missing in GWAS dataset\")\n","\n","    # Read the LD matrix from a file\n","    LD <- as.matrix(utils::read.table(ld_matrix_path,\n","                                      header = TRUE,\n","                                      check.names = FALSE))\n","\n","    if(!all(snpID %in% colnames(LD))) stop(\"Some SNPs missing in LD matrix\")\n","\n","    ## Align data\n","    eqtl_data <- eqtl_data[match(snpID,eqtl_data$variant_id),]\n","    gwas_data <- gwas_data[match(snpID,gwas_data$variant_id),]\n","    LD <- LD[match(snpID,rownames(LD)),match(snpID,colnames(LD))]\n","\n","    eqtl_betas <- as.matrix(eqtl_data[,paste0(\"beta_\",1:nTiss)])\n","    eqtl_se <- as.matrix(eqtl_data[,paste0(\"SE_\",1:nTiss)])\n","    eqtl_pvals <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","    gwas_betas <- gwas_data$beta\n","    gwas_se <- gwas_data$SE\n","\n","    row.names(eqtl_betas) <- row.names(eqtl_se) <- row.names(eqtl_pvals) <- eqtl_data$variant_id\n","\n","    return(list(snpID=snpID,gwas_betas=gwas_betas,gwas_se=gwas_se,\n","                eqtl_betas=eqtl_betas,eqtl_se=eqtl_se,eqtl_pvals=eqtl_pvals,LD=LD))\n","  }, error = function(e) print(e))\n","}\n","\n","# Step 6: Apply the previous setup function:\n","\n","# Step 6.1: Initialize an empty list to store the results:\n","MR_MtRobin_results <- list()\n","\n","# Step 6.2: Define the number of cores to use for the parallelism process:\n","num_cores <- 36\n","\n","# Step 6.3: Loop over each gene in result_selectedIVs:\n","MR_MtRobin_results <- mclapply(names(result_selectedIVs), function(geneID) {\n","\n","  # Step 6.3.1: Get the SNPs associated with each gene:\n","  snpID <- result_selectedIVs[[geneID]]\n","\n","  # Step 6.3.2: Generate a unique file path for the LD matrix of this gene:\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  # Step 6.3.3: Call the MR_MtRobin_setup() function with the parameters for each gene:\n","  return(tryCatch({\n","    MR_MtRobin_setup(geneID = geneID,\n","                     snpID = snpID,\n","                     eqtl_data = eqtl_covid_exposure,\n","                     gwas_data = gwas_covid_outcome,\n","                     ld_matrix_path = ld_matrix_path,\n","                     nTiss = 49)\n","  }, error = function(e) print(e)))\n","}, mc.cores = num_cores)\n","\n","# Step 7: Assign names to the list:\n","names(MR_MtRobin_results) <- names(result_selectedIVs)\n","\n","# Step 8: Save the previous result as a .rds file:\n","saveRDS(MR_MtRobin_results, file = \"/data/long_covid_setup_fs_1.rds\")"],"metadata":{"id":"vSky0ZX9qcEy"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/setup_function_fs_1.R\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"QlE9EakQrjqc"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 14:27:55:\n","   Job Id:             93416546.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      77.91\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 20:20:02\n","   Memory Requested:   300.0GB               Memory Used: 77.45GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:43:17\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"kh6344GsrnmZ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Open the output file:\n","result_setup_fs_1 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_setup_fs_1.rds\")\n","\n","# Filter out errors:\n","result_setup_fs_1 <- result_setup_fs_1[!sapply(result_setup_fs_1, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_setup_fs_1))\n","print(length(unique(result_setup_fs_1)))\n","print(result_setup_fs_1[[1]])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"DhA6OZOer4gl","executionInfo":{"status":"ok","timestamp":1693539825340,"user_tz":-570,"elapsed":11733,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"187cc329-ab33-4369-f8e1-0efe0f218f18"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 4881\n","[1] 4881\n","$snpID\n","[1] \"rs147560086\" \"rs7955986\"   \"rs718389\"    \"rs11571383\"  \"rs2107614\"  \n","[6] \"rs4980935\"   \"rs11064536\"  \"rs117766797\"\n","\n","$gwas_betas\n","[1]  0.0320435 -0.0154484 -0.0713325 -0.0156255 -0.0385448 -0.0151348  0.0651312\n","[8] -0.1373340\n","\n","$gwas_se\n","[1] 0.0386181 0.0307614 0.0374723 0.0667074 0.0328251 0.0408972 0.0422060\n","[8] 0.1174630\n","\n","$eqtl_betas\n","            beta_1 beta_2    beta_3   beta_4 beta_5    beta_6    beta_7 beta_8\n","rs147560086     NA     NA -0.332722 -0.36379     NA -0.428724        NA     NA\n","rs7955986       NA     NA        NA       NA     NA        NA        NA     NA\n","rs718389        NA     NA        NA       NA     NA        NA        NA     NA\n","rs11571383      NA     NA        NA       NA     NA        NA -0.525715     NA\n","rs2107614       NA     NA        NA       NA     NA        NA        NA     NA\n","rs4980935       NA     NA        NA       NA     NA        NA        NA     NA\n","rs11064536      NA     NA        NA       NA     NA        NA        NA     NA\n","rs117766797     NA     NA        NA       NA     NA        NA        NA     NA\n","               beta_9 beta_10 beta_11 beta_12 beta_13 beta_14 beta_15   beta_16\n","rs147560086 -0.362536      NA      NA -0.5128      NA      NA      NA        NA\n","rs7955986          NA      NA      NA      NA      NA      NA      NA        NA\n","rs718389           NA      NA      NA      NA      NA      NA      NA        NA\n","rs11571383         NA      NA      NA      NA      NA      NA      NA        NA\n","rs2107614          NA      NA      NA      NA      NA      NA      NA -0.224019\n","rs4980935          NA      NA      NA      NA      NA      NA      NA        NA\n","rs11064536         NA      NA      NA      NA      NA      NA      NA        NA\n","rs117766797        NA      NA      NA      NA      NA      NA      NA        NA\n","            beta_17  beta_18  beta_19   beta_20   beta_21 beta_22   beta_23\n","rs147560086      NA -0.27895       NA -0.393493 -0.538392      NA -0.380267\n","rs7955986        NA       NA       NA        NA        NA      NA        NA\n","rs718389         NA       NA       NA        NA        NA      NA        NA\n","rs11571383       NA       NA       NA        NA        NA      NA        NA\n","rs2107614        NA       NA       NA        NA        NA      NA        NA\n","rs4980935        NA       NA 0.242572        NA        NA      NA        NA\n","rs11064536       NA       NA       NA        NA        NA      NA        NA\n","rs117766797      NA       NA       NA        NA        NA      NA        NA\n","            beta_24  beta_25 beta_26 beta_27 beta_28   beta_29  beta_30 beta_31\n","rs147560086      NA -0.37429      NA      NA      NA -0.413552       NA      NA\n","rs7955986        NA       NA      NA      NA      NA        NA       NA      NA\n","rs718389         NA       NA      NA      NA      NA        NA       NA      NA\n","rs11571383       NA       NA      NA      NA      NA        NA       NA      NA\n","rs2107614        NA       NA      NA      NA      NA        NA       NA      NA\n","rs4980935        NA       NA      NA      NA      NA        NA       NA      NA\n","rs11064536       NA       NA      NA      NA      NA        NA 0.524096      NA\n","rs117766797      NA       NA      NA      NA      NA        NA       NA      NA\n","            beta_32 beta_33   beta_34   beta_35 beta_36 beta_37 beta_38\n","rs147560086      NA      NA -0.389496 -0.325135      NA      NA      NA\n","rs7955986        NA      NA        NA        NA      NA      NA      NA\n","rs718389         NA      NA        NA        NA      NA      NA      NA\n","rs11571383       NA      NA        NA        NA      NA      NA      NA\n","rs2107614        NA      NA        NA        NA      NA      NA      NA\n","rs4980935        NA      NA        NA        NA      NA      NA      NA\n","rs11064536       NA      NA        NA        NA      NA      NA      NA\n","rs117766797      NA      NA        NA        NA      NA      NA      NA\n","              beta_39 beta_40 beta_41   beta_42 beta_43   beta_44   beta_45\n","rs147560086 -0.348491      NA      NA        NA      NA -0.283741        NA\n","rs7955986          NA      NA      NA        NA      NA        NA        NA\n","rs718389           NA      NA      NA        NA      NA        NA -0.282161\n","rs11571383         NA      NA      NA -0.504512      NA        NA        NA\n","rs2107614          NA      NA      NA        NA      NA        NA        NA\n","rs4980935          NA      NA      NA        NA      NA        NA        NA\n","rs11064536         NA      NA      NA        NA      NA        NA        NA\n","rs117766797        NA      NA      NA        NA      NA        NA        NA\n","             beta_46 beta_47   beta_48 beta_49\n","rs147560086       NA      NA        NA      NA\n","rs7955986   0.243345      NA        NA      NA\n","rs718389          NA      NA        NA      NA\n","rs11571383        NA      NA        NA      NA\n","rs2107614         NA      NA        NA      NA\n","rs4980935         NA      NA        NA      NA\n","rs11064536        NA      NA        NA      NA\n","rs117766797       NA      NA -0.395786      NA\n","\n","$eqtl_se\n","            SE_1 SE_2      SE_3     SE_4 SE_5      SE_6     SE_7 SE_8      SE_9\n","rs147560086   NA   NA 0.0464181 0.029495   NA 0.0261202       NA   NA 0.0550403\n","rs7955986     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs718389      NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs11571383    NA   NA        NA       NA   NA        NA 0.105786   NA        NA\n","rs2107614     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs4980935     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs11064536    NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs117766797   NA   NA        NA       NA   NA        NA       NA   NA        NA\n","            SE_10 SE_11     SE_12 SE_13 SE_14 SE_15     SE_16 SE_17     SE_18\n","rs147560086    NA    NA 0.0567995    NA    NA    NA        NA    NA 0.0775922\n","rs7955986      NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs718389       NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs11571383     NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs2107614      NA    NA        NA    NA    NA    NA 0.0401716    NA        NA\n","rs4980935      NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs11064536     NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs117766797    NA    NA        NA    NA    NA    NA        NA    NA        NA\n","                SE_19     SE_20     SE_21 SE_22     SE_23 SE_24     SE_25 SE_26\n","rs147560086        NA 0.0327279 0.0351663    NA 0.0352046    NA 0.0353582    NA\n","rs7955986          NA        NA        NA    NA        NA    NA        NA    NA\n","rs718389           NA        NA        NA    NA        NA    NA        NA    NA\n","rs11571383         NA        NA        NA    NA        NA    NA        NA    NA\n","rs2107614          NA        NA        NA    NA        NA    NA        NA    NA\n","rs4980935   0.0586861        NA        NA    NA        NA    NA        NA    NA\n","rs11064536         NA        NA        NA    NA        NA    NA        NA    NA\n","rs117766797        NA        NA        NA    NA        NA    NA        NA    NA\n","            SE_27 SE_28     SE_29    SE_30 SE_31 SE_32 SE_33    SE_34     SE_35\n","rs147560086    NA    NA 0.0384595       NA    NA    NA    NA 0.030555 0.0222479\n","rs7955986      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs718389       NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs11571383     NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs2107614      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs4980935      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs11064536     NA    NA        NA 0.136083    NA    NA    NA       NA        NA\n","rs117766797    NA    NA        NA       NA    NA    NA    NA       NA        NA\n","            SE_36 SE_37 SE_38     SE_39 SE_40 SE_41     SE_42 SE_43     SE_44\n","rs147560086    NA    NA    NA 0.0452671    NA    NA        NA    NA 0.0360217\n","rs7955986      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs718389       NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs11571383     NA    NA    NA        NA    NA    NA 0.0733203    NA        NA\n","rs2107614      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs4980935      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs11064536     NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs117766797    NA    NA    NA        NA    NA    NA        NA    NA        NA\n","                SE_45     SE_46 SE_47    SE_48 SE_49\n","rs147560086        NA        NA    NA       NA    NA\n","rs7955986          NA 0.0191166    NA       NA    NA\n","rs718389    0.0359888        NA    NA       NA    NA\n","rs11571383         NA        NA    NA       NA    NA\n","rs2107614          NA        NA    NA       NA    NA\n","rs4980935          NA        NA    NA       NA    NA\n","rs11064536         NA        NA    NA       NA    NA\n","rs117766797        NA        NA    NA 0.107595    NA\n","\n","$eqtl_pvals\n","            pvalue_1 pvalue_2    pvalue_3    pvalue_4 pvalue_5    pvalue_6\n","rs147560086       NA       NA 1.57506e-11 8.15374e-29       NA 5.49137e-49\n","rs7955986         NA       NA          NA          NA       NA          NA\n","rs718389          NA       NA          NA          NA       NA          NA\n","rs11571383        NA       NA          NA          NA       NA          NA\n","rs2107614         NA       NA          NA          NA       NA          NA\n","rs4980935         NA       NA          NA          NA       NA          NA\n","rs11064536        NA       NA          NA          NA       NA          NA\n","rs117766797       NA       NA          NA          NA       NA          NA\n","               pvalue_7 pvalue_8    pvalue_9 pvalue_10 pvalue_11   pvalue_12\n","rs147560086          NA       NA 6.71399e-10        NA        NA 4.40477e-16\n","rs7955986            NA       NA          NA        NA        NA          NA\n","rs718389             NA       NA          NA        NA        NA          NA\n","rs11571383  2.64972e-06       NA          NA        NA        NA          NA\n","rs2107614            NA       NA          NA        NA        NA          NA\n","rs4980935            NA       NA          NA        NA        NA          NA\n","rs11064536           NA       NA          NA        NA        NA          NA\n","rs117766797          NA       NA          NA        NA        NA          NA\n","            pvalue_13 pvalue_14 pvalue_15  pvalue_16 pvalue_17  pvalue_18\n","rs147560086        NA        NA        NA         NA        NA 0.00050353\n","rs7955986          NA        NA        NA         NA        NA         NA\n","rs718389           NA        NA        NA         NA        NA         NA\n","rs11571383         NA        NA        NA         NA        NA         NA\n","rs2107614          NA        NA        NA 1.0067e-07        NA         NA\n","rs4980935          NA        NA        NA         NA        NA         NA\n","rs11064536         NA        NA        NA         NA        NA         NA\n","rs117766797        NA        NA        NA         NA        NA         NA\n","              pvalue_19   pvalue_20   pvalue_21 pvalue_22   pvalue_23 pvalue_24\n","rs147560086          NA 8.26297e-28 3.16236e-42        NA 9.45416e-23        NA\n","rs7955986            NA          NA          NA        NA          NA        NA\n","rs718389             NA          NA          NA        NA          NA        NA\n","rs11571383           NA          NA          NA        NA          NA        NA\n","rs2107614            NA          NA          NA        NA          NA        NA\n","rs4980935   8.06751e-05          NA          NA        NA          NA        NA\n","rs11064536           NA          NA          NA        NA          NA        NA\n","rs117766797          NA          NA          NA        NA          NA        NA\n","              pvalue_25 pvalue_26 pvalue_27 pvalue_28   pvalue_29   pvalue_30\n","rs147560086 3.44003e-22        NA        NA        NA 3.44995e-23          NA\n","rs7955986            NA        NA        NA        NA          NA          NA\n","rs718389             NA        NA        NA        NA          NA          NA\n","rs11571383           NA        NA        NA        NA          NA          NA\n","rs2107614            NA        NA        NA        NA          NA          NA\n","rs4980935            NA        NA        NA        NA          NA          NA\n","rs11064536           NA        NA        NA        NA          NA 0.000347708\n","rs117766797          NA        NA        NA        NA          NA          NA\n","            pvalue_31 pvalue_32 pvalue_33  pvalue_34   pvalue_35 pvalue_36\n","rs147560086        NA        NA        NA 2.6301e-33 4.95626e-40        NA\n","rs7955986          NA        NA        NA         NA          NA        NA\n","rs718389           NA        NA        NA         NA          NA        NA\n","rs11571383         NA        NA        NA         NA          NA        NA\n","rs2107614          NA        NA        NA         NA          NA        NA\n","rs4980935          NA        NA        NA         NA          NA        NA\n","rs11064536         NA        NA        NA         NA          NA        NA\n","rs117766797        NA        NA        NA         NA          NA        NA\n","            pvalue_37 pvalue_38   pvalue_39 pvalue_40 pvalue_41   pvalue_42\n","rs147560086        NA        NA 8.63893e-13        NA        NA          NA\n","rs7955986          NA        NA          NA        NA        NA          NA\n","rs718389           NA        NA          NA        NA        NA          NA\n","rs11571383         NA        NA          NA        NA        NA 2.07196e-10\n","rs2107614          NA        NA          NA        NA        NA          NA\n","rs4980935          NA        NA          NA        NA        NA          NA\n","rs11064536         NA        NA          NA        NA        NA          NA\n","rs117766797        NA        NA          NA        NA        NA          NA\n","            pvalue_43   pvalue_44   pvalue_45   pvalue_46 pvalue_47   pvalue_48\n","rs147560086        NA 8.32165e-14          NA          NA        NA          NA\n","rs7955986          NA          NA          NA 2.21429e-32        NA          NA\n","rs718389           NA          NA 1.07802e-13          NA        NA          NA\n","rs11571383         NA          NA          NA          NA        NA          NA\n","rs2107614          NA          NA          NA          NA        NA          NA\n","rs4980935          NA          NA          NA          NA        NA          NA\n","rs11064536         NA          NA          NA          NA        NA          NA\n","rs117766797        NA          NA          NA          NA        NA 0.000357146\n","            pvalue_49\n","rs147560086        NA\n","rs7955986          NA\n","rs718389           NA\n","rs11571383         NA\n","rs2107614          NA\n","rs4980935          NA\n","rs11064536         NA\n","rs117766797        NA\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n"]}]},{"cell_type":"code","source":["# Mt_Robin function:\n","\n","# Step 1: Load the required libraries:\n","library(base) # R package installed by default\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(Matrix) # R package installed by default\n","library(methods) # R package installed by default\n","library(stats) # R package installed by default\n","library(lme4) # It needs to be installed separately\n","library(data.table) # It needs to be installed separately\n","\n","# Step 2: Main Mt_Robin function:\n","MR_MtRobin <- function(MR_MtRobin_input,pval_thresh=0.001){\n","  snpID <- MR_MtRobin_input$snpID\n","  gwas_betas <- MR_MtRobin_input$gwas_betas\n","  gwas_se <- MR_MtRobin_input$gwas_se\n","  eqtl_betas <- MR_MtRobin_input$eqtl_betas\n","  eqtl_se <- MR_MtRobin_input$eqtl_se\n","  eqtl_pvals <- MR_MtRobin_input$eqtl_pvals\n","  LD <- MR_MtRobin_input$LD\n","\n","  if(ncol(eqtl_betas) != ncol(eqtl_se) | ncol(eqtl_betas) != ncol(eqtl_pvals)){\n","    stop(\"eQTL betas, SE and p-values do not all have same number of columns\")\n","  }\n","\n","  ## determine number of tissues (or conditions/studies/etc.)\n","  nT <- ncol(eqtl_betas)\n","\n","  ## rename columns for standardized processing\n","  colnames(eqtl_betas) <- paste0(\"beta_\",1:nT)\n","  colnames(eqtl_se) <- paste0(\"SE_\",1:nT)\n","  colnames(eqtl_pvals) <- paste0(\"pvalue_\",1:nT)\n","\n","  ## combine eQTL results into data.table\n","  eqtl_res <- data.table::data.table(snpID,eqtl_betas,eqtl_se,eqtl_pvals)\n","\n","  # reshape to long format (each SNP-gene-tissue will be row)\n","  eQTL_res_melt <- data.table::melt(eqtl_res,id.vars=c(\"snpID\"),\n","                                    measure.vars=patterns(\"^beta_\",\"^SE_\",\"^pvalue_\"),value.name=c(\"beta\",\"se\",\"pvalue\"),\n","                                    variable.name=\"tissue\")\n","\n","  ## subset to strong instrumental variables (based on p-value threshold)\n","  eQTL_res_melt_PltThresh <- subset(eQTL_res_melt, pvalue < pval_thresh)\n","\n","  ## set up gwas data for merging\n","  gwas_res <- data.table::data.table(snpID=snpID,gwas_beta=gwas_betas,gwas_se=gwas_se)\n","\n","  ## merge eQTL and GWAS results\n","  data.table::setkey(eQTL_res_melt_PltThresh,snpID)\n","  data.table::setkey(gwas_res,snpID)\n","  merged_res <- data.table::merge.data.table(eQTL_res_melt_PltThresh,gwas_res)\n","\n","  ## set up coefficients for reverse regression\n","  beta_x <- matrix(merged_res$beta,ncol=1)\n","  beta_y <- matrix(merged_res$gwas_beta,ncol=1)\n","\n","  ## standard errors (for weights)\n","  se_x <- matrix(merged_res$se,ncol=1)\n","  se_y <- matrix(merged_res$gwas_se,ncol=1) ## not used by function; used in resampling (consider returning with results)\n","\n","  ## identifiers\n","  snpID <- merged_res$snpID\n","  lme_res <- lme4::lmer(beta_x~(beta_y-1)+(beta_y-1|snpID),weights=1/se_x^2)\n","\n","  ## return results from weighted regression with random slopes (and correlated errors) along with GWAS SE and LD (needed for resampling)\n","  return(list(lme_res=lme_res,gwas_res=gwas_res,LD=LD))\n","}\n","\n","# Step 3: Read the previous file:\n","result_setup <- readRDS(\"/data/long_covid_setup_fs_1.rds\")\n","\n","# Step 4: Apply the main Mt_Robin function using parallel processing:\n","no_cores <- 38\n","\n","cl <- makeCluster(no_cores) # Create a cluster with the number of cores\n","clusterExport(cl, \"MR_MtRobin\") # Export the MR_MtRobin function to all the nodes of the cluster\n","result_list_MtRobin <- parLapply(cl, result_setup, function(MR_MtRobin_input) {\n","  tryCatch({\n","    MR_MtRobin(MR_MtRobin_input = MR_MtRobin_input)\n","  }, error = function(e) print(e))\n","})\n","stopCluster(cl) # Stop the cluster once the computation is done\n","\n","# Step 5: The names of the results will correspond to the gene ids:\n","names(result_list_MtRobin) <- names(result_setup)\n","\n","# Step 6: Save the previous result as a .rds file:\n","saveRDS(result_list_MtRobin, file = \"/data/long_covid_main_fs_1.rds\")"],"metadata":{"id":"XDEaFDaUt6o9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/main_function.sif Rscript /data/main_function_fs_1.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"CHgu_NdcuAf9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 17:11:55:\n","   Job Id:             93440305.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      2.73\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 00:02:51\n","   Memory Requested:   300.0GB               Memory Used: 9.67GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:01:31\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"O-dYpPtou2PR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Open the output file:\n","result_main_fs_1 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_1.rds\")\n","\n","# Filter out errors:\n","result_main_fs_1 <- result_main_fs_1[!sapply(result_main_fs_1, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_main_fs_1))\n","print(length(unique(result_main_fs_1)))\n","print(head(result_main_fs_1))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fe9766LVuD87","executionInfo":{"status":"ok","timestamp":1699504464761,"user_tz":-630,"elapsed":11369,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"6b9255f5-ee12-4191-ffaa-5ea93781842d"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1670\n","[1] 1670\n","$ENSG00000002016\n","$ENSG00000002016$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -25.7147\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 14.411  \n"," Residual         1.807  \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n","0.8438  \n","\n","$ENSG00000002016$gwas_res\n","        snpID  gwas_beta   gwas_se\n","1  rs11064536  0.0651312 0.0422060\n","2  rs11571383 -0.0156255 0.0667074\n","3 rs117766797 -0.1373340 0.1174630\n","4 rs147560086  0.0320435 0.0386181\n","5   rs2107614 -0.0385448 0.0328251\n","6   rs4980935 -0.0151348 0.0408972\n","7    rs718389 -0.0713325 0.0374723\n","8   rs7955986 -0.0154484 0.0307614\n","\n","$ENSG00000002016$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","\n","$ENSG00000005156\n","$ENSG00000005156$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -13.6931\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 4.235   \n"," Residual        3.172   \n","Number of obs: 10, groups:  snpID, 3\n","Fixed Effects:\n","beta_y  \n"," 12.63  \n","\n","$ENSG00000005156$gwas_res\n","        snpID  gwas_beta   gwas_se\n","1  rs12945428 -0.0388416 0.0297673\n","2 rs201016783  0.0540866 0.0713454\n","3   rs3135958  0.0203918 0.0482369\n","\n","$ENSG00000005156$LD\n","             rs12945428    rs3135958  rs201016783\n","rs12945428   1.00000000 -0.295076381  0.020183627\n","rs3135958   -0.29507638  1.000000000 -0.006054471\n","rs201016783  0.02018363 -0.006054471  1.000000000\n","\n","\n","$ENSG00000006282\n","$ENSG00000006282$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -27.3239\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 9.481   \n"," Residual        5.685   \n","Number of obs: 46, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n"," -27.1  \n","\n","$ENSG00000006282$gwas_res\n","      snpID gwas_beta   gwas_se\n","1 rs2306001 0.0146700 0.0488688\n","2 rs9890200 0.0248025 0.0312196\n","\n","$ENSG00000006282$LD\n","          rs9890200 rs2306001\n","rs9890200 1.0000000 0.3686704\n","rs2306001 0.3686704 1.0000000\n","\n","\n","$ENSG00000006744\n","$ENSG00000006744$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -3.3853\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 6.003   \n"," Residual        6.801   \n","Number of obs: 18, groups:  snpID, 4\n","Fixed Effects:\n","beta_y  \n"," 7.592  \n","\n","$ENSG00000006744$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1  rs1044564  0.03076870 0.0305059\n","2 rs11655929  0.03161910 0.0303495\n","3   rs733890 -0.00479602 0.0322052\n","4 rs78840287 -0.03849020 0.0948532\n","\n","$ENSG00000006744$LD\n","              rs1044564    rs733890 rs11655929   rs78840287\n","rs1044564   1.000000000  0.66545097 -0.6626427 -0.007565868\n","rs733890    0.665450966  1.00000000 -0.4997189  0.027633741\n","rs11655929 -0.662642701 -0.49971889  1.0000000  0.034843798\n","rs78840287 -0.007565868  0.02763374  0.0348438  1.000000000\n","\n","\n","$ENSG00000007171\n","$ENSG00000007171$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 12.3579\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 46.825  \n"," Residual         4.258  \n","Number of obs: 31, groups:  snpID, 5\n","Fixed Effects:\n","beta_y  \n","-20.15  \n","\n","$ENSG00000007171$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1  rs2314813 -0.02583820 0.0498335\n","2  rs3751972  0.00873644 0.0340366\n","3   rs624196  0.05470420 0.0324635\n","4 rs72846720  0.11658400 0.0774620\n","5  rs8078340  0.00728667 0.0455412\n","\n","$ENSG00000007171$LD\n","              rs3751972   rs8078340    rs624196    rs2314813  rs72846720\n","rs3751972   1.000000000 -0.20973763 -0.03796279  0.004244027 -0.08622187\n","rs8078340  -0.209737626  1.00000000  0.03757975  0.160333611 -0.06812959\n","rs624196   -0.037962786  0.03757975  1.00000000 -0.039583359  0.20176556\n","rs2314813   0.004244027  0.16033361 -0.03958336  1.000000000 -0.10409667\n","rs72846720 -0.086221874 -0.06812959  0.20176556 -0.104096673  1.00000000\n","\n","\n","$ENSG00000007341\n","$ENSG00000007341$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -1.8083\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 448.357 \n"," Residual          7.687 \n","Number of obs: 12, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n"," 298.1  \n","\n","$ENSG00000007341$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1 rs10745330 0.000842121 0.0295070\n","2   rs954679 0.027804200 0.0302953\n","\n","$ENSG00000007341$LD\n","           rs10745330  rs954679\n","rs10745330  1.0000000 0.6167231\n","rs954679    0.6167231 1.0000000\n","\n","\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_fs_1.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_1_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_1_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_fs1.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"CILNgcMk700D"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_fs_1.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"Kf7wqLL076NV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-06 20:16:00:\n","   Job Id:             107649683.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      656.68\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 135:10:53\n","   Memory Requested:   500.0GB               Memory Used: 117.95GB\n","   Walltime requested: 10:00:00            Walltime Used: 04:33:37\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"BerN69LV799a"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results_2 <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs1.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results_2))\n","print(head(results_2))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"TRbxRlRtv0zX","executionInfo":{"status":"ok","timestamp":1707263476395,"user_tz":-630,"elapsed":697,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"ae4b27f9-5327-44d1-8b45-b932c0a5ebed"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1670    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.884      10000 0.9779379\n","2 ENSG00000005156  0.143      10000 0.5553721\n","3 ENSG00000006282  0.126      10000 0.5367857\n","4 ENSG00000006744  0.140      10000 0.5514520\n","5 ENSG00000007171  0.467      10000 0.8385914\n","6 ENSG00000007341  0.286      10000 0.7280793\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_1.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"3nY8b92wKUOR","executionInfo":{"status":"ok","timestamp":1706597990697,"user_tz":-630,"elapsed":94558,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"eeff171c-3b86-4bc0-f301-e65cca0bdd9f"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 1670 10 \n","\n","          gene_id gene_name      gene_type      beta_y Obs SNPs\n","1 ENSG00000002016     RAD52 protein_coding   0.8437588  23    8\n","2 ENSG00000005156      LIG3 protein_coding  12.6260663  10    3\n","3 ENSG00000006282   SPATA20 protein_coding -27.0963502  46    2\n","4 ENSG00000006744     ELAC2 protein_coding   7.5924973  18    4\n","5 ENSG00000007171      NOS2 protein_coding -20.1542226  31    5\n","6 ENSG00000007341      ST7L protein_coding 298.0706258  12    2\n","                                                                               SNP_IDs\n","1 rs11064536,rs11571383,rs117766797,rs147560086,rs2107614,rs4980935,rs718389,rs7955986\n","2                                                     rs12945428,rs201016783,rs3135958\n","3                                                                  rs2306001,rs9890200\n","4                                             rs1044564,rs11655929,rs733890,rs78840287\n","5                                    rs2314813,rs3751972,rs624196,rs72846720,rs8078340\n","6                                                                  rs10745330,rs954679\n","        REML         SD Residual_SD\n","1 -25.714736  14.411447    1.807167\n","2 -13.693065   4.235270    3.171697\n","3 -27.323910   9.480576    5.684947\n","4  -3.385293   6.003191    6.801396\n","5  12.357902  46.824997    4.257685\n","6  -1.808276 448.357122    7.686964\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_protein_coding_genes_sig_non_sig.txt \n","Dimensions of protein_coding_genes : 748 10 \n","\n","          gene_id gene_name      gene_type      beta_y Obs SNPs\n","1 ENSG00000162994     CLHC1 protein_coding -79145.4801  40    2\n","2 ENSG00000095917     TPSD1 protein_coding   -626.4370  15    2\n","3 ENSG00000169629     RGPD8 protein_coding    392.9212   7    2\n","4 ENSG00000007341      ST7L protein_coding    298.0706  12    2\n","5 ENSG00000183172     SMDT1 protein_coding    278.5482   9    2\n","6 ENSG00000167555    ZNF528 protein_coding   -254.4705  39    1\n","               SNP_IDs       REML          SD Residual_SD\n","1 rs1437391,rs17046778   3.886870 111985.8440    4.429060\n","2 rs2745084,rs28546647 -20.191703    873.4912    2.891943\n","3 rs34510112,rs6704845  -9.279843    562.1549    2.639139\n","4  rs10745330,rs954679  -1.808276    448.3571    7.686964\n","5  rs4822088,rs9611734  -6.552788    538.4289    6.425609\n","6   rs162123,rs7258975 -68.384950      0.0000    2.512113\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# FS1 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_1_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs1.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"m1P1ZftO05wQ","executionInfo":{"status":"ok","timestamp":1709776289286,"user_tz":-630,"elapsed":13164,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"e3df6d28-6a52-495d-9088-6915aa386ade"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1670\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -25.7147\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 14.411  \n"," Residual         1.807  \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n","0.8438  \n","\n","$gwas_res\n","        snpID  gwas_beta   gwas_se\n","1  rs11064536  0.0651312 0.0422060\n","2  rs11571383 -0.0156255 0.0667074\n","3 rs117766797 -0.1373340 0.1174630\n","4 rs147560086  0.0320435 0.0386181\n","5   rs2107614 -0.0385448 0.0328251\n","6   rs4980935 -0.0151348 0.0408972\n","7    rs718389 -0.0713325 0.0374723\n","8   rs7955986 -0.0154484 0.0307614\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 748  10\n","          gene_id gene_name      gene_type      beta_y Obs SNPs\n","1 ENSG00000162994     CLHC1 protein_coding -79145.4801  40    2\n","2 ENSG00000095917     TPSD1 protein_coding   -626.4370  15    2\n","3 ENSG00000169629     RGPD8 protein_coding    392.9212   7    2\n","4 ENSG00000007341      ST7L protein_coding    298.0706  12    2\n","5 ENSG00000183172     SMDT1 protein_coding    278.5482   9    2\n","6 ENSG00000167555    ZNF528 protein_coding   -254.4705  39    1\n","               SNP_IDs       REML          SD Residual_SD\n","1 rs1437391,rs17046778   3.886870 111985.8440    4.429060\n","2 rs2745084,rs28546647 -20.191703    873.4912    2.891943\n","3 rs34510112,rs6704845  -9.279843    562.1549    2.639139\n","4  rs10745330,rs954679  -1.808276    448.3571    7.686964\n","5  rs4822088,rs9611734  -6.552788    538.4289    6.425609\n","6   rs162123,rs7258975 -68.384950      0.0000    2.512113\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1670    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.884      10000 0.9779379\n","2 ENSG00000005156  0.143      10000 0.5553721\n","3 ENSG00000006282  0.126      10000 0.5367857\n","4 ENSG00000006744  0.140      10000 0.5514520\n","5 ENSG00000007171  0.467      10000 0.8385914\n","6 ENSG00000007341  0.286      10000 0.7280793\n","[1] \"Final table (protein coding genes): \"\n","[1] 14 22\n","           gene_id gene_name SNPs\n","1  ENSG00000171160     MORN4    2\n","2  ENSG00000176386     CDC26    2\n","3  ENSG00000132507     EIF5A    2\n","4  ENSG00000146530      VWDE    8\n","5  ENSG00000173540     GMPPB    2\n","6  ENSG00000184983    NDUFA6    2\n","7  ENSG00000113734     BNIP1    7\n","8  ENSG00000090661     CERS4   15\n","9  ENSG00000065457     ADAT1    4\n","10 ENSG00000177025  C19orf18   17\n","11 ENSG00000183336     BOLA2    3\n","12 ENSG00000158825       CDA   14\n","13 ENSG00000139714     MORN3   18\n","14 ENSG00000026950    BTN3A1    9\n","                                                                                                                                                                                                SNPs_IDs\n","1                                                                                                                                                                                   rs12358715,rs7905087\n","2                                                                                                                                                                                  rs10981712,rs10981764\n","3                                                                                                                                                                                   rs62059195,rs7503161\n","4                                                                                                                      rs12673085,rs12699367,rs2287067,rs28811348,rs2966627,rs4721083,rs6971764,rs970035\n","5                                                                                                                                                                                  rs200287431,rs2280405\n","6                                                                                                                                                                                   rs137139,rs139043947\n","7                                                                                                                             rs10476082,rs11385449,rs35813302,rs4868243,rs5745108,rs72814178,rs77146849\n","8                                      rs113738552,rs11880745,rs142392712,rs147441807,rs1965888,rs2336171,rs4542783,rs4804282,rs56120315,rs56184962,rs57062897,rs59099428,rs60585010,rs7260363,rs8100119\n","9                                                                                                                                                             rs11642599,rs35261357,rs4888438,rs59297469\n","10                     .,rs1014767,rs112024224,rs11668314,rs118061825,rs138343123,rs142282855,rs145778607,rs149882742,rs150518835,rs257676,rs292831,rs34640515,rs36115826,rs3764536,rs4801529,rs57797045\n","11                                                                                                                                                                        rs4533290,rs79379134,rs9888967\n","12                                                    rs10753504,rs10916784,rs10916834,rs12026680,rs12045689,rs12135822,rs1253799,rs3131712,rs3215400,rs55665119,rs66731853,rs72649199,rs818199,rs879088\n","13 rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","14                                                                                                        rs114157910,rs12213056,rs13194491,rs2275906,rs3208733,rs3799378,rs56325223,rs7356982,rs9467733\n","   Obs     beta_y        Min     First_Q      Median   Third_Q       Max\n","1    3  34.218202 -0.5582269 -0.52550340 -0.49277994 0.3547653 1.2023106\n","2   18 -30.039865 -1.7931493 -0.71338242 -0.15416886 0.3773474 2.4878981\n","3   20  21.013660 -1.3847656 -0.83600738 -0.47418977 0.2216229 2.5618909\n","4   12  15.508145 -2.8063695 -0.57174743 -0.10704222 0.1810687 0.4180642\n","5   18  13.284320 -1.7327039 -0.70127932  0.31619771 0.9434917 1.3353628\n","6   24   9.789797 -2.9183208 -0.46957241 -0.04249690 0.6506457 1.5058086\n","7   13   8.994436 -2.0411354 -0.57137796  0.13557018 0.5707352 1.3174713\n","8   23  -7.804575 -1.9381654 -0.03024777  0.45603996 0.7985665 2.7709937\n","9   11  -7.713004 -1.5988537 -0.53363730 -0.28594064 0.1299801 2.3368205\n","10  20   7.706716 -2.0921744 -0.88505264 -0.17477279 0.4768810 1.4996117\n","11  21  -6.115419 -2.1062585 -0.64359310 -0.21806788 0.5977871 2.1480464\n","12  26  -6.070831 -0.6375271 -0.29178418  0.07034354 0.6669760 3.1846266\n","13  21   4.893023 -1.2647845 -0.64117122  0.19594240 0.9362305 1.9255321\n","14  22  -4.817862 -2.4698318 -0.12749636  0.14242212 0.6493537 2.1276346\n","         SD_RE Res_SD_RE       Var_RE Res_Var_RE     SE_FE t_value_FE\n","1  0.00000e+00  4.486594 0.000000e+00  20.129522 5.4261924   6.306117\n","2  0.00000e+00  1.577515 0.000000e+00   2.488552 1.2245776 -24.530796\n","3  0.00000e+00  2.500070 0.000000e+00   6.250350 0.9625492  21.831257\n","4  0.00000e+00  4.285012 0.000000e+00  18.361329 1.8111180   8.562747\n","5  2.90556e-09  4.429408 8.442281e-18  19.619659 0.9340473  14.222321\n","6  0.00000e+00  4.670701 0.000000e+00  21.815451 0.6053178  16.172988\n","7  0.00000e+00  3.182195 0.000000e+00  10.126362 1.0026723   8.970464\n","8  0.00000e+00  6.604208 0.000000e+00  43.615568 1.1491150  -6.791814\n","9  0.00000e+00  3.225459 0.000000e+00  10.403585 0.4917438 -15.685005\n","10 0.00000e+00  3.901527 0.000000e+00  15.221914 1.6030804   4.807442\n","11 0.00000e+00  1.600422 0.000000e+00   2.561349 0.2303847 -26.544381\n","12 0.00000e+00  5.545202 0.000000e+00  30.749267 0.8879370  -6.837006\n","13 0.00000e+00  3.759714 0.000000e+00  14.135450 0.9631687   5.080131\n","14 0.00000e+00  2.530855 0.000000e+00   6.405225 0.3331290 -14.462452\n","         REML p_value        FDR   CI_lower   CI_upper\n","1   -9.507130   0.000 0.00000000  23.582865  44.853539\n","2  -36.975906   0.000 0.00000000 -32.440037 -27.639692\n","3  -57.889199   0.000 0.00000000  19.127064  22.900257\n","4    2.721439   0.000 0.00000000  11.958354  19.057937\n","5  -14.996300   0.000 0.00000000  11.453587  15.115053\n","6   16.256488   0.000 0.00000000   8.603374  10.976220\n","7   -4.170912   0.000 0.00000000   7.029199  10.959674\n","8   18.165023   0.000 0.00000000 -10.056841  -5.552310\n","9  -12.929202   0.000 0.00000000  -8.676822  -6.749186\n","10  16.143471   0.001 0.04638889   4.564678  10.848753\n","11 -38.027137   0.000 0.00000000  -6.566973  -5.663865\n","12  10.109554   0.000 0.00000000  -7.811188  -4.330474\n","13  16.245064   0.000 0.00000000   3.005212   6.780834\n","14 -28.354550   0.000 0.00000000  -5.470795  -4.164929\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the last function to retrieve the tissue information and categories for fs1:\n","input_eqtl_path <- \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\"\n","input_mt_robin_path <- \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_protein_coding_genes_sig_full_table.txt\"\n","output_path <- \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_protein_coding_genes_sig_full_table_tissues.txt\"\n","process_eqtl_mt_robin(input_eqtl_path, input_mt_robin_path, output_path)"],"metadata":{"id":"ouQG9thnQ2dc","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714699054595,"user_tz":-570,"elapsed":48571,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"e74f9e61-8386-421a-d32f-bb372bcfe043"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["           gene_id gene_name SNPs\n","1  ENSG00000026950    BTN3A1    9\n","2  ENSG00000065457     ADAT1    4\n","3  ENSG00000090661     CERS4   15\n","4  ENSG00000113734     BNIP1    7\n","5  ENSG00000132507     EIF5A    2\n","6  ENSG00000139714     MORN3   18\n","7  ENSG00000146530      VWDE    8\n","8  ENSG00000158825       CDA   14\n","9  ENSG00000171160     MORN4    2\n","10 ENSG00000173540     GMPPB    2\n","11 ENSG00000176386     CDC26    2\n","12 ENSG00000177025  C19orf18   17\n","13 ENSG00000183336     BOLA2    3\n","14 ENSG00000184983    NDUFA6    2\n","                                                                                                                                                                                                SNPs_IDs\n","1                                                                                                         rs114157910,rs12213056,rs13194491,rs2275906,rs3208733,rs3799378,rs56325223,rs7356982,rs9467733\n","2                                                                                                                                                             rs11642599,rs35261357,rs4888438,rs59297469\n","3                                      rs113738552,rs11880745,rs142392712,rs147441807,rs1965888,rs2336171,rs4542783,rs4804282,rs56120315,rs56184962,rs57062897,rs59099428,rs60585010,rs7260363,rs8100119\n","4                                                                                                                             rs10476082,rs11385449,rs35813302,rs4868243,rs5745108,rs72814178,rs77146849\n","5                                                                                                                                                                                   rs62059195,rs7503161\n","6  rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","7                                                                                                                      rs12673085,rs12699367,rs2287067,rs28811348,rs2966627,rs4721083,rs6971764,rs970035\n","8                                                     rs10753504,rs10916784,rs10916834,rs12026680,rs12045689,rs12135822,rs1253799,rs3131712,rs3215400,rs55665119,rs66731853,rs72649199,rs818199,rs879088\n","9                                                                                                                                                                                   rs12358715,rs7905087\n","10                                                                                                                                                                                 rs200287431,rs2280405\n","11                                                                                                                                                                                 rs10981712,rs10981764\n","12                     .,rs1014767,rs112024224,rs11668314,rs118061825,rs138343123,rs142282855,rs145778607,rs149882742,rs150518835,rs257676,rs292831,rs34640515,rs36115826,rs3764536,rs4801529,rs57797045\n","13                                                                                                                                                                        rs4533290,rs79379134,rs9888967\n","14                                                                                                                                                                                  rs137139,rs139043947\n","   Obs     beta_y        Min     First_Q      Median   Third_Q       Max\n","1   22  -4.817862 -2.4698318 -0.12749636  0.14242212 0.6493537 2.1276346\n","2   11  -7.713004 -1.5988537 -0.53363730 -0.28594064 0.1299801 2.3368205\n","3   23  -7.804575 -1.9381654 -0.03024777  0.45603996 0.7985665 2.7709937\n","4   13   8.994436 -2.0411354 -0.57137796  0.13557018 0.5707352 1.3174713\n","5   20  21.013660 -1.3847656 -0.83600738 -0.47418977 0.2216229 2.5618909\n","6   21   4.893023 -1.2647845 -0.64117122  0.19594240 0.9362305 1.9255321\n","7   12  15.508145 -2.8063695 -0.57174743 -0.10704222 0.1810687 0.4180642\n","8   26  -6.070831 -0.6375271 -0.29178418  0.07034354 0.6669760 3.1846266\n","9    3  34.218202 -0.5582269 -0.52550340 -0.49277994 0.3547653 1.2023106\n","10  18  13.284320 -1.7327039 -0.70127932  0.31619771 0.9434917 1.3353628\n","11  18 -30.039865 -1.7931493 -0.71338242 -0.15416886 0.3773474 2.4878981\n","12  20   7.706716 -2.0921744 -0.88505264 -0.17477279 0.4768810 1.4996117\n","13  21  -6.115419 -2.1062585 -0.64359310 -0.21806788 0.5977871 2.1480464\n","14  24   9.789797 -2.9183208 -0.46957241 -0.04249690 0.6506457 1.5058086\n","         SD_RE Res_SD_RE       Var_RE Res_Var_RE     SE_FE t_value_FE\n","1  0.00000e+00  2.530855 0.000000e+00   6.405225 0.3331290 -14.462452\n","2  0.00000e+00  3.225459 0.000000e+00  10.403585 0.4917438 -15.685005\n","3  0.00000e+00  6.604208 0.000000e+00  43.615568 1.1491150  -6.791814\n","4  0.00000e+00  3.182195 0.000000e+00  10.126362 1.0026723   8.970464\n","5  0.00000e+00  2.500070 0.000000e+00   6.250350 0.9625492  21.831257\n","6  0.00000e+00  3.759714 0.000000e+00  14.135450 0.9631687   5.080131\n","7  0.00000e+00  4.285012 0.000000e+00  18.361329 1.8111180   8.562747\n","8  0.00000e+00  5.545202 0.000000e+00  30.749267 0.8879370  -6.837006\n","9  0.00000e+00  4.486594 0.000000e+00  20.129522 5.4261924   6.306117\n","10 2.90556e-09  4.429408 8.442281e-18  19.619659 0.9340473  14.222321\n","11 0.00000e+00  1.577515 0.000000e+00   2.488552 1.2245776 -24.530796\n","12 0.00000e+00  3.901527 0.000000e+00  15.221914 1.6030804   4.807442\n","13 0.00000e+00  1.600422 0.000000e+00   2.561349 0.2303847 -26.544381\n","14 0.00000e+00  4.670701 0.000000e+00  21.815451 0.6053178  16.172988\n","         REML p_value        FDR   CI_lower   CI_upper\n","1  -28.354550   0.000 0.00000000  -5.470795  -4.164929\n","2  -12.929202   0.000 0.00000000  -8.676822  -6.749186\n","3   18.165023   0.000 0.00000000 -10.056841  -5.552310\n","4   -4.170912   0.000 0.00000000   7.029199  10.959674\n","5  -57.889199   0.000 0.00000000  19.127064  22.900257\n","6   16.245064   0.000 0.00000000   3.005212   6.780834\n","7    2.721439   0.000 0.00000000  11.958354  19.057937\n","8   10.109554   0.000 0.00000000  -7.811188  -4.330474\n","9   -9.507130   0.000 0.00000000  23.582865  44.853539\n","10 -14.996300   0.000 0.00000000  11.453587  15.115053\n","11 -36.975906   0.000 0.00000000 -32.440037 -27.639692\n","12  16.143471   0.001 0.04638889   4.564678  10.848753\n","13 -38.027137   0.000 0.00000000  -6.566973  -5.663865\n","14  16.256488   0.000 0.00000000   8.603374  10.976220\n","                                                                                                                                                                                                                                                                                                                                                                                                                     Tissues\n","1                                                           Subcutaneous adipose,Aorta artery,Coronary artery,Amygdala brain,Anterior cingulate cortex,Cerebellum,Hippocampus,Nucleus accumbens (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Sigmoid colon,Esophagus muscularis,Kidney cortex,Lung,Skeletal muscle,Tibial nerve,Pancreas,Prostate,Not sun-exposed skin (suprapubic),Testis,Whole blood\n","2                                                                                                                                                                                                                                                                    Subcutaneous adipose,Cerebellar hemisphere,Cerebellum,Frontal cortex,Hippocampus,Hypothalamus,Putamen (basal ganglia),Liver,Lung,Skeletal muscle,Testis\n","3                                    Visceral omentum adipose,Tibial artery,Amygdala brain,Anterior cingulate cortex,Caudate,Cerebellar hemisphere,Cortex,Frontal cortex,Hippocampus,Spinal cord,Substantia nigra,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Minor salivary gland,Skeletal muscle,Ovary,Pituitary,Not sun-exposed skin (suprapubic),Testis,Uterus\n","4                                                                                                                                                                                                        Spinal cord,Cultured fibroblasts,EBV-transformed lymphocytes,Sigmoid colon,Esophagus mucosa,Esophagus muscularis,Kidney cortex,Lung,Minor salivary gland,Prostate,Small intestine terminal ileum,Uterus,Whole blood\n","5                                              Subcutaneous adipose,Tibial artery,Amygdala brain,Frontal cortex,Hypothalamus,Nucleus accumbens (basal ganglia),Substantia nigra,Cultured fibroblasts,Esophagus mucosa,Atrial appendage,Minor salivary gland,Skeletal muscle,Tibial nerve,Pancreas,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Thyroid,Whole blood\n","6                                                        Visceral omentum adipose,Adrenal gland,Tibial artery,Cerebellum,Nucleus accumbens (basal ganglia),Spinal cord,Breast mammary tissue,EBV-transformed lymphocytes,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Kidney cortex,Liver,Minor salivary gland,Pancreas,Prostate,Sun-exposed skin (lower leg),Thyroid,Vagina,Whole blood\n","7                                                                                                                                                                                                                                                  Adrenal gland,Caudate,Spinal cord,Atrial appendage,Kidney cortex,Tibial nerve,Pituitary,Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Testis,Thyroid\n","8            Visceral omentum adipose,Adrenal gland,Coronary artery,Tibial artery,Hippocampus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Esophagus mucosa,Atrial appendage,Left ventricle,Kidney cortex,Minor salivary gland,Tibial nerve,Ovary,Pancreas,Pituitary,Prostate,Stomach,Thyroid,Uterus,Vagina,Whole blood\n","9                                                                                                                                                                                                                                                                                                                                                                                Cultured fibroblasts,Left ventricle,Thyroid\n","10                                                                                                Subcutaneous adipose,Adrenal gland,Aorta artery,Coronary artery,Tibial artery,Cerebellum,Hippocampus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Esophagus muscularis,Lung,Minor salivary gland,Skeletal muscle,Tibial nerve,Pancreas,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Stomach\n","11                                                                                                                                                                                    Amygdala brain,Caudate,Cerebellar hemisphere,Cerebellum,Cortex,Frontal cortex,Putamen (basal ganglia),Spinal cord,Substantia nigra,Left ventricle,Liver,Pancreas,Pituitary,Prostate,Sun-exposed skin (lower leg),Spleen,Testis,Thyroid\n","12                                                                               Adrenal gland,Coronary artery,Anterior cingulate cortex,Cerebellum,Hypothalamus,Putamen (basal ganglia),Substantia nigra,Cultured fibroblasts,Sigmoid colon,Gastroesophageal junction,Esophagus mucosa,Minor salivary gland,Tibial nerve,Pituitary,Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Stomach,Uterus,Vagina\n","13                       Subcutaneous adipose,Visceral omentum adipose,Tibial artery,Anterior cingulate cortex,Cerebellar hemisphere,Cerebellum,Breast mammary tissue,EBV-transformed lymphocytes,Transverse colon,Atrial appendage,Kidney cortex,Minor salivary gland,Skeletal muscle,Ovary,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Thyroid,Uterus,Vagina,Whole blood\n","14 Subcutaneous adipose,Adrenal gland,Coronary artery,Amygdala brain,Anterior cingulate cortex,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus muscularis,Left ventricle,Kidney cortex,Liver,Minor salivary gland,Tibial nerve,Pituitary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Stomach,Thyroid,Uterus\n","                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Tissue_Categories\n","1                                                                              Adipose Tissue,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Digestive System,Digestive System,Excretory System,Respiratory System,Muscular System,Brain and Nervous System,Digestive System,Reproductive System,Skin,Reproductive System,Blood\n","2                                                                                                                                                                                                                                                                              Adipose Tissue,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Digestive System,Respiratory System,Muscular System,Reproductive System\n","3                                                 Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Exocrine System,Muscular System,Reproductive System,Endocrine System,Skin,Reproductive System,Reproductive System\n","4                                                                                                                                                                                                                                                                                            Brain and Nervous System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Excretory System,Respiratory System,Exocrine System,Reproductive System,Digestive System,Reproductive System,Blood\n","5                                                                                                                                                Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Cardiovascular System,Exocrine System,Muscular System,Brain and Nervous System,Digestive System,Reproductive System,Skin,Skin,Digestive System,Endocrine System,Blood\n","6                                                                                                                                    Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Digestive System,Exocrine System,Digestive System,Reproductive System,Skin,Endocrine System,Reproductive System,Blood\n","7                                                                                                                                                                                                                                                                                      Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Cardiovascular System,Excretory System,Brain and Nervous System,Endocrine System,Skin,Digestive System,Immune System,Reproductive System,Endocrine System\n","8  Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Cardiovascular System,Cardiovascular System,Excretory System,Exocrine System,Brain and Nervous System,Reproductive System,Digestive System,Endocrine System,Reproductive System,Digestive System,Endocrine System,Reproductive System,Reproductive System,Blood\n","9                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Connective Tissue,Cardiovascular System,Endocrine System\n","10                                                                                                                                                                   Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Digestive System,Respiratory System,Exocrine System,Muscular System,Brain and Nervous System,Digestive System,Skin,Skin,Digestive System\n","11                                                                                                                                    Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Cardiovascular System,Digestive System,Digestive System,Endocrine System,Reproductive System,Skin,Immune System,Reproductive System,Endocrine System\n","12                                                                                                                  Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Digestive System,Digestive System,Exocrine System,Brain and Nervous System,Endocrine System,Skin,Digestive System,Immune System,Digestive System,Reproductive System,Reproductive System\n","13                                                                                                                                                       Adipose Tissue,Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Cardiovascular System,Excretory System,Exocrine System,Muscular System,Reproductive System,Skin,Skin,Digestive System,Endocrine System,Reproductive System,Reproductive System,Blood\n","14                                                                                   Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Digestive System,Exocrine System,Brain and Nervous System,Endocrine System,Reproductive System,Skin,Skin,Digestive System,Immune System,Digestive System,Endocrine System,Reproductive System\n"]}]},{"cell_type":"markdown","source":["### **GWAS2**"],"metadata":{"id":"GDoyKevmqf8o"}},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(matrixStats) # It needs to be installed separately\n","#library(BiocManager) # It needs to be installed separately\n","library(snpStats) # It needs to be installed separately\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(reshape2) # Additional package for melt function\n","\n","# Step 2: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 3: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_2_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 4: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 5: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 6: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 7: Create a variable for the path to the directory to save the LD matrix results:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_2\"\n","\n","# Step 8: Create the directory if it doesn't exist:\n","if (!dir.exists(ld_matrix_dir)) {\n","  dir.create(ld_matrix_dir)\n","}\n","\n","# Step 9: Modify the original select_IV_bs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_bs() --> backward selection starts with all variables and removes one at a time:\n","select_IV_bs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset the eQTL data set for the gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Get pairwise r^2:\n","  LD_r2_pw <- LD_r2\n","  LD_r2_pw[lower.tri(LD_r2_pw,diag=T)] <- NA\n","  LD_r2_pw <- reshape2::melt(LD_r2_pw)\n","  colnames(LD_r2_pw) <- c(\"SNP1\",\"SNP2\",\"r2\")\n","  LD_r2_pw$SNP1 <- as.character(LD_r2_pw$SNP1)\n","  LD_r2_pw$SNP2 <- as.character(LD_r2_pw$SNP2)\n","\n","  ## Drop SNPs paired with themselves and duplicates (resulting from lower triangle, set to NA):\n","  LD_r2_pw <- subset(LD_r2_pw, SNP1 != SNP2 & !is.na(r2))\n","\n","  ## Sort by descending r^2:\n","  LD_r2_pw <- LD_r2_pw[order(LD_r2_pw$r2,decreasing = T),]\n","\n","  ## Check for case all candidate SNPs have pw r^2 > thresh (return SNPs with smallest p-value):\n","  if(min(LD_r2_pw$r2) >=ld_thresh){\n","    return(eqtl_data$variant_id[which.max(eqtl_data$min_p)])\n","  }\n","\n","  while(max(LD_r2_pw$r2)>=ld_thresh){\n","\n","    ## Identify top SNPs in the pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    LD_r2_pw <- subset(LD_r2_pw, SNP1 %in% SNP_pool & SNP2 %in% SNP_pool)\n","    LD_r2_pw_geThresh <- subset(LD_r2_pw, r2>=ld_thresh)\n","\n","    ## Identify SNP pair with highest remaining r^2:\n","    next_pair <- LD_r2_pw[1,]\n","\n","    ## Criteria 1: Drop the SNP with largest number of correlations > threshold with other SNPs:\n","    SNP1_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP1) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP1)\n","    SNP2_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP2) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP2)\n","    if(SNP1_nSNP_geThresh > SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP1\n","    } else if (SNP1_nSNP_geThresh < SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP2\n","    } else{\n","\n","      ## Criteria 2: Drop the SNP with smaller number of tissues with p<0.001:\n","      next_pair_eqtl <- subset(eqtl_data, variant_id %in% c(next_pair$SNP1,next_pair$SNP2))\n","      SNP1_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP1)$nTiss_ltThresh\n","      SNP2_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP2)$nTiss_ltThresh\n","\n","      if(SNP1_nTiss > SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP2\n","      } else if(SNP1_nTiss < SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP1\n","      } else{\n","\n","        ## Criteria 3: Drop the SNP with larger minimum p-value:\n","        next_SNP_drop <- next_pair_eqtl$variant_id[which.max(next_pair_eqtl$min_p)]\n","      }\n","    }\n","    SNP_pool <- SNP_pool[which(SNP_pool != next_SNP_drop)]\n","  }\n","  return(SNP_pool)\n","}\n","\n","# Step 10: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 11:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","  # Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_bs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","  }, error = function(e) print(e))\n","}, mc.cores = num_cores)\n","\n","# Step 12:\n","names(result_list) <- all_genes_vector\n","\n","# Step 13:\n","saveRDS(result_list, file = \"/data/long_covid_IVs_bs_2.rds\")"],"metadata":{"id":"j0cGdv3nqC7b"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/select_function_bs_2.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"eaCfw5nmqDjb"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 04:51:18:\n","   Job Id:             93353705.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      800.16\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 258:44:17\n","   Memory Requested:   500.0GB               Memory Used: 393.32GB\n","   Walltime requested: 30:00:00            Walltime Used: 05:33:24\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"EBxIgiQVqEOB"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Open the output file:\n","result_selectIV_fs_2 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_IVs_fs_2.rds\")\n","\n","# Filter out errors:\n","result_selectIV_fs_2 <- result_selectIV_fs_2[!sapply(result_selectIV_fs_2, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_selectIV_fs_2))\n","print(length(unique(result_selectIV_fs_2)))\n","print(head(result_selectIV_fs_2))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"OOlfUFtJqEdm","executionInfo":{"status":"ok","timestamp":1693540209513,"user_tz":-570,"elapsed":3364,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"69d62f44-cc12-4ef7-c826-4a78bda128f8"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 38574\n","[1] 38480\n","$ENSG00000000003\n"," [1] \"rs8780\"      \"rs1204407\"   \"rs189980168\" \"rs4145090\"   \"rs139645081\"\n"," [6] \"rs112002620\" \"rs113922837\" \"rs17329105\"  \"rs193124579\" \"rs146639579\"\n","[11] \"rs201585484\" \"rs111478058\" \"rs143491698\" \"rs572714672\" \"rs56100956\" \n","[16] \"rs147601806\" \"rs139830751\" \"rs12842251\"  \"rs73630694\" \n","\n","$ENSG00000000005\n"," [1] \"rs6616166\"   \"rs4828038\"   \"rs12006573\"  \"rs73557922\"  \"rs141394856\"\n"," [6] \"rs140468798\" \"rs73555045\"  \"rs5921738\"   \"rs139231094\" \"rs59368006\" \n","[11] \"rs111915382\" \"rs73562536\"  \"rs12832642\"  \"rs201254498\" \"rs186611188\"\n","[16] \"rs113422802\" \"rs796285800\" \"rs180795909\" \"rs1004006\"   \"rs148617015\"\n","[21] \"rs140109110\"\n","\n","$ENSG00000000419\n"," [1] \"rs1292040204\" \"rs79562553\"   \"rs4809840\"    \"rs35201621\"   \"rs35530561\"  \n"," [6] \"rs141892692\"  \"rs944933\"     \"rs11483800\"   \"rs113036491\"  \"rs116892696\" \n","[11] \"rs34483216\"   \"rs60578213\"   \"rs118181672\"  \"rs4811253\"    \"rs11086363\"  \n","[16] \"rs73278571\"   \"rs73131221\"   \"rs78319058\"   \"rs79223715\"   \"rs189417964\" \n","[21] \"rs74422706\"   \"rs112070162\"  \"rs59034704\"   \"rs180710262\"  \"rs577456025\" \n","[26] \"rs117900008\"  \"rs34223230\"   \"rs141885330\"  \"rs78937289\"   \"rs16995896\"  \n","[31] \"rs72020869\"   \"rs142281981\"  \"rs13037515\"   \"rs1883804\"    \"rs17730415\"  \n","[36] \"rs11697209\"   \"rs112566287\"  \"rs149762110\"  \"rs4809839\"    \"rs6021441\"   \n","[41] \"rs75866341\"   \"rs13036755\"   \"rs74177405\"  \n","\n","$ENSG00000000457\n"," [1] \"rs10753796\"  \"rs4656710\"   \"rs115234612\" \"rs3766155\"   \"rs2056980\"  \n"," [6] \"rs4987403\"   \"rs12123695\"  \"rs185963038\" \"rs116684351\" \"rs1076637\"  \n","[11] \"rs7417281\"   \"rs7541615\"   \"rs973608\"    \"rs2982465\"   \"rs113123846\"\n","[16] \"rs4987338\"   \"rs114779274\" \"rs12038549\"  \"rs112361437\"\n","\n","$ENSG00000000460\n"," [1] \"rs35296671\"  \"rs1800016\"   \"rs4656198\"   \"rs3753311\"   \"rs4987391\"  \n"," [6] \"rs180738486\" \"rs4562681\"   \"rs531844\"    \"rs115157324\" \"rs12116424\" \n","[11] \"rs17346300\"  \"rs60843604\"  \"rs12742457\"  \"rs78159465\"  \"rs139679825\"\n","[16] \"rs35662131\" \n","\n","$ENSG00000000938\n"," [1] \"rs3737803\"   \"rs72191132\"  \"rs137984979\" \"rs11273428\"  \"rs75800899\" \n"," [6] \"rs191579330\" \"rs9426420\"   \"rs148432575\" \"rs10430146\"  \"rs139100109\"\n","[11] \"rs6701065\"   \"rs80298610\"  \"rs17361686\"  \"rs484740\"    \"rs114059345\"\n","[16] \"rs72651589\"  \"rs4357550\"   \"rs41284294\"  \"rs138392492\" \"rs148577969\"\n","[21] \"rs151288141\" \"rs71642227\"  \"rs12759594\"  \"rs2076457\"   \"rs141256722\"\n","[26] \"rs76443568\" \n","\n"]}]},{"cell_type":"code","source":["library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_2_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the list of genes-SNPs:\n","result_selectedIVs <- readRDS(\"/data/long_covid_IVs_bs_2.rds\")\n","\n","# Step 4: Create a variable for the path to the LD matrix directory:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_2\"\n","\n","# Step 5: Setup the Mt_Robin method using the following setup function:\n","MR_MtRobin_setup <- function(geneID,\n","                             snpID,\n","                             eqtl_data,\n","                             gwas_data,\n","                             ld_matrix_path,\n","                             nTiss){\n","  tryCatch({\n","    ## Check if the gene is present and subset\n","    if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","    eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","    ## Confirm all SNPs are present in the dataset\n","    if(!all(snpID %in% eqtl_data$variant_id)) stop(\"Some SNPs missing in eQTL dataset\")\n","    if(!all(snpID %in% gwas_data$variant_id)) stop(\"Some SNPs missing in GWAS dataset\")\n","\n","    # Read the LD matrix from a file\n","    LD <- as.matrix(utils::read.table(ld_matrix_path,\n","                                      header = TRUE,\n","                                      check.names = FALSE))\n","\n","    if(!all(snpID %in% colnames(LD))) stop(\"Some SNPs missing in LD matrix\")\n","\n","    ## Align data\n","    eqtl_data <- eqtl_data[match(snpID,eqtl_data$variant_id),]\n","    gwas_data <- gwas_data[match(snpID,gwas_data$variant_id),]\n","    LD <- LD[match(snpID,rownames(LD)),match(snpID,colnames(LD))]\n","\n","    eqtl_betas <- as.matrix(eqtl_data[,paste0(\"beta_\",1:nTiss)])\n","    eqtl_se <- as.matrix(eqtl_data[,paste0(\"SE_\",1:nTiss)])\n","    eqtl_pvals <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","    gwas_betas <- gwas_data$beta\n","    gwas_se <- gwas_data$SE\n","\n","    row.names(eqtl_betas) <- row.names(eqtl_se) <- row.names(eqtl_pvals) <- eqtl_data$variant_id\n","\n","    return(list(snpID=snpID,gwas_betas=gwas_betas,gwas_se=gwas_se,\n","                eqtl_betas=eqtl_betas,eqtl_se=eqtl_se,eqtl_pvals=eqtl_pvals,LD=LD))\n","  }, error = function(e) print(e))\n","}\n","\n","# Step 6: Apply the previous setup function:\n","\n","# Step 6.1: Initialize an empty list to store the results:\n","MR_MtRobin_results <- list()\n","\n","# Step 6.2: Define the number of cores to use for the parallelism process:\n","num_cores <- 36\n","\n","# Step 6.3: Loop over each gene in result_selectedIVs:\n","MR_MtRobin_results <- mclapply(names(result_selectedIVs), function(geneID) {\n","\n","  # Step 6.3.1: Get the SNPs associated with each gene:\n","  snpID <- result_selectedIVs[[geneID]]\n","\n","  # Step 6.3.2: Generate a unique file path for the LD matrix of this gene:\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  # Step 6.3.3: Call the MR_MtRobin_setup() function with the parameters for each gene:\n","  return(tryCatch({\n","    MR_MtRobin_setup(geneID = geneID,\n","                     snpID = snpID,\n","                     eqtl_data = eqtl_covid_exposure,\n","                     gwas_data = gwas_covid_outcome,\n","                     ld_matrix_path = ld_matrix_path,\n","                     nTiss = 49)\n","  }, error = function(e) print(e)))\n","}, mc.cores = num_cores)\n","\n","# Step 7: Assign names to the list:\n","names(MR_MtRobin_results) <- names(result_selectedIVs)\n","\n","# Step 8: Save the previous result as a .rds file:\n","saveRDS(MR_MtRobin_results, file = \"/data/long_covid_setup_bs_2.rds\")"],"metadata":{"id":"352rPJi67sv4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/setup_function_bs_2.R\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"zgbVIN2C70o2"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 16:29:36:\n","   Job Id:             93430390.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      74.34\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 19:47:13\n","   Memory Requested:   300.0GB               Memory Used: 83.56GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:41:18\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"jY1IYtv_74uR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Open the output file:\n","result_setup_fs_2 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_setup_fs_2.rds\")\n","\n","# Filter out errors:\n","result_setup_fs_2 <- result_setup_fs_2[!sapply(result_setup_fs_2, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_setup_fs_2))\n","print(length(unique(result_setup_fs_2)))\n","print(result_setup_fs_2[[1]])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"TiqniwTi76l3","executionInfo":{"status":"ok","timestamp":1693540233872,"user_tz":-570,"elapsed":7902,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"783a019b-d746-4b9b-bbff-2d1d5a497eca"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 4754\n","[1] 4754\n","$snpID\n","[1] \"rs147560086\" \"rs7955986\"   \"rs718389\"    \"rs11571383\"  \"rs2107614\"  \n","[6] \"rs4980935\"   \"rs11064536\"  \"rs117766797\"\n","\n","$gwas_betas\n","[1]  0.021120800 -0.006508970 -0.017204900 -0.045787500 -0.000839245\n","[6] -0.015072100  0.005801520 -0.016338200\n","\n","$gwas_se\n","[1] 0.0258757 0.0206358 0.0234662 0.0569715 0.0220439 0.0229060 0.0281981\n","[8] 0.0628533\n","\n","$eqtl_betas\n","            beta_1 beta_2    beta_3   beta_4 beta_5    beta_6    beta_7 beta_8\n","rs147560086     NA     NA -0.332722 -0.36379     NA -0.428724        NA     NA\n","rs7955986       NA     NA        NA       NA     NA        NA        NA     NA\n","rs718389        NA     NA        NA       NA     NA        NA        NA     NA\n","rs11571383      NA     NA        NA       NA     NA        NA -0.525715     NA\n","rs2107614       NA     NA        NA       NA     NA        NA        NA     NA\n","rs4980935       NA     NA        NA       NA     NA        NA        NA     NA\n","rs11064536      NA     NA        NA       NA     NA        NA        NA     NA\n","rs117766797     NA     NA        NA       NA     NA        NA        NA     NA\n","               beta_9 beta_10 beta_11 beta_12 beta_13 beta_14 beta_15   beta_16\n","rs147560086 -0.362536      NA      NA -0.5128      NA      NA      NA        NA\n","rs7955986          NA      NA      NA      NA      NA      NA      NA        NA\n","rs718389           NA      NA      NA      NA      NA      NA      NA        NA\n","rs11571383         NA      NA      NA      NA      NA      NA      NA        NA\n","rs2107614          NA      NA      NA      NA      NA      NA      NA -0.224019\n","rs4980935          NA      NA      NA      NA      NA      NA      NA        NA\n","rs11064536         NA      NA      NA      NA      NA      NA      NA        NA\n","rs117766797        NA      NA      NA      NA      NA      NA      NA        NA\n","            beta_17  beta_18  beta_19   beta_20   beta_21 beta_22   beta_23\n","rs147560086      NA -0.27895       NA -0.393493 -0.538392      NA -0.380267\n","rs7955986        NA       NA       NA        NA        NA      NA        NA\n","rs718389         NA       NA       NA        NA        NA      NA        NA\n","rs11571383       NA       NA       NA        NA        NA      NA        NA\n","rs2107614        NA       NA       NA        NA        NA      NA        NA\n","rs4980935        NA       NA 0.242572        NA        NA      NA        NA\n","rs11064536       NA       NA       NA        NA        NA      NA        NA\n","rs117766797      NA       NA       NA        NA        NA      NA        NA\n","            beta_24  beta_25 beta_26 beta_27 beta_28   beta_29  beta_30 beta_31\n","rs147560086      NA -0.37429      NA      NA      NA -0.413552       NA      NA\n","rs7955986        NA       NA      NA      NA      NA        NA       NA      NA\n","rs718389         NA       NA      NA      NA      NA        NA       NA      NA\n","rs11571383       NA       NA      NA      NA      NA        NA       NA      NA\n","rs2107614        NA       NA      NA      NA      NA        NA       NA      NA\n","rs4980935        NA       NA      NA      NA      NA        NA       NA      NA\n","rs11064536       NA       NA      NA      NA      NA        NA 0.524096      NA\n","rs117766797      NA       NA      NA      NA      NA        NA       NA      NA\n","            beta_32 beta_33   beta_34   beta_35 beta_36 beta_37 beta_38\n","rs147560086      NA      NA -0.389496 -0.325135      NA      NA      NA\n","rs7955986        NA      NA        NA        NA      NA      NA      NA\n","rs718389         NA      NA        NA        NA      NA      NA      NA\n","rs11571383       NA      NA        NA        NA      NA      NA      NA\n","rs2107614        NA      NA        NA        NA      NA      NA      NA\n","rs4980935        NA      NA        NA        NA      NA      NA      NA\n","rs11064536       NA      NA        NA        NA      NA      NA      NA\n","rs117766797      NA      NA        NA        NA      NA      NA      NA\n","              beta_39 beta_40 beta_41   beta_42 beta_43   beta_44   beta_45\n","rs147560086 -0.348491      NA      NA        NA      NA -0.283741        NA\n","rs7955986          NA      NA      NA        NA      NA        NA        NA\n","rs718389           NA      NA      NA        NA      NA        NA -0.282161\n","rs11571383         NA      NA      NA -0.504512      NA        NA        NA\n","rs2107614          NA      NA      NA        NA      NA        NA        NA\n","rs4980935          NA      NA      NA        NA      NA        NA        NA\n","rs11064536         NA      NA      NA        NA      NA        NA        NA\n","rs117766797        NA      NA      NA        NA      NA        NA        NA\n","             beta_46 beta_47   beta_48 beta_49\n","rs147560086       NA      NA        NA      NA\n","rs7955986   0.243345      NA        NA      NA\n","rs718389          NA      NA        NA      NA\n","rs11571383        NA      NA        NA      NA\n","rs2107614         NA      NA        NA      NA\n","rs4980935         NA      NA        NA      NA\n","rs11064536        NA      NA        NA      NA\n","rs117766797       NA      NA -0.395786      NA\n","\n","$eqtl_se\n","            SE_1 SE_2      SE_3     SE_4 SE_5      SE_6     SE_7 SE_8      SE_9\n","rs147560086   NA   NA 0.0464181 0.029495   NA 0.0261202       NA   NA 0.0550403\n","rs7955986     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs718389      NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs11571383    NA   NA        NA       NA   NA        NA 0.105786   NA        NA\n","rs2107614     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs4980935     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs11064536    NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs117766797   NA   NA        NA       NA   NA        NA       NA   NA        NA\n","            SE_10 SE_11     SE_12 SE_13 SE_14 SE_15     SE_16 SE_17     SE_18\n","rs147560086    NA    NA 0.0567995    NA    NA    NA        NA    NA 0.0775922\n","rs7955986      NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs718389       NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs11571383     NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs2107614      NA    NA        NA    NA    NA    NA 0.0401716    NA        NA\n","rs4980935      NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs11064536     NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs117766797    NA    NA        NA    NA    NA    NA        NA    NA        NA\n","                SE_19     SE_20     SE_21 SE_22     SE_23 SE_24     SE_25 SE_26\n","rs147560086        NA 0.0327279 0.0351663    NA 0.0352046    NA 0.0353582    NA\n","rs7955986          NA        NA        NA    NA        NA    NA        NA    NA\n","rs718389           NA        NA        NA    NA        NA    NA        NA    NA\n","rs11571383         NA        NA        NA    NA        NA    NA        NA    NA\n","rs2107614          NA        NA        NA    NA        NA    NA        NA    NA\n","rs4980935   0.0586861        NA        NA    NA        NA    NA        NA    NA\n","rs11064536         NA        NA        NA    NA        NA    NA        NA    NA\n","rs117766797        NA        NA        NA    NA        NA    NA        NA    NA\n","            SE_27 SE_28     SE_29    SE_30 SE_31 SE_32 SE_33    SE_34     SE_35\n","rs147560086    NA    NA 0.0384595       NA    NA    NA    NA 0.030555 0.0222479\n","rs7955986      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs718389       NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs11571383     NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs2107614      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs4980935      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs11064536     NA    NA        NA 0.136083    NA    NA    NA       NA        NA\n","rs117766797    NA    NA        NA       NA    NA    NA    NA       NA        NA\n","            SE_36 SE_37 SE_38     SE_39 SE_40 SE_41     SE_42 SE_43     SE_44\n","rs147560086    NA    NA    NA 0.0452671    NA    NA        NA    NA 0.0360217\n","rs7955986      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs718389       NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs11571383     NA    NA    NA        NA    NA    NA 0.0733203    NA        NA\n","rs2107614      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs4980935      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs11064536     NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs117766797    NA    NA    NA        NA    NA    NA        NA    NA        NA\n","                SE_45     SE_46 SE_47    SE_48 SE_49\n","rs147560086        NA        NA    NA       NA    NA\n","rs7955986          NA 0.0191166    NA       NA    NA\n","rs718389    0.0359888        NA    NA       NA    NA\n","rs11571383         NA        NA    NA       NA    NA\n","rs2107614          NA        NA    NA       NA    NA\n","rs4980935          NA        NA    NA       NA    NA\n","rs11064536         NA        NA    NA       NA    NA\n","rs117766797        NA        NA    NA 0.107595    NA\n","\n","$eqtl_pvals\n","            pvalue_1 pvalue_2    pvalue_3    pvalue_4 pvalue_5    pvalue_6\n","rs147560086       NA       NA 1.57506e-11 8.15374e-29       NA 5.49137e-49\n","rs7955986         NA       NA          NA          NA       NA          NA\n","rs718389          NA       NA          NA          NA       NA          NA\n","rs11571383        NA       NA          NA          NA       NA          NA\n","rs2107614         NA       NA          NA          NA       NA          NA\n","rs4980935         NA       NA          NA          NA       NA          NA\n","rs11064536        NA       NA          NA          NA       NA          NA\n","rs117766797       NA       NA          NA          NA       NA          NA\n","               pvalue_7 pvalue_8    pvalue_9 pvalue_10 pvalue_11   pvalue_12\n","rs147560086          NA       NA 6.71399e-10        NA        NA 4.40477e-16\n","rs7955986            NA       NA          NA        NA        NA          NA\n","rs718389             NA       NA          NA        NA        NA          NA\n","rs11571383  2.64972e-06       NA          NA        NA        NA          NA\n","rs2107614            NA       NA          NA        NA        NA          NA\n","rs4980935            NA       NA          NA        NA        NA          NA\n","rs11064536           NA       NA          NA        NA        NA          NA\n","rs117766797          NA       NA          NA        NA        NA          NA\n","            pvalue_13 pvalue_14 pvalue_15  pvalue_16 pvalue_17  pvalue_18\n","rs147560086        NA        NA        NA         NA        NA 0.00050353\n","rs7955986          NA        NA        NA         NA        NA         NA\n","rs718389           NA        NA        NA         NA        NA         NA\n","rs11571383         NA        NA        NA         NA        NA         NA\n","rs2107614          NA        NA        NA 1.0067e-07        NA         NA\n","rs4980935          NA        NA        NA         NA        NA         NA\n","rs11064536         NA        NA        NA         NA        NA         NA\n","rs117766797        NA        NA        NA         NA        NA         NA\n","              pvalue_19   pvalue_20   pvalue_21 pvalue_22   pvalue_23 pvalue_24\n","rs147560086          NA 8.26297e-28 3.16236e-42        NA 9.45416e-23        NA\n","rs7955986            NA          NA          NA        NA          NA        NA\n","rs718389             NA          NA          NA        NA          NA        NA\n","rs11571383           NA          NA          NA        NA          NA        NA\n","rs2107614            NA          NA          NA        NA          NA        NA\n","rs4980935   8.06751e-05          NA          NA        NA          NA        NA\n","rs11064536           NA          NA          NA        NA          NA        NA\n","rs117766797          NA          NA          NA        NA          NA        NA\n","              pvalue_25 pvalue_26 pvalue_27 pvalue_28   pvalue_29   pvalue_30\n","rs147560086 3.44003e-22        NA        NA        NA 3.44995e-23          NA\n","rs7955986            NA        NA        NA        NA          NA          NA\n","rs718389             NA        NA        NA        NA          NA          NA\n","rs11571383           NA        NA        NA        NA          NA          NA\n","rs2107614            NA        NA        NA        NA          NA          NA\n","rs4980935            NA        NA        NA        NA          NA          NA\n","rs11064536           NA        NA        NA        NA          NA 0.000347708\n","rs117766797          NA        NA        NA        NA          NA          NA\n","            pvalue_31 pvalue_32 pvalue_33  pvalue_34   pvalue_35 pvalue_36\n","rs147560086        NA        NA        NA 2.6301e-33 4.95626e-40        NA\n","rs7955986          NA        NA        NA         NA          NA        NA\n","rs718389           NA        NA        NA         NA          NA        NA\n","rs11571383         NA        NA        NA         NA          NA        NA\n","rs2107614          NA        NA        NA         NA          NA        NA\n","rs4980935          NA        NA        NA         NA          NA        NA\n","rs11064536         NA        NA        NA         NA          NA        NA\n","rs117766797        NA        NA        NA         NA          NA        NA\n","            pvalue_37 pvalue_38   pvalue_39 pvalue_40 pvalue_41   pvalue_42\n","rs147560086        NA        NA 8.63893e-13        NA        NA          NA\n","rs7955986          NA        NA          NA        NA        NA          NA\n","rs718389           NA        NA          NA        NA        NA          NA\n","rs11571383         NA        NA          NA        NA        NA 2.07196e-10\n","rs2107614          NA        NA          NA        NA        NA          NA\n","rs4980935          NA        NA          NA        NA        NA          NA\n","rs11064536         NA        NA          NA        NA        NA          NA\n","rs117766797        NA        NA          NA        NA        NA          NA\n","            pvalue_43   pvalue_44   pvalue_45   pvalue_46 pvalue_47   pvalue_48\n","rs147560086        NA 8.32165e-14          NA          NA        NA          NA\n","rs7955986          NA          NA          NA 2.21429e-32        NA          NA\n","rs718389           NA          NA 1.07802e-13          NA        NA          NA\n","rs11571383         NA          NA          NA          NA        NA          NA\n","rs2107614          NA          NA          NA          NA        NA          NA\n","rs4980935          NA          NA          NA          NA        NA          NA\n","rs11064536         NA          NA          NA          NA        NA          NA\n","rs117766797        NA          NA          NA          NA        NA 0.000357146\n","            pvalue_49\n","rs147560086        NA\n","rs7955986          NA\n","rs718389           NA\n","rs11571383         NA\n","rs2107614          NA\n","rs4980935          NA\n","rs11064536         NA\n","rs117766797        NA\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n"]}]},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(base) # R package installed by default\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(Matrix) # R package installed by default\n","library(methods) # R package installed by default\n","library(stats) # R package installed by default\n","library(lme4) # It needs to be installed separately\n","library(data.table) # It needs to be installed separately\n","\n","# Step 2: Main Mt_Robin function:\n","MR_MtRobin <- function(MR_MtRobin_input,pval_thresh=0.001){\n","  snpID <- MR_MtRobin_input$snpID\n","  gwas_betas <- MR_MtRobin_input$gwas_betas\n","  gwas_se <- MR_MtRobin_input$gwas_se\n","  eqtl_betas <- MR_MtRobin_input$eqtl_betas\n","  eqtl_se <- MR_MtRobin_input$eqtl_se\n","  eqtl_pvals <- MR_MtRobin_input$eqtl_pvals\n","  LD <- MR_MtRobin_input$LD\n","\n","  if(ncol(eqtl_betas) != ncol(eqtl_se) | ncol(eqtl_betas) != ncol(eqtl_pvals)){\n","    stop(\"eQTL betas, SE and p-values do not all have same number of columns\")\n","  }\n","\n","  ## Determine number of tissues (or conditions/studies/etc.)\n","  nT <- ncol(eqtl_betas)\n","\n","  ## rename columns for standardized processing\n","  colnames(eqtl_betas) <- paste0(\"beta_\",1:nT)\n","  colnames(eqtl_se) <- paste0(\"SE_\",1:nT)\n","  colnames(eqtl_pvals) <- paste0(\"pvalue_\",1:nT)\n","\n","  ## combine eQTL results into data.table\n","  eqtl_res <- data.table::data.table(snpID,eqtl_betas,eqtl_se,eqtl_pvals)\n","\n","  # reshape to long format (each SNP-gene-tissue will be row)\n","  eQTL_res_melt <- data.table::melt(eqtl_res,id.vars=c(\"snpID\"),\n","                                    measure.vars=patterns(\"^beta_\",\"^SE_\",\"^pvalue_\"),value.name=c(\"beta\",\"se\",\"pvalue\"),\n","                                    variable.name=\"tissue\")\n","\n","  ## subset to strong instrumental variables (based on p-value threshold)\n","  eQTL_res_melt_PltThresh <- subset(eQTL_res_melt, pvalue < pval_thresh)\n","\n","  ## set up gwas data for merging\n","  gwas_res <- data.table::data.table(snpID=snpID,gwas_beta=gwas_betas,gwas_se=gwas_se)\n","\n","  ## merge eQTL and GWAS results\n","  data.table::setkey(eQTL_res_melt_PltThresh,snpID)\n","  data.table::setkey(gwas_res,snpID)\n","  merged_res <- data.table::merge.data.table(eQTL_res_melt_PltThresh,gwas_res)\n","\n","  ## set up coefficients for reverse regression\n","  beta_x <- matrix(merged_res$beta,ncol=1)\n","  beta_y <- matrix(merged_res$gwas_beta,ncol=1)\n","\n","  ## standard errors (for weights)\n","  se_x <- matrix(merged_res$se,ncol=1)\n","  se_y <- matrix(merged_res$gwas_se,ncol=1) ## not used by function; used in resampling (consider returning with results)\n","\n","  ## identifiers\n","  snpID <- merged_res$snpID\n","\n","  lme_res <- lme4::lmer(beta_x~(beta_y-1)+(beta_y-1|snpID),weights=1/se_x^2)\n","\n","  ## return results from weighted regression with random slopes (and correlated errors) along with GWAS SE and LD (needed for resampling)\n","  return(list(lme_res=lme_res,gwas_res=gwas_res,LD=LD))\n","}\n","\n","# Step 3: Read the previous file:\n","result_setup <- readRDS(\"/data/long_covid_setup_fs_2.rds\")\n","\n","# Step 4: Apply the main Mt_Robin function using parallel processing:\n","no_cores <- 38\n","\n","cl <- makeCluster(no_cores) # Create a cluster with the number of cores\n","clusterExport(cl, \"MR_MtRobin\") # Export the MR_MtRobin function to all the nodes of the cluster\n","result_list_MtRobin <- parLapply(cl, result_setup, function(MR_MtRobin_input) {\n","  tryCatch({\n","    MR_MtRobin(MR_MtRobin_input = MR_MtRobin_input)\n","  }, error = function(e) print(e))\n","})\n","stopCluster(cl) # Stop the cluster once the computation is done\n","\n","# Step 5: The names of the results will correspond to the gene ids:\n","names(result_list_MtRobin) <- names(result_setup)\n","\n","# Step 6: Save the previous result as a .rds file:\n","saveRDS(result_list_MtRobin, file = \"/data/long_covid_main_fs_2.rds\")"],"metadata":{"id":"asasPGhk9P0t"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/main_function.sif Rscript /data/main_function_fs_2.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"QpvcsR3M9bj_"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 17:34:46:\n","   Job Id:             93443496.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      2.70\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 00:02:51\n","   Memory Requested:   300.0GB               Memory Used: 9.55GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:01:30\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"Zzbf_baM9fpj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Open the output file:\n","result_main_fs_2 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_2.rds\")\n","\n","# Filter out errors:\n","result_main_fs_2 <- result_main_fs_2[!sapply(result_main_fs_2, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_main_fs_2))\n","print(length(unique(result_main_fs_2)))\n","print(head(result_main_fs_2))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"17rLLAUl9r73","executionInfo":{"status":"ok","timestamp":1699507422068,"user_tz":-630,"elapsed":10505,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"093e3b65-615e-4f65-e7ed-66af5089d475"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1624\n","[1] 1624\n","$ENSG00000002016\n","$ENSG00000002016$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -27.2451\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 24.597  \n"," Residual         2.247  \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," 1.193  \n","\n","$ENSG00000002016$gwas_res\n","        snpID    gwas_beta   gwas_se\n","1  rs11064536  0.005801520 0.0281981\n","2  rs11571383 -0.045787500 0.0569715\n","3 rs117766797 -0.016338200 0.0628533\n","4 rs147560086  0.021120800 0.0258757\n","5   rs2107614 -0.000839245 0.0220439\n","6   rs4980935 -0.015072100 0.0229060\n","7    rs718389 -0.017204900 0.0234662\n","8   rs7955986 -0.006508970 0.0206358\n","\n","$ENSG00000002016$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","\n","$ENSG00000005156\n","$ENSG00000005156$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -13.5936\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 2.202   \n"," Residual        3.171   \n","Number of obs: 10, groups:  snpID, 3\n","Fixed Effects:\n","beta_y  \n","  9.75  \n","\n","$ENSG00000005156$gwas_res\n","        snpID  gwas_beta   gwas_se\n","1  rs12945428 -0.0334790 0.0199865\n","2 rs201016783  0.1101960 0.0462604\n","3   rs3135958  0.0523497 0.0323807\n","\n","$ENSG00000005156$LD\n","             rs12945428    rs3135958  rs201016783\n","rs12945428   1.00000000 -0.295076381  0.020183627\n","rs3135958   -0.29507638  1.000000000 -0.006054471\n","rs201016783  0.02018363 -0.006054471  1.000000000\n","\n","\n","$ENSG00000006282\n","$ENSG00000006282$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -26.1961\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 129.340 \n"," Residual          5.685 \n","Number of obs: 46, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n"," 107.3  \n","\n","$ENSG00000006282$gwas_res\n","      snpID   gwas_beta   gwas_se\n","1 rs2306001 -0.01463990 0.0336874\n","2 rs9890200 -0.00404758 0.0213245\n","\n","$ENSG00000006282$LD\n","          rs9890200 rs2306001\n","rs9890200 1.0000000 0.3686704\n","rs2306001 0.3686704 1.0000000\n","\n","\n","$ENSG00000006744\n","$ENSG00000006744$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -3.1412\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 12.172  \n"," Residual         6.323  \n","Number of obs: 18, groups:  snpID, 4\n","Fixed Effects:\n","beta_y  \n","-1.127  \n","\n","$ENSG00000006744$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1  rs1044564  0.0408637 0.0204543\n","2 rs11655929  0.0264804 0.0202880\n","3   rs733890 -0.0243341 0.0212502\n","4 rs78840287  0.0625188 0.0699093\n","\n","$ENSG00000006744$LD\n","              rs1044564    rs733890 rs11655929   rs78840287\n","rs1044564   1.000000000  0.66545097 -0.6626427 -0.007565868\n","rs733890    0.665450966  1.00000000 -0.4997189  0.027633741\n","rs11655929 -0.662642701 -0.49971889  1.0000000  0.034843798\n","rs78840287 -0.007565868  0.02763374  0.0348438  1.000000000\n","\n","\n","$ENSG00000007171\n","$ENSG00000007171$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 11.3893\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 50.686  \n"," Residual         4.262  \n","Number of obs: 31, groups:  snpID, 5\n","Fixed Effects:\n","beta_y  \n","-9.296  \n","\n","$ENSG00000007171$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1  rs2314813 -0.00817650 0.0263767\n","2  rs3751972  0.00878937 0.0230670\n","3   rs624196  0.03713040 0.0221403\n","4 rs72846720  0.05844630 0.0495926\n","5  rs8078340 -0.02882090 0.0304053\n","\n","$ENSG00000007171$LD\n","              rs3751972   rs8078340    rs624196    rs2314813  rs72846720\n","rs3751972   1.000000000 -0.20973763 -0.03796279  0.004244027 -0.08622187\n","rs8078340  -0.209737626  1.00000000  0.03757975  0.160333611 -0.06812959\n","rs624196   -0.037962786  0.03757975  1.00000000 -0.039583359  0.20176556\n","rs2314813   0.004244027  0.16033361 -0.03958336  1.000000000 -0.10409667\n","rs72846720 -0.086221874 -0.06812959  0.20176556 -0.104096673  1.00000000\n","\n","\n","$ENSG00000007341\n","$ENSG00000007341$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -1.4243\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 0.000   \n"," Residual        7.332   \n","Number of obs: 12, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n","-15.16  \n","optimizer (nloptwrap) convergence code: 0 (OK) ; 0 optimizer warnings; 1 lme4 warnings \n","\n","$ENSG00000007341$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1 rs10745330 -0.0345791 0.0197806\n","2   rs954679  0.0369388 0.0202454\n","\n","$ENSG00000007341$LD\n","           rs10745330  rs954679\n","rs10745330  1.0000000 0.6167231\n","rs954679    0.6167231 1.0000000\n","\n","\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_fs_2.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_2_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_2_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_fs2.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"DbzwSiA27UtS"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Option 1: Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_fs_2.R\n","\n","# Option 2: Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"hGIXVouO7YSE"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-07 15:15:59:\n","   Job Id:             107706975.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      630.16\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 130:46:06\n","   Memory Requested:   500.0GB               Memory Used: 116.65GB\n","   Walltime requested: 10:00:00            Walltime Used: 04:22:34\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"0FhOmRfZ6aYP"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs2.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results))\n","print(head(results))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"bmreCSSa6Yrn","executionInfo":{"status":"ok","timestamp":1707281863601,"user_tz":-630,"elapsed":843,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"3a479138-e16e-44f6-e52b-93b926446749"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1624    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.914      10000 0.9861120\n","2 ENSG00000005156  0.083      10000 0.5029552\n","3 ENSG00000006282  0.472      10000 0.8593363\n","4 ENSG00000006744  0.926      10000 0.9861120\n","5 ENSG00000007171  0.769      10000 0.9505140\n","6 ENSG00000007341  0.070      10000 0.4796624\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_2.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kCzH98xT6nsv","executionInfo":{"status":"ok","timestamp":1707282253019,"user_tz":-630,"elapsed":97100,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"32575cc6-ff2f-43a4-d25d-6c28d73d411c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 1624 10 \n","\n","          gene_id gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000002016     RAD52 protein_coding   1.193051  23    8\n","2 ENSG00000005156      LIG3 protein_coding   9.750327  10    3\n","3 ENSG00000006282   SPATA20 protein_coding 107.342074  46    2\n","4 ENSG00000006744     ELAC2 protein_coding  -1.126989  18    4\n","5 ENSG00000007171      NOS2 protein_coding  -9.295643  31    5\n","6 ENSG00000007341      ST7L protein_coding -15.155748  12    1\n","                                                                               SNP_IDs\n","1 rs11064536,rs11571383,rs117766797,rs147560086,rs2107614,rs4980935,rs718389,rs7955986\n","2                                                     rs12945428,rs201016783,rs3135958\n","3                                                                  rs2306001,rs9890200\n","4                                             rs1044564,rs11655929,rs733890,rs78840287\n","5                                    rs2314813,rs3751972,rs624196,rs72846720,rs8078340\n","6                                                                  rs10745330,rs954679\n","        REML         SD Residual_SD\n","1 -27.245061  24.597364    2.247270\n","2 -13.593554   2.201648    3.171135\n","3 -26.196089 129.340321    5.684947\n","4  -3.141167  12.171610    6.322981\n","5  11.389348  50.686085    4.262305\n","6  -1.424301   0.000000    7.331944\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_protein_coding_genes_sig_not_sig.txt \n","Dimensions of protein_coding_genes : 727 10 \n","\n","          gene_id gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000126602     TRAP1 protein_coding  2511.0867  39    2\n","2 ENSG00000176386     CDC26 protein_coding -1366.2795  18    2\n","3 ENSG00000213214  ARHGEF35 protein_coding  -729.5112  38    2\n","4 ENSG00000118197     DDX59 protein_coding  -425.6771   7    2\n","5 ENSG00000234719    NPIPB2 protein_coding   387.3580  48    2\n","6 ENSG00000221923    ZNF880 protein_coding  -376.8372  19    2\n","                SNP_IDs       REML        SD Residual_SD\n","1  rs4786429,rs72770333  15.544909 3499.5968   7.7084648\n","2 rs10981712,rs10981764 -38.133963 1584.4742   1.6183943\n","3  rs12703596,rs1731354 -23.831270  898.6506   2.6323630\n","4  rs34428088,rs7539252 -16.865505  616.4466   0.9685471\n","5    rs393329,rs9928563 -46.819443  753.5127   3.5970610\n","6   rs162127,rs34470614   3.819046  564.9874   6.7658290\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# FS2 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_2_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs2.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"vdz4cKYB7yGI","executionInfo":{"status":"ok","timestamp":1707282275470,"user_tz":-630,"elapsed":22456,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"92f439cd-b9e3-4d9c-ca21-7f614c197e17"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1624\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -27.2451\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 24.597  \n"," Residual         2.247  \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," 1.193  \n","\n","$gwas_res\n","        snpID    gwas_beta   gwas_se\n","1  rs11064536  0.005801520 0.0281981\n","2  rs11571383 -0.045787500 0.0569715\n","3 rs117766797 -0.016338200 0.0628533\n","4 rs147560086  0.021120800 0.0258757\n","5   rs2107614 -0.000839245 0.0220439\n","6   rs4980935 -0.015072100 0.0229060\n","7    rs718389 -0.017204900 0.0234662\n","8   rs7955986 -0.006508970 0.0206358\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 727  10\n","          gene_id gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000126602     TRAP1 protein_coding  2511.0867  39    2\n","2 ENSG00000176386     CDC26 protein_coding -1366.2795  18    2\n","3 ENSG00000213214  ARHGEF35 protein_coding  -729.5112  38    2\n","4 ENSG00000118197     DDX59 protein_coding  -425.6771   7    2\n","5 ENSG00000234719    NPIPB2 protein_coding   387.3580  48    2\n","6 ENSG00000221923    ZNF880 protein_coding  -376.8372  19    2\n","                SNP_IDs       REML        SD Residual_SD\n","1  rs4786429,rs72770333  15.544909 3499.5968   7.7084648\n","2 rs10981712,rs10981764 -38.133963 1584.4742   1.6183943\n","3  rs12703596,rs1731354 -23.831270  898.6506   2.6323630\n","4  rs34428088,rs7539252 -16.865505  616.4466   0.9685471\n","5    rs393329,rs9928563 -46.819443  753.5127   3.5970610\n","6   rs162127,rs34470614   3.819046  564.9874   6.7658290\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1624    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.914      10000 0.9861120\n","2 ENSG00000005156  0.083      10000 0.5029552\n","3 ENSG00000006282  0.472      10000 0.8593363\n","4 ENSG00000006744  0.926      10000 0.9861120\n","5 ENSG00000007171  0.769      10000 0.9505140\n","6 ENSG00000007341  0.070      10000 0.4796624\n","[1] \"Final table (protein coding genes): \"\n","[1] 19 22\n","           gene_id gene_name SNPs\n","1  ENSG00000223510    CDRT15    3\n","2  ENSG00000127399    LRRC61    2\n","3  ENSG00000197134    ZNF257    5\n","4  ENSG00000184361   SPATA32    8\n","5  ENSG00000149089      APIP    4\n","6  ENSG00000205078    SYCE1L    3\n","7  ENSG00000163161     ERCC3    5\n","8  ENSG00000212127   TAS2R14    5\n","9  ENSG00000141076      UTP4    3\n","10 ENSG00000090432      MUL1    5\n","11 ENSG00000147576    ADHFE1    8\n","12 ENSG00000090661     CERS4   15\n","13 ENSG00000142156    COL6A1   22\n","14 ENSG00000177082     WDR73   13\n","15 ENSG00000139714     MORN3   18\n","16 ENSG00000108439      PNPO   15\n","17 ENSG00000068097    HEATR6    9\n","18 ENSG00000089063   TMEM230    6\n","19 ENSG00000107186      MPDZ   18\n","                                                                                                                                                                                                                                      SNPs_IDs\n","1                                                                                                                                                                                                               rs55939784,rs56354007,rs736261\n","2                                                                                                                                                                                                                         rs10247016,rs1962004\n","3                                                                                                                                                                                           rs17554725,rs1969193,rs2262908,rs401424,rs67458228\n","4                                                                                                                                                       rs113819830,rs12450413,rs140252688,rs45469892,rs4792879,rs60420573,rs7208548,rs9908267\n","5                                                                                                                                                                                                   rs11032943,rs11500295,rs2915228,rs74344381\n","6                                                                                                                                                                                                                  rs13331769,rs6564410,rs8398\n","7                                                                                                                                                                                         rs11688763,rs4150476,rs56369397,rs7568261,rs76693372\n","8                                                                                                                                                                                         rs10772397,rs3851583,rs4418898,rs73046589,rs79686186\n","9                                                                                                                                                                                                               rs7200932,rs74886619,rs9941143\n","10                                                                                                                                                                                           rs3831949,rs595980,rs6700034,rs7514270,rs77048043\n","11                                                                                                                                                        rs10481242,rs2357671,rs2357893,rs35450081,rs55878528,rs56256543,rs60640306,rs6472267\n","12                                                                           rs113738552,rs11880745,rs142392712,rs147441807,rs1965888,rs2336171,rs4542783,rs4804282,rs56120315,rs56184962,rs57062897,rs59099428,rs60585010,rs7260363,rs8100119\n","13 rs1053315,rs113777625,rs11700719,rs11701124,rs117025791,rs11702647,rs118063718,rs11908797,rs13048668,rs140953671,rs146698426,rs1980983,rs2277813,rs2481872,rs2839080,rs4819159,rs57078525,rs73159603,rs76536010,rs7671,rs76970926,rs9983378\n","14                                                                                                 rs11259919,rs12900358,rs138567963,rs141005445,rs17532346,rs1960770,rs2137270,rs55702300,rs55736017,rs56053835,rs6496090,rs77590624,rs893162\n","15                                       rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","16                                                                                   rs13353189,rs17250967,rs17617250,rs1986694,rs2001380,rs2002136,rs2325749,rs2737,rs35614613,rs35653190,rs4793671,rs62065840,rs6504745,rs76098486,rs9903904\n","17                                                                                                                                              rs1078378,rs10853017,rs117054060,rs345172,rs7210383,rs72831252,rs76439721,rs79483284,rs9912341\n","18                                                                                                                                                                                rs118000137,rs1741303,rs261350,rs261367,rs4315602,rs73610161\n","19                                      rs10961168,rs111384271,rs111681668,rs117235036,rs1325126,rs1331672,rs139926461,rs141414147,rs143951908,rs2094958,rs3198580,rs34101279,rs35086129,rs35494822,rs71491604,rs72713964,rs75569334,rs7874074\n","   Obs     beta_y        Min     First_Q        Median    Third_Q      Max\n","1   12 -78.995771 -0.4702224 -0.33334823  0.0854493183 0.53056842 2.979938\n","2   23  38.987530 -1.4292932 -0.62405647  0.2907361376 0.70513413 2.097100\n","3   18 -28.142186 -1.6342228 -0.37752860  0.0143747542 0.63510868 2.096204\n","4   10 -25.971831 -0.7205483  0.05256287  0.3541130954 1.18185819 1.907545\n","5   25  23.413555 -2.2094793 -0.50974836  0.1153998777 0.56260308 1.632209\n","6   26  22.204254 -1.7098012 -0.58565930  0.0399800593 0.93934754 1.907866\n","7    8  19.930424 -1.2537539 -0.20104618  0.0009795153 0.99958200 1.762694\n","8   34  19.718143 -2.0751569 -0.44368841  0.1245177080 0.44287329 2.427000\n","9   28 -15.408071 -2.0358161 -0.95094089 -0.4384156460 0.09298467 2.383013\n","10  39 -13.982064 -1.8739525 -0.47638993 -0.2163270417 0.42176447 2.819006\n","11   9 -13.090833 -0.4910621 -0.08443309  0.5996130639 0.72708666 1.350845\n","12  23 -11.881211 -1.9032917 -0.23285150  0.2939892132 0.84937947 2.749017\n","13  28 -11.591537 -2.0437183 -0.60288199  0.1585432130 0.61248945 2.181431\n","14  14 -11.427477 -2.0193880 -0.63287523 -0.0641050103 0.57294771 1.991510\n","15  21   8.048604 -1.7952165 -0.70986299  0.0691898583 0.71505286 2.057158\n","16  18   7.286517 -1.4283710 -0.68523133 -0.0676194005 0.52122443 2.097884\n","17  25   6.630314 -1.6049564 -0.40310399  0.1344778152 0.89791548 2.286848\n","18  34   4.993561 -1.7425428 -0.46226234  0.0896087541 0.48787852 3.835098\n","19  35   4.279047 -2.3092345 -0.93746099 -0.2359915867 0.39022496 1.914908\n","          SD_RE Res_SD_RE       Var_RE Res_Var_RE     SE_FE t_value_FE\n","1  0.000000e+00  4.012238 0.000000e+00  16.098054 5.5372337 -14.266288\n","2  1.371125e-07  4.666925 1.879984e-14  21.780191 2.2673426  17.195253\n","3  0.000000e+00  2.227377 0.000000e+00   4.961208 1.3078812 -21.517387\n","4  0.000000e+00  3.250016 0.000000e+00  10.562603 4.5942163  -5.653158\n","5  6.336802e+00  4.572878 4.015506e+01  20.911211 4.1687798   5.616405\n","6  0.000000e+00  4.255625 0.000000e+00  18.110343 1.0131588  21.915867\n","7  0.000000e+00  2.842270 0.000000e+00   8.078501 1.5677012  12.713152\n","8  0.000000e+00  3.552101 0.000000e+00  12.617425 0.9213138  21.402201\n","9  0.000000e+00  3.377591 0.000000e+00  11.408122 0.9769454 -15.771680\n","10 0.000000e+00  4.656710 0.000000e+00  21.684946 0.8717065 -16.039876\n","11 6.215734e+00  3.466729 3.863535e+01  12.018208 3.2522754  -4.025131\n","12 1.616380e+00  6.679292 2.612685e+00  44.612948 2.2450789  -5.292113\n","13 0.000000e+00  5.795755 0.000000e+00  33.590779 2.2212404  -5.218497\n","14 0.000000e+00  3.578420 0.000000e+00  12.805089 2.1783143  -5.246019\n","15 0.000000e+00  3.764951 0.000000e+00  14.174859 1.5882537   5.067581\n","16 1.238522e+00  3.045930 1.533936e+00   9.277687 1.0893213   6.689043\n","17 0.000000e+00  3.480201 0.000000e+00  12.111801 0.5133983  12.914562\n","18 0.000000e+00  3.551145 0.000000e+00  12.610633 0.2369775  21.071878\n","19 0.000000e+00  4.115457 0.000000e+00  16.936987 0.4740063   9.027404\n","          REML p_value   FDR   CI_lower   CI_upper\n","1  -10.2884289   0.000 0.000 -89.848749 -68.142793\n","2   -9.1158192   0.000 0.000  34.543539  43.431522\n","3  -24.3399413   0.000 0.000 -30.705634 -25.578739\n","4    0.5274341   0.001 0.037 -34.976495 -16.967167\n","5   -4.8693053   0.001 0.037  15.242746  31.584363\n","6   -9.6210169   0.001 0.037  20.218463  24.190045\n","7  -10.1128454   0.000 0.000  16.857730  23.003119\n","8  -25.3596335   0.001 0.037  17.912368  21.523918\n","9    1.5014582   0.000 0.000 -17.322884 -13.493258\n","10  -7.7281383   0.001 0.037 -15.690608 -12.273519\n","11  -5.4690195   0.001 0.037 -19.465293  -6.716374\n","12  17.9653353   0.001 0.037 -16.281566  -7.480857\n","13   9.2308731   0.000 0.000 -15.945168  -7.237905\n","14   4.0712110   0.000 0.000 -15.696973  -7.157982\n","15  15.3032063   0.000 0.000   4.935626  11.161581\n","16 -20.5219401   0.000 0.000   5.151447   9.421587\n","17   4.1759859   0.000 0.000   5.624053   7.636575\n","18 -48.2519897   0.000 0.000   4.529086   5.458037\n","19  17.7184892   0.000 0.000   3.349994   5.208099\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the last function to retrieve the tissue information and categories for fs1:\n","input_eqtl_path <- \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\"\n","input_mt_robin_path <- \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_protein_coding_genes_sig_full_table.txt\"\n","output_path <- \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_protein_coding_genes_sig_full_table_tissues.txt\"\n","process_eqtl_mt_robin(input_eqtl_path, input_mt_robin_path, output_path)"],"metadata":{"id":"E4J3OSQHXfno","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714700593014,"user_tz":-570,"elapsed":57134,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"50cd0f0c-f0c7-402d-c32f-2aa690ab10f7"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["           gene_id gene_name SNPs\n","1  ENSG00000068097    HEATR6    9\n","2  ENSG00000089063   TMEM230    6\n","3  ENSG00000090432      MUL1    5\n","4  ENSG00000090661     CERS4   15\n","5  ENSG00000107186      MPDZ   18\n","6  ENSG00000108439      PNPO   15\n","7  ENSG00000127399    LRRC61    2\n","8  ENSG00000139714     MORN3   18\n","9  ENSG00000141076      UTP4    3\n","10 ENSG00000142156    COL6A1   22\n","11 ENSG00000147576    ADHFE1    8\n","12 ENSG00000149089      APIP    4\n","13 ENSG00000163161     ERCC3    5\n","14 ENSG00000177082     WDR73   13\n","15 ENSG00000184361   SPATA32    8\n","16 ENSG00000197134    ZNF257    5\n","17 ENSG00000205078    SYCE1L    3\n","18 ENSG00000212127   TAS2R14    5\n","19 ENSG00000223510    CDRT15    3\n","                                                                                                                                                                                                                                      SNPs_IDs\n","1                                                                                                                                               rs1078378,rs10853017,rs117054060,rs345172,rs7210383,rs72831252,rs76439721,rs79483284,rs9912341\n","2                                                                                                                                                                                 rs118000137,rs1741303,rs261350,rs261367,rs4315602,rs73610161\n","3                                                                                                                                                                                            rs3831949,rs595980,rs6700034,rs7514270,rs77048043\n","4                                                                            rs113738552,rs11880745,rs142392712,rs147441807,rs1965888,rs2336171,rs4542783,rs4804282,rs56120315,rs56184962,rs57062897,rs59099428,rs60585010,rs7260363,rs8100119\n","5                                       rs10961168,rs111384271,rs111681668,rs117235036,rs1325126,rs1331672,rs139926461,rs141414147,rs143951908,rs2094958,rs3198580,rs34101279,rs35086129,rs35494822,rs71491604,rs72713964,rs75569334,rs7874074\n","6                                                                                    rs13353189,rs17250967,rs17617250,rs1986694,rs2001380,rs2002136,rs2325749,rs2737,rs35614613,rs35653190,rs4793671,rs62065840,rs6504745,rs76098486,rs9903904\n","7                                                                                                                                                                                                                         rs10247016,rs1962004\n","8                                        rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","9                                                                                                                                                                                                               rs7200932,rs74886619,rs9941143\n","10 rs1053315,rs113777625,rs11700719,rs11701124,rs117025791,rs11702647,rs118063718,rs11908797,rs13048668,rs140953671,rs146698426,rs1980983,rs2277813,rs2481872,rs2839080,rs4819159,rs57078525,rs73159603,rs76536010,rs7671,rs76970926,rs9983378\n","11                                                                                                                                                        rs10481242,rs2357671,rs2357893,rs35450081,rs55878528,rs56256543,rs60640306,rs6472267\n","12                                                                                                                                                                                                  rs11032943,rs11500295,rs2915228,rs74344381\n","13                                                                                                                                                                                        rs11688763,rs4150476,rs56369397,rs7568261,rs76693372\n","14                                                                                                 rs11259919,rs12900358,rs138567963,rs141005445,rs17532346,rs1960770,rs2137270,rs55702300,rs55736017,rs56053835,rs6496090,rs77590624,rs893162\n","15                                                                                                                                                      rs113819830,rs12450413,rs140252688,rs45469892,rs4792879,rs60420573,rs7208548,rs9908267\n","16                                                                                                                                                                                          rs17554725,rs1969193,rs2262908,rs401424,rs67458228\n","17                                                                                                                                                                                                                 rs13331769,rs6564410,rs8398\n","18                                                                                                                                                                                        rs10772397,rs3851583,rs4418898,rs73046589,rs79686186\n","19                                                                                                                                                                                                              rs55939784,rs56354007,rs736261\n","   Obs     beta_y        Min     First_Q        Median    Third_Q      Max\n","1   25   6.630314 -1.6049564 -0.40310399  0.1344778152 0.89791548 2.286848\n","2   34   4.993561 -1.7425428 -0.46226234  0.0896087541 0.48787852 3.835098\n","3   39 -13.982064 -1.8739525 -0.47638993 -0.2163270417 0.42176447 2.819006\n","4   23 -11.881211 -1.9032917 -0.23285150  0.2939892132 0.84937947 2.749017\n","5   35   4.279047 -2.3092345 -0.93746099 -0.2359915867 0.39022496 1.914908\n","6   18   7.286517 -1.4283710 -0.68523133 -0.0676194005 0.52122443 2.097884\n","7   23  38.987530 -1.4292932 -0.62405647  0.2907361376 0.70513413 2.097100\n","8   21   8.048604 -1.7952165 -0.70986299  0.0691898583 0.71505286 2.057158\n","9   28 -15.408071 -2.0358161 -0.95094089 -0.4384156460 0.09298467 2.383013\n","10  28 -11.591537 -2.0437183 -0.60288199  0.1585432130 0.61248945 2.181431\n","11   9 -13.090833 -0.4910621 -0.08443309  0.5996130639 0.72708666 1.350845\n","12  25  23.413555 -2.2094793 -0.50974836  0.1153998777 0.56260308 1.632209\n","13   8  19.930424 -1.2537539 -0.20104618  0.0009795153 0.99958200 1.762694\n","14  14 -11.427477 -2.0193880 -0.63287523 -0.0641050103 0.57294771 1.991510\n","15  10 -25.971831 -0.7205483  0.05256287  0.3541130954 1.18185819 1.907545\n","16  18 -28.142186 -1.6342228 -0.37752860  0.0143747542 0.63510868 2.096204\n","17  26  22.204254 -1.7098012 -0.58565930  0.0399800593 0.93934754 1.907866\n","18  34  19.718143 -2.0751569 -0.44368841  0.1245177080 0.44287329 2.427000\n","19  12 -78.995771 -0.4702224 -0.33334823  0.0854493183 0.53056842 2.979938\n","          SD_RE Res_SD_RE       Var_RE Res_Var_RE     SE_FE t_value_FE\n","1  0.000000e+00  3.480201 0.000000e+00  12.111801 0.5133983  12.914562\n","2  0.000000e+00  3.551145 0.000000e+00  12.610633 0.2369775  21.071878\n","3  0.000000e+00  4.656710 0.000000e+00  21.684946 0.8717065 -16.039876\n","4  1.616380e+00  6.679292 2.612685e+00  44.612948 2.2450789  -5.292113\n","5  0.000000e+00  4.115457 0.000000e+00  16.936987 0.4740063   9.027404\n","6  1.238522e+00  3.045930 1.533936e+00   9.277687 1.0893213   6.689043\n","7  1.371125e-07  4.666925 1.879984e-14  21.780191 2.2673426  17.195253\n","8  0.000000e+00  3.764951 0.000000e+00  14.174859 1.5882537   5.067581\n","9  0.000000e+00  3.377591 0.000000e+00  11.408122 0.9769454 -15.771680\n","10 0.000000e+00  5.795755 0.000000e+00  33.590779 2.2212404  -5.218497\n","11 6.215734e+00  3.466729 3.863535e+01  12.018208 3.2522754  -4.025131\n","12 6.336802e+00  4.572878 4.015506e+01  20.911211 4.1687798   5.616405\n","13 0.000000e+00  2.842270 0.000000e+00   8.078501 1.5677012  12.713152\n","14 0.000000e+00  3.578420 0.000000e+00  12.805089 2.1783143  -5.246019\n","15 0.000000e+00  3.250016 0.000000e+00  10.562603 4.5942163  -5.653158\n","16 0.000000e+00  2.227377 0.000000e+00   4.961208 1.3078812 -21.517387\n","17 0.000000e+00  4.255625 0.000000e+00  18.110343 1.0131588  21.915867\n","18 0.000000e+00  3.552101 0.000000e+00  12.617425 0.9213138  21.402201\n","19 0.000000e+00  4.012238 0.000000e+00  16.098054 5.5372337 -14.266288\n","          REML p_value   FDR   CI_lower   CI_upper\n","1    4.1759859   0.000 0.000   5.624053   7.636575\n","2  -48.2519897   0.000 0.000   4.529086   5.458037\n","3   -7.7281383   0.001 0.037 -15.690608 -12.273519\n","4   17.9653353   0.001 0.037 -16.281566  -7.480857\n","5   17.7184892   0.000 0.000   3.349994   5.208099\n","6  -20.5219401   0.000 0.000   5.151447   9.421587\n","7   -9.1158192   0.000 0.000  34.543539  43.431522\n","8   15.3032063   0.000 0.000   4.935626  11.161581\n","9    1.5014582   0.000 0.000 -17.322884 -13.493258\n","10   9.2308731   0.000 0.000 -15.945168  -7.237905\n","11  -5.4690195   0.001 0.037 -19.465293  -6.716374\n","12  -4.8693053   0.001 0.037  15.242746  31.584363\n","13 -10.1128454   0.000 0.000  16.857730  23.003119\n","14   4.0712110   0.000 0.000 -15.696973  -7.157982\n","15   0.5274341   0.001 0.037 -34.976495 -16.967167\n","16 -24.3399413   0.000 0.000 -30.705634 -25.578739\n","17  -9.6210169   0.001 0.037  20.218463  24.190045\n","18 -25.3596335   0.001 0.037  17.912368  21.523918\n","19 -10.2884289   0.000 0.000 -89.848749 -68.142793\n","                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             Tissues\n","1                                                                                                                                                                                                                    Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Coronary artery,Anterior cingulate cortex,Cortex,Frontal cortex,Spinal cord,Breast mammary tissue,EBV-transformed lymphocytes,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Liver,Minor salivary gland,Skeletal muscle,Ovary,Pancreas,Prostate,Not sun-exposed skin (suprapubic),Small intestine terminal ileum,Stomach,Testis,Uterus,Whole blood\n","2                                                                              Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Aorta artery,Coronary artery,Anterior cingulate cortex,Cortex,Hypothalamus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Cultured fibroblasts,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Kidney cortex,Lung,Minor salivary gland,Tibial nerve,Ovary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Spleen,Stomach,Testis,Thyroid,Uterus,Vagina,Whole blood\n","3  Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Aorta artery,Coronary artery,Tibial artery,Cerebellar hemisphere,Cerebellum,Cortex,Frontal cortex,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Cultured fibroblasts,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Left ventricle,Kidney cortex,Liver,Lung,Minor salivary gland,Skeletal muscle,Tibial nerve,Ovary,Prostate,Not sun-exposed skin (suprapubic),Small intestine terminal ileum,Spleen,Stomach,Testis,Thyroid,Uterus,Vagina,Whole blood\n","4                                                                                                                                                                                                                                                            Visceral omentum adipose,Tibial artery,Amygdala brain,Anterior cingulate cortex,Caudate,Cerebellar hemisphere,Cortex,Frontal cortex,Hippocampus,Spinal cord,Substantia nigra,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Minor salivary gland,Skeletal muscle,Ovary,Pituitary,Not sun-exposed skin (suprapubic),Testis,Uterus\n","5                                                                             Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Aorta artery,Coronary artery,Tibial artery,Amygdala brain,Anterior cingulate cortex,Caudate,Cerebellar hemisphere,Cerebellum,Frontal cortex,Hypothalamus,Nucleus accumbens (basal ganglia),Spinal cord,Cultured fibroblasts,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Left ventricle,Kidney cortex,Liver,Lung,Skeletal muscle,Tibial nerve,Ovary,Pancreas,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Thyroid,Uterus,Vagina\n","6                                                                                                                                                                                                                                                                                                        Aorta artery,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Left ventricle,Kidney cortex,Liver,Skeletal muscle,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Thyroid,Uterus\n","7                                                                                                                                                                                                                                    Visceral omentum adipose,Adrenal gland,Aorta artery,Tibial artery,Anterior cingulate cortex,Cerebellar hemisphere,Cerebellum,Cortex,Hippocampus,Hypothalamus,Nucleus accumbens (basal ganglia),Substantia nigra,Cultured fibroblasts,Sigmoid colon,Gastroesophageal junction,Atrial appendage,Lung,Skeletal muscle,Tibial nerve,Not sun-exposed skin (suprapubic),Small intestine terminal ileum,Thyroid,Uterus\n","8                                                                                                                                                                                                                                                                                Visceral omentum adipose,Adrenal gland,Tibial artery,Cerebellum,Nucleus accumbens (basal ganglia),Spinal cord,Breast mammary tissue,EBV-transformed lymphocytes,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Kidney cortex,Liver,Minor salivary gland,Pancreas,Prostate,Sun-exposed skin (lower leg),Thyroid,Vagina,Whole blood\n","9                                                                                                                                                          Subcutaneous adipose,Tibial artery,Amygdala brain,Anterior cingulate cortex,Caudate,Cerebellar hemisphere,Cortex,Hippocampus,Nucleus accumbens (basal ganglia),Substantia nigra,Cultured fibroblasts,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus muscularis,Atrial appendage,Kidney cortex,Lung,Skeletal muscle,Tibial nerve,Ovary,Pituitary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Thyroid\n","10                                                                                                                                                                             Subcutaneous adipose,Aorta artery,Tibial artery,Anterior cingulate cortex,Cerebellar hemisphere,Cerebellum,Frontal cortex,Hypothalamus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Spinal cord,Substantia nigra,Cultured fibroblasts,Transverse colon,Kidney cortex,Lung,Tibial nerve,Ovary,Pancreas,Pituitary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Thyroid,Vagina,Whole blood\n","11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    Nucleus accumbens (basal ganglia),Substantia nigra,EBV-transformed lymphocytes,Transverse colon,Kidney cortex,Minor salivary gland,Tibial nerve,Thyroid,Vagina\n","12                                                                                                                                                                                                                                  Subcutaneous adipose,Aorta artery,Coronary artery,Tibial artery,Amygdala brain,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Left ventricle,Tibial nerve,Ovary,Prostate,Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Testis,Thyroid,Uterus,Vagina\n","13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Amygdala brain,Cerebellum,Atrial appendage,Tibial nerve,Small intestine terminal ileum,Testis,Thyroid,Vagina\n","14                                                                                                                                                                                                                                                                                                                                                                                             Coronary artery,Cerebellar hemisphere,Cerebellum,Hypothalamus,Putamen (basal ganglia),EBV-transformed lymphocytes,Esophagus mucosa,Kidney cortex,Liver,Pituitary,Not sun-exposed skin (suprapubic),Small intestine terminal ileum,Thyroid,Whole blood\n","15                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Adrenal gland,Cerebellum,EBV-transformed lymphocytes,Kidney cortex,Pancreas,Pituitary,Stomach,Testis,Vagina,Whole blood\n","16                                                                                                                                                                                                                                                                                                                                                        Subcutaneous adipose,Aorta artery,Tibial artery,Anterior cingulate cortex,Hippocampus,Putamen (basal ganglia),Breast mammary tissue,Cultured fibroblasts,Esophagus mucosa,Esophagus muscularis,Lung,Tibial nerve,Pancreas,Small intestine terminal ileum,Testis,Thyroid,Uterus,Whole blood\n","17                                                                                                                                                                                                         Subcutaneous adipose,Aorta artery,Coronary artery,Tibial artery,Caudate,Cerebellum,Cortex,Hippocampus,Hypothalamus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Transverse colon,Esophagus mucosa,Esophagus muscularis,Left ventricle,Lung,Skeletal muscle,Tibial nerve,Not sun-exposed skin (suprapubic),Testis,Thyroid,Whole blood\n","18                                                                                        Adrenal gland,Aorta artery,Tibial artery,Anterior cingulate cortex,Cerebellar hemisphere,Cerebellum,Cortex,Hippocampus,Nucleus accumbens (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Liver,Lung,Skeletal muscle,Tibial nerve,Ovary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Stomach,Testis,Thyroid,Uterus,Vagina,Whole blood\n","19                                                                                                                                                                                                                                                                                                                                                                                                                                                Cerebellum,Cultured fibroblasts,Esophagus mucosa,Atrial appendage,Lung,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Stomach,Testis,Thyroid\n","                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Tissue_Categories\n","1                                                                                                                                                                                                                                                                                                                                   Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Exocrine System,Muscular System,Reproductive System,Digestive System,Reproductive System,Skin,Digestive System,Digestive System,Reproductive System,Reproductive System,Blood\n","2                                                                                                                                     Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Respiratory System,Exocrine System,Brain and Nervous System,Reproductive System,Reproductive System,Skin,Skin,Immune System,Digestive System,Reproductive System,Endocrine System,Reproductive System,Reproductive System,Blood\n","3  Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Cardiovascular System,Excretory System,Digestive System,Respiratory System,Exocrine System,Muscular System,Brain and Nervous System,Reproductive System,Reproductive System,Skin,Digestive System,Immune System,Digestive System,Reproductive System,Endocrine System,Reproductive System,Reproductive System,Blood\n","4                                                                                                                                                                                                                                                                                                                          Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Exocrine System,Muscular System,Reproductive System,Endocrine System,Skin,Reproductive System,Reproductive System\n","5                                                                                 Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Digestive System,Cardiovascular System,Cardiovascular System,Excretory System,Digestive System,Respiratory System,Muscular System,Brain and Nervous System,Reproductive System,Digestive System,Reproductive System,Skin,Skin,Digestive System,Immune System,Endocrine System,Reproductive System,Reproductive System\n","6                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Digestive System,Muscular System,Reproductive System,Skin,Skin,Endocrine System,Reproductive System\n","7                                                                                                                                                                                                                                                                                                       Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Digestive System,Cardiovascular System,Respiratory System,Muscular System,Brain and Nervous System,Skin,Digestive System,Endocrine System,Reproductive System\n","8                                                                                                                                                                                                                                                                                                                                                                                                             Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Digestive System,Exocrine System,Digestive System,Reproductive System,Skin,Endocrine System,Reproductive System,Blood\n","9                                                                                                                                                                                                                                           Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Respiratory System,Muscular System,Brain and Nervous System,Reproductive System,Endocrine System,Reproductive System,Skin,Skin,Digestive System,Immune System,Endocrine System\n","10                                                                                                                                                                                                                                         Adipose Tissue,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Excretory System,Respiratory System,Brain and Nervous System,Reproductive System,Digestive System,Endocrine System,Reproductive System,Skin,Skin,Digestive System,Immune System,Endocrine System,Reproductive System,Blood\n","11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Excretory System,Exocrine System,Brain and Nervous System,Endocrine System,Reproductive System\n","12                                                                                                                                                                                                                                                                                                                      Adipose Tissue,Cardiovascular System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Reproductive System,Reproductive System,Skin,Digestive System,Immune System,Reproductive System,Endocrine System,Reproductive System,Reproductive System\n","13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        Brain and Nervous System,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Digestive System,Reproductive System,Endocrine System,Reproductive System\n","14                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Excretory System,Digestive System,Endocrine System,Skin,Digestive System,Endocrine System,Blood\n","15                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Brain and Nervous System,Brain and Nervous System,Immune System,Excretory System,Digestive System,Endocrine System,Digestive System,Reproductive System,Reproductive System,Blood\n","16                                                                                                                                                                                                                                                                                                                                                                                                                                            Adipose Tissue,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Digestive System,Digestive System,Respiratory System,Brain and Nervous System,Digestive System,Digestive System,Reproductive System,Endocrine System,Reproductive System,Blood\n","17                                                                                                                                                                                                                                                                     Adipose Tissue,Cardiovascular System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Respiratory System,Muscular System,Brain and Nervous System,Skin,Reproductive System,Endocrine System,Blood\n","18                                                                                                                                Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Respiratory System,Muscular System,Brain and Nervous System,Reproductive System,Reproductive System,Skin,Skin,Digestive System,Immune System,Digestive System,Reproductive System,Endocrine System,Reproductive System,Reproductive System,Blood\n","19                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Brain and Nervous System,Connective Tissue,Digestive System,Cardiovascular System,Respiratory System,Skin,Skin,Digestive System,Immune System,Digestive System,Reproductive System,Endocrine System\n"]}]},{"cell_type":"markdown","source":["### **GWAS3**"],"metadata":{"id":"RlFoSnH1Apqj"}},{"cell_type":"code","source":["library(matrixStats) # It needs to be installed separately\n","#library(BiocManager) # It needs to be installed separately\n","library(snpStats) # It needs to be installed separately\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_3_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 4: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 5: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 6: Create a variable for the path to the directory to save the LD matrix results:\n","ld_matrix_dir <- \"/data/ld_matrix_fs_3\"\n","\n","# Step 7: Create the directory if it doesn't exist:\n","if (!dir.exists(ld_matrix_dir)) {\n","  dir.create(ld_matrix_dir)\n","}\n","\n","# Step 8: Modify the original select_IV_fs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_fs() function --> Starts with no variables and adds one at a time:\n","select_IV_fs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset eQTL dataset to gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  # pval_mat <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  #sig_neg_eQTL <- as.numeric(apply(betas_sig,1,function(x) any(x < -1e-15)))\n","  #sig_pos_eQTL <- as.numeric(apply(betas_sig,1,function(x) any(x > 1e-15)))\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Select the best SNPs remaining in SNP_pool based on minimum p-value:\n","  selected_snps <- NULL\n","  while(length(SNP_pool) > 1){\n","\n","    ## Identify top SNPs in pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    next_snp_idx <- which.min(eqtl_data$min_p)\n","    next_snp <- eqtl_data$variant_id[next_snp_idx]\n","    selected_snps <-c(selected_snps,next_snp)\n","\n","    ## Identify SNPs not in high LD with top SNPs in the pool:\n","    LD_idx <- which(row.names(LD_r2)==next_snp)\n","    LD_keep_idx <- which(LD_r2[LD_idx,] < ld_thresh)\n","    LD_r2 <- LD_r2[LD_keep_idx,LD_keep_idx]\n","    SNP_pool <- names(LD_keep_idx)\n","  }\n","  selected_snps <- c(selected_snps, SNP_pool)\n","  return(selected_snps)\n","}\n","\n","# Step 9: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 10: Apply the function to each gene id:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","  # Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_fs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","  }, error = function(e) print(e))\n","}, mc.cores = num_cores)\n","\n","# Step 11: Name the elements of the list with corresponding gene ids:\n","names(result_list) <- all_genes_vector\n","\n","# Step 12: Save the previous result as a .rds file:\n","saveRDS(result_list, file = \"/data/long_covid_IVs_fs_3.rds\")"],"metadata":{"id":"1jKcL4zuAt03"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/select_function_fs_3.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"5TrdiGaFKl6U"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 04:50:19:\n","   Job Id:             93353706.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      798.52\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 257:47:58\n","   Memory Requested:   500.0GB               Memory Used: 393.04GB\n","   Walltime requested: 30:00:00            Walltime Used: 05:32:43\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"iSsKivXvKpcQ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_selectIV_fs_3 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_IVs_fs_3.rds\")\n","\n","# Filter out errors:\n","result_selectIV_fs_3 <- result_selectIV_fs_3[!sapply(result_selectIV_fs_3, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_selectIV_fs_3))\n","print(length(unique(result_selectIV_fs_3)))\n","print(head(result_selectIV_fs_3))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jUTKapSiK8fx","executionInfo":{"status":"ok","timestamp":1693540539406,"user_tz":-570,"elapsed":2951,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"4ad0a440-e92d-4686-ba6b-bd366b2d2df0"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 38574\n","[1] 38480\n","$ENSG00000000003\n"," [1] \"rs8780\"      \"rs1204407\"   \"rs189980168\" \"rs4145090\"   \"rs139645081\"\n"," [6] \"rs112002620\" \"rs113922837\" \"rs17329105\"  \"rs193124579\" \"rs146639579\"\n","[11] \"rs201585484\" \"rs111478058\" \"rs143491698\" \"rs572714672\" \"rs56100956\" \n","[16] \"rs147601806\" \"rs139830751\" \"rs12842251\"  \"rs73630694\" \n","\n","$ENSG00000000005\n"," [1] \"rs6616166\"   \"rs4828038\"   \"rs12006573\"  \"rs73557922\"  \"rs141394856\"\n"," [6] \"rs140468798\" \"rs73555045\"  \"rs5921738\"   \"rs139231094\" \"rs59368006\" \n","[11] \"rs111915382\" \"rs73562536\"  \"rs12832642\"  \"rs201254498\" \"rs186611188\"\n","[16] \"rs113422802\" \"rs796285800\" \"rs180795909\" \"rs1004006\"   \"rs148617015\"\n","[21] \"rs140109110\"\n","\n","$ENSG00000000419\n"," [1] \"rs1292040204\" \"rs79562553\"   \"rs4809840\"    \"rs35201621\"   \"rs35530561\"  \n"," [6] \"rs141892692\"  \"rs944933\"     \"rs11483800\"   \"rs113036491\"  \"rs116892696\" \n","[11] \"rs34483216\"   \"rs60578213\"   \"rs118181672\"  \"rs4811253\"    \"rs11086363\"  \n","[16] \"rs73278571\"   \"rs73131221\"   \"rs78319058\"   \"rs79223715\"   \"rs189417964\" \n","[21] \"rs74422706\"   \"rs112070162\"  \"rs59034704\"   \"rs180710262\"  \"rs577456025\" \n","[26] \"rs117900008\"  \"rs34223230\"   \"rs141885330\"  \"rs78937289\"   \"rs16995896\"  \n","[31] \"rs72020869\"   \"rs142281981\"  \"rs13037515\"   \"rs1883804\"    \"rs17730415\"  \n","[36] \"rs11697209\"   \"rs112566287\"  \"rs149762110\"  \"rs4809839\"    \"rs6021441\"   \n","[41] \"rs75866341\"   \"rs13036755\"   \"rs74177405\"  \n","\n","$ENSG00000000457\n"," [1] \"rs10753796\"  \"rs4656710\"   \"rs115234612\" \"rs3766155\"   \"rs2056980\"  \n"," [6] \"rs4987403\"   \"rs12123695\"  \"rs185963038\" \"rs116684351\" \"rs1076637\"  \n","[11] \"rs7417281\"   \"rs7541615\"   \"rs973608\"    \"rs2982465\"   \"rs113123846\"\n","[16] \"rs4987338\"   \"rs114779274\" \"rs12038549\"  \"rs112361437\"\n","\n","$ENSG00000000460\n"," [1] \"rs35296671\"  \"rs1800016\"   \"rs4656198\"   \"rs3753311\"   \"rs4987391\"  \n"," [6] \"rs180738486\" \"rs4562681\"   \"rs531844\"    \"rs115157324\" \"rs12116424\" \n","[11] \"rs17346300\"  \"rs60843604\"  \"rs12742457\"  \"rs78159465\"  \"rs139679825\"\n","[16] \"rs35662131\" \n","\n","$ENSG00000000938\n"," [1] \"rs3737803\"   \"rs72191132\"  \"rs137984979\" \"rs11273428\"  \"rs75800899\" \n"," [6] \"rs191579330\" \"rs9426420\"   \"rs148432575\" \"rs10430146\"  \"rs139100109\"\n","[11] \"rs6701065\"   \"rs80298610\"  \"rs17361686\"  \"rs484740\"    \"rs114059345\"\n","[16] \"rs72651589\"  \"rs4357550\"   \"rs41284294\"  \"rs138392492\" \"rs148577969\"\n","[21] \"rs151288141\" \"rs71642227\"  \"rs12759594\"  \"rs2076457\"   \"rs141256722\"\n","[26] \"rs76443568\" \n","\n"]}]},{"cell_type":"code","source":["library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_3_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the list of genes-SNPs:\n","result_selectedIVs <- readRDS(\"/data/long_covid_IVs_fs_3.rds\")\n","\n","# Step 4: Create a variable for the path to the LD matrix directory:\n","ld_matrix_dir <- \"/data/ld_matrix_fs_3\"\n","\n","# Step 5: Setup the Mt_Robin method using the following setup function:\n","MR_MtRobin_setup <- function(geneID,\n","                             snpID,\n","                             eqtl_data,\n","                             gwas_data,\n","                             ld_matrix_path,\n","                             nTiss){\n","  tryCatch({\n","    ## Check if the gene is present and subset\n","    if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","    eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","    ## Confirm all SNPs are present in the dataset\n","    if(!all(snpID %in% eqtl_data$variant_id)) stop(\"Some SNPs missing in eQTL dataset\")\n","    if(!all(snpID %in% gwas_data$variant_id)) stop(\"Some SNPs missing in GWAS dataset\")\n","\n","    # Read the LD matrix from a file\n","    LD <- as.matrix(utils::read.table(ld_matrix_path,\n","                                      header = TRUE,\n","                                      check.names = FALSE))\n","\n","    if(!all(snpID %in% colnames(LD))) stop(\"Some SNPs missing in LD matrix\")\n","\n","    ## Align data\n","    eqtl_data <- eqtl_data[match(snpID,eqtl_data$variant_id),]\n","    gwas_data <- gwas_data[match(snpID,gwas_data$variant_id),]\n","    LD <- LD[match(snpID,rownames(LD)),match(snpID,colnames(LD))]\n","\n","    eqtl_betas <- as.matrix(eqtl_data[,paste0(\"beta_\",1:nTiss)])\n","    eqtl_se <- as.matrix(eqtl_data[,paste0(\"SE_\",1:nTiss)])\n","    eqtl_pvals <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","    gwas_betas <- gwas_data$beta\n","    gwas_se <- gwas_data$SE\n","\n","    row.names(eqtl_betas) <- row.names(eqtl_se) <- row.names(eqtl_pvals) <- eqtl_data$variant_id\n","\n","    return(list(snpID=snpID,gwas_betas=gwas_betas,gwas_se=gwas_se,\n","                eqtl_betas=eqtl_betas,eqtl_se=eqtl_se,eqtl_pvals=eqtl_pvals,LD=LD))\n","  }, error = function(e) print(e))\n","}\n","\n","# Step 6: Apply the previous setup function:\n","\n","# Step 6.1: Initialize an empty list to store the results:\n","MR_MtRobin_results <- list()\n","\n","# Step 6.2: Define the number of cores to use for the parallelism process:\n","num_cores <- 36\n","\n","# Step 6.3: Loop over each gene in result_selectedIVs:\n","MR_MtRobin_results <- mclapply(names(result_selectedIVs), function(geneID) {\n","\n","  # Step 6.3.1: Get the SNPs associated with each gene:\n","  snpID <- result_selectedIVs[[geneID]]\n","\n","  # Step 6.3.2: Generate a unique file path for the LD matrix of this gene:\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  # Step 6.3.3: Call the MR_MtRobin_setup() function with the parameters for each gene:\n","  return(tryCatch({\n","    MR_MtRobin_setup(geneID = geneID,\n","                     snpID = snpID,\n","                     eqtl_data = eqtl_covid_exposure,\n","                     gwas_data = gwas_covid_outcome,\n","                     ld_matrix_path = ld_matrix_path,\n","                     nTiss = 49)\n","  }, error = function(e) print(e)))\n","}, mc.cores = num_cores)\n","\n","# Step 7: Assign names to the list:\n","names(MR_MtRobin_results) <- names(result_selectedIVs)\n","\n","# Step 8: Save the previous result as a .rds file:\n","saveRDS(MR_MtRobin_results, file = \"/data/long_covid_setup_fs_3.rds\")"],"metadata":{"id":"KBSFQkmFLW5H"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/setup_function_fs_3.R\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"MeOi3c5WLbYp"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 16:29:35:\n","   Job Id:             93430395.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      73.56\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 19:57:44\n","   Memory Requested:   300.0GB               Memory Used: 81.99GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:40:52\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"dJz0i9acUSJR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_setup_fs_3 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_setup_fs_3.rds\")\n","\n","# Filter out errors:\n","result_setup_fs_3 <- result_setup_fs_3[!sapply(result_setup_fs_3, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_setup_fs_3))\n","print(length(unique(result_setup_fs_3)))\n","print(result_setup_fs_3[[1]])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"W9DBXclgLfSr","executionInfo":{"status":"ok","timestamp":1697000309989,"user_tz":-630,"elapsed":6753,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"2aa22087-5305-49a1-8378-1816c901dc1b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 5167\n","[1] 5167\n","$snpID\n","[1] \"rs147560086\" \"rs7955986\"   \"rs718389\"    \"rs11571383\"  \"rs2107614\"  \n","[6] \"rs4980935\"   \"rs11064536\"  \"rs117766797\"\n","\n","$gwas_betas\n","[1]  0.04747050 -0.03200370 -0.06199730 -0.03928230 -0.02811250 -0.00631512\n","[7]  0.07077150 -0.09468080\n","\n","$gwas_se\n","[1] 0.0395567 0.0324808 0.0403831 0.0700993 0.0339167 0.0413246 0.0437185\n","[8] 0.1211550\n","\n","$eqtl_betas\n","            beta_1 beta_2    beta_3   beta_4 beta_5    beta_6    beta_7 beta_8\n","rs147560086     NA     NA -0.332722 -0.36379     NA -0.428724        NA     NA\n","rs7955986       NA     NA        NA       NA     NA        NA        NA     NA\n","rs718389        NA     NA        NA       NA     NA        NA        NA     NA\n","rs11571383      NA     NA        NA       NA     NA        NA -0.525715     NA\n","rs2107614       NA     NA        NA       NA     NA        NA        NA     NA\n","rs4980935       NA     NA        NA       NA     NA        NA        NA     NA\n","rs11064536      NA     NA        NA       NA     NA        NA        NA     NA\n","rs117766797     NA     NA        NA       NA     NA        NA        NA     NA\n","               beta_9 beta_10 beta_11 beta_12 beta_13 beta_14 beta_15   beta_16\n","rs147560086 -0.362536      NA      NA -0.5128      NA      NA      NA        NA\n","rs7955986          NA      NA      NA      NA      NA      NA      NA        NA\n","rs718389           NA      NA      NA      NA      NA      NA      NA        NA\n","rs11571383         NA      NA      NA      NA      NA      NA      NA        NA\n","rs2107614          NA      NA      NA      NA      NA      NA      NA -0.224019\n","rs4980935          NA      NA      NA      NA      NA      NA      NA        NA\n","rs11064536         NA      NA      NA      NA      NA      NA      NA        NA\n","rs117766797        NA      NA      NA      NA      NA      NA      NA        NA\n","            beta_17  beta_18  beta_19   beta_20   beta_21 beta_22   beta_23\n","rs147560086      NA -0.27895       NA -0.393493 -0.538392      NA -0.380267\n","rs7955986        NA       NA       NA        NA        NA      NA        NA\n","rs718389         NA       NA       NA        NA        NA      NA        NA\n","rs11571383       NA       NA       NA        NA        NA      NA        NA\n","rs2107614        NA       NA       NA        NA        NA      NA        NA\n","rs4980935        NA       NA 0.242572        NA        NA      NA        NA\n","rs11064536       NA       NA       NA        NA        NA      NA        NA\n","rs117766797      NA       NA       NA        NA        NA      NA        NA\n","            beta_24  beta_25 beta_26 beta_27 beta_28   beta_29  beta_30 beta_31\n","rs147560086      NA -0.37429      NA      NA      NA -0.413552       NA      NA\n","rs7955986        NA       NA      NA      NA      NA        NA       NA      NA\n","rs718389         NA       NA      NA      NA      NA        NA       NA      NA\n","rs11571383       NA       NA      NA      NA      NA        NA       NA      NA\n","rs2107614        NA       NA      NA      NA      NA        NA       NA      NA\n","rs4980935        NA       NA      NA      NA      NA        NA       NA      NA\n","rs11064536       NA       NA      NA      NA      NA        NA 0.524096      NA\n","rs117766797      NA       NA      NA      NA      NA        NA       NA      NA\n","            beta_32 beta_33   beta_34   beta_35 beta_36 beta_37 beta_38\n","rs147560086      NA      NA -0.389496 -0.325135      NA      NA      NA\n","rs7955986        NA      NA        NA        NA      NA      NA      NA\n","rs718389         NA      NA        NA        NA      NA      NA      NA\n","rs11571383       NA      NA        NA        NA      NA      NA      NA\n","rs2107614        NA      NA        NA        NA      NA      NA      NA\n","rs4980935        NA      NA        NA        NA      NA      NA      NA\n","rs11064536       NA      NA        NA        NA      NA      NA      NA\n","rs117766797      NA      NA        NA        NA      NA      NA      NA\n","              beta_39 beta_40 beta_41   beta_42 beta_43   beta_44   beta_45\n","rs147560086 -0.348491      NA      NA        NA      NA -0.283741        NA\n","rs7955986          NA      NA      NA        NA      NA        NA        NA\n","rs718389           NA      NA      NA        NA      NA        NA -0.282161\n","rs11571383         NA      NA      NA -0.504512      NA        NA        NA\n","rs2107614          NA      NA      NA        NA      NA        NA        NA\n","rs4980935          NA      NA      NA        NA      NA        NA        NA\n","rs11064536         NA      NA      NA        NA      NA        NA        NA\n","rs117766797        NA      NA      NA        NA      NA        NA        NA\n","             beta_46 beta_47   beta_48 beta_49\n","rs147560086       NA      NA        NA      NA\n","rs7955986   0.243345      NA        NA      NA\n","rs718389          NA      NA        NA      NA\n","rs11571383        NA      NA        NA      NA\n","rs2107614         NA      NA        NA      NA\n","rs4980935         NA      NA        NA      NA\n","rs11064536        NA      NA        NA      NA\n","rs117766797       NA      NA -0.395786      NA\n","\n","$eqtl_se\n","            SE_1 SE_2      SE_3     SE_4 SE_5      SE_6     SE_7 SE_8      SE_9\n","rs147560086   NA   NA 0.0464181 0.029495   NA 0.0261202       NA   NA 0.0550403\n","rs7955986     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs718389      NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs11571383    NA   NA        NA       NA   NA        NA 0.105786   NA        NA\n","rs2107614     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs4980935     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs11064536    NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs117766797   NA   NA        NA       NA   NA        NA       NA   NA        NA\n","            SE_10 SE_11     SE_12 SE_13 SE_14 SE_15     SE_16 SE_17     SE_18\n","rs147560086    NA    NA 0.0567995    NA    NA    NA        NA    NA 0.0775922\n","rs7955986      NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs718389       NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs11571383     NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs2107614      NA    NA        NA    NA    NA    NA 0.0401716    NA        NA\n","rs4980935      NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs11064536     NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs117766797    NA    NA        NA    NA    NA    NA        NA    NA        NA\n","                SE_19     SE_20     SE_21 SE_22     SE_23 SE_24     SE_25 SE_26\n","rs147560086        NA 0.0327279 0.0351663    NA 0.0352046    NA 0.0353582    NA\n","rs7955986          NA        NA        NA    NA        NA    NA        NA    NA\n","rs718389           NA        NA        NA    NA        NA    NA        NA    NA\n","rs11571383         NA        NA        NA    NA        NA    NA        NA    NA\n","rs2107614          NA        NA        NA    NA        NA    NA        NA    NA\n","rs4980935   0.0586861        NA        NA    NA        NA    NA        NA    NA\n","rs11064536         NA        NA        NA    NA        NA    NA        NA    NA\n","rs117766797        NA        NA        NA    NA        NA    NA        NA    NA\n","            SE_27 SE_28     SE_29    SE_30 SE_31 SE_32 SE_33    SE_34     SE_35\n","rs147560086    NA    NA 0.0384595       NA    NA    NA    NA 0.030555 0.0222479\n","rs7955986      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs718389       NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs11571383     NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs2107614      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs4980935      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs11064536     NA    NA        NA 0.136083    NA    NA    NA       NA        NA\n","rs117766797    NA    NA        NA       NA    NA    NA    NA       NA        NA\n","            SE_36 SE_37 SE_38     SE_39 SE_40 SE_41     SE_42 SE_43     SE_44\n","rs147560086    NA    NA    NA 0.0452671    NA    NA        NA    NA 0.0360217\n","rs7955986      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs718389       NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs11571383     NA    NA    NA        NA    NA    NA 0.0733203    NA        NA\n","rs2107614      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs4980935      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs11064536     NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs117766797    NA    NA    NA        NA    NA    NA        NA    NA        NA\n","                SE_45     SE_46 SE_47    SE_48 SE_49\n","rs147560086        NA        NA    NA       NA    NA\n","rs7955986          NA 0.0191166    NA       NA    NA\n","rs718389    0.0359888        NA    NA       NA    NA\n","rs11571383         NA        NA    NA       NA    NA\n","rs2107614          NA        NA    NA       NA    NA\n","rs4980935          NA        NA    NA       NA    NA\n","rs11064536         NA        NA    NA       NA    NA\n","rs117766797        NA        NA    NA 0.107595    NA\n","\n","$eqtl_pvals\n","            pvalue_1 pvalue_2    pvalue_3    pvalue_4 pvalue_5    pvalue_6\n","rs147560086       NA       NA 1.57506e-11 8.15374e-29       NA 5.49137e-49\n","rs7955986         NA       NA          NA          NA       NA          NA\n","rs718389          NA       NA          NA          NA       NA          NA\n","rs11571383        NA       NA          NA          NA       NA          NA\n","rs2107614         NA       NA          NA          NA       NA          NA\n","rs4980935         NA       NA          NA          NA       NA          NA\n","rs11064536        NA       NA          NA          NA       NA          NA\n","rs117766797       NA       NA          NA          NA       NA          NA\n","               pvalue_7 pvalue_8    pvalue_9 pvalue_10 pvalue_11   pvalue_12\n","rs147560086          NA       NA 6.71399e-10        NA        NA 4.40477e-16\n","rs7955986            NA       NA          NA        NA        NA          NA\n","rs718389             NA       NA          NA        NA        NA          NA\n","rs11571383  2.64972e-06       NA          NA        NA        NA          NA\n","rs2107614            NA       NA          NA        NA        NA          NA\n","rs4980935            NA       NA          NA        NA        NA          NA\n","rs11064536           NA       NA          NA        NA        NA          NA\n","rs117766797          NA       NA          NA        NA        NA          NA\n","            pvalue_13 pvalue_14 pvalue_15  pvalue_16 pvalue_17  pvalue_18\n","rs147560086        NA        NA        NA         NA        NA 0.00050353\n","rs7955986          NA        NA        NA         NA        NA         NA\n","rs718389           NA        NA        NA         NA        NA         NA\n","rs11571383         NA        NA        NA         NA        NA         NA\n","rs2107614          NA        NA        NA 1.0067e-07        NA         NA\n","rs4980935          NA        NA        NA         NA        NA         NA\n","rs11064536         NA        NA        NA         NA        NA         NA\n","rs117766797        NA        NA        NA         NA        NA         NA\n","              pvalue_19   pvalue_20   pvalue_21 pvalue_22   pvalue_23 pvalue_24\n","rs147560086          NA 8.26297e-28 3.16236e-42        NA 9.45416e-23        NA\n","rs7955986            NA          NA          NA        NA          NA        NA\n","rs718389             NA          NA          NA        NA          NA        NA\n","rs11571383           NA          NA          NA        NA          NA        NA\n","rs2107614            NA          NA          NA        NA          NA        NA\n","rs4980935   8.06751e-05          NA          NA        NA          NA        NA\n","rs11064536           NA          NA          NA        NA          NA        NA\n","rs117766797          NA          NA          NA        NA          NA        NA\n","              pvalue_25 pvalue_26 pvalue_27 pvalue_28   pvalue_29   pvalue_30\n","rs147560086 3.44003e-22        NA        NA        NA 3.44995e-23          NA\n","rs7955986            NA        NA        NA        NA          NA          NA\n","rs718389             NA        NA        NA        NA          NA          NA\n","rs11571383           NA        NA        NA        NA          NA          NA\n","rs2107614            NA        NA        NA        NA          NA          NA\n","rs4980935            NA        NA        NA        NA          NA          NA\n","rs11064536           NA        NA        NA        NA          NA 0.000347708\n","rs117766797          NA        NA        NA        NA          NA          NA\n","            pvalue_31 pvalue_32 pvalue_33  pvalue_34   pvalue_35 pvalue_36\n","rs147560086        NA        NA        NA 2.6301e-33 4.95626e-40        NA\n","rs7955986          NA        NA        NA         NA          NA        NA\n","rs718389           NA        NA        NA         NA          NA        NA\n","rs11571383         NA        NA        NA         NA          NA        NA\n","rs2107614          NA        NA        NA         NA          NA        NA\n","rs4980935          NA        NA        NA         NA          NA        NA\n","rs11064536         NA        NA        NA         NA          NA        NA\n","rs117766797        NA        NA        NA         NA          NA        NA\n","            pvalue_37 pvalue_38   pvalue_39 pvalue_40 pvalue_41   pvalue_42\n","rs147560086        NA        NA 8.63893e-13        NA        NA          NA\n","rs7955986          NA        NA          NA        NA        NA          NA\n","rs718389           NA        NA          NA        NA        NA          NA\n","rs11571383         NA        NA          NA        NA        NA 2.07196e-10\n","rs2107614          NA        NA          NA        NA        NA          NA\n","rs4980935          NA        NA          NA        NA        NA          NA\n","rs11064536         NA        NA          NA        NA        NA          NA\n","rs117766797        NA        NA          NA        NA        NA          NA\n","            pvalue_43   pvalue_44   pvalue_45   pvalue_46 pvalue_47   pvalue_48\n","rs147560086        NA 8.32165e-14          NA          NA        NA          NA\n","rs7955986          NA          NA          NA 2.21429e-32        NA          NA\n","rs718389           NA          NA 1.07802e-13          NA        NA          NA\n","rs11571383         NA          NA          NA          NA        NA          NA\n","rs2107614          NA          NA          NA          NA        NA          NA\n","rs4980935          NA          NA          NA          NA        NA          NA\n","rs11064536         NA          NA          NA          NA        NA          NA\n","rs117766797        NA          NA          NA          NA        NA 0.000357146\n","            pvalue_49\n","rs147560086        NA\n","rs7955986          NA\n","rs718389           NA\n","rs11571383         NA\n","rs2107614          NA\n","rs4980935          NA\n","rs11064536         NA\n","rs117766797        NA\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n"]}]},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(base) # R package installed by default\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(Matrix) # R package installed by default\n","library(methods) # R package installed by default\n","library(stats) # R package installed by default\n","library(lme4) # It needs to be installed separately\n","library(data.table) # It needs to be installed separately\n","\n","# Step 2: Main Mt_Robin function:\n","MR_MtRobin <- function(MR_MtRobin_input,pval_thresh=0.001){\n","  snpID <- MR_MtRobin_input$snpID\n","  gwas_betas <- MR_MtRobin_input$gwas_betas\n","  gwas_se <- MR_MtRobin_input$gwas_se\n","  eqtl_betas <- MR_MtRobin_input$eqtl_betas\n","  eqtl_se <- MR_MtRobin_input$eqtl_se\n","  eqtl_pvals <- MR_MtRobin_input$eqtl_pvals\n","  LD <- MR_MtRobin_input$LD\n","\n","  if(ncol(eqtl_betas) != ncol(eqtl_se) | ncol(eqtl_betas) != ncol(eqtl_pvals)){\n","    stop(\"eQTL betas, SE and p-values do not all have same number of columns\")\n","  }\n","\n","  ## determine number of tissues (or conditions/studies/etc.)\n","  nT <- ncol(eqtl_betas)\n","\n","  ## rename columns for standardized processing\n","  colnames(eqtl_betas) <- paste0(\"beta_\",1:nT)\n","  colnames(eqtl_se) <- paste0(\"SE_\",1:nT)\n","  colnames(eqtl_pvals) <- paste0(\"pvalue_\",1:nT)\n","\n","  ## combine eQTL results into data.table\n","  eqtl_res <- data.table::data.table(snpID,eqtl_betas,eqtl_se,eqtl_pvals)\n","\n","  # reshape to long format (each SNP-gene-tissue will be row)\n","  eQTL_res_melt <- data.table::melt(eqtl_res,id.vars=c(\"snpID\"),\n","                                    measure.vars=patterns(\"^beta_\",\"^SE_\",\"^pvalue_\"),value.name=c(\"beta\",\"se\",\"pvalue\"),\n","                                    variable.name=\"tissue\")\n","\n","  ## subset to strong instrumental variables (based on p-value threshold)\n","  eQTL_res_melt_PltThresh <- subset(eQTL_res_melt, pvalue < pval_thresh)\n","\n","  ## set up gwas data for merging\n","  gwas_res <- data.table::data.table(snpID=snpID,gwas_beta=gwas_betas,gwas_se=gwas_se)\n","\n","  ## merge eQTL and GWAS results\n","  data.table::setkey(eQTL_res_melt_PltThresh,snpID)\n","  data.table::setkey(gwas_res,snpID)\n","  merged_res <- data.table::merge.data.table(eQTL_res_melt_PltThresh,gwas_res)\n","\n","  ## set up coefficients for reverse regression\n","  beta_x <- matrix(merged_res$beta,ncol=1)\n","  beta_y <- matrix(merged_res$gwas_beta,ncol=1)\n","\n","  ## standard errors (for weights)\n","  se_x <- matrix(merged_res$se,ncol=1)\n","  se_y <- matrix(merged_res$gwas_se,ncol=1) ## not used by function; used in resampling (consider returning with results)\n","\n","  ## identifiers\n","  snpID <- merged_res$snpID\n","\n","  lme_res <- lme4::lmer(beta_x~(beta_y-1)+(beta_y-1|snpID),weights=1/se_x^2)\n","\n","  ## return results from weighted regression with random slopes (and correlated errors) along with GWAS SE and LD (needed for resampling)\n","  return(list(lme_res=lme_res,gwas_res=gwas_res,LD=LD))\n","}\n","\n","# Step 3: Read the previous file:\n","result_setup <- readRDS(\"/data/long_covid_setup_fs_3.rds\")\n","\n","# Step 4: Apply the main Mt_Robin function using parallel processing:\n","no_cores <- 38\n","\n","cl <- makeCluster(no_cores) # Create a cluster with the number of cores\n","clusterExport(cl, \"MR_MtRobin\") # Export the MR_MtRobin function to all the nodes of the cluster\n","result_list_MtRobin <- parLapply(cl, result_setup, function(MR_MtRobin_input) {\n","  tryCatch({\n","    MR_MtRobin(MR_MtRobin_input = MR_MtRobin_input)\n","  }, error = function(e) print(e))\n","})\n","stopCluster(cl) # Stop the cluster once the computation is done\n","\n","# Step 5: The names of the results will correspond to the gene ids:\n","names(result_list_MtRobin) <- names(result_setup)\n","\n","# Step 6: Save the previous result as a .rds file:\n","saveRDS(result_list_MtRobin, file = \"/data/long_covid_main_fs_3.rds\")"],"metadata":{"id":"6JA2vSfELihk"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/main_function.sif Rscript /data/main_function_fs_3.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"ZD7Dp5LxLpO4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 17:34:46:\n","   Job Id:             93443498.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      2.73\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 00:02:59\n","   Memory Requested:   300.0GB               Memory Used: 9.72GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:01:31\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"nTBgCZ2MMvZi"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_main_fs_3 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_3.rds\")\n","\n","# Filter out errors:\n","result_main_fs_3 <- result_main_fs_3[!sapply(result_main_fs_3, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_main_fs_3))\n","print(length(unique(result_main_fs_3)))\n","print(head(result_main_fs_3))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"J-Rn3PkmLSJF","executionInfo":{"status":"ok","timestamp":1699424476911,"user_tz":-630,"elapsed":12206,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"3a77060d-11d8-4c18-dec4-c5c7816e0984"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1796\n","[1] 1796\n","$ENSG00000002016\n","$ENSG00000002016$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -28.894\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 8.267   \n"," Residual        1.941   \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," 1.655  \n","\n","$ENSG00000002016$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1  rs11064536  0.07077150 0.0437185\n","2  rs11571383 -0.03928230 0.0700993\n","3 rs117766797 -0.09468080 0.1211550\n","4 rs147560086  0.04747050 0.0395567\n","5   rs2107614 -0.02811250 0.0339167\n","6   rs4980935 -0.00631512 0.0413246\n","7    rs718389 -0.06199730 0.0403831\n","8   rs7955986 -0.03200370 0.0324808\n","\n","$ENSG00000002016$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","\n","$ENSG00000005156\n","$ENSG00000005156$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -7.0476\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 3.583   \n"," Residual        4.728   \n","Number of obs: 10, groups:  snpID, 3\n","Fixed Effects:\n","beta_y  \n"," 9.793  \n","\n","$ENSG00000005156$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1  rs12945428 -0.03365740 0.0306682\n","2 rs201016783  0.13056400 0.0778736\n","3   rs3135958  0.00248115 0.0477995\n","\n","$ENSG00000005156$LD\n","             rs12945428    rs3135958  rs201016783\n","rs12945428   1.00000000 -0.295076381  0.020183627\n","rs3135958   -0.29507638  1.000000000 -0.006054471\n","rs201016783  0.02018363 -0.006054471  1.000000000\n","\n","\n","$ENSG00000006282\n","$ENSG00000006282$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -23.9258\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 19.861  \n"," Residual         5.685  \n","Number of obs: 46, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n","-18.93  \n","\n","$ENSG00000006282$gwas_res\n","      snpID gwas_beta  gwas_se\n","1 rs2306001 0.0488945 0.050476\n","2 rs9890200 0.0244462 0.032195\n","\n","$ENSG00000006282$LD\n","          rs9890200 rs2306001\n","rs9890200 1.0000000 0.3686704\n","rs2306001 0.3686704 1.0000000\n","\n","\n","$ENSG00000006744\n","$ENSG00000006744$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -2.1316\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 7.049   \n"," Residual        6.781   \n","Number of obs: 18, groups:  snpID, 4\n","Fixed Effects:\n","beta_y  \n"," 6.471  \n","\n","$ENSG00000006744$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1  rs1044564  0.03018110 0.0317623\n","2 rs11655929  0.05194220 0.0332599\n","3   rs733890 -0.00789432 0.0327968\n","4 rs78840287 -0.11447800 0.0967152\n","\n","$ENSG00000006744$LD\n","              rs1044564    rs733890 rs11655929   rs78840287\n","rs1044564   1.000000000  0.66545097 -0.6626427 -0.007565868\n","rs733890    0.665450966  1.00000000 -0.4997189  0.027633741\n","rs11655929 -0.662642701 -0.49971889  1.0000000  0.034843798\n","rs78840287 -0.007565868  0.02763374  0.0348438  1.000000000\n","\n","\n","$ENSG00000007171\n","$ENSG00000007171$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 9.2054\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 49.045  \n"," Residual         4.269  \n","Number of obs: 31, groups:  snpID, 5\n","Fixed Effects:\n","beta_y  \n"," 29.39  \n","\n","$ENSG00000007171$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1  rs2314813 -0.00713282 0.0465681\n","2  rs3751972 -0.00791894 0.0365161\n","3   rs624196  0.05032310 0.0342287\n","4 rs72846720  0.09079760 0.0795650\n","5  rs8078340  0.00528333 0.0463574\n","\n","$ENSG00000007171$LD\n","              rs3751972   rs8078340    rs624196    rs2314813  rs72846720\n","rs3751972   1.000000000 -0.20973763 -0.03796279  0.004244027 -0.08622187\n","rs8078340  -0.209737626  1.00000000  0.03757975  0.160333611 -0.06812959\n","rs624196   -0.037962786  0.03757975  1.00000000 -0.039583359  0.20176556\n","rs2314813   0.004244027  0.16033361 -0.03958336  1.000000000 -0.10409667\n","rs72846720 -0.086221874 -0.06812959  0.20176556 -0.104096673  1.00000000\n","\n","\n","$ENSG00000007341\n","$ENSG00000007341$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -1.3777\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 79.922  \n"," Residual         7.687  \n","Number of obs: 12, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n"," -71.5  \n","\n","$ENSG00000007341$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1 rs10745330 -0.0040702 0.0312692\n","2   rs954679  0.0392061 0.0322534\n","\n","$ENSG00000007341$LD\n","           rs10745330  rs954679\n","rs10745330  1.0000000 0.6167231\n","rs954679    0.6167231 1.0000000\n","\n","\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_fs_3.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_3_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_3_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_fs3.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"zPWoH9zfBtC6"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Option 1: Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_fs_3.R\n","\n","# Option 2: Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"Ze3PnnadBvE0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-07 20:15:36:\n","   Job Id:             107707388.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      692.00\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 139:26:08\n","   Memory Requested:   500.0GB               Memory Used: 114.15GB\n","   Walltime requested: 10:00:00            Walltime Used: 04:48:20\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"9mNBNPTIB0KD"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs3.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results))\n","print(head(results))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ihlqo-9gB60H","executionInfo":{"status":"ok","timestamp":1707334331043,"user_tz":-630,"elapsed":1072,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"127ab6f9-fe28-4f21-b0f9-0283ce7abe8e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1796    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.628      10000 0.9095871\n","2 ENSG00000005156  0.165      10000 0.6704525\n","3 ENSG00000006282  0.365      10000 0.8204506\n","4 ENSG00000006744  0.181      10000 0.6901826\n","5 ENSG00000007171  0.289      10000 0.7645310\n","6 ENSG00000007341  0.232      10000 0.7050288\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_3.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mMpFwmnaCNpS","executionInfo":{"status":"ok","timestamp":1707334430413,"user_tz":-630,"elapsed":99371,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"a3fb3078-3f8b-43b8-c47a-3c9be22da1b3"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 1796 10 \n","\n","          gene_id gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000002016     RAD52 protein_coding   1.655462  23    8\n","2 ENSG00000005156      LIG3 protein_coding   9.792645  10    3\n","3 ENSG00000006282   SPATA20 protein_coding -18.930839  46    2\n","4 ENSG00000006744     ELAC2 protein_coding   6.471382  18    4\n","5 ENSG00000007171      NOS2 protein_coding  29.386142  31    5\n","6 ENSG00000007341      ST7L protein_coding -71.497400  12    2\n","                                                                               SNP_IDs\n","1 rs11064536,rs11571383,rs117766797,rs147560086,rs2107614,rs4980935,rs718389,rs7955986\n","2                                                     rs12945428,rs201016783,rs3135958\n","3                                                                  rs2306001,rs9890200\n","4                                             rs1044564,rs11655929,rs733890,rs78840287\n","5                                    rs2314813,rs3751972,rs624196,rs72846720,rs8078340\n","6                                                                  rs10745330,rs954679\n","        REML        SD Residual_SD\n","1 -28.894033  8.266675    1.940726\n","2  -7.047622  3.582992    4.728046\n","3 -23.925755 19.861204    5.684947\n","4  -2.131642  7.049343    6.781474\n","5   9.205360 49.045380    4.268737\n","6  -1.377678 79.921933    7.686964\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_not_sig.txt \n","Dimensions of protein_coding_genes : 825 10 \n","\n","          gene_id gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000160200       CBS protein_coding  817.3963  27    3\n","2 ENSG00000213221      DNLZ protein_coding -670.2225  12    2\n","3 ENSG00000205307     SAP25 protein_coding  312.8933  23    5\n","4 ENSG00000165806     CASP7 protein_coding  300.8689  43    4\n","5 ENSG00000148690  FRA10AC1 protein_coding  286.5442  45    3\n","6 ENSG00000101639    CEP192 protein_coding  276.9824   6    2\n","                                                  SNP_IDs       REML        SD\n","1                            rs234728,rs35879780,rs747568 -24.218945 1279.1168\n","2                                  rs150174514,rs78298678   1.853155  975.3201\n","3 rs112888745,rs117460458,rs118088865,rs7793561,rs7799695   9.028055  575.6981\n","4                rs10553596,rs3127094,rs3814231,rs7918733  -6.485768  330.2137\n","5                          rs10736078,rs1998840,rs7909452  -3.200758  491.0294\n","6                                   rs17659749,rs35790309 -11.996949  251.9709\n","  Residual_SD\n","1    1.650699\n","2    1.592763\n","3    3.777067\n","4    5.109980\n","5    3.481673\n","6    3.665194\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# FS3 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_3_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs3.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"vJRq2aB4Ch1F","executionInfo":{"status":"ok","timestamp":1707334444644,"user_tz":-630,"elapsed":14235,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"ffc80894-3477-4c58-f569-8f4d85b2ba5a"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1796\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -28.894\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 8.267   \n"," Residual        1.941   \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," 1.655  \n","\n","$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1  rs11064536  0.07077150 0.0437185\n","2  rs11571383 -0.03928230 0.0700993\n","3 rs117766797 -0.09468080 0.1211550\n","4 rs147560086  0.04747050 0.0395567\n","5   rs2107614 -0.02811250 0.0339167\n","6   rs4980935 -0.00631512 0.0413246\n","7    rs718389 -0.06199730 0.0403831\n","8   rs7955986 -0.03200370 0.0324808\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 825  10\n","          gene_id gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000160200       CBS protein_coding  817.3963  27    3\n","2 ENSG00000213221      DNLZ protein_coding -670.2225  12    2\n","3 ENSG00000205307     SAP25 protein_coding  312.8933  23    5\n","4 ENSG00000165806     CASP7 protein_coding  300.8689  43    4\n","5 ENSG00000148690  FRA10AC1 protein_coding  286.5442  45    3\n","6 ENSG00000101639    CEP192 protein_coding  276.9824   6    2\n","                                                  SNP_IDs       REML        SD\n","1                            rs234728,rs35879780,rs747568 -24.218945 1279.1168\n","2                                  rs150174514,rs78298678   1.853155  975.3201\n","3 rs112888745,rs117460458,rs118088865,rs7793561,rs7799695   9.028055  575.6981\n","4                rs10553596,rs3127094,rs3814231,rs7918733  -6.485768  330.2137\n","5                          rs10736078,rs1998840,rs7909452  -3.200758  491.0294\n","6                                   rs17659749,rs35790309 -11.996949  251.9709\n","  Residual_SD\n","1    1.650699\n","2    1.592763\n","3    3.777067\n","4    5.109980\n","5    3.481673\n","6    3.665194\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1796    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.628      10000 0.9095871\n","2 ENSG00000005156  0.165      10000 0.6704525\n","3 ENSG00000006282  0.365      10000 0.8204506\n","4 ENSG00000006744  0.181      10000 0.6901826\n","5 ENSG00000007171  0.289      10000 0.7645310\n","6 ENSG00000007341  0.232      10000 0.7050288\n","[1] \"Final table (protein coding genes): \"\n","[1] 10 22\n","           gene_id gene_name SNPs\n","1  ENSG00000231887      PRH1    2\n","2  ENSG00000146530      VWDE    8\n","3  ENSG00000163218   PGLYRP4    3\n","4  ENSG00000180988    OR52N2    4\n","5  ENSG00000188878      FBF1    5\n","6  ENSG00000184983    NDUFA6    2\n","7  ENSG00000163879    DNALI1    9\n","8  ENSG00000158825       CDA   14\n","9  ENSG00000156603     MED19   14\n","10 ENSG00000139714     MORN3   18\n","                                                                                                                                                                                                SNPs_IDs\n","1                                                                                                                                                                                    rs1047712,rs2192139\n","2                                                                                                                      rs12673085,rs12699367,rs2287067,rs28811348,rs2966627,rs4721083,rs6971764,rs970035\n","3                                                                                                                                                                         rs2987769,rs3014859,rs75282098\n","4                                                                                                                                                              rs10838647,rs11500741,rs4758435,rs9630185\n","5                                                                                                                                                  rs1135889,rs115872616,rs140602544,rs8074216,rs9674908\n","6                                                                                                                                                                                   rs137139,rs139043947\n","7                                                                                                       rs10493083,rs11264069,rs114745536,rs12035095,rs12037070,rs17510031,rs215812,rs5773597,rs61778057\n","8                                                     rs10753504,rs10916784,rs10916834,rs12026680,rs12045689,rs12135822,rs1253799,rs3131712,rs3215400,rs55665119,rs66731853,rs72649199,rs818199,rs879088\n","9                                                 rs11229259,rs11601607,rs117394301,rs117732092,rs12801976,rs141274831,rs17151811,rs1783818,rs200563414,rs2244169,rs4938870,rs501738,rs527528,rs80244775\n","10 rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","   Obs     beta_y        Min    First_Q       Median   Third_Q       Max SD_RE\n","1    4  39.268399 -1.1517977 -0.9487258 -0.433631947 0.2470970 0.9470741     0\n","2   12  17.974368 -2.4028586 -0.5267255  0.005083745 0.3245058 0.6611698     0\n","3    4  12.736981 -0.7974389 -0.4267816 -0.245301670 0.2333873 1.4956718     0\n","4    7 -10.971544 -1.4074704 -0.3450274  0.091901147 0.5702460 1.6823007     0\n","5   13   9.887838 -1.5111109 -0.4066475  0.361600262 0.6293731 1.8350816     0\n","6   24   9.537779 -2.9053644 -0.5220021 -0.042367062 0.6476085 1.4989567     0\n","7   17  -8.512086 -1.7251893 -0.8498967  0.191919481 0.6148812 1.9927265     0\n","8   26  -6.629380 -0.8342598 -0.2775603  0.149572559 0.7866489 3.7019786     0\n","9   30   5.155099 -2.4345413 -0.1434968  0.138425502 0.4958109 1.3218241     0\n","10  21   4.800290 -1.2851537 -0.7241708  0.157421562 0.8632107 2.2643836     0\n","   Res_SD_RE Var_RE Res_Var_RE     SE_FE t_value_FE      REML p_value FDR\n","1   5.859655      0  34.335562 7.0046140   5.606076 -4.666248       0   0\n","2   5.028867      0  25.289502 2.5363699   7.086651  5.889571       0   0\n","3   6.391086      0  40.845977 2.3615584   5.393464  0.223643       0   0\n","4   2.810013      0   7.896174 1.2010824  -9.134714 -6.057579       0   0\n","5   4.112469      0  16.912398 0.7944717  12.445803 -5.902892       0   0\n","6   4.691699      0  22.012035 0.5926211  16.094228 16.514186       0   0\n","7   3.179144      0  10.106958 0.7656531 -11.117418 -5.049260       0   0\n","8   6.066736      0  36.805288 1.1215372  -5.910976 14.316605       0   0\n","9   3.237346      0  10.480408 0.5199279   9.915027 -4.079301       0   0\n","10  3.633380      0  13.201452 0.8906453   5.389676 14.966089       0   0\n","     CI_lower  CI_upper\n","1   25.539355 52.997442\n","2   13.003083 22.945653\n","3    8.108326 17.365635\n","4  -13.325665 -8.617422\n","5    8.330673 11.445002\n","6    8.376242 10.699316\n","7  -10.012767 -7.011406\n","8   -8.827593 -4.431167\n","9    4.136040  6.174158\n","10   3.054625  6.545954\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_fs_3.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_3_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_3_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_fs3.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"T-F89C0NOFCd"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Option 1: Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_fs_3.R\n","\n","# Option 2: Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"Hcw3Hp4NOKcW"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-07 20:15:36:\n","   Job Id:             107707388.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      692.00\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 139:26:08\n","   Memory Requested:   500.0GB               Memory Used: 114.15GB\n","   Walltime requested: 10:00:00            Walltime Used: 04:48:20\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"yF90q8tqONF0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs3.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results))\n","print(head(results))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Pjmm0EF1OTw2","executionInfo":{"status":"ok","timestamp":1707354331166,"user_tz":-630,"elapsed":530,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"6313d119-8f96-4d0e-c0ff-fa6ec8f70b12"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1796    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.628      10000 0.9095871\n","2 ENSG00000005156  0.165      10000 0.6704525\n","3 ENSG00000006282  0.365      10000 0.8204506\n","4 ENSG00000006744  0.181      10000 0.6901826\n","5 ENSG00000007171  0.289      10000 0.7645310\n","6 ENSG00000007341  0.232      10000 0.7050288\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_3.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ORWa6kmZN_72","executionInfo":{"status":"ok","timestamp":1707354470263,"user_tz":-630,"elapsed":139099,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"050de5ac-c8ed-45bf-d72a-6bc9a93a7f6b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 1796 10 \n","\n","          gene_id gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000002016     RAD52 protein_coding   1.655462  23    8\n","2 ENSG00000005156      LIG3 protein_coding   9.792645  10    3\n","3 ENSG00000006282   SPATA20 protein_coding -18.930839  46    2\n","4 ENSG00000006744     ELAC2 protein_coding   6.471382  18    4\n","5 ENSG00000007171      NOS2 protein_coding  29.386142  31    5\n","6 ENSG00000007341      ST7L protein_coding -71.497400  12    2\n","                                                                               SNP_IDs\n","1 rs11064536,rs11571383,rs117766797,rs147560086,rs2107614,rs4980935,rs718389,rs7955986\n","2                                                     rs12945428,rs201016783,rs3135958\n","3                                                                  rs2306001,rs9890200\n","4                                             rs1044564,rs11655929,rs733890,rs78840287\n","5                                    rs2314813,rs3751972,rs624196,rs72846720,rs8078340\n","6                                                                  rs10745330,rs954679\n","        REML        SD Residual_SD\n","1 -28.894033  8.266675    1.940726\n","2  -7.047622  3.582992    4.728046\n","3 -23.925755 19.861204    5.684947\n","4  -2.131642  7.049343    6.781474\n","5   9.205360 49.045380    4.268737\n","6  -1.377678 79.921933    7.686964\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_not_sig.txt \n","Dimensions of protein_coding_genes : 825 10 \n","\n","          gene_id gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000160200       CBS protein_coding  817.3963  27    3\n","2 ENSG00000213221      DNLZ protein_coding -670.2225  12    2\n","3 ENSG00000205307     SAP25 protein_coding  312.8933  23    5\n","4 ENSG00000165806     CASP7 protein_coding  300.8689  43    4\n","5 ENSG00000148690  FRA10AC1 protein_coding  286.5442  45    3\n","6 ENSG00000101639    CEP192 protein_coding  276.9824   6    2\n","                                                  SNP_IDs       REML        SD\n","1                            rs234728,rs35879780,rs747568 -24.218945 1279.1168\n","2                                  rs150174514,rs78298678   1.853155  975.3201\n","3 rs112888745,rs117460458,rs118088865,rs7793561,rs7799695   9.028055  575.6981\n","4                rs10553596,rs3127094,rs3814231,rs7918733  -6.485768  330.2137\n","5                          rs10736078,rs1998840,rs7909452  -3.200758  491.0294\n","6                                   rs17659749,rs35790309 -11.996949  251.9709\n","  Residual_SD\n","1    1.650699\n","2    1.592763\n","3    3.777067\n","4    5.109980\n","5    3.481673\n","6    3.665194\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# FS3 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_3_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs3.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1707354491777,"user_tz":-630,"elapsed":21517,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"52b3a9e8-1661-4a72-8f8d-239c78eed44a","id":"FmVdxX2nO-_V"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1796\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -28.894\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 8.267   \n"," Residual        1.941   \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," 1.655  \n","\n","$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1  rs11064536  0.07077150 0.0437185\n","2  rs11571383 -0.03928230 0.0700993\n","3 rs117766797 -0.09468080 0.1211550\n","4 rs147560086  0.04747050 0.0395567\n","5   rs2107614 -0.02811250 0.0339167\n","6   rs4980935 -0.00631512 0.0413246\n","7    rs718389 -0.06199730 0.0403831\n","8   rs7955986 -0.03200370 0.0324808\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 825  10\n","          gene_id gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000160200       CBS protein_coding  817.3963  27    3\n","2 ENSG00000213221      DNLZ protein_coding -670.2225  12    2\n","3 ENSG00000205307     SAP25 protein_coding  312.8933  23    5\n","4 ENSG00000165806     CASP7 protein_coding  300.8689  43    4\n","5 ENSG00000148690  FRA10AC1 protein_coding  286.5442  45    3\n","6 ENSG00000101639    CEP192 protein_coding  276.9824   6    2\n","                                                  SNP_IDs       REML        SD\n","1                            rs234728,rs35879780,rs747568 -24.218945 1279.1168\n","2                                  rs150174514,rs78298678   1.853155  975.3201\n","3 rs112888745,rs117460458,rs118088865,rs7793561,rs7799695   9.028055  575.6981\n","4                rs10553596,rs3127094,rs3814231,rs7918733  -6.485768  330.2137\n","5                          rs10736078,rs1998840,rs7909452  -3.200758  491.0294\n","6                                   rs17659749,rs35790309 -11.996949  251.9709\n","  Residual_SD\n","1    1.650699\n","2    1.592763\n","3    3.777067\n","4    5.109980\n","5    3.481673\n","6    3.665194\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1796    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.628      10000 0.9095871\n","2 ENSG00000005156  0.165      10000 0.6704525\n","3 ENSG00000006282  0.365      10000 0.8204506\n","4 ENSG00000006744  0.181      10000 0.6901826\n","5 ENSG00000007171  0.289      10000 0.7645310\n","6 ENSG00000007341  0.232      10000 0.7050288\n","[1] \"Final table (protein coding genes): \"\n","[1] 10 22\n","           gene_id gene_name SNPs\n","1  ENSG00000231887      PRH1    2\n","2  ENSG00000146530      VWDE    8\n","3  ENSG00000163218   PGLYRP4    3\n","4  ENSG00000180988    OR52N2    4\n","5  ENSG00000188878      FBF1    5\n","6  ENSG00000184983    NDUFA6    2\n","7  ENSG00000163879    DNALI1    9\n","8  ENSG00000158825       CDA   14\n","9  ENSG00000156603     MED19   14\n","10 ENSG00000139714     MORN3   18\n","                                                                                                                                                                                                SNPs_IDs\n","1                                                                                                                                                                                    rs1047712,rs2192139\n","2                                                                                                                      rs12673085,rs12699367,rs2287067,rs28811348,rs2966627,rs4721083,rs6971764,rs970035\n","3                                                                                                                                                                         rs2987769,rs3014859,rs75282098\n","4                                                                                                                                                              rs10838647,rs11500741,rs4758435,rs9630185\n","5                                                                                                                                                  rs1135889,rs115872616,rs140602544,rs8074216,rs9674908\n","6                                                                                                                                                                                   rs137139,rs139043947\n","7                                                                                                       rs10493083,rs11264069,rs114745536,rs12035095,rs12037070,rs17510031,rs215812,rs5773597,rs61778057\n","8                                                     rs10753504,rs10916784,rs10916834,rs12026680,rs12045689,rs12135822,rs1253799,rs3131712,rs3215400,rs55665119,rs66731853,rs72649199,rs818199,rs879088\n","9                                                 rs11229259,rs11601607,rs117394301,rs117732092,rs12801976,rs141274831,rs17151811,rs1783818,rs200563414,rs2244169,rs4938870,rs501738,rs527528,rs80244775\n","10 rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","   Obs     beta_y        Min    First_Q       Median   Third_Q       Max SD_RE\n","1    4  39.268399 -1.1517977 -0.9487258 -0.433631947 0.2470970 0.9470741     0\n","2   12  17.974368 -2.4028586 -0.5267255  0.005083745 0.3245058 0.6611698     0\n","3    4  12.736981 -0.7974389 -0.4267816 -0.245301670 0.2333873 1.4956718     0\n","4    7 -10.971544 -1.4074704 -0.3450274  0.091901147 0.5702460 1.6823007     0\n","5   13   9.887838 -1.5111109 -0.4066475  0.361600262 0.6293731 1.8350816     0\n","6   24   9.537779 -2.9053644 -0.5220021 -0.042367062 0.6476085 1.4989567     0\n","7   17  -8.512086 -1.7251893 -0.8498967  0.191919481 0.6148812 1.9927265     0\n","8   26  -6.629380 -0.8342598 -0.2775603  0.149572559 0.7866489 3.7019786     0\n","9   30   5.155099 -2.4345413 -0.1434968  0.138425502 0.4958109 1.3218241     0\n","10  21   4.800290 -1.2851537 -0.7241708  0.157421562 0.8632107 2.2643836     0\n","   Res_SD_RE Var_RE Res_Var_RE     SE_FE t_value_FE      REML p_value FDR\n","1   5.859655      0  34.335562 7.0046140   5.606076 -4.666248       0   0\n","2   5.028867      0  25.289502 2.5363699   7.086651  5.889571       0   0\n","3   6.391086      0  40.845977 2.3615584   5.393464  0.223643       0   0\n","4   2.810013      0   7.896174 1.2010824  -9.134714 -6.057579       0   0\n","5   4.112469      0  16.912398 0.7944717  12.445803 -5.902892       0   0\n","6   4.691699      0  22.012035 0.5926211  16.094228 16.514186       0   0\n","7   3.179144      0  10.106958 0.7656531 -11.117418 -5.049260       0   0\n","8   6.066736      0  36.805288 1.1215372  -5.910976 14.316605       0   0\n","9   3.237346      0  10.480408 0.5199279   9.915027 -4.079301       0   0\n","10  3.633380      0  13.201452 0.8906453   5.389676 14.966089       0   0\n","     CI_lower  CI_upper\n","1   25.539355 52.997442\n","2   13.003083 22.945653\n","3    8.108326 17.365635\n","4  -13.325665 -8.617422\n","5    8.330673 11.445002\n","6    8.376242 10.699316\n","7  -10.012767 -7.011406\n","8   -8.827593 -4.431167\n","9    4.136040  6.174158\n","10   3.054625  6.545954\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the last function to retrieve the tissue information and categories for fs1:\n","input_eqtl_path <- \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\"\n","input_mt_robin_path <- \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_full_table.txt\"\n","output_path <- \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_full_table_tissues.txt\"\n","process_eqtl_mt_robin(input_eqtl_path, input_mt_robin_path, output_path)"],"metadata":{"id":"kYJr4T8uO-_V","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714699187553,"user_tz":-570,"elapsed":44887,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"05a40fb8-fa08-40ee-b9ee-6af46a64f777"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["           gene_id gene_name SNPs\n","1  ENSG00000139714     MORN3   18\n","2  ENSG00000146530      VWDE    8\n","3  ENSG00000156603     MED19   14\n","4  ENSG00000158825       CDA   14\n","5  ENSG00000163218   PGLYRP4    3\n","6  ENSG00000163879    DNALI1    9\n","7  ENSG00000180988    OR52N2    4\n","8  ENSG00000184983    NDUFA6    2\n","9  ENSG00000188878      FBF1    5\n","10 ENSG00000231887      PRH1    2\n","                                                                                                                                                                                                SNPs_IDs\n","1  rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","2                                                                                                                      rs12673085,rs12699367,rs2287067,rs28811348,rs2966627,rs4721083,rs6971764,rs970035\n","3                                                 rs11229259,rs11601607,rs117394301,rs117732092,rs12801976,rs141274831,rs17151811,rs1783818,rs200563414,rs2244169,rs4938870,rs501738,rs527528,rs80244775\n","4                                                     rs10753504,rs10916784,rs10916834,rs12026680,rs12045689,rs12135822,rs1253799,rs3131712,rs3215400,rs55665119,rs66731853,rs72649199,rs818199,rs879088\n","5                                                                                                                                                                         rs2987769,rs3014859,rs75282098\n","6                                                                                                       rs10493083,rs11264069,rs114745536,rs12035095,rs12037070,rs17510031,rs215812,rs5773597,rs61778057\n","7                                                                                                                                                              rs10838647,rs11500741,rs4758435,rs9630185\n","8                                                                                                                                                                                   rs137139,rs139043947\n","9                                                                                                                                                  rs1135889,rs115872616,rs140602544,rs8074216,rs9674908\n","10                                                                                                                                                                                   rs1047712,rs2192139\n","   Obs     beta_y        Min    First_Q       Median   Third_Q       Max SD_RE\n","1   21   4.800290 -1.2851537 -0.7241708  0.157421562 0.8632107 2.2643836     0\n","2   12  17.974368 -2.4028586 -0.5267255  0.005083745 0.3245058 0.6611698     0\n","3   30   5.155099 -2.4345413 -0.1434968  0.138425502 0.4958109 1.3218241     0\n","4   26  -6.629380 -0.8342598 -0.2775603  0.149572559 0.7866489 3.7019786     0\n","5    4  12.736981 -0.7974389 -0.4267816 -0.245301670 0.2333873 1.4956718     0\n","6   17  -8.512086 -1.7251893 -0.8498967  0.191919481 0.6148812 1.9927265     0\n","7    7 -10.971544 -1.4074704 -0.3450274  0.091901147 0.5702460 1.6823007     0\n","8   24   9.537779 -2.9053644 -0.5220021 -0.042367062 0.6476085 1.4989567     0\n","9   13   9.887838 -1.5111109 -0.4066475  0.361600262 0.6293731 1.8350816     0\n","10   4  39.268399 -1.1517977 -0.9487258 -0.433631947 0.2470970 0.9470741     0\n","   Res_SD_RE Var_RE Res_Var_RE     SE_FE t_value_FE      REML p_value FDR\n","1   3.633380      0  13.201452 0.8906453   5.389676 14.966089       0   0\n","2   5.028867      0  25.289502 2.5363699   7.086651  5.889571       0   0\n","3   3.237346      0  10.480408 0.5199279   9.915027 -4.079301       0   0\n","4   6.066736      0  36.805288 1.1215372  -5.910976 14.316605       0   0\n","5   6.391086      0  40.845977 2.3615584   5.393464  0.223643       0   0\n","6   3.179144      0  10.106958 0.7656531 -11.117418 -5.049260       0   0\n","7   2.810013      0   7.896174 1.2010824  -9.134714 -6.057579       0   0\n","8   4.691699      0  22.012035 0.5926211  16.094228 16.514186       0   0\n","9   4.112469      0  16.912398 0.7944717  12.445803 -5.902892       0   0\n","10  5.859655      0  34.335562 7.0046140   5.606076 -4.666248       0   0\n","     CI_lower  CI_upper\n","1    3.054625  6.545954\n","2   13.003083 22.945653\n","3    4.136040  6.174158\n","4   -8.827593 -4.431167\n","5    8.108326 17.365635\n","6  -10.012767 -7.011406\n","7  -13.325665 -8.617422\n","8    8.376242 10.699316\n","9    8.330673 11.445002\n","10  25.539355 52.997442\n","                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Tissues\n","1                                                                                                                                     Visceral omentum adipose,Adrenal gland,Tibial artery,Cerebellum,Nucleus accumbens (basal ganglia),Spinal cord,Breast mammary tissue,EBV-transformed lymphocytes,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Kidney cortex,Liver,Minor salivary gland,Pancreas,Prostate,Sun-exposed skin (lower leg),Thyroid,Vagina,Whole blood\n","2                                                                                                                                                                                                                                                                                                                               Adrenal gland,Caudate,Spinal cord,Atrial appendage,Kidney cortex,Tibial nerve,Pituitary,Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Testis,Thyroid\n","3  Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Coronary artery,Tibial artery,Amygdala brain,Caudate,Cerebellum,Hippocampus,Hypothalamus,Putamen (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Kidney cortex,Minor salivary gland,Skeletal muscle,Tibial nerve,Ovary,Pituitary,Prostate,Sun-exposed skin (lower leg),Small intestine terminal ileum,Stomach,Testis,Uterus,Whole blood\n","4                                                                                         Visceral omentum adipose,Adrenal gland,Coronary artery,Tibial artery,Hippocampus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Esophagus mucosa,Atrial appendage,Left ventricle,Kidney cortex,Minor salivary gland,Tibial nerve,Ovary,Pancreas,Pituitary,Prostate,Stomach,Thyroid,Uterus,Vagina,Whole blood\n","5                                                                                                                                                                                                                                                                                                                                                                                                                                         EBV-transformed lymphocytes,Esophagus mucosa,Vagina,Whole blood\n","6                                                                                                                                                                                                                                                            Visceral omentum adipose,Aorta artery,Cortex,Substantia nigra,EBV-transformed lymphocytes,Esophagus muscularis,Kidney cortex,Liver,Skeletal muscle,Tibial nerve,Pancreas,Pituitary,Sun-exposed skin (lower leg),Testis,Thyroid,Uterus,Vagina\n","7                                                                                                                                                                                                                                                                                                                                                                                  Subcutaneous adipose,EBV-transformed lymphocytes,Tibial nerve,Small intestine terminal ileum,Spleen,Testis,Whole blood\n","8                                                                               Subcutaneous adipose,Adrenal gland,Coronary artery,Amygdala brain,Anterior cingulate cortex,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus muscularis,Left ventricle,Kidney cortex,Liver,Minor salivary gland,Tibial nerve,Pituitary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Stomach,Thyroid,Uterus\n","9                                                                                                                                                                                                                                                                                                  Frontal cortex,Substantia nigra,Breast mammary tissue,EBV-transformed lymphocytes,Gastroesophageal junction,Esophagus muscularis,Atrial appendage,Tibial nerve,Ovary,Testis,Thyroid,Uterus,Whole blood\n","10                                                                                                                                                                                                                                                                                                                                                                                                                                          Anterior cingulate cortex,Cultured fibroblasts,Testis,Thyroid\n","                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Tissue_Categories\n","1                                                                                                                                                                                                                                Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Digestive System,Exocrine System,Digestive System,Reproductive System,Skin,Endocrine System,Reproductive System,Blood\n","2                                                                                                                                                                                                                                                                                                                                                                                  Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Cardiovascular System,Excretory System,Brain and Nervous System,Endocrine System,Skin,Digestive System,Immune System,Reproductive System,Endocrine System\n","3  Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Excretory System,Exocrine System,Muscular System,Brain and Nervous System,Reproductive System,Endocrine System,Reproductive System,Skin,Digestive System,Digestive System,Reproductive System,Reproductive System,Blood\n","4                                                                                              Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Cardiovascular System,Cardiovascular System,Excretory System,Exocrine System,Brain and Nervous System,Reproductive System,Digestive System,Endocrine System,Reproductive System,Digestive System,Endocrine System,Reproductive System,Reproductive System,Blood\n","5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Immune System,Digestive System,Reproductive System,Blood\n","6                                                                                                                                                                                                                                                                                                  Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Excretory System,Digestive System,Muscular System,Brain and Nervous System,Digestive System,Endocrine System,Skin,Reproductive System,Endocrine System,Reproductive System,Reproductive System\n","7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        Adipose Tissue,Immune System,Brain and Nervous System,Digestive System,Immune System,Reproductive System,Blood\n","8                                                                                                                                                                                Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Digestive System,Exocrine System,Brain and Nervous System,Endocrine System,Reproductive System,Skin,Skin,Digestive System,Immune System,Digestive System,Endocrine System,Reproductive System\n","9                                                                                                                                                                                                                                                                                                                                                                   Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Digestive System,Cardiovascular System,Brain and Nervous System,Reproductive System,Reproductive System,Endocrine System,Reproductive System,Blood\n","10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Brain and Nervous System,Connective Tissue,Reproductive System,Endocrine System\n"]}]},{"cell_type":"markdown","source":["### **GWAS4**"],"metadata":{"id":"HfzIRUVYAsRc"}},{"cell_type":"code","source":["library(matrixStats) # It needs to be installed separately\n","#library(BiocManager) # It needs to be installed separately\n","library(snpStats) # It needs to be installed separately\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_4_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 4: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 5: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 6: Create a variable for the path to the directory to save the LD matrix results:\n","ld_matrix_dir <- \"/data/ld_matrix_fs_4\"\n","\n","# Step 7: Create the directory if it doesn't exist:\n","if (!dir.exists(ld_matrix_dir)) {\n","  dir.create(ld_matrix_dir)\n","}\n","\n","# Step 8: Modify the original select_IV_fs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_fs() function --> Starts with no variables and adds one at a time:\n","select_IV_fs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset eQTL dataset to gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  # pval_mat <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  #sig_neg_eQTL <- as.numeric(apply(betas_sig,1,function(x) any(x < -1e-15)))\n","  #sig_pos_eQTL <- as.numeric(apply(betas_sig,1,function(x) any(x > 1e-15)))\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Select the best SNPs remaining in SNP_pool based on minimum p-value:\n","  selected_snps <- NULL\n","  while(length(SNP_pool) > 1){\n","\n","    ## Identify top SNPs in pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    next_snp_idx <- which.min(eqtl_data$min_p)\n","    next_snp <- eqtl_data$variant_id[next_snp_idx]\n","    selected_snps <-c(selected_snps,next_snp)\n","\n","    ## Identify SNPs not in high LD with top SNPs in the pool:\n","    LD_idx <- which(row.names(LD_r2)==next_snp)\n","    LD_keep_idx <- which(LD_r2[LD_idx,] < ld_thresh)\n","    LD_r2 <- LD_r2[LD_keep_idx,LD_keep_idx]\n","    SNP_pool <- names(LD_keep_idx)\n","  }\n","  selected_snps <- c(selected_snps, SNP_pool)\n","  return(selected_snps)\n","}\n","\n","# Step 9: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 10: Apply the function to each gene id:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","  # Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_fs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","  }, error = function(e) print(e))\n","}, mc.cores = num_cores)\n","\n","# Step 11: Name the elements of the list with corresponding gene ids:\n","names(result_list) <- all_genes_vector\n","\n","# Step 12: Save the previous result as a .rds file:\n","saveRDS(result_list, file = \"/data/long_covid_IVs_fs_4.rds\")"],"metadata":{"id":"g6Y3KhJaeKKT"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/select_function_fs_4.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"6AJnctHWeKRa"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 10:26:27:\n","   Job Id:             93353708.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      800.88\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 258:56:13\n","   Memory Requested:   500.0GB               Memory Used: 393.53GB\n","   Walltime requested: 30:00:00            Walltime Used: 05:33:42\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"eYEswiXTM7zO"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_selectIV_fs_4 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_IVs_fs_4.rds\")\n","\n","# Filter out errors:\n","result_selectIV_fs_4 <- result_selectIV_fs_4[!sapply(result_selectIV_fs_4, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_selectIV_fs_4))\n","print(length(unique(result_selectIV_fs_4)))\n","print(head(result_selectIV_fs_4))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"2-g2-VwkM8ny","executionInfo":{"status":"ok","timestamp":1693540787437,"user_tz":-570,"elapsed":3506,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"a2108033-8708-4cd7-a46e-50de60e9ecac"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 38574\n","[1] 38480\n","$ENSG00000000003\n"," [1] \"rs8780\"      \"rs1204407\"   \"rs189980168\" \"rs4145090\"   \"rs139645081\"\n"," [6] \"rs112002620\" \"rs113922837\" \"rs17329105\"  \"rs193124579\" \"rs146639579\"\n","[11] \"rs201585484\" \"rs111478058\" \"rs143491698\" \"rs572714672\" \"rs56100956\" \n","[16] \"rs147601806\" \"rs139830751\" \"rs12842251\"  \"rs73630694\" \n","\n","$ENSG00000000005\n"," [1] \"rs6616166\"   \"rs4828038\"   \"rs12006573\"  \"rs73557922\"  \"rs141394856\"\n"," [6] \"rs140468798\" \"rs73555045\"  \"rs5921738\"   \"rs139231094\" \"rs59368006\" \n","[11] \"rs111915382\" \"rs73562536\"  \"rs12832642\"  \"rs201254498\" \"rs186611188\"\n","[16] \"rs113422802\" \"rs796285800\" \"rs180795909\" \"rs1004006\"   \"rs148617015\"\n","[21] \"rs140109110\"\n","\n","$ENSG00000000419\n"," [1] \"rs1292040204\" \"rs79562553\"   \"rs4809840\"    \"rs35201621\"   \"rs35530561\"  \n"," [6] \"rs141892692\"  \"rs944933\"     \"rs11483800\"   \"rs113036491\"  \"rs116892696\" \n","[11] \"rs34483216\"   \"rs60578213\"   \"rs118181672\"  \"rs4811253\"    \"rs11086363\"  \n","[16] \"rs73278571\"   \"rs73131221\"   \"rs78319058\"   \"rs79223715\"   \"rs189417964\" \n","[21] \"rs74422706\"   \"rs112070162\"  \"rs59034704\"   \"rs180710262\"  \"rs577456025\" \n","[26] \"rs117900008\"  \"rs34223230\"   \"rs141885330\"  \"rs78937289\"   \"rs16995896\"  \n","[31] \"rs72020869\"   \"rs142281981\"  \"rs13037515\"   \"rs1883804\"    \"rs17730415\"  \n","[36] \"rs11697209\"   \"rs112566287\"  \"rs149762110\"  \"rs4809839\"    \"rs6021441\"   \n","[41] \"rs75866341\"   \"rs13036755\"   \"rs74177405\"  \n","\n","$ENSG00000000457\n"," [1] \"rs10753796\"  \"rs4656710\"   \"rs115234612\" \"rs3766155\"   \"rs2056980\"  \n"," [6] \"rs4987403\"   \"rs12123695\"  \"rs185963038\" \"rs116684351\" \"rs1076637\"  \n","[11] \"rs7417281\"   \"rs7541615\"   \"rs973608\"    \"rs2982465\"   \"rs113123846\"\n","[16] \"rs4987338\"   \"rs114779274\" \"rs12038549\"  \"rs112361437\"\n","\n","$ENSG00000000460\n"," [1] \"rs35296671\"  \"rs1800016\"   \"rs4656198\"   \"rs3753311\"   \"rs4987391\"  \n"," [6] \"rs180738486\" \"rs4562681\"   \"rs531844\"    \"rs115157324\" \"rs12116424\" \n","[11] \"rs17346300\"  \"rs60843604\"  \"rs12742457\"  \"rs78159465\"  \"rs139679825\"\n","[16] \"rs35662131\" \n","\n","$ENSG00000000938\n"," [1] \"rs3737803\"   \"rs72191132\"  \"rs137984979\" \"rs11273428\"  \"rs75800899\" \n"," [6] \"rs191579330\" \"rs9426420\"   \"rs148432575\" \"rs10430146\"  \"rs139100109\"\n","[11] \"rs6701065\"   \"rs80298610\"  \"rs17361686\"  \"rs484740\"    \"rs114059345\"\n","[16] \"rs72651589\"  \"rs4357550\"   \"rs41284294\"  \"rs138392492\" \"rs148577969\"\n","[21] \"rs151288141\" \"rs71642227\"  \"rs12759594\"  \"rs2076457\"   \"rs141256722\"\n","[26] \"rs76443568\" \n","\n"]}]},{"cell_type":"code","source":["library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_4_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the list of genes-SNPs:\n","result_selectedIVs <- readRDS(\"/data/long_covid_IVs_fs_4.rds\")\n","\n","# Step 4: Create a variable for the path to the LD matrix directory:\n","ld_matrix_dir <- \"/data/ld_matrix_fs_4\"\n","\n","# Step 5: Setup the Mt_Robin method using the following setup function:\n","MR_MtRobin_setup <- function(geneID,\n","                             snpID,\n","                             eqtl_data,\n","                             gwas_data,\n","                             ld_matrix_path,\n","                             nTiss){\n","  tryCatch({\n","    ## Check if the gene is present and subset\n","    if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","    eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","    ## Confirm all SNPs are present in the dataset\n","    if(!all(snpID %in% eqtl_data$variant_id)) stop(\"Some SNPs missing in eQTL dataset\")\n","    if(!all(snpID %in% gwas_data$variant_id)) stop(\"Some SNPs missing in GWAS dataset\")\n","\n","    # Read the LD matrix from a file\n","    LD <- as.matrix(utils::read.table(ld_matrix_path,\n","                                      header = TRUE,\n","                                      check.names = FALSE))\n","\n","    if(!all(snpID %in% colnames(LD))) stop(\"Some SNPs missing in LD matrix\")\n","\n","    ## Align data\n","    eqtl_data <- eqtl_data[match(snpID,eqtl_data$variant_id),]\n","    gwas_data <- gwas_data[match(snpID,gwas_data$variant_id),]\n","    LD <- LD[match(snpID,rownames(LD)),match(snpID,colnames(LD))]\n","\n","    eqtl_betas <- as.matrix(eqtl_data[,paste0(\"beta_\",1:nTiss)])\n","    eqtl_se <- as.matrix(eqtl_data[,paste0(\"SE_\",1:nTiss)])\n","    eqtl_pvals <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","    gwas_betas <- gwas_data$beta\n","    gwas_se <- gwas_data$SE\n","\n","    row.names(eqtl_betas) <- row.names(eqtl_se) <- row.names(eqtl_pvals) <- eqtl_data$variant_id\n","\n","    return(list(snpID=snpID,gwas_betas=gwas_betas,gwas_se=gwas_se,\n","                eqtl_betas=eqtl_betas,eqtl_se=eqtl_se,eqtl_pvals=eqtl_pvals,LD=LD))\n","  }, error = function(e) print(e))\n","}\n","\n","# Step 6: Apply the previous setup function:\n","\n","# Step 6.1: Initialize an empty list to store the results:\n","MR_MtRobin_results <- list()\n","\n","# Step 6.2: Define the number of cores to use for the parallelism process:\n","num_cores <- 36\n","\n","# Step 6.3: Loop over each gene in result_selectedIVs:\n","MR_MtRobin_results <- mclapply(names(result_selectedIVs), function(geneID) {\n","\n","  # Step 6.3.1: Get the SNPs associated with each gene:\n","  snpID <- result_selectedIVs[[geneID]]\n","\n","  # Step 6.3.2: Generate a unique file path for the LD matrix of this gene:\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  # Step 6.3.3: Call the MR_MtRobin_setup() function with the parameters for each gene:\n","  return(tryCatch({\n","    MR_MtRobin_setup(geneID = geneID,\n","                     snpID = snpID,\n","                     eqtl_data = eqtl_covid_exposure,\n","                     gwas_data = gwas_covid_outcome,\n","                     ld_matrix_path = ld_matrix_path,\n","                     nTiss = 49)\n","  }, error = function(e) print(e)))\n","}, mc.cores = num_cores)\n","\n","# Step 7: Assign names to the list:\n","names(MR_MtRobin_results) <- names(result_selectedIVs)\n","\n","# Step 8: Save the previous result as a .rds file:\n","saveRDS(MR_MtRobin_results, file = \"/data/long_covid_setup_fs_4.rds\")"],"metadata":{"id":"zMfy_nKaAua3"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup.sif Rscript /data/setup_function_fs_4.R\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"Me6CoWlxM5yp"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 16:32:34:\n","   Job Id:             93430398.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      79.80\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 20:38:10\n","   Memory Requested:   300.0GB               Memory Used: 83.78GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:44:20\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"1NWeVe0tebWS"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_setup_fs_4 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_setup_fs_4.rds\")\n","\n","# Filter out errors:\n","result_setup_fs_4 <- result_setup_fs_4[!sapply(result_setup_fs_4, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_setup_fs_4))\n","print(length(unique(result_setup_fs_4)))\n","print(result_setup_fs_4[[1]])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"-fhljwwaNB9l","executionInfo":{"status":"ok","timestamp":1699417043803,"user_tz":-630,"elapsed":6031,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"859c14e7-a5a7-4e39-92e5-a3a34ff9dd67"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 4958\n","[1] 4958\n","$snpID\n","[1] \"rs147560086\" \"rs7955986\"   \"rs718389\"    \"rs11571383\"  \"rs2107614\"  \n","[6] \"rs4980935\"   \"rs11064536\"  \"rs117766797\"\n","\n","$gwas_betas\n","[1]  0.03468750 -0.02506380 -0.03680740 -0.05198390 -0.00444987 -0.00825649\n","[7]  0.01503010 -0.00869216\n","\n","$gwas_se\n","[1] 0.0287461 0.0233427 0.0272009 0.0617035 0.0245610 0.0261202 0.0315410\n","[8] 0.0734673\n","\n","$eqtl_betas\n","            beta_1 beta_2    beta_3   beta_4 beta_5    beta_6    beta_7 beta_8\n","rs147560086     NA     NA -0.332722 -0.36379     NA -0.428724        NA     NA\n","rs7955986       NA     NA        NA       NA     NA        NA        NA     NA\n","rs718389        NA     NA        NA       NA     NA        NA        NA     NA\n","rs11571383      NA     NA        NA       NA     NA        NA -0.525715     NA\n","rs2107614       NA     NA        NA       NA     NA        NA        NA     NA\n","rs4980935       NA     NA        NA       NA     NA        NA        NA     NA\n","rs11064536      NA     NA        NA       NA     NA        NA        NA     NA\n","rs117766797     NA     NA        NA       NA     NA        NA        NA     NA\n","               beta_9 beta_10 beta_11 beta_12 beta_13 beta_14 beta_15   beta_16\n","rs147560086 -0.362536      NA      NA -0.5128      NA      NA      NA        NA\n","rs7955986          NA      NA      NA      NA      NA      NA      NA        NA\n","rs718389           NA      NA      NA      NA      NA      NA      NA        NA\n","rs11571383         NA      NA      NA      NA      NA      NA      NA        NA\n","rs2107614          NA      NA      NA      NA      NA      NA      NA -0.224019\n","rs4980935          NA      NA      NA      NA      NA      NA      NA        NA\n","rs11064536         NA      NA      NA      NA      NA      NA      NA        NA\n","rs117766797        NA      NA      NA      NA      NA      NA      NA        NA\n","            beta_17  beta_18  beta_19   beta_20   beta_21 beta_22   beta_23\n","rs147560086      NA -0.27895       NA -0.393493 -0.538392      NA -0.380267\n","rs7955986        NA       NA       NA        NA        NA      NA        NA\n","rs718389         NA       NA       NA        NA        NA      NA        NA\n","rs11571383       NA       NA       NA        NA        NA      NA        NA\n","rs2107614        NA       NA       NA        NA        NA      NA        NA\n","rs4980935        NA       NA 0.242572        NA        NA      NA        NA\n","rs11064536       NA       NA       NA        NA        NA      NA        NA\n","rs117766797      NA       NA       NA        NA        NA      NA        NA\n","            beta_24  beta_25 beta_26 beta_27 beta_28   beta_29  beta_30 beta_31\n","rs147560086      NA -0.37429      NA      NA      NA -0.413552       NA      NA\n","rs7955986        NA       NA      NA      NA      NA        NA       NA      NA\n","rs718389         NA       NA      NA      NA      NA        NA       NA      NA\n","rs11571383       NA       NA      NA      NA      NA        NA       NA      NA\n","rs2107614        NA       NA      NA      NA      NA        NA       NA      NA\n","rs4980935        NA       NA      NA      NA      NA        NA       NA      NA\n","rs11064536       NA       NA      NA      NA      NA        NA 0.524096      NA\n","rs117766797      NA       NA      NA      NA      NA        NA       NA      NA\n","            beta_32 beta_33   beta_34   beta_35 beta_36 beta_37 beta_38\n","rs147560086      NA      NA -0.389496 -0.325135      NA      NA      NA\n","rs7955986        NA      NA        NA        NA      NA      NA      NA\n","rs718389         NA      NA        NA        NA      NA      NA      NA\n","rs11571383       NA      NA        NA        NA      NA      NA      NA\n","rs2107614        NA      NA        NA        NA      NA      NA      NA\n","rs4980935        NA      NA        NA        NA      NA      NA      NA\n","rs11064536       NA      NA        NA        NA      NA      NA      NA\n","rs117766797      NA      NA        NA        NA      NA      NA      NA\n","              beta_39 beta_40 beta_41   beta_42 beta_43   beta_44   beta_45\n","rs147560086 -0.348491      NA      NA        NA      NA -0.283741        NA\n","rs7955986          NA      NA      NA        NA      NA        NA        NA\n","rs718389           NA      NA      NA        NA      NA        NA -0.282161\n","rs11571383         NA      NA      NA -0.504512      NA        NA        NA\n","rs2107614          NA      NA      NA        NA      NA        NA        NA\n","rs4980935          NA      NA      NA        NA      NA        NA        NA\n","rs11064536         NA      NA      NA        NA      NA        NA        NA\n","rs117766797        NA      NA      NA        NA      NA        NA        NA\n","             beta_46 beta_47   beta_48 beta_49\n","rs147560086       NA      NA        NA      NA\n","rs7955986   0.243345      NA        NA      NA\n","rs718389          NA      NA        NA      NA\n","rs11571383        NA      NA        NA      NA\n","rs2107614         NA      NA        NA      NA\n","rs4980935         NA      NA        NA      NA\n","rs11064536        NA      NA        NA      NA\n","rs117766797       NA      NA -0.395786      NA\n","\n","$eqtl_se\n","            SE_1 SE_2      SE_3     SE_4 SE_5      SE_6     SE_7 SE_8      SE_9\n","rs147560086   NA   NA 0.0464181 0.029495   NA 0.0261202       NA   NA 0.0550403\n","rs7955986     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs718389      NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs11571383    NA   NA        NA       NA   NA        NA 0.105786   NA        NA\n","rs2107614     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs4980935     NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs11064536    NA   NA        NA       NA   NA        NA       NA   NA        NA\n","rs117766797   NA   NA        NA       NA   NA        NA       NA   NA        NA\n","            SE_10 SE_11     SE_12 SE_13 SE_14 SE_15     SE_16 SE_17     SE_18\n","rs147560086    NA    NA 0.0567995    NA    NA    NA        NA    NA 0.0775922\n","rs7955986      NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs718389       NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs11571383     NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs2107614      NA    NA        NA    NA    NA    NA 0.0401716    NA        NA\n","rs4980935      NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs11064536     NA    NA        NA    NA    NA    NA        NA    NA        NA\n","rs117766797    NA    NA        NA    NA    NA    NA        NA    NA        NA\n","                SE_19     SE_20     SE_21 SE_22     SE_23 SE_24     SE_25 SE_26\n","rs147560086        NA 0.0327279 0.0351663    NA 0.0352046    NA 0.0353582    NA\n","rs7955986          NA        NA        NA    NA        NA    NA        NA    NA\n","rs718389           NA        NA        NA    NA        NA    NA        NA    NA\n","rs11571383         NA        NA        NA    NA        NA    NA        NA    NA\n","rs2107614          NA        NA        NA    NA        NA    NA        NA    NA\n","rs4980935   0.0586861        NA        NA    NA        NA    NA        NA    NA\n","rs11064536         NA        NA        NA    NA        NA    NA        NA    NA\n","rs117766797        NA        NA        NA    NA        NA    NA        NA    NA\n","            SE_27 SE_28     SE_29    SE_30 SE_31 SE_32 SE_33    SE_34     SE_35\n","rs147560086    NA    NA 0.0384595       NA    NA    NA    NA 0.030555 0.0222479\n","rs7955986      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs718389       NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs11571383     NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs2107614      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs4980935      NA    NA        NA       NA    NA    NA    NA       NA        NA\n","rs11064536     NA    NA        NA 0.136083    NA    NA    NA       NA        NA\n","rs117766797    NA    NA        NA       NA    NA    NA    NA       NA        NA\n","            SE_36 SE_37 SE_38     SE_39 SE_40 SE_41     SE_42 SE_43     SE_44\n","rs147560086    NA    NA    NA 0.0452671    NA    NA        NA    NA 0.0360217\n","rs7955986      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs718389       NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs11571383     NA    NA    NA        NA    NA    NA 0.0733203    NA        NA\n","rs2107614      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs4980935      NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs11064536     NA    NA    NA        NA    NA    NA        NA    NA        NA\n","rs117766797    NA    NA    NA        NA    NA    NA        NA    NA        NA\n","                SE_45     SE_46 SE_47    SE_48 SE_49\n","rs147560086        NA        NA    NA       NA    NA\n","rs7955986          NA 0.0191166    NA       NA    NA\n","rs718389    0.0359888        NA    NA       NA    NA\n","rs11571383         NA        NA    NA       NA    NA\n","rs2107614          NA        NA    NA       NA    NA\n","rs4980935          NA        NA    NA       NA    NA\n","rs11064536         NA        NA    NA       NA    NA\n","rs117766797        NA        NA    NA 0.107595    NA\n","\n","$eqtl_pvals\n","            pvalue_1 pvalue_2    pvalue_3    pvalue_4 pvalue_5    pvalue_6\n","rs147560086       NA       NA 1.57506e-11 8.15374e-29       NA 5.49137e-49\n","rs7955986         NA       NA          NA          NA       NA          NA\n","rs718389          NA       NA          NA          NA       NA          NA\n","rs11571383        NA       NA          NA          NA       NA          NA\n","rs2107614         NA       NA          NA          NA       NA          NA\n","rs4980935         NA       NA          NA          NA       NA          NA\n","rs11064536        NA       NA          NA          NA       NA          NA\n","rs117766797       NA       NA          NA          NA       NA          NA\n","               pvalue_7 pvalue_8    pvalue_9 pvalue_10 pvalue_11   pvalue_12\n","rs147560086          NA       NA 6.71399e-10        NA        NA 4.40477e-16\n","rs7955986            NA       NA          NA        NA        NA          NA\n","rs718389             NA       NA          NA        NA        NA          NA\n","rs11571383  2.64972e-06       NA          NA        NA        NA          NA\n","rs2107614            NA       NA          NA        NA        NA          NA\n","rs4980935            NA       NA          NA        NA        NA          NA\n","rs11064536           NA       NA          NA        NA        NA          NA\n","rs117766797          NA       NA          NA        NA        NA          NA\n","            pvalue_13 pvalue_14 pvalue_15  pvalue_16 pvalue_17  pvalue_18\n","rs147560086        NA        NA        NA         NA        NA 0.00050353\n","rs7955986          NA        NA        NA         NA        NA         NA\n","rs718389           NA        NA        NA         NA        NA         NA\n","rs11571383         NA        NA        NA         NA        NA         NA\n","rs2107614          NA        NA        NA 1.0067e-07        NA         NA\n","rs4980935          NA        NA        NA         NA        NA         NA\n","rs11064536         NA        NA        NA         NA        NA         NA\n","rs117766797        NA        NA        NA         NA        NA         NA\n","              pvalue_19   pvalue_20   pvalue_21 pvalue_22   pvalue_23 pvalue_24\n","rs147560086          NA 8.26297e-28 3.16236e-42        NA 9.45416e-23        NA\n","rs7955986            NA          NA          NA        NA          NA        NA\n","rs718389             NA          NA          NA        NA          NA        NA\n","rs11571383           NA          NA          NA        NA          NA        NA\n","rs2107614            NA          NA          NA        NA          NA        NA\n","rs4980935   8.06751e-05          NA          NA        NA          NA        NA\n","rs11064536           NA          NA          NA        NA          NA        NA\n","rs117766797          NA          NA          NA        NA          NA        NA\n","              pvalue_25 pvalue_26 pvalue_27 pvalue_28   pvalue_29   pvalue_30\n","rs147560086 3.44003e-22        NA        NA        NA 3.44995e-23          NA\n","rs7955986            NA        NA        NA        NA          NA          NA\n","rs718389             NA        NA        NA        NA          NA          NA\n","rs11571383           NA        NA        NA        NA          NA          NA\n","rs2107614            NA        NA        NA        NA          NA          NA\n","rs4980935            NA        NA        NA        NA          NA          NA\n","rs11064536           NA        NA        NA        NA          NA 0.000347708\n","rs117766797          NA        NA        NA        NA          NA          NA\n","            pvalue_31 pvalue_32 pvalue_33  pvalue_34   pvalue_35 pvalue_36\n","rs147560086        NA        NA        NA 2.6301e-33 4.95626e-40        NA\n","rs7955986          NA        NA        NA         NA          NA        NA\n","rs718389           NA        NA        NA         NA          NA        NA\n","rs11571383         NA        NA        NA         NA          NA        NA\n","rs2107614          NA        NA        NA         NA          NA        NA\n","rs4980935          NA        NA        NA         NA          NA        NA\n","rs11064536         NA        NA        NA         NA          NA        NA\n","rs117766797        NA        NA        NA         NA          NA        NA\n","            pvalue_37 pvalue_38   pvalue_39 pvalue_40 pvalue_41   pvalue_42\n","rs147560086        NA        NA 8.63893e-13        NA        NA          NA\n","rs7955986          NA        NA          NA        NA        NA          NA\n","rs718389           NA        NA          NA        NA        NA          NA\n","rs11571383         NA        NA          NA        NA        NA 2.07196e-10\n","rs2107614          NA        NA          NA        NA        NA          NA\n","rs4980935          NA        NA          NA        NA        NA          NA\n","rs11064536         NA        NA          NA        NA        NA          NA\n","rs117766797        NA        NA          NA        NA        NA          NA\n","            pvalue_43   pvalue_44   pvalue_45   pvalue_46 pvalue_47   pvalue_48\n","rs147560086        NA 8.32165e-14          NA          NA        NA          NA\n","rs7955986          NA          NA          NA 2.21429e-32        NA          NA\n","rs718389           NA          NA 1.07802e-13          NA        NA          NA\n","rs11571383         NA          NA          NA          NA        NA          NA\n","rs2107614          NA          NA          NA          NA        NA          NA\n","rs4980935          NA          NA          NA          NA        NA          NA\n","rs11064536         NA          NA          NA          NA        NA          NA\n","rs117766797        NA          NA          NA          NA        NA 0.000357146\n","            pvalue_49\n","rs147560086        NA\n","rs7955986          NA\n","rs718389           NA\n","rs11571383         NA\n","rs2107614          NA\n","rs4980935          NA\n","rs11064536         NA\n","rs117766797        NA\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n"]}]},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(base) # R package installed by default\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(Matrix) # R package installed by default\n","library(methods) # R package installed by default\n","library(stats) # R package installed by default\n","library(lme4) # It needs to be installed separately\n","library(data.table) # It needs to be installed separately\n","\n","# Step 2: Main Mt_Robin function:\n","MR_MtRobin <- function(MR_MtRobin_input,pval_thresh=0.001){\n","  snpID <- MR_MtRobin_input$snpID\n","  gwas_betas <- MR_MtRobin_input$gwas_betas\n","  gwas_se <- MR_MtRobin_input$gwas_se\n","  eqtl_betas <- MR_MtRobin_input$eqtl_betas\n","  eqtl_se <- MR_MtRobin_input$eqtl_se\n","  eqtl_pvals <- MR_MtRobin_input$eqtl_pvals\n","  LD <- MR_MtRobin_input$LD\n","\n","  if(ncol(eqtl_betas) != ncol(eqtl_se) | ncol(eqtl_betas) != ncol(eqtl_pvals)){\n","    stop(\"eQTL betas, SE and p-values do not all have same number of columns\")\n","  }\n","\n","  ## determine number of tissues (or conditions/studies/etc.)\n","  nT <- ncol(eqtl_betas)\n","\n","  ## rename columns for standardized processing\n","  colnames(eqtl_betas) <- paste0(\"beta_\",1:nT)\n","  colnames(eqtl_se) <- paste0(\"SE_\",1:nT)\n","  colnames(eqtl_pvals) <- paste0(\"pvalue_\",1:nT)\n","\n","  ## combine eQTL results into data.table\n","  eqtl_res <- data.table::data.table(snpID,eqtl_betas,eqtl_se,eqtl_pvals)\n","\n","  # reshape to long format (each SNP-gene-tissue will be row)\n","  eQTL_res_melt <- data.table::melt(eqtl_res,id.vars=c(\"snpID\"),\n","                                    measure.vars=patterns(\"^beta_\",\"^SE_\",\"^pvalue_\"),value.name=c(\"beta\",\"se\",\"pvalue\"),\n","                                    variable.name=\"tissue\")\n","\n","  ## subset to strong instrumental variables (based on p-value threshold)\n","  eQTL_res_melt_PltThresh <- subset(eQTL_res_melt, pvalue < pval_thresh)\n","\n","  ## set up gwas data for merging\n","  gwas_res <- data.table::data.table(snpID=snpID,gwas_beta=gwas_betas,gwas_se=gwas_se)\n","\n","  ## merge eQTL and GWAS results\n","  data.table::setkey(eQTL_res_melt_PltThresh,snpID)\n","  data.table::setkey(gwas_res,snpID)\n","  merged_res <- data.table::merge.data.table(eQTL_res_melt_PltThresh,gwas_res)\n","\n","  ## set up coefficients for reverse regression\n","  beta_x <- matrix(merged_res$beta,ncol=1)\n","  beta_y <- matrix(merged_res$gwas_beta,ncol=1)\n","\n","  ## standard errors (for weights)\n","  se_x <- matrix(merged_res$se,ncol=1)\n","  se_y <- matrix(merged_res$gwas_se,ncol=1) ## not used by function; used in resampling (consider returning with results)\n","\n","  ## identifiers\n","  snpID <- merged_res$snpID\n","\n","  lme_res <- lme4::lmer(beta_x~(beta_y-1)+(beta_y-1|snpID),weights=1/se_x^2)\n","\n","  ## return results from weighted regression with random slopes (and correlated errors) along with GWAS SE and LD (needed for resampling)\n","  return(list(lme_res=lme_res,gwas_res=gwas_res,LD=LD))\n","}\n","\n","# Step 3: Read the previous file:\n","result_setup <- readRDS(\"/data/long_covid_setup_fs_4.rds\")\n","\n","# Step 4: Apply the main Mt_Robin function using parallel processing:\n","no_cores <- 38\n","\n","cl <- makeCluster(no_cores) # Create a cluster with the number of cores\n","clusterExport(cl, \"MR_MtRobin\") # Export the MR_MtRobin function to all the nodes of the cluster\n","result_list_MtRobin <- parLapply(cl, result_setup, function(MR_MtRobin_input) {\n","  tryCatch({\n","    MR_MtRobin(MR_MtRobin_input = MR_MtRobin_input)\n","  }, error = function(e) print(e))\n","})\n","stopCluster(cl) # Stop the cluster once the computation is done\n","\n","# Step 5: The names of the results will correspond to the gene ids:\n","names(result_list_MtRobin) <- names(result_setup)\n","\n","# Step 6: Save the previous result as a .rds file:\n","saveRDS(result_list_MtRobin, file = \"/data/long_covid_main_fs_4.rds\")"],"metadata":{"id":"hez_WFuDNWtU"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/main_function.sif Rscript /data/main_function_fs_4.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"WvZIPsioNXmq"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 17:34:46:\n","   Job Id:             93443499.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      2.73\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 00:02:56\n","   Memory Requested:   300.0GB               Memory Used: 9.71GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:01:31\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"3A5GUg5_NZaP"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_main_fs_4 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_4.rds\")\n","\n","# Filter out errors:\n","result_main_fs_4 <- result_main_fs_4[!sapply(result_main_fs_4, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_main_fs_4))\n","print(length(unique(result_main_fs_4)))\n","print(head(result_main_fs_4))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mKLEidMsNbEn","executionInfo":{"status":"ok","timestamp":1699424775296,"user_tz":-630,"elapsed":10508,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"7b6282e1-1072-44ed-e95b-f1f4b4c16501"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1724\n","[1] 1724\n","$ENSG00000002016\n","$ENSG00000002016$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -29.6022\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 22.469  \n"," Residual         1.836  \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," 8.039  \n","\n","$ENSG00000002016$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1  rs11064536  0.01503010 0.0315410\n","2  rs11571383 -0.05198390 0.0617035\n","3 rs117766797 -0.00869216 0.0734673\n","4 rs147560086  0.03468750 0.0287461\n","5   rs2107614 -0.00444987 0.0245610\n","6   rs4980935 -0.00825649 0.0261202\n","7    rs718389 -0.03680740 0.0272009\n","8   rs7955986 -0.02506380 0.0233427\n","\n","$ENSG00000002016$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","\n","$ENSG00000005156\n","$ENSG00000005156$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -12.2531\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 3.441   \n"," Residual        3.261   \n","Number of obs: 10, groups:  snpID, 3\n","Fixed Effects:\n","beta_y  \n"," 9.272  \n","\n","$ENSG00000005156$gwas_res\n","        snpID  gwas_beta   gwas_se\n","1  rs12945428 -0.0378619 0.0223006\n","2 rs201016783  0.1432490 0.0547025\n","3   rs3135958  0.0314358 0.0355257\n","\n","$ENSG00000005156$LD\n","             rs12945428    rs3135958  rs201016783\n","rs12945428   1.00000000 -0.295076381  0.020183627\n","rs3135958   -0.29507638  1.000000000 -0.006054471\n","rs201016783  0.02018363 -0.006054471  1.000000000\n","\n","\n","$ENSG00000006282\n","$ENSG00000006282$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -28.6714\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 156.988 \n"," Residual          5.685 \n","Number of obs: 46, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n"," 6.366  \n","\n","$ENSG00000006282$gwas_res\n","      snpID   gwas_beta   gwas_se\n","1 rs2306001 -0.00152670 0.0369697\n","2 rs9890200  0.00847768 0.0237429\n","\n","$ENSG00000006282$LD\n","          rs9890200 rs2306001\n","rs9890200 1.0000000 0.3686704\n","rs2306001 0.3686704 1.0000000\n","\n","\n","$ENSG00000006744\n","$ENSG00000006744$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -2.4272\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 8.392   \n"," Residual        6.335   \n","Number of obs: 18, groups:  snpID, 4\n","Fixed Effects:\n","beta_y  \n","0.8209  \n","\n","$ENSG00000006744$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1  rs1044564  0.0489453 0.0229255\n","2 rs11655929  0.0438521 0.0233704\n","3   rs733890 -0.0408626 0.0236182\n","4 rs78840287 -0.0166731 0.0761180\n","\n","$ENSG00000006744$LD\n","              rs1044564    rs733890 rs11655929   rs78840287\n","rs1044564   1.000000000  0.66545097 -0.6626427 -0.007565868\n","rs733890    0.665450966  1.00000000 -0.4997189  0.027633741\n","rs11655929 -0.662642701 -0.49971889  1.0000000  0.034843798\n","rs78840287 -0.007565868  0.02763374  0.0348438  1.000000000\n","\n","\n","$ENSG00000007171\n","$ENSG00000007171$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 19.4874\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 206.624 \n"," Residual          4.283 \n","Number of obs: 31, groups:  snpID, 5\n","Fixed Effects:\n","beta_y  \n","-77.48  \n","\n","$ENSG00000007171$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1  rs2314813 -0.00906011 0.0295782\n","2  rs3751972  0.00186134 0.0262152\n","3   rs624196  0.04924490 0.0250560\n","4 rs72846720  0.07148670 0.0564068\n","5  rs8078340 -0.01735770 0.0338649\n","\n","$ENSG00000007171$LD\n","              rs3751972   rs8078340    rs624196    rs2314813  rs72846720\n","rs3751972   1.000000000 -0.20973763 -0.03796279  0.004244027 -0.08622187\n","rs8078340  -0.209737626  1.00000000  0.03757975  0.160333611 -0.06812959\n","rs624196   -0.037962786  0.03757975  1.00000000 -0.039583359  0.20176556\n","rs2314813   0.004244027  0.16033361 -0.03958336  1.000000000 -0.10409667\n","rs72846720 -0.086221874 -0.06812959  0.20176556 -0.104096673  1.00000000\n","\n","\n","$ENSG00000007341\n","$ENSG00000007341$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -1.8099\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 0.000   \n"," Residual        7.379   \n","Number of obs: 12, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n","-19.68  \n","optimizer (nloptwrap) convergence code: 0 (OK) ; 0 optimizer warnings; 1 lme4 warnings \n","\n","$ENSG00000007341$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1 rs10745330 -0.0264908 0.0223919\n","2   rs954679  0.0409259 0.0230021\n","\n","$ENSG00000007341$LD\n","           rs10745330  rs954679\n","rs10745330  1.0000000 0.6167231\n","rs954679    0.6167231 1.0000000\n","\n","\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_fs_4.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_4_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_fs_4_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_fs4.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"2FUsFxxFPULQ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Option 1: Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_fs_4.R\n","\n","# Option 2: Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"kIADdXLTP4sr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-08 12:27:08:\n","   Job Id:             107797348.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      446.90\n","   NCPUs Requested:    24                     NCPUs Used: 24\n","                                           CPU Time Used: 131:46:19\n","   Memory Requested:   200.0GB               Memory Used: 58.68GB\n","   Walltime requested: 10:00:00            Walltime Used: 06:12:25\n","   JobFS requested:    100.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"NqSpcYGJP5-D"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs4.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results))\n","print(head(results))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"y6GO9VyFQLJf","executionInfo":{"status":"ok","timestamp":1707356196348,"user_tz":-630,"elapsed":709,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"170f5626-6fb4-46e6-ea80-3f842dffd286"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 1724    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.404      10000 0.8213396\n","2 ENSG00000005156  0.141      10000 0.6241316\n","3 ENSG00000006282  0.988      10000 0.9976805\n","4 ENSG00000006744  0.931      10000 0.9817077\n","5 ENSG00000007171  0.514      10000 0.8611623\n","6 ENSG00000007341  0.071      10000 0.4684075\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_4.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"dZAKoU5xQIU4","executionInfo":{"status":"ok","timestamp":1707356305142,"user_tz":-630,"elapsed":98505,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"b3caee16-ec06-437f-e70f-f5281cc8db7c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 1724 10 \n","\n","          gene_id gene_name      gene_type      beta_y Obs SNPs\n","1 ENSG00000002016     RAD52 protein_coding   8.0391110  23    8\n","2 ENSG00000005156      LIG3 protein_coding   9.2715580  10    3\n","3 ENSG00000006282   SPATA20 protein_coding   6.3657734  46    2\n","4 ENSG00000006744     ELAC2 protein_coding   0.8208629  18    4\n","5 ENSG00000007171      NOS2 protein_coding -77.4824431  31    5\n","6 ENSG00000007341      ST7L protein_coding -19.6826570  12    1\n","                                                                               SNP_IDs\n","1 rs11064536,rs11571383,rs117766797,rs147560086,rs2107614,rs4980935,rs718389,rs7955986\n","2                                                     rs12945428,rs201016783,rs3135958\n","3                                                                  rs2306001,rs9890200\n","4                                             rs1044564,rs11655929,rs733890,rs78840287\n","5                                    rs2314813,rs3751972,rs624196,rs72846720,rs8078340\n","6                                                                  rs10745330,rs954679\n","        REML         SD Residual_SD\n","1 -29.602237  22.469403    1.835676\n","2 -12.253073   3.440798    3.260563\n","3 -28.671406 156.988429    5.684947\n","4  -2.427229   8.391515    6.335188\n","5  19.487357 206.624246    4.282789\n","6  -1.809889   0.000000    7.378557\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_protein_coding_genes_sig_not_sig.txt \n","Dimensions of protein_coding_genes : 775 10 \n","\n","          gene_id gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000103351    CLUAP1 protein_coding -2337.9091  48    2\n","2 ENSG00000237541  HLA-DQA2 protein_coding   837.0902   9    2\n","3 ENSG00000162994     CLHC1 protein_coding  -623.6641  40    2\n","4 ENSG00000166435     XRRA1 protein_coding  -500.2091  19    2\n","5 ENSG00000107959    PITRM1 protein_coding  -431.0385  30    6\n","6 ENSG00000166275    BORCS7 protein_coding  -419.5582  35    1\n","                                                        SNP_IDs       REML\n","1                                          rs67733706,rs8054140  -2.228960\n","2                                           rs9272362,rs9274554 -11.459793\n","3                                          rs1437391,rs17046778   3.335454\n","4                                         rs146081906,rs2165162 -45.650082\n","5 rs3740612,rs4880597,rs4880598,rs61837966,rs76088740,rs9423701  23.216029\n","6                                          rs11191421,rs7896519 -35.605323\n","         SD Residual_SD\n","1 3266.9578    3.935313\n","2 1105.4211    3.166610\n","3  869.1305    4.429066\n","4  682.4462    2.416524\n","5  732.3133    2.964579\n","6    0.0000    3.823279\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# FS4 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_fs_4_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_fs4.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"h_6LZ3-vQIU4","executionInfo":{"status":"ok","timestamp":1707356331414,"user_tz":-630,"elapsed":26276,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"03ac8064-f220-4e96-d827-7f55b65f70a8"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1724\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -29.6022\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 22.469  \n"," Residual         1.836  \n","Number of obs: 23, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," 8.039  \n","\n","$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1  rs11064536  0.01503010 0.0315410\n","2  rs11571383 -0.05198390 0.0617035\n","3 rs117766797 -0.00869216 0.0734673\n","4 rs147560086  0.03468750 0.0287461\n","5   rs2107614 -0.00444987 0.0245610\n","6   rs4980935 -0.00825649 0.0261202\n","7    rs718389 -0.03680740 0.0272009\n","8   rs7955986 -0.02506380 0.0233427\n","\n","$LD\n","            rs147560086   rs7955986    rs718389  rs11571383   rs2107614\n","rs147560086  1.00000000  0.57848413 -0.19239180  0.66694928 -0.29808391\n","rs7955986    0.57848413  1.00000000 -0.35065828  0.41826699 -0.19482499\n","rs718389    -0.19239180 -0.35065828  1.00000000  0.03742702  0.47355084\n","rs11571383   0.66694928  0.41826699  0.03742702  1.00000000 -0.18100269\n","rs2107614   -0.29808391 -0.19482499  0.47355084 -0.18100269  1.00000000\n","rs4980935   -0.06971404 -0.07400750  0.05993201 -0.03527063  0.02505299\n","rs11064536  -0.20068438 -0.31351943  0.43005106 -0.14059500  0.54526711\n","rs117766797  0.08540422  0.03130702 -0.03831378  0.08108809 -0.04931958\n","              rs4980935  rs11064536 rs117766797\n","rs147560086 -0.06971404 -0.20068438  0.08540422\n","rs7955986   -0.07400750 -0.31351943  0.03130702\n","rs718389     0.05993201  0.43005106 -0.03831378\n","rs11571383  -0.03527063 -0.14059500  0.08108809\n","rs2107614    0.02505299  0.54526711 -0.04931958\n","rs4980935    1.00000000  0.04499302 -0.01181004\n","rs11064536   0.04499302  1.00000000  0.04037438\n","rs117766797 -0.01181004  0.04037438  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 775  10\n","          gene_id gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000103351    CLUAP1 protein_coding -2337.9091  48    2\n","2 ENSG00000237541  HLA-DQA2 protein_coding   837.0902   9    2\n","3 ENSG00000162994     CLHC1 protein_coding  -623.6641  40    2\n","4 ENSG00000166435     XRRA1 protein_coding  -500.2091  19    2\n","5 ENSG00000107959    PITRM1 protein_coding  -431.0385  30    6\n","6 ENSG00000166275    BORCS7 protein_coding  -419.5582  35    1\n","                                                        SNP_IDs       REML\n","1                                          rs67733706,rs8054140  -2.228960\n","2                                           rs9272362,rs9274554 -11.459793\n","3                                          rs1437391,rs17046778   3.335454\n","4                                         rs146081906,rs2165162 -45.650082\n","5 rs3740612,rs4880597,rs4880598,rs61837966,rs76088740,rs9423701  23.216029\n","6                                          rs11191421,rs7896519 -35.605323\n","         SD Residual_SD\n","1 3266.9578    3.935313\n","2 1105.4211    3.166610\n","3  869.1305    4.429066\n","4  682.4462    2.416524\n","5  732.3133    2.964579\n","6    0.0000    3.823279\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 1724    4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000002016  0.404      10000 0.8213396\n","2 ENSG00000005156  0.141      10000 0.6241316\n","3 ENSG00000006282  0.988      10000 0.9976805\n","4 ENSG00000006744  0.931      10000 0.9817077\n","5 ENSG00000007171  0.514      10000 0.8611623\n","6 ENSG00000007341  0.071      10000 0.4684075\n","[1] \"Final table (protein coding genes): \"\n","[1] 24 22\n","           gene_id gene_name SNPs\n","1  ENSG00000089335    ZNF302    2\n","2  ENSG00000171160     MORN4    2\n","3  ENSG00000149089      APIP    4\n","4  ENSG00000205078    SYCE1L    3\n","5  ENSG00000171928    TVP23B    5\n","6  ENSG00000171495    MROH2B    5\n","7  ENSG00000212127   TAS2R14    5\n","8  ENSG00000127399    LRRC61    2\n","9  ENSG00000163161     ERCC3    5\n","10 ENSG00000197013    ZNF429   10\n","11 ENSG00000164002      EXO5   10\n","12 ENSG00000138381    ASNSD1    5\n","13 ENSG00000177082     WDR73   13\n","14 ENSG00000125818     PSMF1    9\n","15 ENSG00000170915     PAQR8   24\n","16 ENSG00000139714     MORN3   18\n","17 ENSG00000108439      PNPO   15\n","18 ENSG00000068097    HEATR6    9\n","19 ENSG00000213246   SUPT4H1   12\n","20 ENSG00000089063   TMEM230    6\n","21 ENSG00000138495     COX17   18\n","22 ENSG00000132305      IMMT    7\n","23 ENSG00000177409    SAMD9L    9\n","24 ENSG00000215251   FASTKD5   22\n","                                                                                                                                                                                                                                                         SNPs_IDs\n","1                                                                                                                                                                                                                                           rs10411245,rs12984680\n","2                                                                                                                                                                                                                                            rs12358715,rs7905087\n","3                                                                                                                                                                                                                      rs11032943,rs11500295,rs2915228,rs74344381\n","4                                                                                                                                                                                                                                     rs13331769,rs6564410,rs8398\n","5                                                                                                                                                                                                            rs10153316,rs10852854,rs2472694,rs56156915,rs8081355\n","6                                                                                                                                                                                                            rs144186443,rs1450664,rs2329437,rs62361684,rs7704504\n","7                                                                                                                                                                                                            rs10772397,rs3851583,rs4418898,rs73046589,rs79686186\n","8                                                                                                                                                                                                                                            rs10247016,rs1962004\n","9                                                                                                                                                                                                            rs11688763,rs4150476,rs56369397,rs7568261,rs76693372\n","10                                                                                                                                                    rs111883431,rs148721166,rs2681382,rs28898308,rs4808279,rs62111479,rs6511254,rs7255865,rs73029504,rs78951000\n","11                                                                                                                                                      rs10157181,rs141713679,rs2025898,rs2810584,rs2982502,rs34762587,rs4562610,rs4660419,rs72941545,rs75831615\n","12                                                                                                                                                                                                          rs1132360,rs1233258,rs145793357,rs75105901,rs77445225\n","13                                                                                                                    rs11259919,rs12900358,rs138567963,rs141005445,rs17532346,rs1960770,rs2137270,rs55702300,rs55736017,rs56053835,rs6496090,rs77590624,rs893162\n","14                                                                                                                                                                   rs17716261,rs200929224,rs2072963,rs4473461,rs4813030,rs6074148,rs6077678,rs6514113,rs7260693\n","15 rs12191519,rs12216249,rs137994093,rs143803123,rs144276820,rs150818476,rs17246535,rs2144700,rs4276501,rs4712001,rs4715302,rs537288,rs55706232,rs567084,rs57078108,rs655113,rs6927270,rs74542499,rs75614796,rs75731219,rs77855336,rs78833977,rs9367480,rs9382100\n","16                                                          rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","17                                                                                                      rs13353189,rs17250967,rs17617250,rs1986694,rs2001380,rs2002136,rs2325749,rs2737,rs35614613,rs35653190,rs4793671,rs62065840,rs6504745,rs76098486,rs9903904\n","18                                                                                                                                                                 rs1078378,rs10853017,rs117054060,rs345172,rs7210383,rs72831252,rs76439721,rs79483284,rs9912341\n","19                                                                                                                                    rs11656836,rs13414,rs2585845,rs2680702,rs33963229,rs35104446,rs35205300,rs3785494,rs72831237,rs76728508,rs8078707,rs9905604\n","20                                                                                                                                                                                                   rs118000137,rs1741303,rs261350,rs261367,rs4315602,rs73610161\n","21                                                                rs13076819,rs13095802,rs138002173,rs139795340,rs140197056,rs141938700,rs148919020,rs1500082,rs1797827,rs201745461,rs2733787,rs34321347,rs4687869,rs6768044,rs8007,rs9828996,rs9835803,rs9860092\n","22                                                                                                                                                                                        rs10210905,rs111227732,rs2218035,rs4832257,rs6749439,rs6750263,rs719023\n","23                                                                                                                                                             rs10236444,rs117613553,rs16868514,rs200034704,rs34050019,rs4729089,rs6956577,rs75504870,rs80278808\n","24                             rs1008405,rs10720,rs13042500,rs138333601,rs145060338,rs148434527,rs17222895,rs241602,rs2422728,rs2422906,rs56258620,rs6037486,rs6049283,rs6049382,rs6133015,rs678963,rs74272866,rs75181535,rs76360286,rs763996,rs77477947,rs857253\n","   Obs     beta_y        Min    First_Q      Median     Third_Q       Max\n","1    4 108.395616 -1.2310657 -0.1421597  0.26583595 0.522444259 1.1571883\n","2    3  44.842821 -0.6522094 -0.4840181 -0.31582692 0.449308962 1.2144448\n","3   25  42.901637 -2.5971741 -0.2874125  0.16352816 0.564183317 1.4560425\n","4   26  21.828754 -1.8127381 -0.6512678 -0.13397391 0.882081584 1.8445297\n","5    7 -20.165050 -1.2883320 -0.5453597 -0.01873335 0.723422169 0.9528934\n","6    6 -18.198135 -0.6787485 -0.2435237  0.58920470 1.197002465 1.3294060\n","7   34  16.985311 -2.2725631 -0.4371969 -0.01349638 0.523851749 2.7735361\n","8   23  16.860555 -1.4953306 -0.4520904  0.27959911 0.689543612 2.1201987\n","9    8  16.849151 -1.5671358 -0.5853061 -0.25068861 0.924882090 1.0652016\n","10  15 -12.661000 -1.2273008 -0.5769362 -0.25686372 0.178671819 2.4603151\n","11  18 -11.518163 -1.7549619 -0.3773185  0.36994740 0.888245500 1.5766878\n","12  26  11.205282 -1.5214965 -0.4094987  0.28041614 0.996589762 1.8997320\n","13  14 -10.785899 -1.1658313 -0.5732226  0.09554979 0.829307627 2.3436740\n","14  14  -9.738410 -1.8610767 -0.4268299  0.37750067 0.883871589 1.4624853\n","15  36  -8.439665 -1.8449843 -0.4933283  0.02499944 0.691036283 2.7145070\n","16  21   6.975234 -1.6672811 -0.3596133 -0.06280085 1.143162329 1.8670928\n","17  18   6.496182 -2.7618740 -0.7146895 -0.19984635 0.634666119 1.6543852\n","18  25   6.355959 -1.5105865 -0.4444658  0.07962101 0.745696539 2.4227008\n","19  29  -5.408228 -1.6044894 -0.8995017 -0.46439995 0.008495112 2.4786752\n","20  34   5.247125 -1.7091221 -0.4243438  0.01723788 0.496954524 3.8883121\n","21  25   5.088389 -1.7566601 -1.1275672  0.06629142 0.568414845 1.7199696\n","22  14  -4.884465 -2.0566415 -0.5117281 -0.19905661 0.289248582 1.6242742\n","23  23  -4.688474 -1.2296495 -0.4059432  0.24317127 0.631188719 3.0688620\n","24  29   2.666513 -1.8493992 -0.8639803 -0.07187184 0.847729187 1.5292046\n","        SD_RE Res_SD_RE       Var_RE Res_Var_RE      SE_FE t_value_FE\n","1   0.0000000  1.926134   0.00000000   3.709993  6.6732088  16.243402\n","2   0.0000000  4.277636   0.00000000  18.298171  6.7643603   6.629277\n","3  26.3011962  4.405787 691.75292020  19.410960 13.9934867   3.065829\n","4   0.0000000  4.226442   0.00000000  17.862811  0.9888432  22.075041\n","5   3.4552061  2.003381  11.93844897   4.013536  2.3639139  -8.530366\n","6   0.1065803  1.342744   0.01135936   1.802961  1.3552533 -13.427848\n","7   0.0000000  3.148761   0.00000000   9.914696  0.6981423  24.329297\n","8   0.0000000  4.570423   0.00000000  20.888766  0.9588012  17.585037\n","9   0.0000000  3.108471   0.00000000   9.662590  1.4556541  11.574969\n","10  0.0000000  3.809331   0.00000000  14.511005  1.2173741 -10.400254\n","11  0.0000000  2.713818   0.00000000   7.364806  0.9341322 -12.330335\n","12  0.0000000  3.724465   0.00000000  13.871641  0.6650530  16.848706\n","13  0.0000000  2.790834   0.00000000   7.788756  1.4729943  -7.322431\n","14  0.0000000  5.770784   0.00000000  33.301950  1.8820247  -5.174432\n","15  1.9578708  4.166621   3.83325793  17.360734  1.4010448  -6.023836\n","16  0.0000000  4.030495   0.00000000  16.244889  1.5652179   4.456398\n","17  0.0000000  3.651913   0.00000000  13.336471  1.0831416   5.997537\n","18  0.0000000  3.362791   0.00000000  11.308361  0.4732972  13.429107\n","19  0.0000000  3.552700   0.00000000  12.621681  0.6671984  -8.105877\n","20  0.0000000  3.534380   0.00000000  12.491840  0.2477485  21.179244\n","21  0.0000000  3.686559   0.00000000  13.590718  0.6943118   7.328680\n","22  0.0000000  3.891861   0.00000000  15.146579  0.5218604  -9.359716\n","23  0.3656696  2.699301   0.13371425   7.286227  0.4220885 -11.107798\n","24  0.0000000  3.220701   0.00000000  10.372915  0.5138098   5.189688\n","          REML p_value   FDR   CI_lower   CI_upper\n","1  -14.4559687   0.000 0.000  95.316126 121.475105\n","2  -10.2341487   0.000 0.000  31.584675  58.100967\n","3   -3.2137862   0.001 0.041  15.474403  70.328871\n","4   -9.9302505   0.000 0.000  19.890622  23.766887\n","5  -17.2250862   0.001 0.041 -24.798321 -15.531778\n","6  -11.6432848   0.000 0.000 -20.854431 -15.541838\n","7  -33.0009402   0.000 0.000  15.616952  18.353670\n","8   -8.3556151   0.000 0.000  14.981305  18.739805\n","9   -8.5320920   0.000 0.000  13.996069  19.702233\n","10  -4.9744345   0.000 0.000 -15.047053 -10.274947\n","11 -15.4116479   0.000 0.000 -13.349062  -9.687264\n","12 -24.7466893   0.000 0.000   9.901778  12.508786\n","13  -2.1065426   0.000 0.000 -13.672968  -7.898831\n","14  -1.4322585   0.000 0.000 -13.427178  -6.049641\n","15   5.3031360   0.001 0.041 -11.185712  -5.693617\n","16  18.1949029   0.000 0.000   3.907407  10.043061\n","17 -15.4714937   0.000 0.000   4.373224   8.619139\n","18   2.6226949   0.000 0.000   5.428297   7.283622\n","19 -23.7918077   0.001 0.041  -6.715937  -4.100520\n","20 -48.6626875   0.000 0.000   4.761538   5.732712\n","21   0.8595335   0.000 0.000   3.727538   6.449240\n","22 -15.6043413   0.000 0.000  -5.907311  -3.861618\n","23   3.6010951   0.000 0.000  -5.515767  -3.861180\n","24 -15.1311012   0.001 0.041   1.659445   3.673580\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the last function to retrieve the tissue information and categories for fs1:\n","input_eqtl_path <- \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\"\n","input_mt_robin_path <- \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_protein_coding_genes_sig_full_table.txt\"\n","output_path <- \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_protein_coding_genes_sig_full_table_tissues.txt\"\n","process_eqtl_mt_robin(input_eqtl_path, input_mt_robin_path, output_path)"],"metadata":{"id":"hHrWzv32QIU4","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714699241644,"user_tz":-570,"elapsed":54108,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"63a0c15f-cc0f-424d-fb11-ca538256c8b6"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["           gene_id gene_name SNPs\n","1  ENSG00000068097    HEATR6    9\n","2  ENSG00000089063   TMEM230    6\n","3  ENSG00000089335    ZNF302    2\n","4  ENSG00000108439      PNPO   15\n","5  ENSG00000125818     PSMF1    9\n","6  ENSG00000127399    LRRC61    2\n","7  ENSG00000132305      IMMT    7\n","8  ENSG00000138381    ASNSD1    5\n","9  ENSG00000138495     COX17   18\n","10 ENSG00000139714     MORN3   18\n","11 ENSG00000149089      APIP    4\n","12 ENSG00000163161     ERCC3    5\n","13 ENSG00000164002      EXO5   10\n","14 ENSG00000170915     PAQR8   24\n","15 ENSG00000171160     MORN4    2\n","16 ENSG00000171495    MROH2B    5\n","17 ENSG00000171928    TVP23B    5\n","18 ENSG00000177082     WDR73   13\n","19 ENSG00000177409    SAMD9L    9\n","20 ENSG00000197013    ZNF429   10\n","21 ENSG00000205078    SYCE1L    3\n","22 ENSG00000212127   TAS2R14    5\n","23 ENSG00000213246   SUPT4H1   12\n","24 ENSG00000215251   FASTKD5   22\n","                                                                                                                                                                                                                                                         SNPs_IDs\n","1                                                                                                                                                                  rs1078378,rs10853017,rs117054060,rs345172,rs7210383,rs72831252,rs76439721,rs79483284,rs9912341\n","2                                                                                                                                                                                                    rs118000137,rs1741303,rs261350,rs261367,rs4315602,rs73610161\n","3                                                                                                                                                                                                                                           rs10411245,rs12984680\n","4                                                                                                       rs13353189,rs17250967,rs17617250,rs1986694,rs2001380,rs2002136,rs2325749,rs2737,rs35614613,rs35653190,rs4793671,rs62065840,rs6504745,rs76098486,rs9903904\n","5                                                                                                                                                                    rs17716261,rs200929224,rs2072963,rs4473461,rs4813030,rs6074148,rs6077678,rs6514113,rs7260693\n","6                                                                                                                                                                                                                                            rs10247016,rs1962004\n","7                                                                                                                                                                                         rs10210905,rs111227732,rs2218035,rs4832257,rs6749439,rs6750263,rs719023\n","8                                                                                                                                                                                                           rs1132360,rs1233258,rs145793357,rs75105901,rs77445225\n","9                                                                 rs13076819,rs13095802,rs138002173,rs139795340,rs140197056,rs141938700,rs148919020,rs1500082,rs1797827,rs201745461,rs2733787,rs34321347,rs4687869,rs6768044,rs8007,rs9828996,rs9835803,rs9860092\n","10                                                          rs10522094,rs112537090,rs113467236,rs11502393,rs138851658,rs144181641,rs146337807,rs148311487,rs148754753,rs208291,rs2263271,rs2458404,rs28421662,rs58423269,rs63496760,rs7313015,rs7486943,rs7957569\n","11                                                                                                                                                                                                                     rs11032943,rs11500295,rs2915228,rs74344381\n","12                                                                                                                                                                                                           rs11688763,rs4150476,rs56369397,rs7568261,rs76693372\n","13                                                                                                                                                      rs10157181,rs141713679,rs2025898,rs2810584,rs2982502,rs34762587,rs4562610,rs4660419,rs72941545,rs75831615\n","14 rs12191519,rs12216249,rs137994093,rs143803123,rs144276820,rs150818476,rs17246535,rs2144700,rs4276501,rs4712001,rs4715302,rs537288,rs55706232,rs567084,rs57078108,rs655113,rs6927270,rs74542499,rs75614796,rs75731219,rs77855336,rs78833977,rs9367480,rs9382100\n","15                                                                                                                                                                                                                                           rs12358715,rs7905087\n","16                                                                                                                                                                                                           rs144186443,rs1450664,rs2329437,rs62361684,rs7704504\n","17                                                                                                                                                                                                           rs10153316,rs10852854,rs2472694,rs56156915,rs8081355\n","18                                                                                                                    rs11259919,rs12900358,rs138567963,rs141005445,rs17532346,rs1960770,rs2137270,rs55702300,rs55736017,rs56053835,rs6496090,rs77590624,rs893162\n","19                                                                                                                                                             rs10236444,rs117613553,rs16868514,rs200034704,rs34050019,rs4729089,rs6956577,rs75504870,rs80278808\n","20                                                                                                                                                    rs111883431,rs148721166,rs2681382,rs28898308,rs4808279,rs62111479,rs6511254,rs7255865,rs73029504,rs78951000\n","21                                                                                                                                                                                                                                    rs13331769,rs6564410,rs8398\n","22                                                                                                                                                                                                           rs10772397,rs3851583,rs4418898,rs73046589,rs79686186\n","23                                                                                                                                    rs11656836,rs13414,rs2585845,rs2680702,rs33963229,rs35104446,rs35205300,rs3785494,rs72831237,rs76728508,rs8078707,rs9905604\n","24                             rs1008405,rs10720,rs13042500,rs138333601,rs145060338,rs148434527,rs17222895,rs241602,rs2422728,rs2422906,rs56258620,rs6037486,rs6049283,rs6049382,rs6133015,rs678963,rs74272866,rs75181535,rs76360286,rs763996,rs77477947,rs857253\n","   Obs     beta_y        Min    First_Q      Median     Third_Q       Max\n","1   25   6.355959 -1.5105865 -0.4444658  0.07962101 0.745696539 2.4227008\n","2   34   5.247125 -1.7091221 -0.4243438  0.01723788 0.496954524 3.8883121\n","3    4 108.395616 -1.2310657 -0.1421597  0.26583595 0.522444259 1.1571883\n","4   18   6.496182 -2.7618740 -0.7146895 -0.19984635 0.634666119 1.6543852\n","5   14  -9.738410 -1.8610767 -0.4268299  0.37750067 0.883871589 1.4624853\n","6   23  16.860555 -1.4953306 -0.4520904  0.27959911 0.689543612 2.1201987\n","7   14  -4.884465 -2.0566415 -0.5117281 -0.19905661 0.289248582 1.6242742\n","8   26  11.205282 -1.5214965 -0.4094987  0.28041614 0.996589762 1.8997320\n","9   25   5.088389 -1.7566601 -1.1275672  0.06629142 0.568414845 1.7199696\n","10  21   6.975234 -1.6672811 -0.3596133 -0.06280085 1.143162329 1.8670928\n","11  25  42.901637 -2.5971741 -0.2874125  0.16352816 0.564183317 1.4560425\n","12   8  16.849151 -1.5671358 -0.5853061 -0.25068861 0.924882090 1.0652016\n","13  18 -11.518163 -1.7549619 -0.3773185  0.36994740 0.888245500 1.5766878\n","14  36  -8.439665 -1.8449843 -0.4933283  0.02499944 0.691036283 2.7145070\n","15   3  44.842821 -0.6522094 -0.4840181 -0.31582692 0.449308962 1.2144448\n","16   6 -18.198135 -0.6787485 -0.2435237  0.58920470 1.197002465 1.3294060\n","17   7 -20.165050 -1.2883320 -0.5453597 -0.01873335 0.723422169 0.9528934\n","18  14 -10.785899 -1.1658313 -0.5732226  0.09554979 0.829307627 2.3436740\n","19  23  -4.688474 -1.2296495 -0.4059432  0.24317127 0.631188719 3.0688620\n","20  15 -12.661000 -1.2273008 -0.5769362 -0.25686372 0.178671819 2.4603151\n","21  26  21.828754 -1.8127381 -0.6512678 -0.13397391 0.882081584 1.8445297\n","22  34  16.985311 -2.2725631 -0.4371969 -0.01349638 0.523851749 2.7735361\n","23  29  -5.408228 -1.6044894 -0.8995017 -0.46439995 0.008495112 2.4786752\n","24  29   2.666513 -1.8493992 -0.8639803 -0.07187184 0.847729187 1.5292046\n","        SD_RE Res_SD_RE       Var_RE Res_Var_RE      SE_FE t_value_FE\n","1   0.0000000  3.362791   0.00000000  11.308361  0.4732972  13.429107\n","2   0.0000000  3.534380   0.00000000  12.491840  0.2477485  21.179244\n","3   0.0000000  1.926134   0.00000000   3.709993  6.6732088  16.243402\n","4   0.0000000  3.651913   0.00000000  13.336471  1.0831416   5.997537\n","5   0.0000000  5.770784   0.00000000  33.301950  1.8820247  -5.174432\n","6   0.0000000  4.570423   0.00000000  20.888766  0.9588012  17.585037\n","7   0.0000000  3.891861   0.00000000  15.146579  0.5218604  -9.359716\n","8   0.0000000  3.724465   0.00000000  13.871641  0.6650530  16.848706\n","9   0.0000000  3.686559   0.00000000  13.590718  0.6943118   7.328680\n","10  0.0000000  4.030495   0.00000000  16.244889  1.5652179   4.456398\n","11 26.3011962  4.405787 691.75292020  19.410960 13.9934867   3.065829\n","12  0.0000000  3.108471   0.00000000   9.662590  1.4556541  11.574969\n","13  0.0000000  2.713818   0.00000000   7.364806  0.9341322 -12.330335\n","14  1.9578708  4.166621   3.83325793  17.360734  1.4010448  -6.023836\n","15  0.0000000  4.277636   0.00000000  18.298171  6.7643603   6.629277\n","16  0.1065803  1.342744   0.01135936   1.802961  1.3552533 -13.427848\n","17  3.4552061  2.003381  11.93844897   4.013536  2.3639139  -8.530366\n","18  0.0000000  2.790834   0.00000000   7.788756  1.4729943  -7.322431\n","19  0.3656696  2.699301   0.13371425   7.286227  0.4220885 -11.107798\n","20  0.0000000  3.809331   0.00000000  14.511005  1.2173741 -10.400254\n","21  0.0000000  4.226442   0.00000000  17.862811  0.9888432  22.075041\n","22  0.0000000  3.148761   0.00000000   9.914696  0.6981423  24.329297\n","23  0.0000000  3.552700   0.00000000  12.621681  0.6671984  -8.105877\n","24  0.0000000  3.220701   0.00000000  10.372915  0.5138098   5.189688\n","          REML p_value   FDR   CI_lower   CI_upper\n","1    2.6226949   0.000 0.000   5.428297   7.283622\n","2  -48.6626875   0.000 0.000   4.761538   5.732712\n","3  -14.4559687   0.000 0.000  95.316126 121.475105\n","4  -15.4714937   0.000 0.000   4.373224   8.619139\n","5   -1.4322585   0.000 0.000 -13.427178  -6.049641\n","6   -8.3556151   0.000 0.000  14.981305  18.739805\n","7  -15.6043413   0.000 0.000  -5.907311  -3.861618\n","8  -24.7466893   0.000 0.000   9.901778  12.508786\n","9    0.8595335   0.000 0.000   3.727538   6.449240\n","10  18.1949029   0.000 0.000   3.907407  10.043061\n","11  -3.2137862   0.001 0.041  15.474403  70.328871\n","12  -8.5320920   0.000 0.000  13.996069  19.702233\n","13 -15.4116479   0.000 0.000 -13.349062  -9.687264\n","14   5.3031360   0.001 0.041 -11.185712  -5.693617\n","15 -10.2341487   0.000 0.000  31.584675  58.100967\n","16 -11.6432848   0.000 0.000 -20.854431 -15.541838\n","17 -17.2250862   0.001 0.041 -24.798321 -15.531778\n","18  -2.1065426   0.000 0.000 -13.672968  -7.898831\n","19   3.6010951   0.000 0.000  -5.515767  -3.861180\n","20  -4.9744345   0.000 0.000 -15.047053 -10.274947\n","21  -9.9302505   0.000 0.000  19.890622  23.766887\n","22 -33.0009402   0.000 0.000  15.616952  18.353670\n","23 -23.7918077   0.001 0.041  -6.715937  -4.100520\n","24 -15.1311012   0.001 0.041   1.659445   3.673580\n","                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         Tissues\n","1                                                                                                                                                                                                Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Coronary artery,Anterior cingulate cortex,Cortex,Frontal cortex,Spinal cord,Breast mammary tissue,EBV-transformed lymphocytes,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Liver,Minor salivary gland,Skeletal muscle,Ovary,Pancreas,Prostate,Not sun-exposed skin (suprapubic),Small intestine terminal ileum,Stomach,Testis,Uterus,Whole blood\n","2                                                          Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Aorta artery,Coronary artery,Anterior cingulate cortex,Cortex,Hypothalamus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Cultured fibroblasts,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Kidney cortex,Lung,Minor salivary gland,Tibial nerve,Ovary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Spleen,Stomach,Testis,Thyroid,Uterus,Vagina,Whole blood\n","3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Coronary artery,Spinal cord,Esophagus mucosa,Esophagus muscularis\n","4                                                                                                                                                                                                                                                                                    Aorta artery,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Left ventricle,Kidney cortex,Liver,Skeletal muscle,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Thyroid,Uterus\n","5                                                                                                                                                                                                                                                                                                                                                                                                                         Visceral omentum adipose,Tibial artery,Amygdala brain,Cortex,Hippocampus,Spinal cord,Substantia nigra,Cultured fibroblasts,EBV-transformed lymphocytes,Tibial nerve,Pancreas,Testis,Vagina,Whole blood\n","6                                                                                                                                                                                                                Visceral omentum adipose,Adrenal gland,Aorta artery,Tibial artery,Anterior cingulate cortex,Cerebellar hemisphere,Cerebellum,Cortex,Hippocampus,Hypothalamus,Nucleus accumbens (basal ganglia),Substantia nigra,Cultured fibroblasts,Sigmoid colon,Gastroesophageal junction,Atrial appendage,Lung,Skeletal muscle,Tibial nerve,Not sun-exposed skin (suprapubic),Small intestine terminal ileum,Thyroid,Uterus\n","7                                                                                                                                                                                                                                                                                                                                                  Subcutaneous adipose,Adrenal gland,Coronary artery,Amygdala brain,Anterior cingulate cortex,Hypothalamus,Putamen (basal ganglia),Transverse colon,Gastroesophageal junction,Esophagus muscularis,Liver,Ovary,Not sun-exposed skin (suprapubic),Small intestine terminal ileum\n","8                                                                                                                                                                                                  Subcutaneous adipose,Tibial artery,Amygdala brain,Anterior cingulate cortex,Caudate,Cerebellum,Cortex,Hippocampus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Substantia nigra,Breast mammary tissue,Sigmoid colon,Esophagus mucosa,Esophagus muscularis,Kidney cortex,Liver,Lung,Skeletal muscle,Tibial nerve,Pituitary,Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Stomach,Whole blood\n","9                                                                                                                                                                                                         Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Aorta artery,Amygdala brain,Anterior cingulate cortex,Cortex,Hippocampus,Hypothalamus,Substantia nigra,Cultured fibroblasts,Sigmoid colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Left ventricle,Kidney cortex,Liver,Minor salivary gland,Pancreas,Sun-exposed skin (lower leg),Spleen,Thyroid,Uterus,Whole blood\n","10                                                                                                                                                                                                                                                           Visceral omentum adipose,Adrenal gland,Tibial artery,Cerebellum,Nucleus accumbens (basal ganglia),Spinal cord,Breast mammary tissue,EBV-transformed lymphocytes,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Kidney cortex,Liver,Minor salivary gland,Pancreas,Prostate,Sun-exposed skin (lower leg),Thyroid,Vagina,Whole blood\n","11                                                                                                                                                                                                              Subcutaneous adipose,Aorta artery,Coronary artery,Tibial artery,Amygdala brain,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Left ventricle,Tibial nerve,Ovary,Prostate,Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Testis,Thyroid,Uterus,Vagina\n","12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Amygdala brain,Cerebellum,Atrial appendage,Tibial nerve,Small intestine terminal ileum,Testis,Thyroid,Vagina\n","13                                                                                                                                                                                                                                                                                                                                                                       Subcutaneous adipose,Adrenal gland,Tibial artery,Amygdala brain,Caudate,Cerebellum,Frontal cortex,Hypothalamus,Spinal cord,Cultured fibroblasts,Kidney cortex,Lung,Tibial nerve,Pituitary,Not sun-exposed skin (suprapubic),Stomach,Thyroid,Whole blood\n","14 Subcutaneous adipose,Visceral omentum adipose,Adrenal gland,Aorta artery,Coronary artery,Amygdala brain,Anterior cingulate cortex,Cerebellar hemisphere,Cerebellum,Hypothalamus,Nucleus accumbens (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Atrial appendage,Left ventricle,Kidney cortex,Lung,Minor salivary gland,Ovary,Pancreas,Pituitary,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Spleen,Testis,Thyroid,Uterus,Whole blood\n","15                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Cultured fibroblasts,Left ventricle,Thyroid\n","16                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Visceral omentum adipose,Sigmoid colon,Transverse colon,Liver,Small intestine terminal ileum,Testis\n","17                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             Anterior cingulate cortex,Caudate,Cultured fibroblasts,Esophagus muscularis,Stomach,Testis,Vagina\n","18                                                                                                                                                                                                                                                                                                                                                                         Coronary artery,Cerebellar hemisphere,Cerebellum,Hypothalamus,Putamen (basal ganglia),EBV-transformed lymphocytes,Esophagus mucosa,Kidney cortex,Liver,Pituitary,Not sun-exposed skin (suprapubic),Small intestine terminal ileum,Thyroid,Whole blood\n","19                                                                                                                                                                                                                                  Subcutaneous adipose,Visceral omentum adipose,Tibial artery,Cerebellar hemisphere,Cerebellum,Hypothalamus,Putamen (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Gastroesophageal junction,Esophagus mucosa,Kidney cortex,Liver,Lung,Tibial nerve,Prostate,Not sun-exposed skin (suprapubic),Spleen,Thyroid,Whole blood\n","20                                                                                                                                                                                                                                                                                                                                                                             Aorta artery,Tibial artery,Amygdala brain,Frontal cortex,Hippocampus,Putamen (basal ganglia),Substantia nigra,Sigmoid colon,Gastroesophageal junction,Esophagus muscularis,Atrial appendage,Liver,Sun-exposed skin (lower leg),Testis,Whole blood\n","21                                                                                                                                                                                     Subcutaneous adipose,Aorta artery,Coronary artery,Tibial artery,Caudate,Cerebellum,Cortex,Hippocampus,Hypothalamus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Transverse colon,Esophagus mucosa,Esophagus muscularis,Left ventricle,Lung,Skeletal muscle,Tibial nerve,Not sun-exposed skin (suprapubic),Testis,Thyroid,Whole blood\n","22                                                                    Adrenal gland,Aorta artery,Tibial artery,Anterior cingulate cortex,Cerebellar hemisphere,Cerebellum,Cortex,Hippocampus,Nucleus accumbens (basal ganglia),Spinal cord,Substantia nigra,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Liver,Lung,Skeletal muscle,Tibial nerve,Ovary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Stomach,Testis,Thyroid,Uterus,Vagina,Whole blood\n","23                                                                                                 Visceral omentum adipose,Adrenal gland,Aorta artery,Coronary artery,Amygdala brain,Caudate,Cerebellar hemisphere,Cortex,Hippocampus,Hypothalamus,Nucleus accumbens (basal ganglia),Spinal cord,Breast mammary tissue,Cultured fibroblasts,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus muscularis,Atrial appendage,Skeletal muscle,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Small intestine terminal ileum,Spleen,Thyroid,Vagina,Whole blood\n","24                                                                                                Subcutaneous adipose,Visceral omentum adipose,Aorta artery,Coronary artery,Anterior cingulate cortex,Cerebellar hemisphere,Cerebellum,Cortex,Hippocampus,Nucleus accumbens (basal ganglia),Putamen (basal ganglia),Substantia nigra,Breast mammary tissue,EBV-transformed lymphocytes,Sigmoid colon,Transverse colon,Gastroesophageal junction,Esophagus mucosa,Esophagus muscularis,Kidney cortex,Liver,Lung,Tibial nerve,Ovary,Pituitary,Prostate,Not sun-exposed skin (suprapubic),Sun-exposed skin (lower leg),Whole blood\n","                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Tissue_Categories\n","1                                                                                                                                                                                                                              Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Exocrine System,Muscular System,Reproductive System,Digestive System,Reproductive System,Skin,Digestive System,Digestive System,Reproductive System,Reproductive System,Blood\n","2                                Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Respiratory System,Exocrine System,Brain and Nervous System,Reproductive System,Reproductive System,Skin,Skin,Immune System,Digestive System,Reproductive System,Endocrine System,Reproductive System,Reproductive System,Blood\n","3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Cardiovascular System,Brain and Nervous System,Digestive System,Digestive System\n","4                                                                                                                                                                                                                                                                                                                                                                         Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Digestive System,Muscular System,Reproductive System,Skin,Skin,Endocrine System,Reproductive System\n","5                                                                                                                                                                                                                                                                                                                                                                                                     Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Immune System,Brain and Nervous System,Digestive System,Reproductive System,Reproductive System,Blood\n","6                                                                                                                                                                                                  Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Digestive System,Cardiovascular System,Respiratory System,Muscular System,Brain and Nervous System,Skin,Digestive System,Endocrine System,Reproductive System\n","7                                                                                                                                                                                                                                                                                                                                                                                                               Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Digestive System,Digestive System,Digestive System,Digestive System,Reproductive System,Skin,Digestive System\n","8                                                                                                                                                                            Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Digestive System,Digestive System,Digestive System,Excretory System,Digestive System,Respiratory System,Muscular System,Brain and Nervous System,Endocrine System,Skin,Digestive System,Immune System,Digestive System,Blood\n","9                                                                                                                                                                                         Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Cardiovascular System,Excretory System,Digestive System,Exocrine System,Digestive System,Skin,Immune System,Endocrine System,Reproductive System,Blood\n","10                                                                                                                                                                                                                                                                                                       Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Excretory System,Digestive System,Exocrine System,Digestive System,Reproductive System,Skin,Endocrine System,Reproductive System,Blood\n","11                                                                                                                                                                                                                 Adipose Tissue,Cardiovascular System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Reproductive System,Reproductive System,Skin,Digestive System,Immune System,Reproductive System,Endocrine System,Reproductive System,Reproductive System\n","12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Brain and Nervous System,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Digestive System,Reproductive System,Endocrine System,Reproductive System\n","13                                                                                                                                                                                                                                                                                                                             Adipose Tissue,Brain and Nervous System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Connective Tissue,Excretory System,Respiratory System,Brain and Nervous System,Endocrine System,Skin,Digestive System,Endocrine System,Blood\n","14 Adipose Tissue,Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Cardiovascular System,Excretory System,Respiratory System,Exocrine System,Reproductive System,Digestive System,Endocrine System,Skin,Skin,Immune System,Reproductive System,Endocrine System,Reproductive System,Blood\n","15                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Connective Tissue,Cardiovascular System,Endocrine System\n","16                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Adipose Tissue,Digestive System,Digestive System,Digestive System,Digestive System,Reproductive System\n","17                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Brain and Nervous System,Brain and Nervous System,Connective Tissue,Digestive System,Digestive System,Reproductive System,Reproductive System\n","18                                                                                                                                                                                                                                                                                                                                                                                                                                     Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Immune System,Digestive System,Excretory System,Digestive System,Endocrine System,Skin,Digestive System,Endocrine System,Blood\n","19                                                                                                                                                                                                                                                      Adipose Tissue,Adipose Tissue,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Excretory System,Digestive System,Respiratory System,Brain and Nervous System,Reproductive System,Skin,Immune System,Endocrine System,Blood\n","20                                                                                                                                                                                                                                                                                                                                                                                            Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Digestive System,Skin,Reproductive System,Blood\n","21                                                                                                                                                                Adipose Tissue,Cardiovascular System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Respiratory System,Muscular System,Brain and Nervous System,Skin,Reproductive System,Endocrine System,Blood\n","22                           Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Respiratory System,Muscular System,Brain and Nervous System,Reproductive System,Reproductive System,Skin,Skin,Digestive System,Immune System,Digestive System,Reproductive System,Endocrine System,Reproductive System,Reproductive System,Blood\n","23                                                                                                                                Adipose Tissue,Brain and Nervous System,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Connective Tissue,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Cardiovascular System,Muscular System,Reproductive System,Skin,Skin,Digestive System,Immune System,Endocrine System,Reproductive System,Blood\n","24                                                                                                                                  Adipose Tissue,Adipose Tissue,Cardiovascular System,Cardiovascular System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Brain and Nervous System,Exocrine System,Immune System,Digestive System,Digestive System,Digestive System,Digestive System,Digestive System,Excretory System,Digestive System,Respiratory System,Brain and Nervous System,Reproductive System,Endocrine System,Reproductive System,Skin,Skin,Blood\n"]}]},{"cell_type":"markdown","source":["### **All GWAS**"],"metadata":{"id":"zaL5l8kgkYYG"}},{"cell_type":"code","source":["%%R\n","\n","# Install and load dplyr:\n","#install.packages(\"dplyr\")\n","#library(dplyr)\n","\n","# Install and load purrr:\n","#install.packages(\"purrr\")\n","#library(purrr)\n","\n","# Read the GWAS datasets\n","gwas1 <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs1_protein_coding_genes_sig_full_table_tissues.txt\", header=TRUE, sep=\"\\t\")\n","gwas2 <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs2_protein_coding_genes_sig_full_table_tissues.txt\", header=TRUE, sep=\"\\t\")\n","gwas3 <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs3_protein_coding_genes_sig_full_table_tissues.txt\", header=TRUE, sep=\"\\t\")\n","gwas4 <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/fs4_protein_coding_genes_sig_full_table_tissues.txt\", header=TRUE, sep=\"\\t\")\n","\n","# Combine all tables into one, allowing for repeated gene_ids if other values vary:\n","all_gwas <- bind_rows(gwas1, gwas2, gwas3, gwas4)\n","\n","# Count duplicates for gene_id:\n","gene_id_counts <- all_gwas %>%\n","  group_by(gene_id) %>%\n","  summarise(count = n(), .groups = 'drop')\n","duplicates_only <- filter(gene_id_counts, count > 1)\n","print(duplicates_only)\n","\n","# Print the dimensions and the head of the combined dataset:\n","print(\"Combined causal genes with repeated gene_id:\")\n","glimpse(all_gwas)\n","\n","# Export combined datasets:\n","write.table(all_gwas, \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/all_gwas_causal_repeated_genes.txt\", sep=\"\\t\", row.names=FALSE, col.names=TRUE, quote=FALSE)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SVdzkSZZjnwA","executionInfo":{"status":"ok","timestamp":1714707056364,"user_tz":-570,"elapsed":294,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"ddeb1d90-5d2f-4db2-f417-9b2615a12ca2"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["# A tibble: 15 × 2\n","   gene_id         count\n","   <chr>           <int>\n"," 1 ENSG00000068097     2\n"," 2 ENSG00000089063     2\n"," 3 ENSG00000090661     2\n"," 4 ENSG00000108439     2\n"," 5 ENSG00000127399     2\n"," 6 ENSG00000139714     4\n"," 7 ENSG00000146530     2\n"," 8 ENSG00000149089     2\n"," 9 ENSG00000158825     2\n","10 ENSG00000163161     2\n","11 ENSG00000171160     2\n","12 ENSG00000177082     2\n","13 ENSG00000184983     2\n","14 ENSG00000205078     2\n","15 ENSG00000212127     2\n","[1] \"Combined causal genes with repeated gene_id:\"\n","Rows: 67\n","Columns: 24\n","$ gene_id           <chr> \"ENSG00000026950\", \"ENSG00000065457\", \"ENSG000000906…\n","$ gene_name         <chr> \"BTN3A1\", \"ADAT1\", \"CERS4\", \"BNIP1\", \"EIF5A\", \"MORN3…\n","$ SNPs              <int> 9, 4, 15, 7, 2, 18, 8, 14, 2, 2, 2, 17, 3, 2, 9, 6, …\n","$ SNPs_IDs          <chr> \"rs114157910,rs12213056,rs13194491,rs2275906,rs32087…\n","$ Obs               <int> 22, 11, 23, 13, 20, 21, 12, 26, 3, 18, 18, 20, 21, 2…\n","$ beta_y            <dbl> -4.817862, -7.713004, -7.804575, 8.994436, 21.013660…\n","$ Min               <dbl> -2.4698318, -1.5988537, -1.9381654, -2.0411354, -1.3…\n","$ First_Q           <dbl> -0.12749636, -0.53363730, -0.03024777, -0.57137796, …\n","$ Median            <dbl> 0.14242212, -0.28594064, 0.45603996, 0.13557018, -0.…\n","$ Third_Q           <dbl> 0.64935373, 0.12998014, 0.79856650, 0.57073516, 0.22…\n","$ Max               <dbl> 2.1276346, 2.3368205, 2.7709937, 1.3174713, 2.561890…\n","$ SD_RE             <dbl> 0.000000e+00, 0.000000e+00, 0.000000e+00, 0.000000e+…\n","$ Res_SD_RE         <dbl> 2.530855, 3.225459, 6.604208, 3.182195, 2.500070, 3.…\n","$ Var_RE            <dbl> 0.000000e+00, 0.000000e+00, 0.000000e+00, 0.000000e+…\n","$ Res_Var_RE        <dbl> 6.405225, 10.403585, 43.615568, 10.126362, 6.250350,…\n","$ SE_FE             <dbl> 0.3331290, 0.4917438, 1.1491150, 1.0026723, 0.962549…\n","$ t_value_FE        <dbl> -14.462452, -15.685005, -6.791814, 8.970464, 21.8312…\n","$ REML              <dbl> -28.354550, -12.929202, 18.165023, -4.170912, -57.88…\n","$ p_value           <dbl> 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.0…\n","$ FDR               <dbl> 0.00000000, 0.00000000, 0.00000000, 0.00000000, 0.00…\n","$ CI_lower          <dbl> -5.470795, -8.676822, -10.056841, 7.029199, 19.12706…\n","$ CI_upper          <dbl> -4.164929, -6.749186, -5.552310, 10.959674, 22.90025…\n","$ Tissues           <chr> \"Subcutaneous adipose,Aorta artery,Coronary artery,A…\n","$ Tissue_Categories <chr> \"Adipose Tissue,Cardiovascular System,Cardiovascular…\n"]}]},{"cell_type":"markdown","source":["## **backward_selection**"],"metadata":{"id":"Irw3ijKRciNT"}},{"cell_type":"markdown","source":["### **GWAS1**"],"metadata":{"id":"yK6pFEdJdMLT"}},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(matrixStats) # It needs to be installed separately\n","#library(BiocManager) # It needs to be installed separately\n","library(snpStats) # It needs to be installed separately\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(reshape2) # Additional package for melt function\n","\n","# Step 2: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 3: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_1_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 4: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 5: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 6: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 7: Create a variable for the path to the directory to save the LD matrix results:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_1\"\n","\n","# Step 8: Create the directory if it doesn't exist:\n","if (!dir.exists(ld_matrix_dir)) {\n","  dir.create(ld_matrix_dir)\n","}\n","\n","# Step 9: Modify the original select_IV_fs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_bs() --> backward selection starts with all variables and removes one at a time:\n","select_IV_bs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset the eQTL data set for the gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Get pairwise r^2:\n","  LD_r2_pw <- LD_r2\n","  LD_r2_pw[lower.tri(LD_r2_pw,diag=T)] <- NA\n","  LD_r2_pw <- reshape2::melt(LD_r2_pw)\n","  colnames(LD_r2_pw) <- c(\"SNP1\",\"SNP2\",\"r2\")\n","  LD_r2_pw$SNP1 <- as.character(LD_r2_pw$SNP1)\n","  LD_r2_pw$SNP2 <- as.character(LD_r2_pw$SNP2)\n","\n","  ## Drop SNPs paired with themselves and duplicates (resulting from lower triangle, set to NA):\n","  LD_r2_pw <- subset(LD_r2_pw, SNP1 != SNP2 & !is.na(r2))\n","\n","  ## Sort by descending r^2:\n","  LD_r2_pw <- LD_r2_pw[order(LD_r2_pw$r2,decreasing = T),]\n","\n","  ## Check for case all candidate SNPs have pw r^2 > thresh (return SNPs with smallest p-value):\n","  if(min(LD_r2_pw$r2) >=ld_thresh){\n","    return(eqtl_data$variant_id[which.max(eqtl_data$min_p)])\n","  }\n","\n","  while(max(LD_r2_pw$r2)>=ld_thresh){\n","\n","    ## Identify top SNPs in the pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    LD_r2_pw <- subset(LD_r2_pw, SNP1 %in% SNP_pool & SNP2 %in% SNP_pool)\n","    LD_r2_pw_geThresh <- subset(LD_r2_pw, r2>=ld_thresh)\n","\n","    ## Identify SNP pair with highest remaining r^2:\n","    next_pair <- LD_r2_pw[1,]\n","\n","    ## Criteria 1: Drop the SNP with largest number of correlations > threshold with other SNPs:\n","    SNP1_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP1) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP1)\n","    SNP2_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP2) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP2)\n","    if(SNP1_nSNP_geThresh > SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP1\n","    } else if (SNP1_nSNP_geThresh < SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP2\n","    } else{\n","\n","      ## Criteria 2: Drop the SNP with smaller number of tissues with p<0.001:\n","      next_pair_eqtl <- subset(eqtl_data, variant_id %in% c(next_pair$SNP1,next_pair$SNP2))\n","      SNP1_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP1)$nTiss_ltThresh\n","      SNP2_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP2)$nTiss_ltThresh\n","\n","      if(SNP1_nTiss > SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP2\n","      } else if(SNP1_nTiss < SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP1\n","      } else{\n","\n","        ## Criteria 3: Drop the SNP with larger minimum p-value:\n","        next_SNP_drop <- next_pair_eqtl$variant_id[which.max(next_pair_eqtl$min_p)]\n","      }\n","    }\n","    SNP_pool <- SNP_pool[which(SNP_pool != next_SNP_drop)]\n","  }\n","  return(SNP_pool)\n","}\n","\n","# Step 10: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 11:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","  # Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_fs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","  }, error = function(e) print(e))\n","}, mc.cores = num_cores)\n","\n","# Step 12:\n","names(result_list) <- all_genes_vector\n","\n","# Step 13:\n","saveRDS(result_list, file = \"/data/long_covid_IVs_bs_1.rds\")"],"metadata":{"id":"lF9gop-lc7cT"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup_bs.sif Rscript /data/select_function_bs_1.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"oeSjZyP9cn2H"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-24 23:39:34:\n","   Job Id:             93444366.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      866.28\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 277:35:00\n","   Memory Requested:   500.0GB               Memory Used: 388.77GB\n","   Walltime requested: 30:00:00            Walltime Used: 06:00:57\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"Ooj9nzONdGSj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_selectIV_bs_1 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_IVs_bs_1.rds\")\n","\n","# Filter out errors:\n","result_selectIV_bs_1 <- result_selectIV_bs_1[!sapply(result_selectIV_bs_1, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_selectIV_bs_1))\n","print(length(unique(result_selectIV_bs_1)))\n","print(head(result_selectIV_bs_1))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"rufH2MHddTKM","executionInfo":{"status":"ok","timestamp":1693365679672,"user_tz":-570,"elapsed":1765,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"97fb573a-21ad-4342-80bc-477bb9297019"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 13809\n","[1] 13737\n","$ENSG00000002079\n","[1] \"rs36168039\"  \"rs151109487\" \"rs41258334\"  \"rs28497153\"  \"rs7457787\"  \n","[6] \"rs73157578\" \n","\n","$ENSG00000004846\n"," [1] \"rs58540449\"  \"rs79448643\"  \"rs114103500\" \"rs111613242\" \"rs76325633\" \n"," [6] \"rs117337555\" \"rs73682327\"  \"rs11981446\"  \"rs7801178\"   \"rs17455533\" \n","[11] \"rs2108258\"   \"rs77254950\"  \"rs139614792\"\n","\n","$ENSG00000004939\n"," [1] \"rs2880712\"   \"rs9747287\"   \"rs117705553\" \"rs79555595\"  \"rs78230173\" \n"," [6] \"rs114651584\" \"rs34637106\"  \"rs7213794\"   \"rs144535065\" \"rs142313238\"\n","[11] \"rs76004777\"  \"rs9944407\"   \"rs34260468\"  \"rs142437085\" \"rs8070867\"  \n","[16] \"rs730818\"    \"rs200108084\" \"rs1728174\"   \"rs140638631\" \"rs4793150\"  \n","[21] \"rs76480869\"  \"rs147752346\" \"rs62080669\"  \"rs9895862\"   \"rs7222401\"  \n","[26] \"rs540551010\" \"rs9890834\"   \"rs76498976\"  \"rs138050677\"\n","\n","$ENSG00000005073\n"," [1] \"rs60642981\"  \"rs2525760\"   \"rs10273618\"  \"rs4722617\"   \"rs138820646\"\n"," [6] \"rs117987507\" \"rs17501111\"  \"rs10251237\"  \"rs147214544\" \"rs6947348\"  \n","[11] \"rs71539542\"  \"rs3735570\"   \"rs71521781\"  \"rs201583312\" \"rs17153901\" \n","[16] \"rs11975130\"  \"rs17153696\"  \"rs112074869\"\n","\n","$ENSG00000005961\n"," [1] \"rs3024272\"   \"rs11867362\"  \"rs74456742\"  \"rs143274822\" \"rs145006591\"\n"," [6] \"rs147997011\" \"rs62640053\"  \"rs4793096\"   \"rs9908627\"   \"rs34001726\" \n","[11] \"rs575873\"    \"rs111516722\" \"rs74679777\"  \"rs79337476\"  \"rs139626122\"\n","[16] \"rs7223523\"   \"rs57934049\"  \"rs72833109\"  \"rs11574815\"  \"rs66838809\" \n","[21] \"rs577074947\" \"rs144701507\" \"rs2003467\"   \"rs113149437\" \"rs148582322\"\n","[26] \"rs12373134\"  \"rs9898671\"   \"rs117944176\" \"rs7214661\"   \"rs765149778\"\n","[31] \"rs118013624\" \"rs17882143\"  \"rs12940906\"  \"rs879835344\" \"rs73318401\" \n","[36] \"rs228766\"    \"rs112435729\" \"rs111420989\" \"rs181883277\" \"rs850733\"   \n","[41] \"rs17846855\" \n","\n","$ENSG00000006059\n","[1] \"rs73301945\"  \"rs8079152\"   \"rs180938378\" \"rs35074247\"  \"rs148663776\"\n","[6] \"rs200586273\"\n","\n"]}]},{"cell_type":"code","source":["library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_1_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the list of genes-SNPs:\n","result_selectedIVs <- readRDS(\"/data/long_covid_IVs_bs_1.rds\")\n","\n","# Step 4: Create a variable for the path to the LD matrix directory:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_1\"\n","\n","# Step 5: Setup the Mt_Robin method using the following setup function:\n","MR_MtRobin_setup <- function(geneID,\n","                             snpID,\n","                             eqtl_data,\n","                             gwas_data,\n","                             ld_matrix_path,\n","                             nTiss){\n","  tryCatch({\n","    ## Check if the gene is present and subset\n","    if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","    eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","    ## Confirm all SNPs are present in the dataset\n","    if(!all(snpID %in% eqtl_data$variant_id)) stop(\"Some SNPs missing in eQTL dataset\")\n","    if(!all(snpID %in% gwas_data$variant_id)) stop(\"Some SNPs missing in GWAS dataset\")\n","\n","    # Read the LD matrix from a file\n","    LD <- as.matrix(utils::read.table(ld_matrix_path,\n","                                      header = TRUE,\n","                                      check.names = FALSE))\n","\n","    if(!all(snpID %in% colnames(LD))) stop(\"Some SNPs missing in LD matrix\")\n","\n","    ## Align data\n","    eqtl_data <- eqtl_data[match(snpID,eqtl_data$variant_id),]\n","    gwas_data <- gwas_data[match(snpID,gwas_data$variant_id),]\n","    LD <- LD[match(snpID,rownames(LD)),match(snpID,colnames(LD))]\n","\n","    eqtl_betas <- as.matrix(eqtl_data[,paste0(\"beta_\",1:nTiss)])\n","    eqtl_se <- as.matrix(eqtl_data[,paste0(\"SE_\",1:nTiss)])\n","    eqtl_pvals <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","    gwas_betas <- gwas_data$beta\n","    gwas_se <- gwas_data$SE\n","\n","    row.names(eqtl_betas) <- row.names(eqtl_se) <- row.names(eqtl_pvals) <- eqtl_data$variant_id\n","\n","    return(list(snpID=snpID,gwas_betas=gwas_betas,gwas_se=gwas_se,\n","                eqtl_betas=eqtl_betas,eqtl_se=eqtl_se,eqtl_pvals=eqtl_pvals,LD=LD))\n","  }, error = function(e) print(e))\n","}\n","\n","# Step 6: Apply the previous setup function:\n","\n","# Step 6.1: Initialize an empty list to store the results:\n","MR_MtRobin_results <- list()\n","\n","# Step 6.2: Define the number of cores to use for the parallelism process:\n","num_cores <- 36\n","\n","# Step 6.3: Loop over each gene in result_selectedIVs:\n","MR_MtRobin_results <- mclapply(names(result_selectedIVs), function(geneID) {\n","\n","  # Step 6.3.1: Get the SNPs associated with each gene:\n","  snpID <- result_selectedIVs[[geneID]]\n","\n","  # Step 6.3.2: Generate a unique file path for the LD matrix of this gene:\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  # Step 6.3.3: Call the MR_MtRobin_setup() function with the parameters for each gene:\n","  return(tryCatch({\n","    MR_MtRobin_setup(geneID = geneID,\n","                     snpID = snpID,\n","                     eqtl_data = eqtl_covid_exposure,\n","                     gwas_data = gwas_covid_outcome,\n","                     ld_matrix_path = ld_matrix_path,\n","                     nTiss = 49)\n","  }, error = function(e) print(e)))\n","}, mc.cores = num_cores)\n","\n","# Step 7: Assign names to the list:\n","names(MR_MtRobin_results) <- names(result_selectedIVs)\n","\n","# Step 8: Save the previous result as a .rds file:\n","saveRDS(MR_MtRobin_results, file = \"/data/long_covid_setup_bs_1.rds\")"],"metadata":{"id":"tJP36QPB6NmG"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup_bs.sif Rscript /data/setup_function_bs_1.R\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"B-RQYWYm6Y5H"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-25 08:50:01:\n","   Job Id:             93510073.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      48.57\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 12:49:52\n","   Memory Requested:   300.0GB               Memory Used: 76.49GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:26:59\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"1ByZJ9Cu6ZuG"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_setup_bs_1 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_setup_bs_1.rds\")\n","\n","# Filter out errors:\n","result_setup_bs_1 <- result_setup_bs_1[!sapply(result_setup_bs_1, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_setup_bs_1))\n","print(length(unique(result_setup_bs_1)))\n","print(result_setup_bs_1[[1]])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ER00bx6m6din","executionInfo":{"status":"ok","timestamp":1693541166197,"user_tz":-570,"elapsed":15023,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"262e6141-2c48-48e8-b3c6-36eab40b7f5f"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 2635\n","[1] 2635\n","$snpID\n"," [1] \"rs2206271\"   \"rs35305915\"  \"rs72901027\"  \"rs9367397\"   \"rs61660288\" \n"," [6] \"rs141707622\" \"rs2477488\"   \"rs13202872\"  \"rs62406441\"  \"rs17667036\" \n","[11] \"rs75397901\"  \"rs10948573\"  \"rs4715208\"   \"rs571503\"   \n","\n","$gwas_betas\n"," [1]  0.05003660 -0.03369060 -0.00541575  0.04018010 -0.06373460  0.05997710\n"," [7]  0.05336300  0.02460510 -0.02075750  0.00391041 -0.06437320 -0.00690805\n","[13]  0.03944380 -0.00698292\n","\n","$gwas_se\n"," [1] 0.0318188 0.0446311 0.0464996 0.0627267 0.0469640 0.0901525 0.0297989\n"," [8] 0.0340206 0.0360749 0.0320274 0.1337120 0.0321628 0.0345040 0.0299366\n","\n","$eqtl_betas\n","              beta_1 beta_2 beta_3 beta_4 beta_5 beta_6 beta_7 beta_8 beta_9\n","rs2206271   0.160091     NA     NA     NA     NA     NA     NA     NA     NA\n","rs35305915        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs72901027        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs9367397         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs61660288        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs141707622       NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs2477488         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs13202872        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs62406441        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs17667036        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs75397901        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs10948573        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs4715208         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs571503          NA     NA     NA     NA     NA     NA     NA     NA     NA\n","             beta_10  beta_11 beta_12 beta_13 beta_14 beta_15 beta_16 beta_17\n","rs2206271         NA       NA      NA      NA      NA      NA      NA      NA\n","rs35305915  0.184274       NA      NA      NA      NA      NA      NA      NA\n","rs72901027        NA -0.26909      NA      NA      NA      NA      NA      NA\n","rs9367397         NA       NA      NA      NA      NA      NA      NA      NA\n","rs61660288        NA       NA      NA      NA      NA      NA      NA      NA\n","rs141707622       NA       NA      NA      NA      NA      NA      NA      NA\n","rs2477488         NA       NA      NA      NA      NA      NA      NA      NA\n","rs13202872        NA       NA      NA      NA      NA      NA      NA      NA\n","rs62406441        NA       NA      NA      NA      NA      NA      NA      NA\n","rs17667036        NA       NA      NA      NA      NA      NA      NA      NA\n","rs75397901        NA       NA      NA      NA      NA      NA      NA      NA\n","rs10948573        NA       NA      NA      NA      NA      NA      NA      NA\n","rs4715208         NA       NA      NA      NA      NA      NA      NA      NA\n","rs571503          NA       NA      NA      NA      NA      NA      NA      NA\n","             beta_18   beta_19  beta_20 beta_21 beta_22   beta_23 beta_24\n","rs2206271         NA        NA       NA      NA      NA        NA      NA\n","rs35305915        NA        NA       NA      NA      NA        NA      NA\n","rs72901027        NA        NA       NA      NA      NA        NA      NA\n","rs9367397   0.829401        NA       NA      NA      NA        NA      NA\n","rs61660288        NA -0.833804       NA      NA      NA        NA      NA\n","rs141707622       NA        NA -0.21223      NA      NA        NA      NA\n","rs2477488         NA        NA       NA      NA      NA -0.353026      NA\n","rs13202872        NA        NA       NA      NA      NA        NA      NA\n","rs62406441        NA        NA       NA      NA      NA        NA      NA\n","rs17667036        NA        NA       NA      NA      NA        NA      NA\n","rs75397901        NA        NA       NA      NA      NA        NA      NA\n","rs10948573        NA        NA       NA      NA      NA        NA      NA\n","rs4715208         NA        NA       NA      NA      NA        NA      NA\n","rs571503          NA        NA       NA      NA      NA        NA      NA\n","            beta_25   beta_26 beta_27 beta_28 beta_29  beta_30 beta_31 beta_32\n","rs2206271        NA        NA      NA      NA      NA       NA      NA      NA\n","rs35305915       NA        NA      NA      NA      NA       NA      NA      NA\n","rs72901027       NA        NA      NA      NA      NA       NA      NA      NA\n","rs9367397        NA        NA      NA      NA      NA       NA      NA      NA\n","rs61660288       NA        NA      NA      NA      NA       NA      NA      NA\n","rs141707622      NA        NA      NA      NA      NA       NA      NA      NA\n","rs2477488        NA        NA      NA      NA      NA       NA      NA      NA\n","rs13202872       NA -0.380743      NA      NA      NA       NA      NA      NA\n","rs62406441       NA        NA      NA      NA      NA 0.497656      NA      NA\n","rs17667036       NA        NA      NA      NA      NA       NA      NA      NA\n","rs75397901       NA        NA      NA      NA      NA       NA      NA      NA\n","rs10948573       NA        NA      NA      NA      NA       NA      NA      NA\n","rs4715208        NA        NA      NA      NA      NA       NA      NA      NA\n","rs571503         NA        NA      NA      NA      NA       NA      NA      NA\n","              beta_33 beta_34  beta_35 beta_36 beta_37 beta_38 beta_39\n","rs2206271          NA      NA 0.225935      NA      NA      NA      NA\n","rs35305915         NA      NA       NA      NA      NA      NA      NA\n","rs72901027         NA      NA       NA      NA      NA      NA      NA\n","rs9367397          NA      NA       NA      NA      NA      NA      NA\n","rs61660288         NA      NA       NA      NA      NA      NA      NA\n","rs141707622        NA      NA       NA      NA      NA      NA      NA\n","rs2477488          NA      NA       NA      NA      NA      NA      NA\n","rs13202872         NA      NA       NA      NA      NA      NA      NA\n","rs62406441         NA      NA       NA      NA      NA      NA      NA\n","rs17667036  -0.422984      NA       NA      NA      NA      NA      NA\n","rs75397901         NA      NA       NA      NA      NA      NA 1.06673\n","rs10948573         NA      NA       NA      NA      NA      NA      NA\n","rs4715208          NA      NA       NA      NA      NA      NA      NA\n","rs571503           NA      NA       NA      NA      NA      NA      NA\n","              beta_40   beta_41 beta_42 beta_43 beta_44   beta_45 beta_46\n","rs2206271          NA        NA      NA      NA      NA        NA      NA\n","rs35305915         NA        NA      NA      NA      NA        NA      NA\n","rs72901027         NA        NA      NA      NA      NA        NA      NA\n","rs9367397          NA        NA      NA      NA      NA        NA      NA\n","rs61660288         NA        NA      NA      NA      NA        NA      NA\n","rs141707622        NA        NA      NA      NA      NA        NA      NA\n","rs2477488          NA        NA      NA      NA      NA        NA      NA\n","rs13202872         NA        NA      NA      NA      NA        NA      NA\n","rs62406441         NA        NA      NA      NA      NA        NA      NA\n","rs17667036         NA        NA      NA      NA      NA        NA      NA\n","rs75397901         NA        NA      NA      NA      NA        NA      NA\n","rs10948573  -0.171584        NA      NA      NA      NA        NA      NA\n","rs4715208          NA -0.205251      NA      NA      NA        NA      NA\n","rs571503           NA        NA      NA      NA      NA -0.243484      NA\n","            beta_47 beta_48 beta_49\n","rs2206271        NA      NA      NA\n","rs35305915       NA      NA      NA\n","rs72901027       NA      NA      NA\n","rs9367397        NA      NA      NA\n","rs61660288       NA      NA      NA\n","rs141707622      NA      NA      NA\n","rs2477488        NA      NA      NA\n","rs13202872       NA      NA      NA\n","rs62406441       NA      NA      NA\n","rs17667036       NA      NA      NA\n","rs75397901       NA      NA      NA\n","rs10948573       NA      NA      NA\n","rs4715208        NA      NA      NA\n","rs571503         NA      NA      NA\n","\n","$eqtl_se\n","                SE_1 SE_2 SE_3 SE_4 SE_5 SE_6 SE_7 SE_8 SE_9     SE_10\n","rs2206271   0.046181   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs35305915        NA   NA   NA   NA   NA   NA   NA   NA   NA 0.0473785\n","rs72901027        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs9367397         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs61660288        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs141707622       NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs2477488         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs13202872        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs62406441        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs17667036        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs75397901        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs10948573        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs4715208         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs571503          NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","                SE_11 SE_12 SE_13 SE_14 SE_15 SE_16 SE_17    SE_18    SE_19\n","rs2206271          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs35305915         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs72901027  0.0612013    NA    NA    NA    NA    NA    NA       NA       NA\n","rs9367397          NA    NA    NA    NA    NA    NA    NA 0.234393       NA\n","rs61660288         NA    NA    NA    NA    NA    NA    NA       NA 0.225191\n","rs141707622        NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs2477488          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs13202872         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs62406441         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs17667036         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs75397901         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs10948573         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs4715208          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs571503           NA    NA    NA    NA    NA    NA    NA       NA       NA\n","                SE_20 SE_21 SE_22     SE_23 SE_24 SE_25     SE_26 SE_27 SE_28\n","rs2206271          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs35305915         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs72901027         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs9367397          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs61660288         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs141707622 0.0565131    NA    NA        NA    NA    NA        NA    NA    NA\n","rs2477488          NA    NA    NA 0.0549176    NA    NA        NA    NA    NA\n","rs13202872         NA    NA    NA        NA    NA    NA 0.0399825    NA    NA\n","rs62406441         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs17667036         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs75397901         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs10948573         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs4715208          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs571503           NA    NA    NA        NA    NA    NA        NA    NA    NA\n","            SE_29    SE_30 SE_31 SE_32     SE_33 SE_34     SE_35 SE_36 SE_37\n","rs2206271      NA       NA    NA    NA        NA    NA 0.0414627    NA    NA\n","rs35305915     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs72901027     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs9367397      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs61660288     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs141707622    NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs2477488      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs13202872     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs62406441     NA 0.109507    NA    NA        NA    NA        NA    NA    NA\n","rs17667036     NA       NA    NA    NA 0.0931801    NA        NA    NA    NA\n","rs75397901     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs10948573     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs4715208      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs571503       NA       NA    NA    NA        NA    NA        NA    NA    NA\n","            SE_38    SE_39     SE_40     SE_41 SE_42 SE_43 SE_44    SE_45 SE_46\n","rs2206271      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs35305915     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs72901027     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs9367397      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs61660288     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs141707622    NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs2477488      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs13202872     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs62406441     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs17667036     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs75397901     NA 0.306409        NA        NA    NA    NA    NA       NA    NA\n","rs10948573     NA       NA 0.0387342        NA    NA    NA    NA       NA    NA\n","rs4715208      NA       NA        NA 0.0365271    NA    NA    NA       NA    NA\n","rs571503       NA       NA        NA        NA    NA    NA    NA 0.037926    NA\n","            SE_47 SE_48 SE_49\n","rs2206271      NA    NA    NA\n","rs35305915     NA    NA    NA\n","rs72901027     NA    NA    NA\n","rs9367397      NA    NA    NA\n","rs61660288     NA    NA    NA\n","rs141707622    NA    NA    NA\n","rs2477488      NA    NA    NA\n","rs13202872     NA    NA    NA\n","rs62406441     NA    NA    NA\n","rs17667036     NA    NA    NA\n","rs75397901     NA    NA    NA\n","rs10948573     NA    NA    NA\n","rs4715208      NA    NA    NA\n","rs571503       NA    NA    NA\n","\n","$eqtl_pvals\n","               pvalue_1 pvalue_2 pvalue_3 pvalue_4 pvalue_5 pvalue_6 pvalue_7\n","rs2206271   0.000571487       NA       NA       NA       NA       NA       NA\n","rs35305915           NA       NA       NA       NA       NA       NA       NA\n","rs72901027           NA       NA       NA       NA       NA       NA       NA\n","rs9367397            NA       NA       NA       NA       NA       NA       NA\n","rs61660288           NA       NA       NA       NA       NA       NA       NA\n","rs141707622          NA       NA       NA       NA       NA       NA       NA\n","rs2477488            NA       NA       NA       NA       NA       NA       NA\n","rs13202872           NA       NA       NA       NA       NA       NA       NA\n","rs62406441           NA       NA       NA       NA       NA       NA       NA\n","rs17667036           NA       NA       NA       NA       NA       NA       NA\n","rs75397901           NA       NA       NA       NA       NA       NA       NA\n","rs10948573           NA       NA       NA       NA       NA       NA       NA\n","rs4715208            NA       NA       NA       NA       NA       NA       NA\n","rs571503             NA       NA       NA       NA       NA       NA       NA\n","            pvalue_8 pvalue_9   pvalue_10  pvalue_11 pvalue_12 pvalue_13\n","rs2206271         NA       NA          NA         NA        NA        NA\n","rs35305915        NA       NA 0.000157047         NA        NA        NA\n","rs72901027        NA       NA          NA 1.9369e-05        NA        NA\n","rs9367397         NA       NA          NA         NA        NA        NA\n","rs61660288        NA       NA          NA         NA        NA        NA\n","rs141707622       NA       NA          NA         NA        NA        NA\n","rs2477488         NA       NA          NA         NA        NA        NA\n","rs13202872        NA       NA          NA         NA        NA        NA\n","rs62406441        NA       NA          NA         NA        NA        NA\n","rs17667036        NA       NA          NA         NA        NA        NA\n","rs75397901        NA       NA          NA         NA        NA        NA\n","rs10948573        NA       NA          NA         NA        NA        NA\n","rs4715208         NA       NA          NA         NA        NA        NA\n","rs571503          NA       NA          NA         NA        NA        NA\n","            pvalue_14 pvalue_15 pvalue_16 pvalue_17   pvalue_18   pvalue_19\n","rs2206271          NA        NA        NA        NA          NA          NA\n","rs35305915         NA        NA        NA        NA          NA          NA\n","rs72901027         NA        NA        NA        NA          NA          NA\n","rs9367397          NA        NA        NA        NA 0.000609782          NA\n","rs61660288         NA        NA        NA        NA          NA 0.000369504\n","rs141707622        NA        NA        NA        NA          NA          NA\n","rs2477488          NA        NA        NA        NA          NA          NA\n","rs13202872         NA        NA        NA        NA          NA          NA\n","rs62406441         NA        NA        NA        NA          NA          NA\n","rs17667036         NA        NA        NA        NA          NA          NA\n","rs75397901         NA        NA        NA        NA          NA          NA\n","rs10948573         NA        NA        NA        NA          NA          NA\n","rs4715208          NA        NA        NA        NA          NA          NA\n","rs571503           NA        NA        NA        NA          NA          NA\n","              pvalue_20 pvalue_21 pvalue_22   pvalue_23 pvalue_24 pvalue_25\n","rs2206271            NA        NA        NA          NA        NA        NA\n","rs35305915           NA        NA        NA          NA        NA        NA\n","rs72901027           NA        NA        NA          NA        NA        NA\n","rs9367397            NA        NA        NA          NA        NA        NA\n","rs61660288           NA        NA        NA          NA        NA        NA\n","rs141707622 0.000204908        NA        NA          NA        NA        NA\n","rs2477488            NA        NA        NA 6.03875e-10        NA        NA\n","rs13202872           NA        NA        NA          NA        NA        NA\n","rs62406441           NA        NA        NA          NA        NA        NA\n","rs17667036           NA        NA        NA          NA        NA        NA\n","rs75397901           NA        NA        NA          NA        NA        NA\n","rs10948573           NA        NA        NA          NA        NA        NA\n","rs4715208            NA        NA        NA          NA        NA        NA\n","rs571503             NA        NA        NA          NA        NA        NA\n","              pvalue_26 pvalue_27 pvalue_28 pvalue_29   pvalue_30 pvalue_31\n","rs2206271            NA        NA        NA        NA          NA        NA\n","rs35305915           NA        NA        NA        NA          NA        NA\n","rs72901027           NA        NA        NA        NA          NA        NA\n","rs9367397            NA        NA        NA        NA          NA        NA\n","rs61660288           NA        NA        NA        NA          NA        NA\n","rs141707622          NA        NA        NA        NA          NA        NA\n","rs2477488            NA        NA        NA        NA          NA        NA\n","rs13202872  1.27032e-19        NA        NA        NA          NA        NA\n","rs62406441           NA        NA        NA        NA 3.72902e-05        NA\n","rs17667036           NA        NA        NA        NA          NA        NA\n","rs75397901           NA        NA        NA        NA          NA        NA\n","rs10948573           NA        NA        NA        NA          NA        NA\n","rs4715208            NA        NA        NA        NA          NA        NA\n","rs571503             NA        NA        NA        NA          NA        NA\n","            pvalue_32   pvalue_33 pvalue_34  pvalue_35 pvalue_36 pvalue_37\n","rs2206271          NA          NA        NA 8.2487e-08        NA        NA\n","rs35305915         NA          NA        NA         NA        NA        NA\n","rs72901027         NA          NA        NA         NA        NA        NA\n","rs9367397          NA          NA        NA         NA        NA        NA\n","rs61660288         NA          NA        NA         NA        NA        NA\n","rs141707622        NA          NA        NA         NA        NA        NA\n","rs2477488          NA          NA        NA         NA        NA        NA\n","rs13202872         NA          NA        NA         NA        NA        NA\n","rs62406441         NA          NA        NA         NA        NA        NA\n","rs17667036         NA 1.35883e-05        NA         NA        NA        NA\n","rs75397901         NA          NA        NA         NA        NA        NA\n","rs10948573         NA          NA        NA         NA        NA        NA\n","rs4715208          NA          NA        NA         NA        NA        NA\n","rs571503           NA          NA        NA         NA        NA        NA\n","            pvalue_38   pvalue_39   pvalue_40   pvalue_41 pvalue_42 pvalue_43\n","rs2206271          NA          NA          NA          NA        NA        NA\n","rs35305915         NA          NA          NA          NA        NA        NA\n","rs72901027         NA          NA          NA          NA        NA        NA\n","rs9367397          NA          NA          NA          NA        NA        NA\n","rs61660288         NA          NA          NA          NA        NA        NA\n","rs141707622        NA          NA          NA          NA        NA        NA\n","rs2477488          NA          NA          NA          NA        NA        NA\n","rs13202872         NA          NA          NA          NA        NA        NA\n","rs62406441         NA          NA          NA          NA        NA        NA\n","rs17667036         NA          NA          NA          NA        NA        NA\n","rs75397901         NA 0.000625214          NA          NA        NA        NA\n","rs10948573         NA          NA 1.18765e-05          NA        NA        NA\n","rs4715208          NA          NA          NA 3.09081e-08        NA        NA\n","rs571503           NA          NA          NA          NA        NA        NA\n","            pvalue_44   pvalue_45 pvalue_46 pvalue_47 pvalue_48 pvalue_49\n","rs2206271          NA          NA        NA        NA        NA        NA\n","rs35305915         NA          NA        NA        NA        NA        NA\n","rs72901027         NA          NA        NA        NA        NA        NA\n","rs9367397          NA          NA        NA        NA        NA        NA\n","rs61660288         NA          NA        NA        NA        NA        NA\n","rs141707622        NA          NA        NA        NA        NA        NA\n","rs2477488          NA          NA        NA        NA        NA        NA\n","rs13202872         NA          NA        NA        NA        NA        NA\n","rs62406441         NA          NA        NA        NA        NA        NA\n","rs17667036         NA          NA        NA        NA        NA        NA\n","rs75397901         NA          NA        NA        NA        NA        NA\n","rs10948573         NA          NA        NA        NA        NA        NA\n","rs4715208          NA          NA        NA        NA        NA        NA\n","rs571503           NA 6.19974e-10        NA        NA        NA        NA\n","\n","$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n"]}]},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(base) # R package installed by default\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(Matrix) # R package installed by default\n","library(methods) # R package installed by default\n","library(stats) # R package installed by default\n","library(lme4) # It needs to be installed separately\n","library(data.table) # It needs to be installed separately\n","\n","# Step 2: Main Mt_Robin function:\n","MR_MtRobin <- function(MR_MtRobin_input,pval_thresh=0.001){\n","  snpID <- MR_MtRobin_input$snpID\n","  gwas_betas <- MR_MtRobin_input$gwas_betas\n","  gwas_se <- MR_MtRobin_input$gwas_se\n","  eqtl_betas <- MR_MtRobin_input$eqtl_betas\n","  eqtl_se <- MR_MtRobin_input$eqtl_se\n","  eqtl_pvals <- MR_MtRobin_input$eqtl_pvals\n","  LD <- MR_MtRobin_input$LD\n","\n","  if(ncol(eqtl_betas) != ncol(eqtl_se) | ncol(eqtl_betas) != ncol(eqtl_pvals)){\n","    stop(\"eQTL betas, SE and p-values do not all have same number of columns\")\n","  }\n","\n","  ## determine number of tissues (or conditions/studies/etc.)\n","  nT <- ncol(eqtl_betas)\n","\n","  ## rename columns for standardized processing\n","  colnames(eqtl_betas) <- paste0(\"beta_\",1:nT)\n","  colnames(eqtl_se) <- paste0(\"SE_\",1:nT)\n","  colnames(eqtl_pvals) <- paste0(\"pvalue_\",1:nT)\n","\n","  ## combine eQTL results into data.table\n","  eqtl_res <- data.table::data.table(snpID,eqtl_betas,eqtl_se,eqtl_pvals)\n","\n","  # reshape to long format (each SNP-gene-tissue will be row)\n","  eQTL_res_melt <- data.table::melt(eqtl_res,id.vars=c(\"snpID\"),\n","                                    measure.vars=patterns(\"^beta_\",\"^SE_\",\"^pvalue_\"),value.name=c(\"beta\",\"se\",\"pvalue\"),\n","                                    variable.name=\"tissue\")\n","\n","  ## subset to strong instrumental variables (based on p-value threshold)\n","  eQTL_res_melt_PltThresh <- subset(eQTL_res_melt, pvalue < pval_thresh)\n","\n","  ## set up gwas data for merging\n","  gwas_res <- data.table::data.table(snpID=snpID,gwas_beta=gwas_betas,gwas_se=gwas_se)\n","\n","  ## merge eQTL and GWAS results\n","  data.table::setkey(eQTL_res_melt_PltThresh,snpID)\n","  data.table::setkey(gwas_res,snpID)\n","  merged_res <- data.table::merge.data.table(eQTL_res_melt_PltThresh,gwas_res)\n","\n","  ## set up coefficients for reverse regression\n","  beta_x <- matrix(merged_res$beta,ncol=1)\n","  beta_y <- matrix(merged_res$gwas_beta,ncol=1)\n","\n","  ## standard errors (for weights)\n","  se_x <- matrix(merged_res$se,ncol=1)\n","  se_y <- matrix(merged_res$gwas_se,ncol=1) ## not used by function; used in resampling (consider returning with results)\n","\n","  ## identifiers\n","  snpID <- merged_res$snpID\n","\n","  lme_res <- lme4::lmer(beta_x~(beta_y-1)+(beta_y-1|snpID),weights=1/se_x^2)\n","\n","  ## return results from weighted regression with random slopes (and correlated errors) along with GWAS SE and LD (needed for resampling)\n","  return(list(lme_res=lme_res,gwas_res=gwas_res,LD=LD))\n","}\n","\n","# Step 3: Read the previous file:\n","result_setup <- readRDS(\"/data/long_covid_setup_bs_1.rds\")\n","\n","# Step 4: Apply the main Mt_Robin function using parallel processing:\n","no_cores <- 36\n","\n","cl <- makeCluster(no_cores) # Create a cluster with the number of cores\n","clusterExport(cl, \"MR_MtRobin\") # Export the MR_MtRobin function to all the nodes of the cluster\n","result_list_MtRobin <- parLapply(cl, result_setup, function(MR_MtRobin_input) {\n","  tryCatch({\n","    MR_MtRobin(MR_MtRobin_input = MR_MtRobin_input)\n","  }, error = function(e) print(e))\n","})\n","stopCluster(cl) # Stop the cluster once the computation is done\n","\n","# Step 5: The names of the results will correspond to the gene ids:\n","names(result_list_MtRobin) <- names(result_setup)\n","\n","# Step 6: Save the previous result as a .rds file:\n","saveRDS(result_list_MtRobin, file = \"/data/long_covid_main_bs_1.rds\")"],"metadata":{"id":"HhaJqIki6z6C"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/main_function.sif Rscript /data/main_function_bs_1.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"ouFcEVuH64t9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-30 10:28:34:\n","   Job Id:             93872797.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      2.25\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 00:01:48\n","   Memory Requested:   300.0GB               Memory Used: 8.24GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:01:15\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"7U9YL3qi67PV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_main_bs_1 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_1.rds\")\n","\n","# Filter out errors:\n","result_main_bs_1 <- result_main_bs_1[!sapply(result_main_bs_1, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_main_bs_1))\n","print(length(unique(result_main_bs_1)))\n","print(head(result_main_bs_1))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"vlsKieRu6-Xd","executionInfo":{"status":"ok","timestamp":1699486804407,"user_tz":-630,"elapsed":8213,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"cd440a87-d5c1-4a85-ac43-3dcece903005"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 141\n","[1] 141\n","$ENSG00000008196\n","$ENSG00000008196$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 6.9579\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 4.105   \n"," Residual        4.354   \n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-2.835  \n","\n","$ENSG00000008196$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs10948573 -0.00690805 0.0321628\n","2   rs13202872  0.02460510 0.0340206\n","3  rs141707622  0.05997710 0.0901525\n","4   rs17667036  0.00391041 0.0320274\n","5    rs2206271  0.05003660 0.0318188\n","6    rs2477488  0.05336300 0.0297989\n","7   rs35305915 -0.03369060 0.0446311\n","8    rs4715208  0.03944380 0.0345040\n","9     rs571503 -0.00698292 0.0299366\n","10  rs61660288 -0.06373460 0.0469640\n","11  rs62406441 -0.02075750 0.0360749\n","12  rs72901027 -0.00541575 0.0464996\n","13  rs75397901 -0.06437320 0.1337120\n","14   rs9367397  0.04018010 0.0627267\n","\n","$ENSG00000008196$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n","\n","$ENSG00000068781\n","$ENSG00000068781$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -26.9628\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 91.834  \n"," Residual         2.485  \n","Number of obs: 29, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n"," 50.95  \n","\n","$ENSG00000068781$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1  rs2140551 0.00700694 0.0305766\n","2 rs62135453 0.05087680 0.0355061\n","\n","$ENSG00000068781$LD\n","             rs2140551  rs62135453\n","rs2140551   1.00000000 -0.01964898\n","rs62135453 -0.01964898  1.00000000\n","\n","\n","$ENSG00000117148\n","$ENSG00000117148$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -5.9672\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 12.42155\n"," Residual         0.07669\n","Number of obs: 5, groups:  snpID, 4\n","Fixed Effects:\n","beta_y  \n","-5.827  \n","optimizer (nloptwrap) convergence code: 0 (OK) ; 0 optimizer warnings; 2 lme4 warnings \n","\n","$ENSG00000117148$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1 rs17362289  0.0424304 0.0629552\n","2 rs34519427 -0.1603040 0.0711101\n","3  rs4141983  0.0161873 0.0314816\n","4   rs478698 -0.0340591 0.0489300\n","\n","$ENSG00000117148$LD\n","              rs478698   rs17362289  rs4141983   rs34519427\n","rs478698    1.00000000  0.038323461 0.09328150 -0.070593355\n","rs17362289  0.03832346  1.000000000 0.06398760 -0.006481147\n","rs4141983   0.09328150  0.063987598 1.00000000  0.059969317\n","rs34519427 -0.07059335 -0.006481147 0.05996932  1.000000000\n","\n","\n","$ENSG00000122965\n","$ENSG00000122965$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -19.8232\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 3.580   \n"," Residual        4.126   \n","Number of obs: 49, groups:  snpID, 10\n","Fixed Effects:\n","beta_y  \n","-3.068  \n","\n","$ENSG00000122965$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs11066483 -0.03463070 0.0406681\n","2   rs11066771  0.01548470 0.0305767\n","3   rs11615428 -0.03437580 0.0403508\n","4  rs116976487  0.35533300 0.1416620\n","5   rs12369142  0.01584240 0.0361321\n","6  rs147783556 -0.04846690 0.0565352\n","7    rs1870352  0.00867405 0.0359426\n","8    rs4767149 -0.03329080 0.0353610\n","9    rs7977045 -0.05162570 0.0326507\n","10  rs80290311  0.05767070 0.0903344\n","\n","$ENSG00000122965$LD\n","               rs7977045  rs147783556    rs1870352   rs11066483 rs116976487\n","rs7977045    1.000000000 -0.206712295  0.081873465 -0.008454639 -0.05750897\n","rs147783556 -0.206712295  1.000000000 -0.007292528  0.035328443  0.05280153\n","rs1870352    0.081873465 -0.007292528  1.000000000  0.026063407 -0.04414327\n","rs11066483  -0.008454639  0.035328443  0.026063407  1.000000000  0.01662551\n","rs116976487 -0.057508966  0.052801529 -0.044143270  0.016625511  1.00000000\n","rs4767149    0.171418783 -0.138617529  0.020116315 -0.088660989 -0.01959316\n","rs12369142  -0.415600450  0.345253496 -0.076423313  0.007411732  0.06514148\n","rs11615428   0.505037949 -0.121173142  0.242169899 -0.020699028 -0.05879359\n","rs11066771  -0.616799801  0.336137832 -0.064410973  0.009494328  0.04829664\n","rs80290311   0.068027570 -0.084604531  0.208119182  0.063108318 -0.02321683\n","              rs4767149   rs12369142  rs11615428   rs11066771  rs80290311\n","rs7977045    0.17141878 -0.415600450  0.50503795 -0.616799801  0.06802757\n","rs147783556 -0.13861753  0.345253496 -0.12117314  0.336137832 -0.08460453\n","rs1870352    0.02011631 -0.076423313  0.24216990 -0.064410973  0.20811918\n","rs11066483  -0.08866099  0.007411732 -0.02069903  0.009494328  0.06310832\n","rs116976487 -0.01959316  0.065141483 -0.05879359  0.048296638 -0.02321683\n","rs4767149    1.00000000 -0.162747818  0.57376537 -0.288171436 -0.05362295\n","rs12369142  -0.16274782  1.000000000 -0.23656439  0.686835278 -0.08084646\n","rs11615428   0.57376537 -0.236564386  1.00000000 -0.332376400  0.23255788\n","rs11066771  -0.28817144  0.686835278 -0.33237640  1.000000000 -0.06798849\n","rs80290311  -0.05362295 -0.080846462  0.23255788 -0.067988491  1.00000000\n","\n","\n","$ENSG00000126266\n","$ENSG00000126266$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -27.6854\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 10.044  \n"," Residual         2.051  \n","Number of obs: 34, groups:  snpID, 5\n","Fixed Effects:\n","beta_y  \n","-12.39  \n","\n","$ENSG00000126266$gwas_res\n","        snpID  gwas_beta   gwas_se\n","1  rs10411704 0.03463980 0.0373223\n","2 rs111874634 0.11213500 0.0369274\n","3  rs12985354 0.10129600 0.0411598\n","4   rs1978013 0.02440630 0.0384278\n","5    rs886107 0.00846602 0.0319109\n","\n","$ENSG00000126266$LD\n","               rs886107 rs10411704  rs12985354 rs111874634  rs1978013\n","rs886107     1.00000000  0.3399258 -0.08398409  -0.1155209 -0.0597597\n","rs10411704   0.33992576  1.0000000 -0.26894148  -0.1770735 -0.3166438\n","rs12985354  -0.08398409 -0.2689415  1.00000000   0.6253324  0.3332107\n","rs111874634 -0.11552092 -0.1770735  0.62533240   1.0000000  0.3076256\n","rs1978013   -0.05975970 -0.3166438  0.33321066   0.3076256  1.0000000\n","\n","\n","$ENSG00000136696\n","$ENSG00000136696$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -1.5063\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 11.862  \n"," Residual         3.168  \n","Number of obs: 5, groups:  snpID, 3\n","Fixed Effects:\n","beta_y  \n","  7.65  \n","\n","$ENSG00000136696$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1 rs11123156 -0.0259520 0.0339096\n","2  rs5833482  0.0693244 0.0652153\n","3 rs76336796 -0.2391140 0.1606870\n","\n","$ENSG00000136696$LD\n","             rs11123156    rs5833482    rs76336796\n","rs11123156  1.000000000 0.0174996062 -0.0011326973\n","rs5833482   0.017499606 1.0000000000  0.0007714084\n","rs76336796 -0.001132697 0.0007714084  1.0000000000\n","\n","\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_bs_1.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_bs_1_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_bs_1_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_bs1.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"7mri1a-jNn9n"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Option 1: Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_bs_1.R\n","\n","# Option 2: Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"_uKx5J5ZNycM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-08 13:49:52:\n","   Job Id:             107817285.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      49.06\n","   NCPUs Requested:    24                     NCPUs Used: 24\n","                                           CPU Time Used: 08:59:16\n","   Memory Requested:   200.0GB               Memory Used: 19.22GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:40:53\n","   JobFS requested:    100.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"UrXtMZBhN04o"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_bs1.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results))\n","print(head(results))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1707370678568,"user_tz":-630,"elapsed":1315,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"919acfd0-a28a-45dc-f5ef-619e805ff1c7","id":"DaAl5MX9Nl01"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 141   4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000008196  0.250      10000 0.7663043\n","2 ENSG00000068781  0.592      10000 0.8100849\n","3 ENSG00000117148  0.453      10000 0.8025000\n","4 ENSG00000122965  0.074      10000 0.5301600\n","5 ENSG00000126266  0.079      10000 0.5301600\n","6 ENSG00000136696  0.510      10000 0.8025000\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs1_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs1_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_1.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1707371153884,"user_tz":-630,"elapsed":83115,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"0ddb8e9d-58b5-4fae-d34e-48b7fc96468e","id":"nt3WcDmTNl02"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs1_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 141 10 \n","\n","          gene_id     gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000008196        TFAP2B protein_coding  -2.835254  15   14\n","2 ENSG00000068781 STON1-GTF2A1L protein_coding  50.950920  29    2\n","3 ENSG00000117148         ACTL8 protein_coding  -5.827370   5    4\n","4 ENSG00000122965         RBM19 protein_coding  -3.067708  49   10\n","5 ENSG00000126266         FFAR1 protein_coding -12.391244  34    5\n","6 ENSG00000136696         IL36B protein_coding   7.650008   5    3\n","                                                                                                                                               SNP_IDs\n","1 rs10948573,rs13202872,rs141707622,rs17667036,rs2206271,rs2477488,rs35305915,rs4715208,rs571503,rs61660288,rs62406441,rs72901027,rs75397901,rs9367397\n","2                                                                                                                                 rs2140551,rs62135453\n","3                                                                                                             rs17362289,rs34519427,rs4141983,rs478698\n","4                                         rs11066483,rs11066771,rs11615428,rs116976487,rs12369142,rs147783556,rs1870352,rs4767149,rs7977045,rs80290311\n","5                                                                                                 rs10411704,rs111874634,rs12985354,rs1978013,rs886107\n","6                                                                                                                      rs11123156,rs5833482,rs76336796\n","        REML        SD Residual_SD\n","1   6.957870  4.105041  4.35409472\n","2 -26.962847 91.834134  2.48485128\n","3  -5.967169 12.421552  0.07669446\n","4 -19.823192  3.579981  4.12600230\n","5 -27.685426 10.043613  2.05140489\n","6  -1.506291 11.862344  3.16764475\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs1_protein_coding_genes_sig_not_sig.txt \n","Dimensions of protein_coding_genes : 22 10 \n","\n","          gene_id     gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000242441       GTF2A1L protein_coding  71.75526  47    2\n","2 ENSG00000068781 STON1-GTF2A1L protein_coding  50.95092  29    2\n","3 ENSG00000229637         PRAC2 protein_coding -43.43297   5    2\n","4 ENSG00000204345       CD300LD protein_coding -40.85786   4    2\n","5 ENSG00000196228       SULT1C3 protein_coding -29.52289   4    3\n","6 ENSG00000214513          NOTO protein_coding  29.34327   3    2\n","                        SNP_IDs       REML       SD Residual_SD\n","1          rs2140551,rs35125065 -40.721896 97.81211  3.35812212\n","2          rs2140551,rs62135453 -26.962847 91.83413  2.48485128\n","3          rs28613067,rs3110599  -9.755527 37.55823  1.54652397\n","4           rs559743,rs59362027  -3.627453 89.22671  3.68822360\n","5 rs1466211,rs2123693,rs4141932  -8.057989 18.44038  1.05664457\n","6        rs11894720,rs143097048 -12.934041 17.74705  0.02932076\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# BS1 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_1_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs1_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_bs1.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs1_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1707371672840,"user_tz":-630,"elapsed":5084,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"0f57dc98-56dc-4189-ffd7-501661fde174","id":"wJKQk1ArNl02"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 141\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 6.9579\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 4.105   \n"," Residual        4.354   \n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-2.835  \n","\n","$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs10948573 -0.00690805 0.0321628\n","2   rs13202872  0.02460510 0.0340206\n","3  rs141707622  0.05997710 0.0901525\n","4   rs17667036  0.00391041 0.0320274\n","5    rs2206271  0.05003660 0.0318188\n","6    rs2477488  0.05336300 0.0297989\n","7   rs35305915 -0.03369060 0.0446311\n","8    rs4715208  0.03944380 0.0345040\n","9     rs571503 -0.00698292 0.0299366\n","10  rs61660288 -0.06373460 0.0469640\n","11  rs62406441 -0.02075750 0.0360749\n","12  rs72901027 -0.00541575 0.0464996\n","13  rs75397901 -0.06437320 0.1337120\n","14   rs9367397  0.04018010 0.0627267\n","\n","$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 22 10\n","          gene_id     gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000242441       GTF2A1L protein_coding  71.75526  47    2\n","2 ENSG00000068781 STON1-GTF2A1L protein_coding  50.95092  29    2\n","3 ENSG00000229637         PRAC2 protein_coding -43.43297   5    2\n","4 ENSG00000204345       CD300LD protein_coding -40.85786   4    2\n","5 ENSG00000196228       SULT1C3 protein_coding -29.52289   4    3\n","6 ENSG00000214513          NOTO protein_coding  29.34327   3    2\n","                        SNP_IDs       REML       SD Residual_SD\n","1          rs2140551,rs35125065 -40.721896 97.81211  3.35812212\n","2          rs2140551,rs62135453 -26.962847 91.83413  2.48485128\n","3          rs28613067,rs3110599  -9.755527 37.55823  1.54652397\n","4           rs559743,rs59362027  -3.627453 89.22671  3.68822360\n","5 rs1466211,rs2123693,rs4141932  -8.057989 18.44038  1.05664457\n","6        rs11894720,rs143097048 -12.934041 17.74705  0.02932076\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 141   4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000008196  0.250      10000 0.7663043\n","2 ENSG00000068781  0.592      10000 0.8100849\n","3 ENSG00000117148  0.453      10000 0.8025000\n","4 ENSG00000122965  0.074      10000 0.5301600\n","5 ENSG00000126266  0.079      10000 0.5301600\n","6 ENSG00000136696  0.510      10000 0.8025000\n","[1] \"Final table (protein coding genes): \"\n","[1]  0 22\n"," [1] gene_id    gene_name  SNPs       SNPs_IDs   Obs        beta_y    \n"," [7] Min        First_Q    Median     Third_Q    Max        SD_RE     \n","[13] Res_SD_RE  Var_RE     Res_Var_RE SE_FE      t_value_FE REML      \n","[19] p_value    FDR        CI_lower   CI_upper  \n","<0 rows> (or 0-length row.names)\n"]}]},{"cell_type":"markdown","source":["**Obs.:** No significant protein coding genes"],"metadata":{"id":"LYhUnuGkSTIA"}},{"cell_type":"markdown","source":["### **GWAS2**"],"metadata":{"id":"hNuXzBPD70Gc"}},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(matrixStats) # It needs to be installed separately\n","#library(BiocManager) # It needs to be installed separately\n","library(snpStats) # It needs to be installed separately\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(reshape2) # Additional package for melt function\n","\n","# Step 2: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 3: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_2_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 4: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 5: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 6: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 7: Create a variable for the path to the directory to save the LD matrix results:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_2\"\n","\n","# Step 8: Create the directory if it doesn't exist:\n","if (!dir.exists(ld_matrix_dir)) {\n","  dir.create(ld_matrix_dir)\n","}\n","\n","# Step 9: Modify the original select_IV_bs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_bs() --> backward selection starts with all variables and removes one at a time:\n","select_IV_bs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset the eQTL data set for the gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Get pairwise r^2:\n","  LD_r2_pw <- LD_r2\n","  LD_r2_pw[lower.tri(LD_r2_pw,diag=T)] <- NA\n","  LD_r2_pw <- reshape2::melt(LD_r2_pw)\n","  colnames(LD_r2_pw) <- c(\"SNP1\",\"SNP2\",\"r2\")\n","  LD_r2_pw$SNP1 <- as.character(LD_r2_pw$SNP1)\n","  LD_r2_pw$SNP2 <- as.character(LD_r2_pw$SNP2)\n","\n","  ## Drop SNPs paired with themselves and duplicates (resulting from lower triangle, set to NA):\n","  LD_r2_pw <- subset(LD_r2_pw, SNP1 != SNP2 & !is.na(r2))\n","\n","  ## Sort by descending r^2:\n","  LD_r2_pw <- LD_r2_pw[order(LD_r2_pw$r2,decreasing = T),]\n","\n","  ## Check for case all candidate SNPs have pw r^2 > thresh (return SNPs with smallest p-value):\n","  if(min(LD_r2_pw$r2) >=ld_thresh){\n","    return(eqtl_data$variant_id[which.max(eqtl_data$min_p)])\n","  }\n","\n","  while(max(LD_r2_pw$r2)>=ld_thresh){\n","\n","    ## Identify top SNPs in the pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    LD_r2_pw <- subset(LD_r2_pw, SNP1 %in% SNP_pool & SNP2 %in% SNP_pool)\n","    LD_r2_pw_geThresh <- subset(LD_r2_pw, r2>=ld_thresh)\n","\n","    ## Identify SNP pair with highest remaining r^2:\n","    next_pair <- LD_r2_pw[1,]\n","\n","    ## Criteria 1: Drop the SNP with largest number of correlations > threshold with other SNPs:\n","    SNP1_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP1) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP1)\n","    SNP2_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP2) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP2)\n","    if(SNP1_nSNP_geThresh > SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP1\n","    } else if (SNP1_nSNP_geThresh < SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP2\n","    } else{\n","\n","      ## Criteria 2: Drop the SNP with smaller number of tissues with p<0.001:\n","      next_pair_eqtl <- subset(eqtl_data, variant_id %in% c(next_pair$SNP1,next_pair$SNP2))\n","      SNP1_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP1)$nTiss_ltThresh\n","      SNP2_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP2)$nTiss_ltThresh\n","\n","      if(SNP1_nTiss > SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP2\n","      } else if(SNP1_nTiss < SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP1\n","      } else{\n","\n","        ## Criteria 3: Drop the SNP with larger minimum p-value:\n","        next_SNP_drop <- next_pair_eqtl$variant_id[which.max(next_pair_eqtl$min_p)]\n","      }\n","    }\n","    SNP_pool <- SNP_pool[which(SNP_pool != next_SNP_drop)]\n","  }\n","  return(SNP_pool)\n","}\n","\n","# Step 10: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 11:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","  # Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_bs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","  }, error = function(e) print(e))\n","}, mc.cores = num_cores)\n","\n","# Step 12:\n","names(result_list) <- all_genes_vector\n","\n","# Step 13:\n","saveRDS(result_list, file = \"/data/long_covid_IVs_bs_2.rds\")"],"metadata":{"id":"H0sc9PaO80dv"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup_bs.sif Rscript /data/select_function_bs_2.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"5DU1S6Bk85YM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-25 14:10:06:\n","   Job Id:             93508970.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      874.08\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 277:39:51\n","   Memory Requested:   500.0GB               Memory Used: 382.3GB\n","   Walltime requested: 30:00:00            Walltime Used: 06:04:12\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"m915tsGO87oF"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_selectIV_bs_2 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_IVs_bs_2.rds\")\n","\n","# Filter out errors:\n","result_selectIV_bs_2 <- result_selectIV_bs_2[!sapply(result_selectIV_bs_2, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_selectIV_bs_2))\n","print(length(unique(result_selectIV_bs_2)))\n","print(head(result_selectIV_bs_2))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Cjuqqd3y9Ehb","executionInfo":{"status":"ok","timestamp":1693541867218,"user_tz":-570,"elapsed":3883,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"df303bcf-b61e-494a-802b-32ac94cb4197"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 13809\n","[1] 13737\n","$ENSG00000002079\n","[1] \"rs36168039\"  \"rs151109487\" \"rs41258334\"  \"rs28497153\"  \"rs7457787\"  \n","[6] \"rs73157578\" \n","\n","$ENSG00000004846\n"," [1] \"rs58540449\"  \"rs79448643\"  \"rs114103500\" \"rs111613242\" \"rs76325633\" \n"," [6] \"rs117337555\" \"rs73682327\"  \"rs11981446\"  \"rs7801178\"   \"rs17455533\" \n","[11] \"rs2108258\"   \"rs77254950\"  \"rs139614792\"\n","\n","$ENSG00000004939\n"," [1] \"rs2880712\"   \"rs9747287\"   \"rs117705553\" \"rs79555595\"  \"rs78230173\" \n"," [6] \"rs114651584\" \"rs34637106\"  \"rs7213794\"   \"rs144535065\" \"rs142313238\"\n","[11] \"rs76004777\"  \"rs9944407\"   \"rs34260468\"  \"rs142437085\" \"rs8070867\"  \n","[16] \"rs730818\"    \"rs200108084\" \"rs1728174\"   \"rs140638631\" \"rs4793150\"  \n","[21] \"rs76480869\"  \"rs147752346\" \"rs62080669\"  \"rs9895862\"   \"rs7222401\"  \n","[26] \"rs540551010\" \"rs9890834\"   \"rs76498976\"  \"rs138050677\"\n","\n","$ENSG00000005073\n"," [1] \"rs60642981\"  \"rs2525760\"   \"rs10273618\"  \"rs4722617\"   \"rs138820646\"\n"," [6] \"rs117987507\" \"rs17501111\"  \"rs10251237\"  \"rs147214544\" \"rs6947348\"  \n","[11] \"rs71539542\"  \"rs3735570\"   \"rs71521781\"  \"rs201583312\" \"rs17153901\" \n","[16] \"rs11975130\"  \"rs17153696\"  \"rs112074869\"\n","\n","$ENSG00000005961\n"," [1] \"rs3024272\"   \"rs11867362\"  \"rs74456742\"  \"rs143274822\" \"rs145006591\"\n"," [6] \"rs147997011\" \"rs62640053\"  \"rs4793096\"   \"rs9908627\"   \"rs34001726\" \n","[11] \"rs575873\"    \"rs111516722\" \"rs74679777\"  \"rs79337476\"  \"rs139626122\"\n","[16] \"rs7223523\"   \"rs57934049\"  \"rs72833109\"  \"rs11574815\"  \"rs66838809\" \n","[21] \"rs577074947\" \"rs144701507\" \"rs2003467\"   \"rs113149437\" \"rs148582322\"\n","[26] \"rs12373134\"  \"rs9898671\"   \"rs117944176\" \"rs7214661\"   \"rs765149778\"\n","[31] \"rs118013624\" \"rs17882143\"  \"rs12940906\"  \"rs879835344\" \"rs73318401\" \n","[36] \"rs228766\"    \"rs112435729\" \"rs111420989\" \"rs181883277\" \"rs850733\"   \n","[41] \"rs17846855\" \n","\n","$ENSG00000006059\n","[1] \"rs73301945\"  \"rs8079152\"   \"rs180938378\" \"rs35074247\"  \"rs148663776\"\n","[6] \"rs200586273\"\n","\n"]}]},{"cell_type":"code","source":["library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_2_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the list of genes-SNPs:\n","result_selectedIVs <- readRDS(\"/data/long_covid_IVs_bs_2.rds\")\n","\n","# Step 4: Create a variable for the path to the LD matrix directory:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_2\"\n","\n","# Step 5: Setup the Mt_Robin method using the following setup function:\n","MR_MtRobin_setup <- function(geneID,\n","                             snpID,\n","                             eqtl_data,\n","                             gwas_data,\n","                             ld_matrix_path,\n","                             nTiss){\n","  tryCatch({\n","    ## Check if the gene is present and subset\n","    if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","    eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","    ## Confirm all SNPs are present in the dataset\n","    if(!all(snpID %in% eqtl_data$variant_id)) stop(\"Some SNPs missing in eQTL dataset\")\n","    if(!all(snpID %in% gwas_data$variant_id)) stop(\"Some SNPs missing in GWAS dataset\")\n","\n","    # Read the LD matrix from a file\n","    LD <- as.matrix(utils::read.table(ld_matrix_path,\n","                                      header = TRUE,\n","                                      check.names = FALSE))\n","\n","    if(!all(snpID %in% colnames(LD))) stop(\"Some SNPs missing in LD matrix\")\n","\n","    ## Align data\n","    eqtl_data <- eqtl_data[match(snpID,eqtl_data$variant_id),]\n","    gwas_data <- gwas_data[match(snpID,gwas_data$variant_id),]\n","    LD <- LD[match(snpID,rownames(LD)),match(snpID,colnames(LD))]\n","\n","    eqtl_betas <- as.matrix(eqtl_data[,paste0(\"beta_\",1:nTiss)])\n","    eqtl_se <- as.matrix(eqtl_data[,paste0(\"SE_\",1:nTiss)])\n","    eqtl_pvals <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","    gwas_betas <- gwas_data$beta\n","    gwas_se <- gwas_data$SE\n","\n","    row.names(eqtl_betas) <- row.names(eqtl_se) <- row.names(eqtl_pvals) <- eqtl_data$variant_id\n","\n","    return(list(snpID=snpID,gwas_betas=gwas_betas,gwas_se=gwas_se,\n","                eqtl_betas=eqtl_betas,eqtl_se=eqtl_se,eqtl_pvals=eqtl_pvals,LD=LD))\n","  }, error = function(e) print(e))\n","}\n","\n","# Step 6: Apply the previous setup function:\n","\n","# Step 6.1: Initialize an empty list to store the results:\n","MR_MtRobin_results <- list()\n","\n","# Step 6.2: Define the number of cores to use for the parallelism process:\n","num_cores <- 36\n","\n","# Step 6.3: Loop over each gene in result_selectedIVs:\n","MR_MtRobin_results <- mclapply(names(result_selectedIVs), function(geneID) {\n","\n","  # Step 6.3.1: Get the SNPs associated with each gene:\n","  snpID <- result_selectedIVs[[geneID]]\n","\n","  # Step 6.3.2: Generate a unique file path for the LD matrix of this gene:\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  # Step 6.3.3: Call the MR_MtRobin_setup() function with the parameters for each gene:\n","  return(tryCatch({\n","    MR_MtRobin_setup(geneID = geneID,\n","                     snpID = snpID,\n","                     eqtl_data = eqtl_covid_exposure,\n","                     gwas_data = gwas_covid_outcome,\n","                     ld_matrix_path = ld_matrix_path,\n","                     nTiss = 49)\n","  }, error = function(e) print(e)))\n","}, mc.cores = num_cores)\n","\n","# Step 7: Assign names to the list:\n","names(MR_MtRobin_results) <- names(result_selectedIVs)\n","\n","# Step 8: Save the previous result as a .rds file:\n","saveRDS(MR_MtRobin_results, file = \"/data/long_covid_setup_bs_2.rds\")"],"metadata":{"id":"9MbfikRI9cbI"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup_bs.sif Rscript /data/setup_function_bs_2.R\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"t_Z5cdA79iqM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-29 17:00:33:\n","   Job Id:             93819200.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      47.70\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 13:22:08\n","   Memory Requested:   300.0GB               Memory Used: 75.35GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:26:30\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"iG78llwC9kvK"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_setup_bs_2 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_setup_bs_2.rds\")\n","\n","# Filter out errors:\n","result_setup_bs_2 <- result_setup_bs_2[!sapply(result_setup_bs_2, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_setup_bs_2))\n","print(length(unique(result_setup_bs_2)))\n","print(result_setup_bs_2[[1]])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yqDxGWnJ9ny6","executionInfo":{"status":"ok","timestamp":1693541927549,"user_tz":-570,"elapsed":13637,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"d3c6c0c6-e9e9-4f22-c54f-92fd71df1030"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 2565\n","[1] 2565\n","$snpID\n"," [1] \"rs2206271\"   \"rs35305915\"  \"rs72901027\"  \"rs9367397\"   \"rs61660288\" \n"," [6] \"rs141707622\" \"rs2477488\"   \"rs13202872\"  \"rs62406441\"  \"rs17667036\" \n","[11] \"rs75397901\"  \"rs10948573\"  \"rs4715208\"   \"rs571503\"   \n","\n","$gwas_betas\n"," [1]  0.03032140 -0.03624490 -0.01994180 -0.00767773 -0.03838880  0.08991210\n"," [7]  0.03898050  0.01695140  0.01588890 -0.02039250 -0.07066480  0.01248640\n","[13]  0.03617890  0.01188490\n","\n","$gwas_se\n"," [1] 0.0210871 0.0272673 0.0303522 0.0418627 0.0329034 0.0595761 0.0199385\n"," [8] 0.0228978 0.0242326 0.0213210 0.0829379 0.0214767 0.0236948 0.0200321\n","\n","$eqtl_betas\n","              beta_1 beta_2 beta_3 beta_4 beta_5 beta_6 beta_7 beta_8 beta_9\n","rs2206271   0.160091     NA     NA     NA     NA     NA     NA     NA     NA\n","rs35305915        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs72901027        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs9367397         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs61660288        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs141707622       NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs2477488         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs13202872        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs62406441        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs17667036        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs75397901        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs10948573        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs4715208         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs571503          NA     NA     NA     NA     NA     NA     NA     NA     NA\n","             beta_10  beta_11 beta_12 beta_13 beta_14 beta_15 beta_16 beta_17\n","rs2206271         NA       NA      NA      NA      NA      NA      NA      NA\n","rs35305915  0.184274       NA      NA      NA      NA      NA      NA      NA\n","rs72901027        NA -0.26909      NA      NA      NA      NA      NA      NA\n","rs9367397         NA       NA      NA      NA      NA      NA      NA      NA\n","rs61660288        NA       NA      NA      NA      NA      NA      NA      NA\n","rs141707622       NA       NA      NA      NA      NA      NA      NA      NA\n","rs2477488         NA       NA      NA      NA      NA      NA      NA      NA\n","rs13202872        NA       NA      NA      NA      NA      NA      NA      NA\n","rs62406441        NA       NA      NA      NA      NA      NA      NA      NA\n","rs17667036        NA       NA      NA      NA      NA      NA      NA      NA\n","rs75397901        NA       NA      NA      NA      NA      NA      NA      NA\n","rs10948573        NA       NA      NA      NA      NA      NA      NA      NA\n","rs4715208         NA       NA      NA      NA      NA      NA      NA      NA\n","rs571503          NA       NA      NA      NA      NA      NA      NA      NA\n","             beta_18   beta_19  beta_20 beta_21 beta_22   beta_23 beta_24\n","rs2206271         NA        NA       NA      NA      NA        NA      NA\n","rs35305915        NA        NA       NA      NA      NA        NA      NA\n","rs72901027        NA        NA       NA      NA      NA        NA      NA\n","rs9367397   0.829401        NA       NA      NA      NA        NA      NA\n","rs61660288        NA -0.833804       NA      NA      NA        NA      NA\n","rs141707622       NA        NA -0.21223      NA      NA        NA      NA\n","rs2477488         NA        NA       NA      NA      NA -0.353026      NA\n","rs13202872        NA        NA       NA      NA      NA        NA      NA\n","rs62406441        NA        NA       NA      NA      NA        NA      NA\n","rs17667036        NA        NA       NA      NA      NA        NA      NA\n","rs75397901        NA        NA       NA      NA      NA        NA      NA\n","rs10948573        NA        NA       NA      NA      NA        NA      NA\n","rs4715208         NA        NA       NA      NA      NA        NA      NA\n","rs571503          NA        NA       NA      NA      NA        NA      NA\n","            beta_25   beta_26 beta_27 beta_28 beta_29  beta_30 beta_31 beta_32\n","rs2206271        NA        NA      NA      NA      NA       NA      NA      NA\n","rs35305915       NA        NA      NA      NA      NA       NA      NA      NA\n","rs72901027       NA        NA      NA      NA      NA       NA      NA      NA\n","rs9367397        NA        NA      NA      NA      NA       NA      NA      NA\n","rs61660288       NA        NA      NA      NA      NA       NA      NA      NA\n","rs141707622      NA        NA      NA      NA      NA       NA      NA      NA\n","rs2477488        NA        NA      NA      NA      NA       NA      NA      NA\n","rs13202872       NA -0.380743      NA      NA      NA       NA      NA      NA\n","rs62406441       NA        NA      NA      NA      NA 0.497656      NA      NA\n","rs17667036       NA        NA      NA      NA      NA       NA      NA      NA\n","rs75397901       NA        NA      NA      NA      NA       NA      NA      NA\n","rs10948573       NA        NA      NA      NA      NA       NA      NA      NA\n","rs4715208        NA        NA      NA      NA      NA       NA      NA      NA\n","rs571503         NA        NA      NA      NA      NA       NA      NA      NA\n","              beta_33 beta_34  beta_35 beta_36 beta_37 beta_38 beta_39\n","rs2206271          NA      NA 0.225935      NA      NA      NA      NA\n","rs35305915         NA      NA       NA      NA      NA      NA      NA\n","rs72901027         NA      NA       NA      NA      NA      NA      NA\n","rs9367397          NA      NA       NA      NA      NA      NA      NA\n","rs61660288         NA      NA       NA      NA      NA      NA      NA\n","rs141707622        NA      NA       NA      NA      NA      NA      NA\n","rs2477488          NA      NA       NA      NA      NA      NA      NA\n","rs13202872         NA      NA       NA      NA      NA      NA      NA\n","rs62406441         NA      NA       NA      NA      NA      NA      NA\n","rs17667036  -0.422984      NA       NA      NA      NA      NA      NA\n","rs75397901         NA      NA       NA      NA      NA      NA 1.06673\n","rs10948573         NA      NA       NA      NA      NA      NA      NA\n","rs4715208          NA      NA       NA      NA      NA      NA      NA\n","rs571503           NA      NA       NA      NA      NA      NA      NA\n","              beta_40   beta_41 beta_42 beta_43 beta_44   beta_45 beta_46\n","rs2206271          NA        NA      NA      NA      NA        NA      NA\n","rs35305915         NA        NA      NA      NA      NA        NA      NA\n","rs72901027         NA        NA      NA      NA      NA        NA      NA\n","rs9367397          NA        NA      NA      NA      NA        NA      NA\n","rs61660288         NA        NA      NA      NA      NA        NA      NA\n","rs141707622        NA        NA      NA      NA      NA        NA      NA\n","rs2477488          NA        NA      NA      NA      NA        NA      NA\n","rs13202872         NA        NA      NA      NA      NA        NA      NA\n","rs62406441         NA        NA      NA      NA      NA        NA      NA\n","rs17667036         NA        NA      NA      NA      NA        NA      NA\n","rs75397901         NA        NA      NA      NA      NA        NA      NA\n","rs10948573  -0.171584        NA      NA      NA      NA        NA      NA\n","rs4715208          NA -0.205251      NA      NA      NA        NA      NA\n","rs571503           NA        NA      NA      NA      NA -0.243484      NA\n","            beta_47 beta_48 beta_49\n","rs2206271        NA      NA      NA\n","rs35305915       NA      NA      NA\n","rs72901027       NA      NA      NA\n","rs9367397        NA      NA      NA\n","rs61660288       NA      NA      NA\n","rs141707622      NA      NA      NA\n","rs2477488        NA      NA      NA\n","rs13202872       NA      NA      NA\n","rs62406441       NA      NA      NA\n","rs17667036       NA      NA      NA\n","rs75397901       NA      NA      NA\n","rs10948573       NA      NA      NA\n","rs4715208        NA      NA      NA\n","rs571503         NA      NA      NA\n","\n","$eqtl_se\n","                SE_1 SE_2 SE_3 SE_4 SE_5 SE_6 SE_7 SE_8 SE_9     SE_10\n","rs2206271   0.046181   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs35305915        NA   NA   NA   NA   NA   NA   NA   NA   NA 0.0473785\n","rs72901027        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs9367397         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs61660288        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs141707622       NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs2477488         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs13202872        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs62406441        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs17667036        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs75397901        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs10948573        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs4715208         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs571503          NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","                SE_11 SE_12 SE_13 SE_14 SE_15 SE_16 SE_17    SE_18    SE_19\n","rs2206271          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs35305915         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs72901027  0.0612013    NA    NA    NA    NA    NA    NA       NA       NA\n","rs9367397          NA    NA    NA    NA    NA    NA    NA 0.234393       NA\n","rs61660288         NA    NA    NA    NA    NA    NA    NA       NA 0.225191\n","rs141707622        NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs2477488          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs13202872         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs62406441         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs17667036         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs75397901         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs10948573         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs4715208          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs571503           NA    NA    NA    NA    NA    NA    NA       NA       NA\n","                SE_20 SE_21 SE_22     SE_23 SE_24 SE_25     SE_26 SE_27 SE_28\n","rs2206271          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs35305915         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs72901027         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs9367397          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs61660288         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs141707622 0.0565131    NA    NA        NA    NA    NA        NA    NA    NA\n","rs2477488          NA    NA    NA 0.0549176    NA    NA        NA    NA    NA\n","rs13202872         NA    NA    NA        NA    NA    NA 0.0399825    NA    NA\n","rs62406441         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs17667036         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs75397901         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs10948573         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs4715208          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs571503           NA    NA    NA        NA    NA    NA        NA    NA    NA\n","            SE_29    SE_30 SE_31 SE_32     SE_33 SE_34     SE_35 SE_36 SE_37\n","rs2206271      NA       NA    NA    NA        NA    NA 0.0414627    NA    NA\n","rs35305915     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs72901027     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs9367397      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs61660288     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs141707622    NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs2477488      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs13202872     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs62406441     NA 0.109507    NA    NA        NA    NA        NA    NA    NA\n","rs17667036     NA       NA    NA    NA 0.0931801    NA        NA    NA    NA\n","rs75397901     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs10948573     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs4715208      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs571503       NA       NA    NA    NA        NA    NA        NA    NA    NA\n","            SE_38    SE_39     SE_40     SE_41 SE_42 SE_43 SE_44    SE_45 SE_46\n","rs2206271      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs35305915     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs72901027     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs9367397      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs61660288     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs141707622    NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs2477488      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs13202872     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs62406441     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs17667036     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs75397901     NA 0.306409        NA        NA    NA    NA    NA       NA    NA\n","rs10948573     NA       NA 0.0387342        NA    NA    NA    NA       NA    NA\n","rs4715208      NA       NA        NA 0.0365271    NA    NA    NA       NA    NA\n","rs571503       NA       NA        NA        NA    NA    NA    NA 0.037926    NA\n","            SE_47 SE_48 SE_49\n","rs2206271      NA    NA    NA\n","rs35305915     NA    NA    NA\n","rs72901027     NA    NA    NA\n","rs9367397      NA    NA    NA\n","rs61660288     NA    NA    NA\n","rs141707622    NA    NA    NA\n","rs2477488      NA    NA    NA\n","rs13202872     NA    NA    NA\n","rs62406441     NA    NA    NA\n","rs17667036     NA    NA    NA\n","rs75397901     NA    NA    NA\n","rs10948573     NA    NA    NA\n","rs4715208      NA    NA    NA\n","rs571503       NA    NA    NA\n","\n","$eqtl_pvals\n","               pvalue_1 pvalue_2 pvalue_3 pvalue_4 pvalue_5 pvalue_6 pvalue_7\n","rs2206271   0.000571487       NA       NA       NA       NA       NA       NA\n","rs35305915           NA       NA       NA       NA       NA       NA       NA\n","rs72901027           NA       NA       NA       NA       NA       NA       NA\n","rs9367397            NA       NA       NA       NA       NA       NA       NA\n","rs61660288           NA       NA       NA       NA       NA       NA       NA\n","rs141707622          NA       NA       NA       NA       NA       NA       NA\n","rs2477488            NA       NA       NA       NA       NA       NA       NA\n","rs13202872           NA       NA       NA       NA       NA       NA       NA\n","rs62406441           NA       NA       NA       NA       NA       NA       NA\n","rs17667036           NA       NA       NA       NA       NA       NA       NA\n","rs75397901           NA       NA       NA       NA       NA       NA       NA\n","rs10948573           NA       NA       NA       NA       NA       NA       NA\n","rs4715208            NA       NA       NA       NA       NA       NA       NA\n","rs571503             NA       NA       NA       NA       NA       NA       NA\n","            pvalue_8 pvalue_9   pvalue_10  pvalue_11 pvalue_12 pvalue_13\n","rs2206271         NA       NA          NA         NA        NA        NA\n","rs35305915        NA       NA 0.000157047         NA        NA        NA\n","rs72901027        NA       NA          NA 1.9369e-05        NA        NA\n","rs9367397         NA       NA          NA         NA        NA        NA\n","rs61660288        NA       NA          NA         NA        NA        NA\n","rs141707622       NA       NA          NA         NA        NA        NA\n","rs2477488         NA       NA          NA         NA        NA        NA\n","rs13202872        NA       NA          NA         NA        NA        NA\n","rs62406441        NA       NA          NA         NA        NA        NA\n","rs17667036        NA       NA          NA         NA        NA        NA\n","rs75397901        NA       NA          NA         NA        NA        NA\n","rs10948573        NA       NA          NA         NA        NA        NA\n","rs4715208         NA       NA          NA         NA        NA        NA\n","rs571503          NA       NA          NA         NA        NA        NA\n","            pvalue_14 pvalue_15 pvalue_16 pvalue_17   pvalue_18   pvalue_19\n","rs2206271          NA        NA        NA        NA          NA          NA\n","rs35305915         NA        NA        NA        NA          NA          NA\n","rs72901027         NA        NA        NA        NA          NA          NA\n","rs9367397          NA        NA        NA        NA 0.000609782          NA\n","rs61660288         NA        NA        NA        NA          NA 0.000369504\n","rs141707622        NA        NA        NA        NA          NA          NA\n","rs2477488          NA        NA        NA        NA          NA          NA\n","rs13202872         NA        NA        NA        NA          NA          NA\n","rs62406441         NA        NA        NA        NA          NA          NA\n","rs17667036         NA        NA        NA        NA          NA          NA\n","rs75397901         NA        NA        NA        NA          NA          NA\n","rs10948573         NA        NA        NA        NA          NA          NA\n","rs4715208          NA        NA        NA        NA          NA          NA\n","rs571503           NA        NA        NA        NA          NA          NA\n","              pvalue_20 pvalue_21 pvalue_22   pvalue_23 pvalue_24 pvalue_25\n","rs2206271            NA        NA        NA          NA        NA        NA\n","rs35305915           NA        NA        NA          NA        NA        NA\n","rs72901027           NA        NA        NA          NA        NA        NA\n","rs9367397            NA        NA        NA          NA        NA        NA\n","rs61660288           NA        NA        NA          NA        NA        NA\n","rs141707622 0.000204908        NA        NA          NA        NA        NA\n","rs2477488            NA        NA        NA 6.03875e-10        NA        NA\n","rs13202872           NA        NA        NA          NA        NA        NA\n","rs62406441           NA        NA        NA          NA        NA        NA\n","rs17667036           NA        NA        NA          NA        NA        NA\n","rs75397901           NA        NA        NA          NA        NA        NA\n","rs10948573           NA        NA        NA          NA        NA        NA\n","rs4715208            NA        NA        NA          NA        NA        NA\n","rs571503             NA        NA        NA          NA        NA        NA\n","              pvalue_26 pvalue_27 pvalue_28 pvalue_29   pvalue_30 pvalue_31\n","rs2206271            NA        NA        NA        NA          NA        NA\n","rs35305915           NA        NA        NA        NA          NA        NA\n","rs72901027           NA        NA        NA        NA          NA        NA\n","rs9367397            NA        NA        NA        NA          NA        NA\n","rs61660288           NA        NA        NA        NA          NA        NA\n","rs141707622          NA        NA        NA        NA          NA        NA\n","rs2477488            NA        NA        NA        NA          NA        NA\n","rs13202872  1.27032e-19        NA        NA        NA          NA        NA\n","rs62406441           NA        NA        NA        NA 3.72902e-05        NA\n","rs17667036           NA        NA        NA        NA          NA        NA\n","rs75397901           NA        NA        NA        NA          NA        NA\n","rs10948573           NA        NA        NA        NA          NA        NA\n","rs4715208            NA        NA        NA        NA          NA        NA\n","rs571503             NA        NA        NA        NA          NA        NA\n","            pvalue_32   pvalue_33 pvalue_34  pvalue_35 pvalue_36 pvalue_37\n","rs2206271          NA          NA        NA 8.2487e-08        NA        NA\n","rs35305915         NA          NA        NA         NA        NA        NA\n","rs72901027         NA          NA        NA         NA        NA        NA\n","rs9367397          NA          NA        NA         NA        NA        NA\n","rs61660288         NA          NA        NA         NA        NA        NA\n","rs141707622        NA          NA        NA         NA        NA        NA\n","rs2477488          NA          NA        NA         NA        NA        NA\n","rs13202872         NA          NA        NA         NA        NA        NA\n","rs62406441         NA          NA        NA         NA        NA        NA\n","rs17667036         NA 1.35883e-05        NA         NA        NA        NA\n","rs75397901         NA          NA        NA         NA        NA        NA\n","rs10948573         NA          NA        NA         NA        NA        NA\n","rs4715208          NA          NA        NA         NA        NA        NA\n","rs571503           NA          NA        NA         NA        NA        NA\n","            pvalue_38   pvalue_39   pvalue_40   pvalue_41 pvalue_42 pvalue_43\n","rs2206271          NA          NA          NA          NA        NA        NA\n","rs35305915         NA          NA          NA          NA        NA        NA\n","rs72901027         NA          NA          NA          NA        NA        NA\n","rs9367397          NA          NA          NA          NA        NA        NA\n","rs61660288         NA          NA          NA          NA        NA        NA\n","rs141707622        NA          NA          NA          NA        NA        NA\n","rs2477488          NA          NA -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n"]}]},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(base) # R package installed by default\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(Matrix) # R package installed by default\n","library(methods) # R package installed by default\n","library(stats) # R package installed by default\n","library(lme4) # It needs to be installed separately\n","library(data.table) # It needs to be installed separately\n","\n","# Step 2: Main Mt_Robin function:\n","MR_MtRobin <- function(MR_MtRobin_input,pval_thresh=0.001){\n","  snpID <- MR_MtRobin_input$snpID\n","  gwas_betas <- MR_MtRobin_input$gwas_betas\n","  gwas_se <- MR_MtRobin_input$gwas_se\n","  eqtl_betas <- MR_MtRobin_input$eqtl_betas\n","  eqtl_se <- MR_MtRobin_input$eqtl_se\n","  eqtl_pvals <- MR_MtRobin_input$eqtl_pvals\n","  LD <- MR_MtRobin_input$LD\n","\n","  if(ncol(eqtl_betas) != ncol(eqtl_se) | ncol(eqtl_betas) != ncol(eqtl_pvals)){\n","    stop(\"eQTL betas, SE and p-values do not all have same number of columns\")\n","  }\n","\n","  ## determine number of tissues (or conditions/studies/etc.)\n","  nT <- ncol(eqtl_betas)\n","\n","  ## rename columns for standardized processing\n","  colnames(eqtl_betas) <- paste0(\"beta_\",1:nT)\n","  colnames(eqtl_se) <- paste0(\"SE_\",1:nT)\n","  colnames(eqtl_pvals) <- paste0(\"pvalue_\",1:nT)\n","\n","  ## combine eQTL results into data.table\n","  eqtl_res <- data.table::data.table(snpID,eqtl_betas,eqtl_se,eqtl_pvals)\n","\n","  # reshape to long format (each SNP-gene-tissue will be row)\n","  eQTL_res_melt <- data.table::melt(eqtl_res,id.vars=c(\"snpID\"),\n","                                    measure.vars=patterns(\"^beta_\",\"^SE_\",\"^pvalue_\"),value.name=c(\"beta\",\"se\",\"pvalue\"),\n","                                    variable.name=\"tissue\")\n","\n","  ## subset to strong instrumental variables (based on p-value threshold)\n","  eQTL_res_melt_PltThresh <- subset(eQTL_res_melt, pvalue < pval_thresh)\n","\n","  ## set up gwas data for merging\n","  gwas_res <- data.table::data.table(snpID=snpID,gwas_beta=gwas_betas,gwas_se=gwas_se)\n","\n","  ## merge eQTL and GWAS results\n","  data.table::setkey(eQTL_res_melt_PltThresh,snpID)\n","  data.table::setkey(gwas_res,snpID)\n","  merged_res <- data.table::merge.data.table(eQTL_res_melt_PltThresh,gwas_res)\n","\n","  ## set up coefficients for reverse regression\n","  beta_x <- matrix(merged_res$beta,ncol=1)\n","  beta_y <- matrix(merged_res$gwas_beta,ncol=1)\n","\n","  ## standard errors (for weights)\n","  se_x <- matrix(merged_res$se,ncol=1)\n","  se_y <- matrix(merged_res$gwas_se,ncol=1) ## not used by function; used in resampling (consider returning with results)\n","\n","  ## identifiers\n","  snpID <- merged_res$snpID\n","\n","  lme_res <- lme4::lmer(beta_x~(beta_y-1)+(beta_y-1|snpID),weights=1/se_x^2)\n","\n","  ## return results from weighted regression with random slopes (and correlated errors) along with GWAS SE and LD (needed for resampling)\n","  return(list(lme_res=lme_res,gwas_res=gwas_res,LD=LD))\n","}\n","\n","# Step 3: Read the previous file:\n","result_setup <- readRDS(\"/data/long_covid_setup_bs_2.rds\")\n","\n","# Step 4: Apply the main Mt_Robin function using parallel processing:\n","no_cores <- 36\n","\n","cl <- makeCluster(no_cores) # Create a cluster with the number of cores\n","clusterExport(cl, \"MR_MtRobin\") # Export the MR_MtRobin function to all the nodes of the cluster\n","result_list_MtRobin <- parLapply(cl, result_setup, function(MR_MtRobin_input) {\n","  tryCatch({\n","    MR_MtRobin(MR_MtRobin_input = MR_MtRobin_input)\n","  }, error = function(e) print(e))\n","})\n","stopCluster(cl) # Stop the cluster once the computation is done\n","\n","# Step 5: The names of the results will correspond to the gene ids:\n","names(result_list_MtRobin) <- names(result_setup)\n","\n","# Step 6: Save the previous result as a .rds file:\n","saveRDS(result_list_MtRobin, file = \"/data/long_covid_main_bs_2.rds\")"],"metadata":{"id":"ATlDNNCX9tly"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/main_function.sif Rscript /data/main_function_bs_2.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"ytq1JBrz9w7b"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-30 10:28:34:\n","   Job Id:             93872798.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      2.25\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 00:01:54\n","   Memory Requested:   300.0GB               Memory Used: 8.17GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:01:15\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"ue0BZYsa90rZ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_main_bs_2 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_2.rds\")\n","\n","# Filter out errors:\n","result_main_bs_2 <- result_main_bs_2[!sapply(result_main_bs_2, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_main_bs_2))\n","print(length(unique(result_main_bs_2)))\n","print(head(result_main_bs_2))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"3l1z24eR92pq","executionInfo":{"status":"ok","timestamp":1699487475394,"user_tz":-630,"elapsed":11106,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"dd5ae7d2-2cd8-42b9-df89-c84b0d997571"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 136\n","[1] 136\n","$ENSG00000008196\n","$ENSG00000008196$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 7.0839\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 3.142   \n"," Residual        4.646   \n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-3.156  \n","\n","$ENSG00000008196$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs10948573  0.01248640 0.0214767\n","2   rs13202872  0.01695140 0.0228978\n","3  rs141707622  0.08991210 0.0595761\n","4   rs17667036 -0.02039250 0.0213210\n","5    rs2206271  0.03032140 0.0210871\n","6    rs2477488  0.03898050 0.0199385\n","7   rs35305915 -0.03624490 0.0272673\n","8    rs4715208  0.03617890 0.0236948\n","9     rs571503  0.01188490 0.0200321\n","10  rs61660288 -0.03838880 0.0329034\n","11  rs62406441  0.01588890 0.0242326\n","12  rs72901027 -0.01994180 0.0303522\n","13  rs75397901 -0.07066480 0.0829379\n","14   rs9367397 -0.00767773 0.0418627\n","\n","$ENSG00000008196$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n","\n","$ENSG00000068781\n","$ENSG00000068781$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -27.8914\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 94.346  \n"," Residual         2.485  \n","Number of obs: 29, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n","  42.5  \n","\n","$ENSG00000068781$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1  rs2140551 0.00743663 0.0202450\n","2 rs62135453 0.02932580 0.0230477\n","\n","$ENSG00000068781$LD\n","             rs2140551  rs62135453\n","rs2140551   1.00000000 -0.01964898\n","rs62135453 -0.01964898  1.00000000\n","\n","\n","$ENSG00000112494\n","$ENSG00000112494$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 12.0344\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 64.578  \n"," Residual         2.648  \n","Number of obs: 9, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," -10.7  \n","\n","$ENSG00000112494$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1 rs117351628 -0.04548170 0.0734231\n","2 rs141514782  0.01295240 0.0544539\n","3 rs149028144 -0.07969640 0.1646450\n","4   rs1736600  0.00377374 0.0253032\n","5   rs3093009  0.00567016 0.0243088\n","6  rs55714481 -0.00797426 0.0571727\n","7  rs62438461  0.01412560 0.0220008\n","8   rs7766490 -0.00951564 0.0287087\n","\n","$ENSG00000112494$LD\n","            rs117351628  rs62438461    rs55714481 rs149028144   rs141514782\n","rs117351628  1.00000000  0.04888374  0.0720036892 -0.03624723  0.0452490319\n","rs62438461   0.04888374  1.00000000  0.0278140914  0.01209383 -0.0555439065\n","rs55714481   0.07200369  0.02781409  1.0000000000 -0.03467921 -0.0001274082\n","rs149028144 -0.03624723  0.01209383 -0.0346792109  1.00000000 -0.0138260032\n","rs141514782  0.04524903 -0.05554391 -0.0001274082 -0.01382600  1.0000000000\n","rs1736600   -0.05817461 -0.40929121  0.0077447039 -0.03015317 -0.0306809936\n","rs7766490   -0.01363783 -0.11982939 -0.0290568889 -0.01900376  0.0434365367\n","rs3093009   -0.00377826 -0.14430582 -0.0042937091 -0.03177324 -0.0506560837\n","               rs1736600   rs7766490    rs3093009\n","rs117351628 -0.058174609 -0.01363783 -0.003778260\n","rs62438461  -0.409291211 -0.11982939 -0.144305816\n","rs55714481   0.007744704 -0.02905689 -0.004293709\n","rs149028144 -0.030153172 -0.01900376 -0.031773236\n","rs141514782 -0.030680994  0.04343654 -0.050656084\n","rs1736600    1.000000000  0.12600742 -0.006354755\n","rs7766490    0.126007425  1.00000000 -0.218761737\n","rs3093009   -0.006354755 -0.21876174  1.000000000\n","\n","\n","$ENSG00000117148\n","$ENSG00000117148$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -9.685\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 13.49128\n"," Residual         0.07668\n","Number of obs: 5, groups:  snpID, 4\n","Fixed Effects:\n","beta_y  \n","-13.39  \n","optimizer (nloptwrap) convergence code: 0 (OK) ; 0 optimizer warnings; 2 lme4 warnings \n","\n","$ENSG00000117148$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1 rs17362289 -0.0274071 0.0407594\n","2 rs34519427 -0.0603449 0.0459342\n","3  rs4141983  0.0107789 0.0211415\n","4   rs478698 -0.0255845 0.0276974\n","\n","$ENSG00000117148$LD\n","              rs478698   rs17362289  rs4141983   rs34519427\n","rs478698    1.00000000  0.038323461 0.09328150 -0.070593355\n","rs17362289  0.03832346  1.000000000 0.06398760 -0.006481147\n","rs4141983   0.09328150  0.063987598 1.00000000  0.059969317\n","rs34519427 -0.07059335 -0.006481147 0.05996932  1.000000000\n","\n","\n","$ENSG00000122965\n","$ENSG00000122965$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -21.2899\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 7.116   \n"," Residual        4.123   \n","Number of obs: 49, groups:  snpID, 10\n","Fixed Effects:\n","beta_y  \n"," -4.92  \n","\n","$ENSG00000122965$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs11066483 -0.03364800 0.0231780\n","2   rs11066771  0.01562870 0.0204274\n","3   rs11615428 -0.01946250 0.0275930\n","4  rs116976487  0.11123300 0.0833088\n","5   rs12369142 -0.00737929 0.0236129\n","6  rs147783556 -0.01742800 0.0496787\n","7    rs1870352  0.01775050 0.0237349\n","8    rs4767149 -0.00544283 0.0237422\n","9    rs7977045 -0.02908660 0.0212106\n","10  rs80290311  0.03401190 0.0585321\n","\n","$ENSG00000122965$LD\n","               rs7977045  rs147783556    rs1870352   rs11066483 rs116976487\n","rs7977045    1.000000000 -0.206712295  0.081873465 -0.008454639 -0.05750897\n","rs147783556 -0.206712295  1.000000000 -0.007292528  0.035328443  0.05280153\n","rs1870352    0.081873465 -0.007292528  1.000000000  0.026063407 -0.04414327\n","rs11066483  -0.008454639  0.035328443  0.026063407  1.000000000  0.01662551\n","rs116976487 -0.057508966  0.052801529 -0.044143270  0.016625511  1.00000000\n","rs4767149    0.171418783 -0.138617529  0.020116315 -0.088660989 -0.01959316\n","rs12369142  -0.415600450  0.345253496 -0.076423313  0.007411732  0.06514148\n","rs11615428   0.505037949 -0.121173142  0.242169899 -0.020699028 -0.05879359\n","rs11066771  -0.616799801  0.336137832 -0.064410973  0.009494328  0.04829664\n","rs80290311   0.068027570 -0.084604531  0.208119182  0.063108318 -0.02321683\n","              rs4767149   rs12369142  rs11615428   rs11066771  rs80290311\n","rs7977045    0.17141878 -0.415600450  0.50503795 -0.616799801  0.06802757\n","rs147783556 -0.13861753  0.345253496 -0.12117314  0.336137832 -0.08460453\n","rs1870352    0.02011631 -0.076423313  0.24216990 -0.064410973  0.20811918\n","rs11066483  -0.08866099  0.007411732 -0.02069903  0.009494328  0.06310832\n","rs116976487 -0.01959316  0.065141483 -0.05879359  0.048296638 -0.02321683\n","rs4767149    1.00000000 -0.162747818  0.57376537 -0.288171436 -0.05362295\n","rs12369142  -0.16274782  1.000000000 -0.23656439  0.686835278 -0.08084646\n","rs11615428   0.57376537 -0.236564386  1.00000000 -0.332376400  0.23255788\n","rs11066771  -0.28817144  0.686835278 -0.33237640  1.000000000 -0.06798849\n","rs80290311  -0.05362295 -0.080846462  0.23255788 -0.067988491  1.00000000\n","\n","\n","$ENSG00000126266\n","$ENSG00000126266$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -22.5545\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 40.720  \n"," Residual         2.064  \n","Number of obs: 34, groups:  snpID, 5\n","Fixed Effects:\n","beta_y  \n","-14.93  \n","\n","$ENSG00000126266$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1  rs10411704  0.01137440 0.0251932\n","2 rs111874634  0.08879660 0.0249608\n","3  rs12985354  0.05152460 0.0237920\n","4   rs1978013 -0.00283115 0.0224835\n","5    rs886107  0.03085760 0.0211264\n","\n","$ENSG00000126266$LD\n","               rs886107 rs10411704  rs12985354 rs111874634  rs1978013\n","rs886107     1.00000000  0.3399258 -0.08398409  -0.1155209 -0.0597597\n","rs10411704   0.33992576  1.0000000 -0.26894148  -0.1770735 -0.3166438\n","rs12985354  -0.08398409 -0.2689415  1.00000000   0.6253324  0.3332107\n","rs111874634 -0.11552092 -0.1770735  0.62533240   1.0000000  0.3076256\n","rs1978013   -0.05975970 -0.3166438  0.33321066   0.3076256  1.0000000\n","\n","\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_bs_2.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_bs_2_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_bs_2_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_bs2.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"jG7EIikSVllj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Option 1: Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_bs_2.R\n","\n","# Option 2: Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"MG27DMtIVlsU"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-08 17:00:12:\n","   Job Id:             107828865.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      32.26\n","   NCPUs Requested:    24                     NCPUs Used: 24\n","                                           CPU Time Used: 09:35:47\n","   Memory Requested:   200.0GB               Memory Used: 19.28GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:26:53\n","   JobFS requested:    100.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"ZzsOH_pIVlyh"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_bs2.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results))\n","print(head(results))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1707372661960,"user_tz":-630,"elapsed":1041,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"afcf95c2-3a8d-4bea-9071-4d3ec6fffd08","id":"dlqY8W4TVmEk"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 136   4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000008196  0.192      10000 0.6866341\n","2 ENSG00000068781  0.638      10000 0.9640889\n","3 ENSG00000112494  0.751      10000 0.9825047\n","4 ENSG00000117148  0.164      10000 0.6760000\n","5 ENSG00000122965  0.144      10000 0.6712258\n","6 ENSG00000126266  0.563      10000 0.9380909\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs2_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs2_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_2.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1707372766311,"user_tz":-630,"elapsed":62226,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"9193b967-864a-49ba-afa2-557de2b5a5ec","id":"dA3xdhtwVmEl"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs2_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 136 10 \n","\n","          gene_id     gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000008196        TFAP2B protein_coding  -3.156101  15   14\n","2 ENSG00000068781 STON1-GTF2A1L protein_coding  42.502211  29    2\n","3 ENSG00000112494        UNC93A protein_coding -10.701137   9    8\n","4 ENSG00000117148         ACTL8 protein_coding -13.389779   5    4\n","5 ENSG00000122965         RBM19 protein_coding  -4.920307  49   10\n","6 ENSG00000126266         FFAR1 protein_coding -14.933544  34    5\n","                                                                                                                                               SNP_IDs\n","1 rs10948573,rs13202872,rs141707622,rs17667036,rs2206271,rs2477488,rs35305915,rs4715208,rs571503,rs61660288,rs62406441,rs72901027,rs75397901,rs9367397\n","2                                                                                                                                 rs2140551,rs62135453\n","3                                                              rs117351628,rs141514782,rs149028144,rs1736600,rs3093009,rs55714481,rs62438461,rs7766490\n","4                                                                                                             rs17362289,rs34519427,rs4141983,rs478698\n","5                                         rs11066483,rs11066771,rs11615428,rs116976487,rs12369142,rs147783556,rs1870352,rs4767149,rs7977045,rs80290311\n","6                                                                                                 rs10411704,rs111874634,rs12985354,rs1978013,rs886107\n","        REML        SD Residual_SD\n","1   7.083873  3.142405   4.6457574\n","2 -27.891363 94.345594   2.4848513\n","3  12.034437 64.578412   2.6480766\n","4  -9.684990 13.491284   0.0766843\n","5 -21.289889  7.116341   4.1232523\n","6 -22.554451 40.719687   2.0643492\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs2_protein_coding_genes_sig_not_sig.txt \n","Dimensions of protein_coding_genes : 23 10 \n","\n","          gene_id     gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000204345       CD300LD protein_coding 164.99060   4    2\n","2 ENSG00000278685        IQCA1L protein_coding  61.32528  12    4\n","3 ENSG00000182986        ZNF320 protein_coding -50.40044  49    3\n","4 ENSG00000242441       GTF2A1L protein_coding  49.54092  47    2\n","5 ENSG00000186226         LCE1E protein_coding -45.55492   4    2\n","6 ENSG00000068781 STON1-GTF2A1L protein_coding  42.50221  29    2\n","                                       SNP_IDs       REML        SD Residual_SD\n","1                          rs559743,rs59362027  -3.597073 232.14295    3.688224\n","2 rs1317223808,rs56291076,rs62481488,rs6977933   6.542085  84.72842    2.047178\n","3              rs117367674,rs3810115,rs8110062 -52.462965  37.74996    3.520868\n","4                         rs2140551,rs35125065 -45.465114 118.53117    3.358122\n","5                        rs41268482,rs41268484 -12.435179  10.04729    1.448435\n","6                         rs2140551,rs62135453 -27.891363  94.34559    2.484851\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# BS2 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_2_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs2_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_bs2.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs2_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1707372770206,"user_tz":-630,"elapsed":3902,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"f563a196-c0f9-487d-e301-a6a070068196","id":"8vozXQTTVmEl"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 136\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 7.0839\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 3.142   \n"," Residual        4.646   \n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-3.156  \n","\n","$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs10948573  0.01248640 0.0214767\n","2   rs13202872  0.01695140 0.0228978\n","3  rs141707622  0.08991210 0.0595761\n","4   rs17667036 -0.02039250 0.0213210\n","5    rs2206271  0.03032140 0.0210871\n","6    rs2477488  0.03898050 0.0199385\n","7   rs35305915 -0.03624490 0.0272673\n","8    rs4715208  0.03617890 0.0236948\n","9     rs571503  0.01188490 0.0200321\n","10  rs61660288 -0.03838880 0.0329034\n","11  rs62406441  0.01588890 0.0242326\n","12  rs72901027 -0.01994180 0.0303522\n","13  rs75397901 -0.07066480 0.0829379\n","14   rs9367397 -0.00767773 0.0418627\n","\n","$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 23 10\n","          gene_id     gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000204345       CD300LD protein_coding 164.99060   4    2\n","2 ENSG00000278685        IQCA1L protein_coding  61.32528  12    4\n","3 ENSG00000182986        ZNF320 protein_coding -50.40044  49    3\n","4 ENSG00000242441       GTF2A1L protein_coding  49.54092  47    2\n","5 ENSG00000186226         LCE1E protein_coding -45.55492   4    2\n","6 ENSG00000068781 STON1-GTF2A1L protein_coding  42.50221  29    2\n","                                       SNP_IDs       REML        SD Residual_SD\n","1                          rs559743,rs59362027  -3.597073 232.14295    3.688224\n","2 rs1317223808,rs56291076,rs62481488,rs6977933   6.542085  84.72842    2.047178\n","3              rs117367674,rs3810115,rs8110062 -52.462965  37.74996    3.520868\n","4                         rs2140551,rs35125065 -45.465114 118.53117    3.358122\n","5                        rs41268482,rs41268484 -12.435179  10.04729    1.448435\n","6                         rs2140551,rs62135453 -27.891363  94.34559    2.484851\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 136   4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000008196  0.192      10000 0.6866341\n","2 ENSG00000068781  0.638      10000 0.9640889\n","3 ENSG00000112494  0.751      10000 0.9825047\n","4 ENSG00000117148  0.164      10000 0.6760000\n","5 ENSG00000122965  0.144      10000 0.6712258\n","6 ENSG00000126266  0.563      10000 0.9380909\n","[1] \"Final table (protein coding genes): \"\n","[1]  0 22\n"," [1] gene_id    gene_name  SNPs       SNPs_IDs   Obs        beta_y    \n"," [7] Min        First_Q    Median     Third_Q    Max        SD_RE     \n","[13] Res_SD_RE  Var_RE     Res_Var_RE SE_FE      t_value_FE REML      \n","[19] p_value    FDR        CI_lower   CI_upper  \n","<0 rows> (or 0-length row.names)\n"]}]},{"cell_type":"markdown","source":["**Obs.:** No significant protein coding genes"],"metadata":{"id":"gXebMJ_pmjXN"}},{"cell_type":"markdown","source":["### **GWAS3**"],"metadata":{"id":"RynmX8_I762d"}},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(matrixStats) # It needs to be installed separately\n","#library(BiocManager) # It needs to be installed separately\n","library(snpStats) # It needs to be installed separately\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(reshape2) # Additional package for melt function\n","\n","# Step 2: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 3: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_3_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 4: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 5: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 6: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 7: Create a variable for the path to the directory to save the LD matrix results:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_3\"\n","\n","# Step 8: Create the directory if it doesn't exist:\n","if (!dir.exists(ld_matrix_dir)) {\n","  dir.create(ld_matrix_dir)\n","}\n","\n","# Step 9: Modify the original select_IV_bs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_bs() --> backward selection starts with all variables and removes one at a time:\n","select_IV_bs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset the eQTL data set for the gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Get pairwise r^2:\n","  LD_r2_pw <- LD_r2\n","  LD_r2_pw[lower.tri(LD_r2_pw,diag=T)] <- NA\n","  LD_r2_pw <- reshape2::melt(LD_r2_pw)\n","  colnames(LD_r2_pw) <- c(\"SNP1\",\"SNP2\",\"r2\")\n","  LD_r2_pw$SNP1 <- as.character(LD_r2_pw$SNP1)\n","  LD_r2_pw$SNP2 <- as.character(LD_r2_pw$SNP2)\n","\n","  ## Drop SNPs paired with themselves and duplicates (resulting from lower triangle, set to NA):\n","  LD_r2_pw <- subset(LD_r2_pw, SNP1 != SNP2 & !is.na(r2))\n","\n","  ## Sort by descending r^2:\n","  LD_r2_pw <- LD_r2_pw[order(LD_r2_pw$r2,decreasing = T),]\n","\n","  ## Check for case all candidate SNPs have pw r^2 > thresh (return SNPs with smallest p-value):\n","  if(min(LD_r2_pw$r2) >=ld_thresh){\n","    return(eqtl_data$variant_id[which.max(eqtl_data$min_p)])\n","  }\n","\n","  while(max(LD_r2_pw$r2)>=ld_thresh){\n","\n","    ## Identify top SNPs in the pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    LD_r2_pw <- subset(LD_r2_pw, SNP1 %in% SNP_pool & SNP2 %in% SNP_pool)\n","    LD_r2_pw_geThresh <- subset(LD_r2_pw, r2>=ld_thresh)\n","\n","    ## Identify SNP pair with highest remaining r^2:\n","    next_pair <- LD_r2_pw[1,]\n","\n","    ## Criteria 1: Drop the SNP with largest number of correlations > threshold with other SNPs:\n","    SNP1_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP1) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP1)\n","    SNP2_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP2) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP2)\n","    if(SNP1_nSNP_geThresh > SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP1\n","    } else if (SNP1_nSNP_geThresh < SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP2\n","    } else{\n","\n","      ## Criteria 2: Drop the SNP with smaller number of tissues with p<0.001:\n","      next_pair_eqtl <- subset(eqtl_data, variant_id %in% c(next_pair$SNP1,next_pair$SNP2))\n","      SNP1_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP1)$nTiss_ltThresh\n","      SNP2_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP2)$nTiss_ltThresh\n","\n","      if(SNP1_nTiss > SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP2\n","      } else if(SNP1_nTiss < SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP1\n","      } else{\n","\n","        ## Criteria 3: Drop the SNP with larger minimum p-value:\n","        next_SNP_drop <- next_pair_eqtl$variant_id[which.max(next_pair_eqtl$min_p)]\n","      }\n","    }\n","    SNP_pool <- SNP_pool[which(SNP_pool != next_SNP_drop)]\n","  }\n","  return(SNP_pool)\n","}\n","\n","# Step 10: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 11:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","  # Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_bs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","  }, error = function(e) print(e))\n","}, mc.cores = num_cores)\n","\n","# Step 12:\n","names(result_list) <- all_genes_vector\n","\n","# Step 13:\n","saveRDS(result_list, file = \"/data/long_covid_IVs_bs_3.rds\")"],"metadata":{"id":"eF48G2T3fhLX"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup_bs.sif Rscript /data/select_function_bs_3.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"xI-lLph7fh69"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-25 20:12:47:\n","   Job Id:             93508988.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      863.04\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 276:35:18\n","   Memory Requested:   500.0GB               Memory Used: 389.23GB\n","   Walltime requested: 30:00:00            Walltime Used: 05:59:36\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"aDIMtarvfm6s"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_selectIV_bs_3 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_IVs_bs_3.rds\")\n","\n","# Filter out errors:\n","result_selectIV_bs_3 <- result_selectIV_bs_3[!sapply(result_selectIV_bs_3, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_selectIV_bs_3))\n","print(length(unique(result_selectIV_bs_3)))\n","print(head(result_selectIV_bs_3))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"GScdLBLqfmUe","executionInfo":{"status":"ok","timestamp":1693551346144,"user_tz":-570,"elapsed":4073,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"4cc076c7-95ae-4928-d2fe-ecd2343fad04"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 13690\n","[1] 13619\n","$ENSG00000002079\n","[1] \"rs36168039\"  \"rs151109487\" \"rs41258334\"  \"rs28497153\"  \"rs7457787\"  \n","[6] \"rs73157578\" \n","\n","$ENSG00000004846\n"," [1] \"rs58540449\"  \"rs79448643\"  \"rs114103500\" \"rs111613242\" \"rs76325633\" \n"," [6] \"rs117337555\" \"rs73682327\"  \"rs11981446\"  \"rs7801178\"   \"rs17455533\" \n","[11] \"rs2108258\"   \"rs77254950\"  \"rs139614792\"\n","\n","$ENSG00000004939\n"," [1] \"rs2880712\"   \"rs9747287\"   \"rs117705553\" \"rs79555595\"  \"rs78230173\" \n"," [6] \"rs114651584\" \"rs34637106\"  \"rs7213794\"   \"rs144535065\" \"rs142313238\"\n","[11] \"rs76004777\"  \"rs9944407\"   \"rs34260468\"  \"rs142437085\" \"rs8070867\"  \n","[16] \"rs730818\"    \"rs200108084\" \"rs1728174\"   \"rs140638631\" \"rs4793150\"  \n","[21] \"rs76480869\"  \"rs147752346\" \"rs62080669\"  \"rs9895862\"   \"rs7222401\"  \n","[26] \"rs540551010\" \"rs9890834\"   \"rs76498976\"  \"rs138050677\"\n","\n","$ENSG00000005073\n"," [1] \"rs60642981\"  \"rs2525760\"   \"rs10273618\"  \"rs4722617\"   \"rs138820646\"\n"," [6] \"rs117987507\" \"rs17501111\"  \"rs10251237\"  \"rs147214544\" \"rs6947348\"  \n","[11] \"rs71539542\"  \"rs3735570\"   \"rs71521781\"  \"rs201583312\" \"rs17153901\" \n","[16] \"rs11975130\"  \"rs17153696\"  \"rs112074869\"\n","\n","$ENSG00000005961\n"," [1] \"rs3024272\"   \"rs11867362\"  \"rs74456742\"  \"rs143274822\" \"rs145006591\"\n"," [6] \"rs147997011\" \"rs62640053\"  \"rs4793096\"   \"rs9908627\"   \"rs34001726\" \n","[11] \"rs575873\"    \"rs111516722\" \"rs74679777\"  \"rs79337476\"  \"rs139626122\"\n","[16] \"rs7223523\"   \"rs57934049\"  \"rs72833109\"  \"rs11574815\"  \"rs66838809\" \n","[21] \"rs577074947\" \"rs144701507\" \"rs2003467\"   \"rs113149437\" \"rs148582322\"\n","[26] \"rs12373134\"  \"rs9898671\"   \"rs117944176\" \"rs7214661\"   \"rs765149778\"\n","[31] \"rs118013624\" \"rs17882143\"  \"rs12940906\"  \"rs879835344\" \"rs73318401\" \n","[36] \"rs228766\"    \"rs112435729\" \"rs111420989\" \"rs181883277\" \"rs850733\"   \n","[41] \"rs17846855\" \n","\n","$ENSG00000006059\n","[1] \"rs73301945\"  \"rs8079152\"   \"rs180938378\" \"rs35074247\"  \"rs148663776\"\n","[6] \"rs200586273\"\n","\n"]}]},{"cell_type":"code","source":["library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_3_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the list of genes-SNPs:\n","result_selectedIVs <- readRDS(\"/data/long_covid_IVs_bs_3.rds\")\n","\n","# Step 4: Create a variable for the path to the LD matrix directory:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_3\"\n","\n","# Step 5: Setup the Mt_Robin method using the following setup function:\n","MR_MtRobin_setup <- function(geneID,\n","                             snpID,\n","                             eqtl_data,\n","                             gwas_data,\n","                             ld_matrix_path,\n","                             nTiss){\n","  tryCatch({\n","    ## Check if the gene is present and subset\n","    if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","    eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","    ## Confirm all SNPs are present in the dataset\n","    if(!all(snpID %in% eqtl_data$variant_id)) stop(\"Some SNPs missing in eQTL dataset\")\n","    if(!all(snpID %in% gwas_data$variant_id)) stop(\"Some SNPs missing in GWAS dataset\")\n","\n","    # Read the LD matrix from a file\n","    LD <- as.matrix(utils::read.table(ld_matrix_path,\n","                                      header = TRUE,\n","                                      check.names = FALSE))\n","\n","    if(!all(snpID %in% colnames(LD))) stop(\"Some SNPs missing in LD matrix\")\n","\n","    ## Align data\n","    eqtl_data <- eqtl_data[match(snpID,eqtl_data$variant_id),]\n","    gwas_data <- gwas_data[match(snpID,gwas_data$variant_id),]\n","    LD <- LD[match(snpID,rownames(LD)),match(snpID,colnames(LD))]\n","\n","    eqtl_betas <- as.matrix(eqtl_data[,paste0(\"beta_\",1:nTiss)])\n","    eqtl_se <- as.matrix(eqtl_data[,paste0(\"SE_\",1:nTiss)])\n","    eqtl_pvals <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","    gwas_betas <- gwas_data$beta\n","    gwas_se <- gwas_data$SE\n","\n","    row.names(eqtl_betas) <- row.names(eqtl_se) <- row.names(eqtl_pvals) <- eqtl_data$variant_id\n","\n","    return(list(snpID=snpID,gwas_betas=gwas_betas,gwas_se=gwas_se,\n","                eqtl_betas=eqtl_betas,eqtl_se=eqtl_se,eqtl_pvals=eqtl_pvals,LD=LD))\n","  }, error = function(e) print(e))\n","}\n","\n","# Step 6: Apply the previous setup function:\n","\n","# Step 6.1: Initialize an empty list to store the results:\n","MR_MtRobin_results <- list()\n","\n","# Step 6.2: Define the number of cores to use for the parallelism process:\n","num_cores <- 36\n","\n","# Step 6.3: Loop over each gene in result_selectedIVs:\n","MR_MtRobin_results <- mclapply(names(result_selectedIVs), function(geneID) {\n","\n","  # Step 6.3.1: Get the SNPs associated with each gene:\n","  snpID <- result_selectedIVs[[geneID]]\n","\n","  # Step 6.3.2: Generate a unique file path for the LD matrix of this gene:\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  # Step 6.3.3: Call the MR_MtRobin_setup() function with the parameters for each gene:\n","  return(tryCatch({\n","    MR_MtRobin_setup(geneID = geneID,\n","                     snpID = snpID,\n","                     eqtl_data = eqtl_covid_exposure,\n","                     gwas_data = gwas_covid_outcome,\n","                     ld_matrix_path = ld_matrix_path,\n","                     nTiss = 49)\n","  }, error = function(e) print(e)))\n","}, mc.cores = num_cores)\n","\n","# Step 7: Assign names to the list:\n","names(MR_MtRobin_results) <- names(result_selectedIVs)\n","\n","# Step 8: Save the previous result as a .rds file:\n","saveRDS(MR_MtRobin_results, file = \"/data/long_covid_setup_bs_3.rds\")"],"metadata":{"id":"tC8i3vOJfvnG"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup_bs.sif Rscript /data/setup_function_bs_3.R\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"5PO45Yjcf-qR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-29 17:00:33:\n","   Job Id:             93819302.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      47.01\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 12:27:09\n","   Memory Requested:   300.0GB               Memory Used: 77.53GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:26:07\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"V8wZifzGgCC1"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_setup_bs_3 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_setup_bs_3.rds\")\n","\n","# Filter out errors:\n","result_setup_bs_3 <- result_setup_bs_3[!sapply(result_setup_bs_3, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_setup_bs_3))\n","print(length(unique(result_setup_bs_3)))\n","print(result_setup_bs_3[[1]])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mZ89avnWgFnL","executionInfo":{"status":"ok","timestamp":1693551371923,"user_tz":-570,"elapsed":14764,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"1a5e97d3-fe2e-45a1-8cab-6a80c48efac9"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 2723\n","[1] 2723\n","$snpID\n"," [1] \"rs2206271\"   \"rs35305915\"  \"rs72901027\"  \"rs9367397\"   \"rs61660288\" \n"," [6] \"rs141707622\" \"rs2477488\"   \"rs13202872\"  \"rs62406441\"  \"rs17667036\" \n","[11] \"rs75397901\"  \"rs10948573\"  \"rs4715208\"   \"rs571503\"   \n","\n","$gwas_betas\n"," [1]  0.03582120 -0.00993813 -0.01632840  0.03942670 -0.03288230  0.05918180\n"," [7]  0.04048180  0.00774911 -0.03468580  0.00966651 -0.10708000 -0.01167420\n","[13]  0.02275550 -0.00592643\n","\n","$gwas_se\n"," [1] 0.0327991 0.0460413 0.0493336 0.0654533 0.0466355 0.0920342 0.0311052\n"," [8] 0.0349034 0.0381658 0.0338888 0.1320330 0.0338025 0.0350512 0.0311419\n","\n","$eqtl_betas\n","              beta_1 beta_2 beta_3 beta_4 beta_5 beta_6 beta_7 beta_8 beta_9\n","rs2206271   0.160091     NA     NA     NA     NA     NA     NA     NA     NA\n","rs35305915        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs72901027        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs9367397         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs61660288        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs141707622       NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs2477488         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs13202872        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs62406441        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs17667036        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs75397901        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs10948573        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs4715208         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs571503          NA     NA     NA     NA     NA     NA     NA     NA     NA\n","             beta_10  beta_11 beta_12 beta_13 beta_14 beta_15 beta_16 beta_17\n","rs2206271         NA       NA      NA      NA      NA      NA      NA      NA\n","rs35305915  0.184274       NA      NA      NA      NA      NA      NA      NA\n","rs72901027        NA -0.26909      NA      NA      NA      NA      NA      NA\n","rs9367397         NA       NA      NA      NA      NA      NA      NA      NA\n","rs61660288        NA       NA      NA      NA      NA      NA      NA      NA\n","rs141707622       NA       NA      NA      NA      NA      NA      NA      NA\n","rs2477488         NA       NA      NA      NA      NA      NA      NA      NA\n","rs13202872        NA       NA      NA      NA      NA      NA      NA      NA\n","rs62406441        NA       NA      NA      NA      NA      NA      NA      NA\n","rs17667036        NA       NA      NA      NA      NA      NA      NA      NA\n","rs75397901        NA       NA      NA      NA      NA      NA      NA      NA\n","rs10948573        NA       NA      NA      NA      NA      NA      NA      NA\n","rs4715208         NA       NA      NA      NA      NA      NA      NA      NA\n","rs571503          NA       NA      NA      NA      NA      NA      NA      NA\n","             beta_18   beta_19  beta_20 beta_21 beta_22   beta_23 beta_24\n","rs2206271         NA        NA       NA      NA      NA        NA      NA\n","rs35305915        NA        NA       NA      NA      NA        NA      NA\n","rs72901027        NA        NA       NA      NA      NA        NA      NA\n","rs9367397   0.829401        NA       NA      NA      NA        NA      NA\n","rs61660288        NA -0.833804       NA      NA      NA        NA      NA\n","rs141707622       NA        NA -0.21223      NA      NA        NA      NA\n","rs2477488         NA        NA       NA      NA      NA -0.353026      NA\n","rs13202872        NA        NA       NA      NA      NA        NA      NA\n","rs62406441        NA        NA       NA      NA      NA        NA      NA\n","rs17667036        NA        NA       NA      NA      NA        NA      NA\n","rs75397901        NA        NA       NA      NA      NA        NA      NA\n","rs10948573        NA        NA       NA      NA      NA        NA      NA\n","rs4715208         NA        NA       NA      NA      NA        NA      NA\n","rs571503          NA        NA       NA      NA      NA        NA      NA\n","            beta_25   beta_26 beta_27 beta_28 beta_29  beta_30 beta_31 beta_32\n","rs2206271        NA        NA      NA      NA      NA       NA      NA      NA\n","rs35305915       NA        NA      NA      NA      NA       NA      NA      NA\n","rs72901027       NA        NA      NA      NA      NA       NA      NA      NA\n","rs9367397        NA        NA      NA      NA      NA       NA      NA      NA\n","rs61660288       NA        NA      NA      NA      NA       NA      NA      NA\n","rs141707622      NA        NA      NA      NA      NA       NA      NA      NA\n","rs2477488        NA        NA      NA      NA      NA       NA      NA      NA\n","rs13202872       NA -0.380743      NA      NA      NA       NA      NA      NA\n","rs62406441       NA        NA      NA      NA      NA 0.497656      NA      NA\n","rs17667036       NA        NA      NA      NA      NA       NA      NA      NA\n","rs75397901       NA        NA      NA      NA      NA       NA      NA      NA\n","rs10948573       NA        NA      NA      NA      NA       NA      NA      NA\n","rs4715208        NA        NA      NA      NA      NA       NA      NA      NA\n","rs571503         NA        NA      NA      NA      NA       NA      NA      NA\n","              beta_33 beta_34  beta_35 beta_36 beta_37 beta_38 beta_39\n","rs2206271          NA      NA 0.225935      NA      NA      NA      NA\n","rs35305915         NA      NA       NA      NA      NA      NA      NA\n","rs72901027         NA      NA       NA      NA      NA      NA      NA\n","rs9367397          NA      NA       NA      NA      NA      NA      NA\n","rs61660288         NA      NA       NA      NA      NA      NA      NA\n","rs141707622        NA      NA       NA      NA      NA      NA      NA\n","rs2477488          NA      NA       NA      NA      NA      NA      NA\n","rs13202872         NA      NA       NA      NA      NA      NA      NA\n","rs62406441         NA      NA       NA      NA      NA      NA      NA\n","rs17667036  -0.422984      NA       NA      NA      NA      NA      NA\n","rs75397901         NA      NA       NA      NA      NA      NA 1.06673\n","rs10948573         NA      NA       NA      NA      NA      NA      NA\n","rs4715208          NA      NA       NA      NA      NA      NA      NA\n","rs571503           NA      NA       NA      NA      NA      NA      NA\n","              beta_40   beta_41 beta_42 beta_43 beta_44   beta_45 beta_46\n","rs2206271          NA        NA      NA      NA      NA        NA      NA\n","rs35305915         NA        NA      NA      NA      NA        NA      NA\n","rs72901027         NA        NA      NA      NA      NA        NA      NA\n","rs9367397          NA        NA      NA      NA      NA        NA      NA\n","rs61660288         NA        NA      NA      NA      NA        NA      NA\n","rs141707622        NA        NA      NA      NA      NA        NA      NA\n","rs2477488          NA        NA      NA      NA      NA        NA      NA\n","rs13202872         NA        NA      NA      NA      NA        NA      NA\n","rs62406441         NA        NA      NA      NA      NA        NA      NA\n","rs17667036         NA        NA      NA      NA      NA        NA      NA\n","rs75397901         NA        NA      NA      NA      NA        NA      NA\n","rs10948573  -0.171584        NA      NA      NA      NA        NA      NA\n","rs4715208          NA -0.205251      NA      NA      NA        NA      NA\n","rs571503           NA        NA      NA      NA      NA -0.243484      NA\n","            beta_47 beta_48 beta_49\n","rs2206271        NA      NA      NA\n","rs35305915       NA      NA      NA\n","rs72901027       NA      NA      NA\n","rs9367397        NA      NA      NA\n","rs61660288       NA      NA      NA\n","rs141707622      NA      NA      NA\n","rs2477488        NA      NA      NA\n","rs13202872       NA      NA      NA\n","rs62406441       NA      NA      NA\n","rs17667036       NA      NA      NA\n","rs75397901       NA      NA      NA\n","rs10948573       NA      NA      NA\n","rs4715208        NA      NA      NA\n","rs571503         NA      NA      NA\n","\n","$eqtl_se\n","                SE_1 SE_2 SE_3 SE_4 SE_5 SE_6 SE_7 SE_8 SE_9     SE_10\n","rs2206271   0.046181   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs35305915        NA   NA   NA   NA   NA   NA   NA   NA   NA 0.0473785\n","rs72901027        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs9367397         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs61660288        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs141707622       NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs2477488         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs13202872        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs62406441        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs17667036        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs75397901        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs10948573        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs4715208         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs571503          NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","                SE_11 SE_12 SE_13 SE_14 SE_15 SE_16 SE_17    SE_18    SE_19\n","rs2206271          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs35305915         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs72901027  0.0612013    NA    NA    NA    NA    NA    NA       NA       NA\n","rs9367397          NA    NA    NA    NA    NA    NA    NA 0.234393       NA\n","rs61660288         NA    NA    NA    NA    NA    NA    NA       NA 0.225191\n","rs141707622        NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs2477488          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs13202872         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs62406441         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs17667036         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs75397901         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs10948573         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs4715208          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs571503           NA    NA    NA    NA    NA    NA    NA       NA       NA\n","                SE_20 SE_21 SE_22     SE_23 SE_24 SE_25     SE_26 SE_27 SE_28\n","rs2206271          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs35305915         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs72901027         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs9367397          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs61660288         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs141707622 0.0565131    NA    NA        NA    NA    NA        NA    NA    NA\n","rs2477488          NA    NA    NA 0.0549176    NA    NA        NA    NA    NA\n","rs13202872         NA    NA    NA        NA    NA    NA 0.0399825    NA    NA\n","rs62406441         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs17667036         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs75397901         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs10948573         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs4715208          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs571503           NA    NA    NA        NA    NA    NA        NA    NA    NA\n","            SE_29    SE_30 SE_31 SE_32     SE_33 SE_34     SE_35 SE_36 SE_37\n","rs2206271      NA       NA    NA    NA        NA    NA 0.0414627    NA    NA\n","rs35305915     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs72901027     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs9367397      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs61660288     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs141707622    NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs2477488      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs13202872     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs62406441     NA 0.109507    NA    NA        NA    NA        NA    NA    NA\n","rs17667036     NA       NA    NA    NA 0.0931801    NA        NA    NA    NA\n","rs75397901     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs10948573     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs4715208      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs571503       NA       NA    NA    NA        NA    NA        NA    NA    NA\n","            SE_38    SE_39     SE_40     SE_41 SE_42 SE_43 SE_44    SE_45 SE_46\n","rs2206271      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs35305915     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs72901027     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs9367397      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs61660288     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs141707622    NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs2477488      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs13202872     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs62406441     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs17667036     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs75397901     NA 0.306409        NA        NA    NA    NA    NA       NA    NA\n","rs10948573     NA       NA 0.0387342        NA    NA    NA    NA       NA    NA\n","rs4715208      NA       NA        NA 0.0365271    NA    NA    NA       NA    NA\n","rs571503       NA       NA        NA        NA    NA    NA    NA 0.037926    NA\n","            SE_47 SE_48 SE_49\n","rs2206271      NA    NA    NA\n","rs35305915     NA    NA    NA\n","rs72901027     NA    NA    NA\n","rs9367397      NA    NA    NA\n","rs61660288     NA    NA    NA\n","rs141707622    NA    NA    NA\n","rs2477488      NA    NA    NA\n","rs13202872     NA    NA    NA\n","rs62406441     NA    NA    NA\n","rs17667036     NA    NA    NA\n","rs75397901     NA    NA    NA\n","rs10948573     NA    NA    NA\n","rs4715208      NA    NA    NA\n","rs571503       NA    NA    NA\n","\n","$eqtl_pvals\n","               pvalue_1 pvalue_2 pvalue_3 pvalue_4 pvalue_5 pvalue_6 pvalue_7\n","rs2206271   0.000571487       NA       NA       NA       NA       NA       NA\n","rs35305915           NA       NA       NA       NA       NA       NA       NA\n","rs72901027           NA       NA       NA       NA       NA       NA       NA\n","rs9367397            NA       NA       NA       NA       NA       NA       NA\n","rs61660288           NA       NA       NA       NA       NA       NA       NA\n","rs141707622          NA       NA       NA       NA       NA       NA       NA\n","rs2477488            NA       NA       NA       NA       NA       NA       NA\n","rs13202872           NA       NA       NA       NA       NA       NA       NA\n","rs62406441           NA       NA       NA       NA       NA       NA       NA\n","rs17667036           NA       NA       NA       NA       NA       NA       NA\n","rs75397901           NA       NA       NA       NA       NA       NA       NA\n","rs10948573           NA       NA       NA       NA       NA       NA       NA\n","rs4715208            NA       NA       NA       NA       NA       NA       NA\n","rs571503             NA       NA       NA       NA       NA       NA       NA\n","            pvalue_8 pvalue_9   pvalue_10  pvalue_11 pvalue_12 pvalue_13\n","rs2206271         NA       NA          NA         NA        NA        NA\n","rs35305915        NA       NA 0.000157047         NA        NA        NA\n","rs72901027        NA       NA          NA 1.9369e-05        NA        NA\n","rs9367397         NA       NA          NA         NA        NA        NA\n","rs61660288        NA       NA          NA         NA        NA        NA\n","rs141707622       NA       NA          NA         NA        NA        NA\n","rs2477488         NA       NA          NA         NA        NA        NA\n","rs13202872        NA       NA          NA         NA        NA        NA\n","rs62406441        NA       NA          NA         NA        NA        NA\n","rs17667036        NA       NA          NA         NA        NA        NA\n","rs75397901        NA       NA          NA         NA        NA        NA\n","rs10948573        NA       NA          NA         NA        NA        NA\n","rs4715208         NA       NA          NA         NA        NA        NA\n","rs571503          NA       NA          NA         NA        NA        NA\n","            pvalue_14 pvalue_15 pvalue_16 pvalue_17   pvalue_18   pvalue_19\n","rs2206271          NA        NA        NA        NA          NA          NA\n","rs35305915         NA        NA        NA        NA          NA          NA\n","rs72901027         NA        NA        NA        NA          NA          NA\n","rs9367397          NA        NA        NA        NA 0.000609782          NA\n","rs61660288         NA        NA        NA        NA          NA 0.000369504\n","rs141707622        NA        NA        NA        NA          NA          NA\n","rs2477488          NA        NA        NA        NA          NA          NA\n","rs13202872         NA        NA        NA        NA          NA          NA\n","rs62406441         NA        NA        NA        NA          NA          NA\n","rs17667036         NA        NA        NA        NA          NA          NA\n","rs75397901         NA        NA        NA        NA          NA          NA\n","rs10948573         NA        NA        NA        NA          NA          NA\n","rs4715208          NA        NA        NA        NA          NA          NA\n","rs571503           NA        NA        NA        NA          NA          NA\n","              pvalue_20 pvalue_21 pvalue_22   pvalue_23 pvalue_24 pvalue_25\n","rs2206271            NA        NA        NA          NA        NA        NA\n","rs35305915           NA        NA        NA          NA        NA        NA\n","rs72901027           NA        NA        NA          NA        NA        NA\n","rs9367397            NA        NA        NA          NA        NA        NA\n","rs61660288           NA        NA        NA          NA        NA        NA\n","rs141707622 0.000204908        NA        NA          NA        NA        NA\n","rs2477488            NA        NA        NA 6.03875e-10        NA        NA\n","rs13202872           NA        NA        NA          NA        NA        NA\n","rs62406441           NA        NA        NA          NA        NA        NA\n","rs17667036           NA        NA        NA          NA        NA        NA\n","rs75397901           NA        NA        NA          NA        NA        NA\n","rs10948573           NA        NA        NA          NA        NA        NA\n","rs4715208            NA        NA        NA          NA        NA        NA\n","rs571503             NA        NA        NA          NA        NA        NA\n","              pvalue_26 pvalue_27 pvalue_28 pvalue_29   pvalue_30 pvalue_31\n","rs2206271            NA        NA        NA        NA          NA        NA\n","rs35305915           NA        NA        NA        NA          NA        NA\n","rs72901027           NA        NA        NA        NA          NA        NA\n","rs9367397            NA        NA        NA        NA          NA        NA\n","rs61660288           NA        NA        NA        NA          NA        NA\n","rs141707622          NA        NA        NA        NA          NA        NA\n","rs2477488            NA        NA        NA        NA          NA        NA\n","rs13202872  1.27032e-19        NA        NA        NA          NA        NA\n","rs62406441           NA        NA        NA        NA 3.72902e-05        NA\n","rs17667036           NA        NA        NA        NA          NA        NA\n","rs75397901           NA        NA        NA        NA          NA        NA\n","rs10948573           NA        NA        NA        NA          NA        NA\n","rs4715208            NA        NA        NA        NA          NA        NA\n","rs571503             NA        NA        NA        NA          NA        NA\n","            pvalue_32   pvalue_33 pvalue_34  pvalue_35 pvalue_36 pvalue_37\n","rs2206271          NA          NA        NA 8.2487e-08        NA        NA\n","rs35305915         NA          NA        NA         NA        NA        NA\n","rs72901027         NA          NA        NA         NA        NA        NA\n","rs9367397          NA          NA        NA         NA        NA        NA\n","rs61660288         NA          NA        NA         NA        NA -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n"]}]},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(base) # R package installed by default\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(Matrix) # R package installed by default\n","library(methods) # R package installed by default\n","library(stats) # R package installed by default\n","library(lme4) # It needs to be installed separately\n","library(data.table) # It needs to be installed separately\n","\n","# Step 2: Main Mt_Robin function:\n","MR_MtRobin <- function(MR_MtRobin_input,pval_thresh=0.001){\n","  snpID <- MR_MtRobin_input$snpID\n","  gwas_betas <- MR_MtRobin_input$gwas_betas\n","  gwas_se <- MR_MtRobin_input$gwas_se\n","  eqtl_betas <- MR_MtRobin_input$eqtl_betas\n","  eqtl_se <- MR_MtRobin_input$eqtl_se\n","  eqtl_pvals <- MR_MtRobin_input$eqtl_pvals\n","  LD <- MR_MtRobin_input$LD\n","\n","  if(ncol(eqtl_betas) != ncol(eqtl_se) | ncol(eqtl_betas) != ncol(eqtl_pvals)){\n","    stop(\"eQTL betas, SE and p-values do not all have same number of columns\")\n","  }\n","\n","  ## determine number of tissues (or conditions/studies/etc.)\n","  nT <- ncol(eqtl_betas)\n","\n","  ## rename columns for standardized processing\n","  colnames(eqtl_betas) <- paste0(\"beta_\",1:nT)\n","  colnames(eqtl_se) <- paste0(\"SE_\",1:nT)\n","  colnames(eqtl_pvals) <- paste0(\"pvalue_\",1:nT)\n","\n","  ## combine eQTL results into data.table\n","  eqtl_res <- data.table::data.table(snpID,eqtl_betas,eqtl_se,eqtl_pvals)\n","\n","  # reshape to long format (each SNP-gene-tissue will be row)\n","  eQTL_res_melt <- data.table::melt(eqtl_res,id.vars=c(\"snpID\"),\n","                                    measure.vars=patterns(\"^beta_\",\"^SE_\",\"^pvalue_\"),value.name=c(\"beta\",\"se\",\"pvalue\"),\n","                                    variable.name=\"tissue\")\n","\n","  ## subset to strong instrumental variables (based on p-value threshold)\n","  eQTL_res_melt_PltThresh <- subset(eQTL_res_melt, pvalue < pval_thresh)\n","\n","  ## set up gwas data for merging\n","  gwas_res <- data.table::data.table(snpID=snpID,gwas_beta=gwas_betas,gwas_se=gwas_se)\n","\n","  ## merge eQTL and GWAS results\n","  data.table::setkey(eQTL_res_melt_PltThresh,snpID)\n","  data.table::setkey(gwas_res,snpID)\n","  merged_res <- data.table::merge.data.table(eQTL_res_melt_PltThresh,gwas_res)\n","\n","  ## set up coefficients for reverse regression\n","  beta_x <- matrix(merged_res$beta,ncol=1)\n","  beta_y <- matrix(merged_res$gwas_beta,ncol=1)\n","\n","  ## standard errors (for weights)\n","  se_x <- matrix(merged_res$se,ncol=1)\n","  se_y <- matrix(merged_res$gwas_se,ncol=1) ## not used by function; used in resampling (consider returning with results)\n","\n","  ## identifiers\n","  snpID <- merged_res$snpID\n","\n","  lme_res <- lme4::lmer(beta_x~(beta_y-1)+(beta_y-1|snpID),weights=1/se_x^2)\n","\n","  ## return results from weighted regression with random slopes (and correlated errors) along with GWAS SE and LD (needed for resampling)\n","  return(list(lme_res=lme_res,gwas_res=gwas_res,LD=LD))\n","}\n","\n","# Step 3: Read the previous file:\n","result_setup <- readRDS(\"/data/long_covid_setup_bs_3.rds\")\n","\n","# Step 4: Apply the main Mt_Robin function using parallel processing:\n","no_cores <- 36\n","\n","cl <- makeCluster(no_cores) # Create a cluster with the number of cores\n","clusterExport(cl, \"MR_MtRobin\") # Export the MR_MtRobin function to all the nodes of the cluster\n","result_list_MtRobin <- parLapply(cl, result_setup, function(MR_MtRobin_input) {\n","  tryCatch({\n","    MR_MtRobin(MR_MtRobin_input = MR_MtRobin_input)\n","  }, error = function(e) print(e))\n","})\n","stopCluster(cl) # Stop the cluster once the computation is done\n","\n","# Step 5: The names of the results will correspond to the gene ids:\n","names(result_list_MtRobin) <- names(result_setup)\n","\n","# Step 6: Save the previous result as a .rds file:\n","saveRDS(result_list_MtRobin, file = \"/data/long_covid_main_bs_3.rds\")"],"metadata":{"id":"W5PIreUDgM4i"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/main_function.sif Rscript /data/main_function_bs_3.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"Lh2JTJhPgQcg"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-30 10:29:30:\n","   Job Id:             93872802.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      2.25\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 00:01:51\n","   Memory Requested:   300.0GB               Memory Used: 8.18GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:01:15\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"LO4VkdEIgTF0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["\n","%%R\n","# Open the output file:\n","result_main_bs_3 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_3.rds\")\n","\n","# Filter out errors:\n","result_main_bs_3 <- result_main_bs_3[!sapply(result_main_bs_3, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_main_bs_3))\n","print(length(unique(result_main_bs_3)))\n","print(head(result_main_bs_3))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"EgS6j0dbgVVe","executionInfo":{"status":"ok","timestamp":1699487819227,"user_tz":-630,"elapsed":6846,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"979caca4-528e-4609-85cd-a4df1a3d18af"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 148\n","[1] 148\n","$ENSG00000008196\n","$ENSG00000008196$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 7.7591\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 4.398   \n"," Residual        4.800   \n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-2.306  \n","\n","$ENSG00000008196$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs10948573 -0.01167420 0.0338025\n","2   rs13202872  0.00774911 0.0349034\n","3  rs141707622  0.05918180 0.0920342\n","4   rs17667036  0.00966651 0.0338888\n","5    rs2206271  0.03582120 0.0327991\n","6    rs2477488  0.04048180 0.0311052\n","7   rs35305915 -0.00993813 0.0460413\n","8    rs4715208  0.02275550 0.0350512\n","9     rs571503 -0.00592643 0.0311419\n","10  rs61660288 -0.03288230 0.0466355\n","11  rs62406441 -0.03468580 0.0381658\n","12  rs72901027 -0.01632840 0.0493336\n","13  rs75397901 -0.10708000 0.1320330\n","14   rs9367397  0.03942670 0.0654533\n","\n","$ENSG00000008196$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n","\n","$ENSG00000068781\n","$ENSG00000068781$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -27.6828\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 300.534 \n"," Residual          2.485 \n","Number of obs: 29, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n","-229.6  \n","\n","$ENSG00000068781$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1  rs2140551 -0.00183636 0.0309906\n","2 rs62135453  0.04138600 0.0382088\n","\n","$ENSG00000068781$LD\n","             rs2140551  rs62135453\n","rs2140551   1.00000000 -0.01964898\n","rs62135453 -0.01964898  1.00000000\n","\n","\n","$ENSG00000117148\n","$ENSG00000117148$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 0.6445\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 0.000   \n"," Residual        4.956   \n","Number of obs: 5, groups:  snpID, 4\n","Fixed Effects:\n","beta_y  \n"," 1.752  \n","optimizer (nloptwrap) convergence code: 0 (OK) ; 0 optimizer warnings; 1 lme4 warnings \n","\n","$ENSG00000117148$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1 rs17362289  0.01606660 0.0629402\n","2 rs34519427 -0.18365300 0.0777237\n","3  rs4141983  0.00845027 0.0333419\n","4   rs478698  0.02554140 0.0513414\n","\n","$ENSG00000117148$LD\n","              rs478698   rs17362289  rs4141983   rs34519427\n","rs478698    1.00000000  0.038323461 0.09328150 -0.070593355\n","rs17362289  0.03832346  1.000000000 0.06398760 -0.006481147\n","rs4141983   0.09328150  0.063987598 1.00000000  0.059969317\n","rs34519427 -0.07059335 -0.006481147 0.05996932  1.000000000\n","\n","\n","$ENSG00000122965\n","$ENSG00000122965$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -20.6603\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 4.111   \n"," Residual        4.146   \n","Number of obs: 49, groups:  snpID, 10\n","Fixed Effects:\n","beta_y  \n"," -3.35  \n","\n","$ENSG00000122965$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs11066483 -0.00387644 0.0389476\n","2   rs11066771  0.00454756 0.0336891\n","3   rs11615428 -0.03382150 0.0425866\n","4  rs116976487  0.24272500 0.1275380\n","5   rs12369142  0.00235131 0.0373624\n","6  rs147783556 -0.03469870 0.0608770\n","7    rs1870352 -0.01742830 0.0344593\n","8    rs4767149 -0.01477370 0.0374337\n","9    rs7977045 -0.04868170 0.0333267\n","10  rs80290311  0.04279210 0.0980738\n","\n","$ENSG00000122965$LD\n","               rs7977045  rs147783556    rs1870352   rs11066483 rs116976487\n","rs7977045    1.000000000 -0.206712295  0.081873465 -0.008454639 -0.05750897\n","rs147783556 -0.206712295  1.000000000 -0.007292528  0.035328443  0.05280153\n","rs1870352    0.081873465 -0.007292528  1.000000000  0.026063407 -0.04414327\n","rs11066483  -0.008454639  0.035328443  0.026063407  1.000000000  0.01662551\n","rs116976487 -0.057508966  0.052801529 -0.044143270  0.016625511  1.00000000\n","rs4767149    0.171418783 -0.138617529  0.020116315 -0.088660989 -0.01959316\n","rs12369142  -0.415600450  0.345253496 -0.076423313  0.007411732  0.06514148\n","rs11615428   0.505037949 -0.121173142  0.242169899 -0.020699028 -0.05879359\n","rs11066771  -0.616799801  0.336137832 -0.064410973  0.009494328  0.04829664\n","rs80290311   0.068027570 -0.084604531  0.208119182  0.063108318 -0.02321683\n","              rs4767149   rs12369142  rs11615428   rs11066771  rs80290311\n","rs7977045    0.17141878 -0.415600450  0.50503795 -0.616799801  0.06802757\n","rs147783556 -0.13861753  0.345253496 -0.12117314  0.336137832 -0.08460453\n","rs1870352    0.02011631 -0.076423313  0.24216990 -0.064410973  0.20811918\n","rs11066483  -0.08866099  0.007411732 -0.02069903  0.009494328  0.06310832\n","rs116976487 -0.01959316  0.065141483 -0.05879359  0.048296638 -0.02321683\n","rs4767149    1.00000000 -0.162747818  0.57376537 -0.288171436 -0.05362295\n","rs12369142  -0.16274782  1.000000000 -0.23656439  0.686835278 -0.08084646\n","rs11615428   0.57376537 -0.236564386  1.00000000 -0.332376400  0.23255788\n","rs11066771  -0.28817144  0.686835278 -0.33237640  1.000000000 -0.06798849\n","rs80290311  -0.05362295 -0.080846462  0.23255788 -0.067988491  1.00000000\n","\n","\n","$ENSG00000126266\n","$ENSG00000126266$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -29.3867\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 6.004   \n"," Residual        2.043   \n","Number of obs: 34, groups:  snpID, 5\n","Fixed Effects:\n","beta_y  \n","-9.199  \n","\n","$ENSG00000126266$gwas_res\n","        snpID gwas_beta   gwas_se\n","1  rs10411704 0.0488451 0.0385657\n","2 rs111874634 0.0895402 0.0378156\n","3  rs12985354 0.0831858 0.0407590\n","4   rs1978013 0.0407456 0.0417681\n","5    rs886107 0.0253548 0.0348123\n","\n","$ENSG00000126266$LD\n","               rs886107 rs10411704  rs12985354 rs111874634  rs1978013\n","rs886107     1.00000000  0.3399258 -0.08398409  -0.1155209 -0.0597597\n","rs10411704   0.33992576  1.0000000 -0.26894148  -0.1770735 -0.3166438\n","rs12985354  -0.08398409 -0.2689415  1.00000000   0.6253324  0.3332107\n","rs111874634 -0.11552092 -0.1770735  0.62533240   1.0000000  0.3076256\n","rs1978013   -0.05975970 -0.3166438  0.33321066   0.3076256  1.0000000\n","\n","\n","$ENSG00000136696\n","$ENSG00000136696$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -2.5164\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 13.369  \n"," Residual         3.074  \n","Number of obs: 5, groups:  snpID, 3\n","Fixed Effects:\n","beta_y  \n"," 9.789  \n","\n","$ENSG00000136696$gwas_res\n","       snpID  gwas_beta   gwas_se\n","1 rs11123156 -0.0234100 0.0337001\n","2  rs5833482  0.0371556 0.0732351\n","3 rs76336796 -0.2330180 0.1692370\n","\n","$ENSG00000136696$LD\n","             rs11123156    rs5833482    rs76336796\n","rs11123156  1.000000000 0.0174996062 -0.0011326973\n","rs5833482   0.017499606 1.0000000000  0.0007714084\n","rs76336796 -0.001132697 0.0007714084  1.0000000000\n","\n","\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_bs_3.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_bs_3_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_bs_3_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_bs3.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"UC1H1vWRXIfM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Option 1: Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_bs_3.R\n","\n","# Option 2: Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"1EQ0mSSvXIlg"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-08 17:48:59:\n","   Job Id:             107834555.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      47.26\n","   NCPUs Requested:    24                     NCPUs Used: 24\n","                                           CPU Time Used: 09:15:37\n","   Memory Requested:   200.0GB               Memory Used: 19.38GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:39:23\n","   JobFS requested:    100.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"58alucCMXVx5"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_bs3.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results))\n","print(head(results))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"KD7G-YU4XIs5","executionInfo":{"status":"ok","timestamp":1707376910022,"user_tz":-630,"elapsed":913,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"65873b9d-52bf-4963-9c40-baff7251cd5d"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 148   4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000008196  0.449      10000 0.8968095\n","2 ENSG00000068781  0.483      10000 0.8968095\n","3 ENSG00000117148  0.451      10000 0.8968095\n","4 ENSG00000122965  0.118      10000 0.8902424\n","5 ENSG00000126266  0.056      10000 0.6375385\n","6 ENSG00000136696  0.456      10000 0.8968095\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs3_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs3_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_3.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"drDEuPXaXNKA","executionInfo":{"status":"ok","timestamp":1707376983382,"user_tz":-630,"elapsed":66615,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"987426ee-0074-4601-dd04-510bbd591c9c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs3_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 148 10 \n","\n","          gene_id     gene_name      gene_type      beta_y Obs SNPs\n","1 ENSG00000008196        TFAP2B protein_coding   -2.306357  15   14\n","2 ENSG00000068781 STON1-GTF2A1L protein_coding -229.601291  29    2\n","3 ENSG00000117148         ACTL8 protein_coding    1.751772   5    1\n","4 ENSG00000122965         RBM19 protein_coding   -3.349624  49   10\n","5 ENSG00000126266         FFAR1 protein_coding   -9.198547  34    5\n","6 ENSG00000136696         IL36B protein_coding    9.788681   5    3\n","                                                                                                                                               SNP_IDs\n","1 rs10948573,rs13202872,rs141707622,rs17667036,rs2206271,rs2477488,rs35305915,rs4715208,rs571503,rs61660288,rs62406441,rs72901027,rs75397901,rs9367397\n","2                                                                                                                                 rs2140551,rs62135453\n","3                                                                                                             rs17362289,rs34519427,rs4141983,rs478698\n","4                                         rs11066483,rs11066771,rs11615428,rs116976487,rs12369142,rs147783556,rs1870352,rs4767149,rs7977045,rs80290311\n","5                                                                                                 rs10411704,rs111874634,rs12985354,rs1978013,rs886107\n","6                                                                                                                      rs11123156,rs5833482,rs76336796\n","         REML         SD Residual_SD\n","1   7.7591132   4.398339    4.799717\n","2 -27.6828028 300.534109    2.484851\n","3   0.6445003   0.000000    4.956173\n","4 -20.6602974   4.111295    4.146190\n","5 -29.3866737   6.004307    2.043088\n","6  -2.5164149  13.368638    3.074319\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs3_protein_coding_genes_sig_not_sig.txt \n","Dimensions of protein_coding_genes : 26 10 \n","\n","          gene_id     gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000242441       GTF2A1L protein_coding -267.22504  47    2\n","2 ENSG00000068781 STON1-GTF2A1L protein_coding -229.60129  29    2\n","3 ENSG00000214513          NOTO protein_coding  100.98850   3    2\n","4 ENSG00000204345       CD300LD protein_coding  -33.41752   4    2\n","5 ENSG00000196228       SULT1C3 protein_coding  -30.06853   4    3\n","6 ENSG00000186226         LCE1E protein_coding   18.58882   4    2\n","                        SNP_IDs       REML         SD Residual_SD\n","1          rs2140551,rs35125065 -41.089248 382.496562   3.3581221\n","2          rs2140551,rs62135453 -27.682803 300.534109   2.4848511\n","3        rs11894720,rs143097048 -12.455569 114.981439   0.0293210\n","4           rs559743,rs59362027  -6.167855 173.870580   3.6882236\n","5 rs1466211,rs2123693,rs4141932  -8.807226  16.547754   0.9981473\n","6         rs41268482,rs41268484  -9.212975   9.972971   1.4484354\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# BS3 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_3_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs3_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_bs3.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs3_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"UgK_ynCsXQTo","executionInfo":{"status":"ok","timestamp":1707376988022,"user_tz":-630,"elapsed":4643,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"b3e00b81-8bd3-4811-c683-c29fab1a95de"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 148\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 7.7591\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 4.398   \n"," Residual        4.800   \n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-2.306  \n","\n","$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs10948573 -0.01167420 0.0338025\n","2   rs13202872  0.00774911 0.0349034\n","3  rs141707622  0.05918180 0.0920342\n","4   rs17667036  0.00966651 0.0338888\n","5    rs2206271  0.03582120 0.0327991\n","6    rs2477488  0.04048180 0.0311052\n","7   rs35305915 -0.00993813 0.0460413\n","8    rs4715208  0.02275550 0.0350512\n","9     rs571503 -0.00592643 0.0311419\n","10  rs61660288 -0.03288230 0.0466355\n","11  rs62406441 -0.03468580 0.0381658\n","12  rs72901027 -0.01632840 0.0493336\n","13  rs75397901 -0.10708000 0.1320330\n","14   rs9367397  0.03942670 0.0654533\n","\n","$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 26 10\n","          gene_id     gene_name      gene_type     beta_y Obs SNPs\n","1 ENSG00000242441       GTF2A1L protein_coding -267.22504  47    2\n","2 ENSG00000068781 STON1-GTF2A1L protein_coding -229.60129  29    2\n","3 ENSG00000214513          NOTO protein_coding  100.98850   3    2\n","4 ENSG00000204345       CD300LD protein_coding  -33.41752   4    2\n","5 ENSG00000196228       SULT1C3 protein_coding  -30.06853   4    3\n","6 ENSG00000186226         LCE1E protein_coding   18.58882   4    2\n","                        SNP_IDs       REML         SD Residual_SD\n","1          rs2140551,rs35125065 -41.089248 382.496562   3.3581221\n","2          rs2140551,rs62135453 -27.682803 300.534109   2.4848511\n","3        rs11894720,rs143097048 -12.455569 114.981439   0.0293210\n","4           rs559743,rs59362027  -6.167855 173.870580   3.6882236\n","5 rs1466211,rs2123693,rs4141932  -8.807226  16.547754   0.9981473\n","6         rs41268482,rs41268484  -9.212975   9.972971   1.4484354\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 148   4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000008196  0.449      10000 0.8968095\n","2 ENSG00000068781  0.483      10000 0.8968095\n","3 ENSG00000117148  0.451      10000 0.8968095\n","4 ENSG00000122965  0.118      10000 0.8902424\n","5 ENSG00000126266  0.056      10000 0.6375385\n","6 ENSG00000136696  0.456      10000 0.8968095\n","[1] \"Final table (protein coding genes): \"\n","[1]  0 22\n"," [1] gene_id    gene_name  SNPs       SNPs_IDs   Obs        beta_y    \n"," [7] Min        First_Q    Median     Third_Q    Max        SD_RE     \n","[13] Res_SD_RE  Var_RE     Res_Var_RE SE_FE      t_value_FE REML      \n","[19] p_value    FDR        CI_lower   CI_upper  \n","<0 rows> (or 0-length row.names)\n"]}]},{"cell_type":"markdown","source":["**Obs.:** No significant protein coding genes"],"metadata":{"id":"749SbWpNmk3Y"}},{"cell_type":"markdown","source":["### **GWAS4**"],"metadata":{"id":"IYCPEXSs7-TT"}},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(matrixStats) # It needs to be installed separately\n","#library(BiocManager) # It needs to be installed separately\n","library(snpStats) # It needs to be installed separately\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(reshape2) # Additional package for melt function\n","\n","# Step 2: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 3: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_4_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 4: Open the genes input data set:\n","all_genes <- read.table(\"/data/all_genes_110.txt\",\n","                        header = TRUE,\n","                        sep = \"\",\n","                        stringsAsFactors = FALSE)\n","\n","# Step 5: Convert the previous list of genes into a vector:\n","all_genes_vector <- all_genes$gene_id\n","\n","# Step 6: Save the paths of the GTEx files:\n","gtex_covid_bed <- \"/data/WGS_GTEx_long_covid_rsID.bed\"\n","gtex_covid_bim <- \"/data/WGS_GTEx_long_covid_rsID.bim\"\n","gtex_covid_fam <- \"/data/WGS_GTEx_long_covid_rsID.fam\"\n","\n","# Step 7: Create a variable for the path to the directory to save the LD matrix results:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_4\"\n","\n","# Step 8: Create the directory if it doesn't exist:\n","if (!dir.exists(ld_matrix_dir)) {\n","  dir.create(ld_matrix_dir)\n","}\n","\n","# Step 9: Modify the original select_IV_bs() function to accept NAs from the eQTL and calculate the LD matrix inside:\n","# select_IV_bs() --> backward selection starts with all variables and removes one at a time:\n","select_IV_bs <- function(geneID,\n","                         eqtl_data,\n","                         nTiss,\n","                         bed_file,\n","                         bim_file,\n","                         fam_file,\n","                         ld_matrix_path,\n","                         ld_thresh=0.5,\n","                         pval_thresh=0.001,\n","                         nTiss_thresh=3){\n","\n","  ## Subset the eQTL data set for the gene of interest:\n","  if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","  eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","  pval_mat <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  eqtl_data$min_p <- matrixStats::rowMins(pval_mat, na.rm = TRUE) # Mod1: I added na.rm = TRUE\n","  eqtl_data$nTissThresh_p <- apply(pval_mat, 1, function(x) sort(x)[nTiss_thresh])\n","  eqtl_data <- subset(eqtl_data, nTissThresh_p < pval_thresh)\n","\n","  ## Ensure same sign for tissues with p < pval_thresh:\n","  betas <- as.matrix(subset(eqtl_data,select=paste0(\"beta_\",1:nTiss)))\n","  pvals <- as.matrix(subset(eqtl_data,select=paste0(\"pvalue_\",1:nTiss)))\n","  pvals_lt001 <- pvals<0.001\n","  betas_sig <- betas*pvals_lt001\n","  sig_neg_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x < -1e-15)))) # Mod2: I added is.na()\n","  sig_pos_eQTL <- as.numeric(apply(betas_sig, 1, function(x) any((!is.na(x) & x > 1e-15)))) # Mod3: I added is.na()\n","  eqtl_data$pos_sig_eQTL <- sig_pos_eQTL\n","  eqtl_data$neg_sig_eQTL <- sig_neg_eQTL\n","  eqtl_data <- subset(eqtl_data, !(pos_sig_eQTL==1 & neg_sig_eQTL==1))\n","\n","  if(nrow(eqtl_data)==0) stop(paste0(\"No cis-SNPs for gene with p<\",pval_thresh,\" in at least \",nTiss_thresh,\" tissues.\"))\n","\n","  ## Obtain a set of candidate SNPs:\n","  SNP_pool <- eqtl_data$variant_id\n","  if(length(SNP_pool)==1){\n","    return(SNP_pool)\n","  }\n","\n","  # Calculate the LD matrix:\n","  snp_data <- read.plink(bed_file, bim_file, fam_file)\n","  snp_matrix <- snp_data$genotypes\n","  numeric_matrix <- as(snp_matrix, \"numeric\")\n","\n","  # Restrict LD matrix to SNPs in the data set:\n","  df_subset <- numeric_matrix[, colnames(numeric_matrix) %in% SNP_pool]\n","  LD <- cor(df_subset, use=\"pairwise.complete\")\n","  LD_r2 <- LD^2\n","\n","  # Export a .txt file of the LD matrix:\n","  utils::write.table(LD,\n","                     file = ld_matrix_path,\n","                     sep = \" \",\n","                     row.names = TRUE,\n","                     col.names = TRUE,\n","                     quote = FALSE)\n","\n","  ## Get pairwise r^2:\n","  LD_r2_pw <- LD_r2\n","  LD_r2_pw[lower.tri(LD_r2_pw,diag=T)] <- NA\n","  LD_r2_pw <- reshape2::melt(LD_r2_pw)\n","  colnames(LD_r2_pw) <- c(\"SNP1\",\"SNP2\",\"r2\")\n","  LD_r2_pw$SNP1 <- as.character(LD_r2_pw$SNP1)\n","  LD_r2_pw$SNP2 <- as.character(LD_r2_pw$SNP2)\n","\n","  ## Drop SNPs paired with themselves and duplicates (resulting from lower triangle, set to NA):\n","  LD_r2_pw <- subset(LD_r2_pw, SNP1 != SNP2 & !is.na(r2))\n","\n","  ## Sort by descending r^2:\n","  LD_r2_pw <- LD_r2_pw[order(LD_r2_pw$r2,decreasing = T),]\n","\n","  ## Check for case all candidate SNPs have pw r^2 > thresh (return SNPs with smallest p-value):\n","  if(min(LD_r2_pw$r2) >=ld_thresh){\n","    return(eqtl_data$variant_id[which.max(eqtl_data$min_p)])\n","  }\n","\n","  while(max(LD_r2_pw$r2)>=ld_thresh){\n","\n","    ## Identify top SNPs in the pool:\n","    eqtl_data <- subset(eqtl_data, variant_id %in% SNP_pool)\n","    LD_r2_pw <- subset(LD_r2_pw, SNP1 %in% SNP_pool & SNP2 %in% SNP_pool)\n","    LD_r2_pw_geThresh <- subset(LD_r2_pw, r2>=ld_thresh)\n","\n","    ## Identify SNP pair with highest remaining r^2:\n","    next_pair <- LD_r2_pw[1,]\n","\n","    ## Criteria 1: Drop the SNP with largest number of correlations > threshold with other SNPs:\n","    SNP1_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP1) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP1)\n","    SNP2_nSNP_geThresh <- sum(LD_r2_pw_geThresh$SNP1 == next_pair$SNP2) + sum(LD_r2_pw_geThresh$SNP2 == next_pair$SNP2)\n","    if(SNP1_nSNP_geThresh > SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP1\n","    } else if (SNP1_nSNP_geThresh < SNP2_nSNP_geThresh){\n","      next_SNP_drop <- next_pair$SNP2\n","    } else{\n","\n","      ## Criteria 2: Drop the SNP with smaller number of tissues with p<0.001:\n","      next_pair_eqtl <- subset(eqtl_data, variant_id %in% c(next_pair$SNP1,next_pair$SNP2))\n","      SNP1_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP1)$nTiss_ltThresh\n","      SNP2_nTiss <- subset(next_pair_eqtl,variant_id==next_pair$SNP2)$nTiss_ltThresh\n","\n","      if(SNP1_nTiss > SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP2\n","      } else if(SNP1_nTiss < SNP2_nTiss){\n","        next_SNP_drop <- next_pair$SNP1\n","      } else{\n","\n","        ## Criteria 3: Drop the SNP with larger minimum p-value:\n","        next_SNP_drop <- next_pair_eqtl$variant_id[which.max(next_pair_eqtl$min_p)]\n","      }\n","    }\n","    SNP_pool <- SNP_pool[which(SNP_pool != next_SNP_drop)]\n","  }\n","  return(SNP_pool)\n","}\n","\n","# Step 10: Define the number of cores to use for the parallelism process:\n","num_cores <- 48\n","\n","# Step 11:\n","result_list <- mclapply(all_genes_vector, function(geneID) {\n","  # Generate a unique file path for the LD matrix of this gene\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  tryCatch({\n","    select_IV_bs(geneID = geneID,\n","                 eqtl_data = eqtl_covid_exposure,\n","                 nTiss = 49,\n","                 bed_file = gtex_covid_bed,\n","                 bim_file = gtex_covid_bim,\n","                 fam_file = gtex_covid_fam,\n","                 ld_matrix_path = ld_matrix_path,\n","                 ld_thresh = 0.5,\n","                 pval_thresh = 0.001,\n","                 nTiss_thresh = 1)\n","  }, error = function(e) print(e))\n","}, mc.cores = num_cores)\n","\n","# Step 12:\n","names(result_list) <- all_genes_vector\n","\n","# Step 13:\n","saveRDS(result_list, file = \"/data/long_covid_IVs_bs_4.rds\")"],"metadata":{"id":"gyQGVKZsh7Tb"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=30:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup_bs.sif Rscript /data/select_function_bs_4.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"9u6J_Qr8h-sV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-26 02:15:44:\n","   Job Id:             93551120.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      865.48\n","   NCPUs Requested:    48                     NCPUs Used: 48\n","                                           CPU Time Used: 278:35:30\n","   Memory Requested:   500.0GB               Memory Used: 388.01GB\n","   Walltime requested: 30:00:00            Walltime Used: 06:00:37\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"4oPCp8gviA5s"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_selectIV_bs_4 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_IVs_bs_4.rds\")\n","\n","# Filter out errors:\n","result_selectIV_bs_4 <- result_selectIV_bs_4[!sapply(result_selectIV_bs_4, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_selectIV_bs_4))\n","print(length(unique(result_selectIV_bs_4)))\n","print(head(result_selectIV_bs_4))"],"metadata":{"id":"gMmPBBVPiEDC","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1697000874162,"user_tz":-630,"elapsed":7466,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"24dadd24-6ff8-4e7a-97f7-af58a99147db"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 13809\n","[1] 13737\n","$ENSG00000002079\n","[1] \"rs36168039\"  \"rs151109487\" \"rs41258334\"  \"rs28497153\"  \"rs7457787\"  \n","[6] \"rs73157578\" \n","\n","$ENSG00000004846\n"," [1] \"rs58540449\"  \"rs79448643\"  \"rs114103500\" \"rs111613242\" \"rs76325633\" \n"," [6] \"rs117337555\" \"rs73682327\"  \"rs11981446\"  \"rs7801178\"   \"rs17455533\" \n","[11] \"rs2108258\"   \"rs77254950\"  \"rs139614792\"\n","\n","$ENSG00000004939\n"," [1] \"rs2880712\"   \"rs9747287\"   \"rs117705553\" \"rs79555595\"  \"rs78230173\" \n"," [6] \"rs114651584\" \"rs34637106\"  \"rs7213794\"   \"rs144535065\" \"rs142313238\"\n","[11] \"rs76004777\"  \"rs9944407\"   \"rs34260468\"  \"rs142437085\" \"rs8070867\"  \n","[16] \"rs730818\"    \"rs200108084\" \"rs1728174\"   \"rs140638631\" \"rs4793150\"  \n","[21] \"rs76480869\"  \"rs147752346\" \"rs62080669\"  \"rs9895862\"   \"rs7222401\"  \n","[26] \"rs540551010\" \"rs9890834\"   \"rs76498976\"  \"rs138050677\"\n","\n","$ENSG00000005073\n"," [1] \"rs60642981\"  \"rs2525760\"   \"rs10273618\"  \"rs4722617\"   \"rs138820646\"\n"," [6] \"rs117987507\" \"rs17501111\"  \"rs10251237\"  \"rs147214544\" \"rs6947348\"  \n","[11] \"rs71539542\"  \"rs3735570\"   \"rs71521781\"  \"rs201583312\" \"rs17153901\" \n","[16] \"rs11975130\"  \"rs17153696\"  \"rs112074869\"\n","\n","$ENSG00000005961\n"," [1] \"rs3024272\"   \"rs11867362\"  \"rs74456742\"  \"rs143274822\" \"rs145006591\"\n"," [6] \"rs147997011\" \"rs62640053\"  \"rs4793096\"   \"rs9908627\"   \"rs34001726\" \n","[11] \"rs575873\"    \"rs111516722\" \"rs74679777\"  \"rs79337476\"  \"rs139626122\"\n","[16] \"rs7223523\"   \"rs57934049\"  \"rs72833109\"  \"rs11574815\"  \"rs66838809\" \n","[21] \"rs577074947\" \"rs144701507\" \"rs2003467\"   \"rs113149437\" \"rs148582322\"\n","[26] \"rs12373134\"  \"rs9898671\"   \"rs117944176\" \"rs7214661\"   \"rs765149778\"\n","[31] \"rs118013624\" \"rs17882143\"  \"rs12940906\"  \"rs879835344\" \"rs73318401\" \n","[36] \"rs228766\"    \"rs112435729\" \"rs111420989\" \"rs181883277\" \"rs850733\"   \n","[41] \"rs17846855\" \n","\n","$ENSG00000006059\n","[1] \"rs73301945\"  \"rs8079152\"   \"rs180938378\" \"rs35074247\"  \"rs148663776\"\n","[6] \"rs200586273\"\n","\n"]}]},{"cell_type":"code","source":["library(parallel) # R package installed by default\n","\n","# Step 1: Open the eQTL file:\n","eqtl_covid_exposure <- read.table(\"/data/eqtl_all_tissues_110.txt\",\n","                                  header = TRUE,\n","                                  sep = \"\",\n","                                  stringsAsFactors = FALSE)\n","\n","# Step 2: Open the GWAS input data set:\n","gwas_covid_outcome <- read.table(\"/data/long_covid_gwas_4_subset.txt\",\n","                                 header = TRUE,\n","                                 sep = \"\",\n","                                 stringsAsFactors = FALSE)\n","\n","# Step 3: Open the list of genes-SNPs:\n","result_selectedIVs <- readRDS(\"/data/long_covid_IVs_bs_4.rds\")\n","\n","# Step 4: Create a variable for the path to the LD matrix directory:\n","ld_matrix_dir <- \"/data/ld_matrix_bs_4\"\n","\n","# Step 5: Setup the Mt_Robin method using the following setup function:\n","MR_MtRobin_setup <- function(geneID,\n","                             snpID,\n","                             eqtl_data,\n","                             gwas_data,\n","                             ld_matrix_path,\n","                             nTiss){\n","  tryCatch({\n","    ## Check if the gene is present and subset\n","    if(!(geneID %in% eqtl_data$gene_id)) stop(\"Gene is missing in eQTL dataset\")\n","    eqtl_data <- subset(eqtl_data, gene_id == geneID)\n","\n","    ## Confirm all SNPs are present in the dataset\n","    if(!all(snpID %in% eqtl_data$variant_id)) stop(\"Some SNPs missing in eQTL dataset\")\n","    if(!all(snpID %in% gwas_data$variant_id)) stop(\"Some SNPs missing in GWAS dataset\")\n","\n","    # Read the LD matrix from a file\n","    LD <- as.matrix(utils::read.table(ld_matrix_path,\n","                                      header = TRUE,\n","                                      check.names = FALSE))\n","\n","    if(!all(snpID %in% colnames(LD))) stop(\"Some SNPs missing in LD matrix\")\n","\n","    ## Align data\n","    eqtl_data <- eqtl_data[match(snpID,eqtl_data$variant_id),]\n","    gwas_data <- gwas_data[match(snpID,gwas_data$variant_id),]\n","    LD <- LD[match(snpID,rownames(LD)),match(snpID,colnames(LD))]\n","\n","    eqtl_betas <- as.matrix(eqtl_data[,paste0(\"beta_\",1:nTiss)])\n","    eqtl_se <- as.matrix(eqtl_data[,paste0(\"SE_\",1:nTiss)])\n","    eqtl_pvals <- as.matrix(eqtl_data[,paste0(\"pvalue_\",1:nTiss)])\n","    gwas_betas <- gwas_data$beta\n","    gwas_se <- gwas_data$SE\n","\n","    row.names(eqtl_betas) <- row.names(eqtl_se) <- row.names(eqtl_pvals) <- eqtl_data$variant_id\n","\n","    return(list(snpID=snpID,gwas_betas=gwas_betas,gwas_se=gwas_se,\n","                eqtl_betas=eqtl_betas,eqtl_se=eqtl_se,eqtl_pvals=eqtl_pvals,LD=LD))\n","  }, error = function(e) print(e))\n","}\n","\n","# Step 6: Apply the previous setup function:\n","\n","# Step 6.1: Initialize an empty list to store the results:\n","MR_MtRobin_results <- list()\n","\n","# Step 6.2: Define the number of cores to use for the parallelism process:\n","num_cores <- 36\n","\n","# Step 6.3: Loop over each gene in result_selectedIVs:\n","MR_MtRobin_results <- mclapply(names(result_selectedIVs), function(geneID) {\n","\n","  # Step 6.3.1: Get the SNPs associated with each gene:\n","  snpID <- result_selectedIVs[[geneID]]\n","\n","  # Step 6.3.2: Generate a unique file path for the LD matrix of this gene:\n","  ld_matrix_path <- file.path(ld_matrix_dir, paste0(\"ld_matrix_\", geneID, \".txt\"))\n","\n","  # Step 6.3.3: Call the MR_MtRobin_setup() function with the parameters for each gene:\n","  return(tryCatch({\n","    MR_MtRobin_setup(geneID = geneID,\n","                     snpID = snpID,\n","                     eqtl_data = eqtl_covid_exposure,\n","                     gwas_data = gwas_covid_outcome,\n","                     ld_matrix_path = ld_matrix_path,\n","                     nTiss = 49)\n","  }, error = function(e) print(e)))\n","}, mc.cores = num_cores)\n","\n","# Step 7: Assign names to the list:\n","names(MR_MtRobin_results) <- names(result_selectedIVs)\n","\n","# Step 8: Save the previous result as a .rds file:\n","saveRDS(MR_MtRobin_results, file = \"/data/long_covid_setup_bs_4.rds\")"],"metadata":{"id":"BhgTZIUniT59"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/select_setup_bs.sif Rscript /data/setup_function_bs_4.R\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"thxurG8WiUvq"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-29 17:02:20:\n","   Job Id:             93819310.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      50.10\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 12:54:56\n","   Memory Requested:   300.0GB               Memory Used: 80.8GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:27:50\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"UbFY1NcCjX_j"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_setup_bs_4 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_setup_bs_4.rds\")\n","\n","# Filter out errors:\n","result_setup_bs_4 <- result_setup_bs_4[!sapply(result_setup_bs_4, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_setup_bs_4))\n","print(length(unique(result_setup_bs_4)))\n","print(result_setup_bs_4[[1]])"],"metadata":{"id":"SYhENfeSjvZB","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1697000886759,"user_tz":-630,"elapsed":12600,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"4f5c6049-c66e-4bfe-c94a-a542cf8d7003"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 2636\n","[1] 2636\n","$snpID\n"," [1] \"rs2206271\"   \"rs35305915\"  \"rs72901027\"  \"rs9367397\"   \"rs61660288\" \n"," [6] \"rs141707622\" \"rs2477488\"   \"rs13202872\"  \"rs62406441\"  \"rs17667036\" \n","[11] \"rs75397901\"  \"rs10948573\"  \"rs4715208\"   \"rs571503\"   \n","\n","$gwas_betas\n"," [1]  0.02738150 -0.03549410 -0.01634200  0.00006950 -0.03065830  0.12363800\n"," [7]  0.03828220  0.01946290  0.00770182 -0.01505850 -0.09779650  0.00386748\n","[13]  0.02011210 -0.00302458\n","\n","$gwas_se\n"," [1] 0.0236648 0.0311378 0.0344939 0.0471503 0.0354960 0.0676588 0.0224505\n"," [8] 0.0255439 0.0275417 0.0241452 0.0923557 0.0242689 0.0261452 0.0224935\n","\n","$eqtl_betas\n","              beta_1 beta_2 beta_3 beta_4 beta_5 beta_6 beta_7 beta_8 beta_9\n","rs2206271   0.160091     NA     NA     NA     NA     NA     NA     NA     NA\n","rs35305915        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs72901027        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs9367397         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs61660288        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs141707622       NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs2477488         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs13202872        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs62406441        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs17667036        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs75397901        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs10948573        NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs4715208         NA     NA     NA     NA     NA     NA     NA     NA     NA\n","rs571503          NA     NA     NA     NA     NA     NA     NA     NA     NA\n","             beta_10  beta_11 beta_12 beta_13 beta_14 beta_15 beta_16 beta_17\n","rs2206271         NA       NA      NA      NA      NA      NA      NA      NA\n","rs35305915  0.184274       NA      NA      NA      NA      NA      NA      NA\n","rs72901027        NA -0.26909      NA      NA      NA      NA      NA      NA\n","rs9367397         NA       NA      NA      NA      NA      NA      NA      NA\n","rs61660288        NA       NA      NA      NA      NA      NA      NA      NA\n","rs141707622       NA       NA      NA      NA      NA      NA      NA      NA\n","rs2477488         NA       NA      NA      NA      NA      NA      NA      NA\n","rs13202872        NA       NA      NA      NA      NA      NA      NA      NA\n","rs62406441        NA       NA      NA      NA      NA      NA      NA      NA\n","rs17667036        NA       NA      NA      NA      NA      NA      NA      NA\n","rs75397901        NA       NA      NA      NA      NA      NA      NA      NA\n","rs10948573        NA       NA      NA      NA      NA      NA      NA      NA\n","rs4715208         NA       NA      NA      NA      NA      NA      NA      NA\n","rs571503          NA       NA      NA      NA      NA      NA      NA      NA\n","             beta_18   beta_19  beta_20 beta_21 beta_22   beta_23 beta_24\n","rs2206271         NA        NA       NA      NA      NA        NA      NA\n","rs35305915        NA        NA       NA      NA      NA        NA      NA\n","rs72901027        NA        NA       NA      NA      NA        NA      NA\n","rs9367397   0.829401        NA       NA      NA      NA        NA      NA\n","rs61660288        NA -0.833804       NA      NA      NA        NA      NA\n","rs141707622       NA        NA -0.21223      NA      NA        NA      NA\n","rs2477488         NA        NA       NA      NA      NA -0.353026      NA\n","rs13202872        NA        NA       NA      NA      NA        NA      NA\n","rs62406441        NA        NA       NA      NA      NA        NA      NA\n","rs17667036        NA        NA       NA      NA      NA        NA      NA\n","rs75397901        NA        NA       NA      NA      NA        NA      NA\n","rs10948573        NA        NA       NA      NA      NA        NA      NA\n","rs4715208         NA        NA       NA      NA      NA        NA      NA\n","rs571503          NA        NA       NA      NA      NA        NA      NA\n","            beta_25   beta_26 beta_27 beta_28 beta_29  beta_30 beta_31 beta_32\n","rs2206271        NA        NA      NA      NA      NA       NA      NA      NA\n","rs35305915       NA        NA      NA      NA      NA       NA      NA      NA\n","rs72901027       NA        NA      NA      NA      NA       NA      NA      NA\n","rs9367397        NA        NA      NA      NA      NA       NA      NA      NA\n","rs61660288       NA        NA      NA      NA      NA       NA      NA      NA\n","rs141707622      NA        NA      NA      NA      NA       NA      NA      NA\n","rs2477488        NA        NA      NA      NA      NA       NA      NA      NA\n","rs13202872       NA -0.380743      NA      NA      NA       NA      NA      NA\n","rs62406441       NA        NA      NA      NA      NA 0.497656      NA      NA\n","rs17667036       NA        NA      NA      NA      NA       NA      NA      NA\n","rs75397901       NA        NA      NA      NA      NA       NA      NA      NA\n","rs10948573       NA        NA      NA      NA      NA       NA      NA      NA\n","rs4715208        NA        NA      NA      NA      NA       NA      NA      NA\n","rs571503         NA        NA      NA      NA      NA       NA      NA      NA\n","              beta_33 beta_34  beta_35 beta_36 beta_37 beta_38 beta_39\n","rs2206271          NA      NA 0.225935      NA      NA      NA      NA\n","rs35305915         NA      NA       NA      NA      NA      NA      NA\n","rs72901027         NA      NA       NA      NA      NA      NA      NA\n","rs9367397          NA      NA       NA      NA      NA      NA      NA\n","rs61660288         NA      NA       NA      NA      NA      NA      NA\n","rs141707622        NA      NA       NA      NA      NA      NA      NA\n","rs2477488          NA      NA       NA      NA      NA      NA      NA\n","rs13202872         NA      NA       NA      NA      NA      NA      NA\n","rs62406441         NA      NA       NA      NA      NA      NA      NA\n","rs17667036  -0.422984      NA       NA      NA      NA      NA      NA\n","rs75397901         NA      NA       NA      NA      NA      NA 1.06673\n","rs10948573         NA      NA       NA      NA      NA      NA      NA\n","rs4715208          NA      NA       NA      NA      NA      NA      NA\n","rs571503           NA      NA       NA      NA      NA      NA      NA\n","              beta_40   beta_41 beta_42 beta_43 beta_44   beta_45 beta_46\n","rs2206271          NA        NA      NA      NA      NA        NA      NA\n","rs35305915         NA        NA      NA      NA      NA        NA      NA\n","rs72901027         NA        NA      NA      NA      NA        NA      NA\n","rs9367397          NA        NA      NA      NA      NA        NA      NA\n","rs61660288         NA        NA      NA      NA      NA        NA      NA\n","rs141707622        NA        NA      NA      NA      NA        NA      NA\n","rs2477488          NA        NA      NA      NA      NA        NA      NA\n","rs13202872         NA        NA      NA      NA      NA        NA      NA\n","rs62406441         NA        NA      NA      NA      NA        NA      NA\n","rs17667036         NA        NA      NA      NA      NA        NA      NA\n","rs75397901         NA        NA      NA      NA      NA        NA      NA\n","rs10948573  -0.171584        NA      NA      NA      NA        NA      NA\n","rs4715208          NA -0.205251      NA      NA      NA        NA      NA\n","rs571503           NA        NA      NA      NA      NA -0.243484      NA\n","            beta_47 beta_48 beta_49\n","rs2206271        NA      NA      NA\n","rs35305915       NA      NA      NA\n","rs72901027       NA      NA      NA\n","rs9367397        NA      NA      NA\n","rs61660288       NA      NA      NA\n","rs141707622      NA      NA      NA\n","rs2477488        NA      NA      NA\n","rs13202872       NA      NA      NA\n","rs62406441       NA      NA      NA\n","rs17667036       NA      NA      NA\n","rs75397901       NA      NA      NA\n","rs10948573       NA      NA      NA\n","rs4715208        NA      NA      NA\n","rs571503         NA      NA      NA\n","\n","$eqtl_se\n","                SE_1 SE_2 SE_3 SE_4 SE_5 SE_6 SE_7 SE_8 SE_9     SE_10\n","rs2206271   0.046181   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs35305915        NA   NA   NA   NA   NA   NA   NA   NA   NA 0.0473785\n","rs72901027        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs9367397         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs61660288        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs141707622       NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs2477488         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs13202872        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs62406441        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs17667036        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs75397901        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs10948573        NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs4715208         NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","rs571503          NA   NA   NA   NA   NA   NA   NA   NA   NA        NA\n","                SE_11 SE_12 SE_13 SE_14 SE_15 SE_16 SE_17    SE_18    SE_19\n","rs2206271          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs35305915         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs72901027  0.0612013    NA    NA    NA    NA    NA    NA       NA       NA\n","rs9367397          NA    NA    NA    NA    NA    NA    NA 0.234393       NA\n","rs61660288         NA    NA    NA    NA    NA    NA    NA       NA 0.225191\n","rs141707622        NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs2477488          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs13202872         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs62406441         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs17667036         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs75397901         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs10948573         NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs4715208          NA    NA    NA    NA    NA    NA    NA       NA       NA\n","rs571503           NA    NA    NA    NA    NA    NA    NA       NA       NA\n","                SE_20 SE_21 SE_22     SE_23 SE_24 SE_25     SE_26 SE_27 SE_28\n","rs2206271          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs35305915         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs72901027         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs9367397          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs61660288         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs141707622 0.0565131    NA    NA        NA    NA    NA        NA    NA    NA\n","rs2477488          NA    NA    NA 0.0549176    NA    NA        NA    NA    NA\n","rs13202872         NA    NA    NA        NA    NA    NA 0.0399825    NA    NA\n","rs62406441         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs17667036         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs75397901         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs10948573         NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs4715208          NA    NA    NA        NA    NA    NA        NA    NA    NA\n","rs571503           NA    NA    NA        NA    NA    NA        NA    NA    NA\n","            SE_29    SE_30 SE_31 SE_32     SE_33 SE_34     SE_35 SE_36 SE_37\n","rs2206271      NA       NA    NA    NA        NA    NA 0.0414627    NA    NA\n","rs35305915     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs72901027     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs9367397      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs61660288     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs141707622    NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs2477488      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs13202872     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs62406441     NA 0.109507    NA    NA        NA    NA        NA    NA    NA\n","rs17667036     NA       NA    NA    NA 0.0931801    NA        NA    NA    NA\n","rs75397901     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs10948573     NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs4715208      NA       NA    NA    NA        NA    NA        NA    NA    NA\n","rs571503       NA       NA    NA    NA        NA    NA        NA    NA    NA\n","            SE_38    SE_39     SE_40     SE_41 SE_42 SE_43 SE_44    SE_45 SE_46\n","rs2206271      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs35305915     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs72901027     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs9367397      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs61660288     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs141707622    NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs2477488      NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs13202872     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs62406441     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs17667036     NA       NA        NA        NA    NA    NA    NA       NA    NA\n","rs75397901     NA 0.306409        NA        NA    NA    NA    NA       NA    NA\n","rs10948573     NA       NA 0.0387342        NA    NA    NA    NA       NA    NA\n","rs4715208      NA       NA        NA 0.0365271    NA    NA    NA       NA    NA\n","rs571503       NA       NA        NA        NA    NA    NA    NA 0.037926    NA\n","            SE_47 SE_48 SE_49\n","rs2206271      NA    NA    NA\n","rs35305915     NA    NA    NA\n","rs72901027     NA    NA    NA\n","rs9367397      NA    NA    NA\n","rs61660288     NA    NA    NA\n","rs141707622    NA    NA    NA\n","rs2477488      NA    NA    NA\n","rs13202872     NA    NA    NA\n","rs62406441     NA    NA    NA\n","rs17667036     NA    NA    NA\n","rs75397901     NA    NA    NA\n","rs10948573     NA    NA    NA\n","rs4715208      NA    NA    NA\n","rs571503       NA    NA    NA\n","\n","$eqtl_pvals\n","               pvalue_1 pvalue_2 pvalue_3 pvalue_4 pvalue_5 pvalue_6 pvalue_7\n","rs2206271   0.000571487       NA       NA       NA       NA       NA       NA\n","rs35305915           NA       NA       NA       NA       NA       NA       NA\n","rs72901027           NA       NA       NA       NA       NA       NA       NA\n","rs9367397            NA       NA       NA       NA       NA       NA       NA\n","rs61660288           NA       NA       NA       NA       NA       NA       NA\n","rs141707622          NA       NA       NA       NA       NA       NA       NA\n","rs2477488            NA       NA       NA       NA       NA       NA       NA\n","rs13202872           NA       NA       NA       NA       NA       NA       NA\n","rs62406441           NA       NA       NA       NA       NA       NA       NA\n","rs17667036           NA       NA       NA       NA       NA       NA       NA\n","rs75397901           NA       NA       NA       NA       NA       NA       NA\n","rs10948573           NA       NA       NA       NA       NA       NA       NA\n","rs4715208            NA       NA       NA       NA       NA       NA       NA\n","rs571503             NA       NA       NA       NA       NA       NA       NA\n","            pvalue_8 pvalue_9   pvalue_10  pvalue_11 pvalue_12 pvalue_13\n","rs2206271         NA       NA          NA         NA        NA        NA\n","rs35305915        NA       NA 0.000157047         NA        NA        NA\n","rs72901027        NA       NA          NA 1.9369e-05        NA        NA\n","rs9367397         NA       NA          NA         NA        NA        NA\n","rs61660288        NA       NA          NA         NA        NA        NA\n","rs141707622       NA       NA          NA         NA        NA        NA\n","rs2477488         NA       NA          NA         NA        NA        NA\n","rs13202872        NA       NA          NA         NA        NA        NA\n","rs62406441        NA       NA          NA         NA        NA        NA\n","rs17667036        NA       NA          NA         NA        NA        NA\n","rs75397901        NA       NA          NA         NA        NA        NA\n","rs10948573        NA       NA          NA         NA        NA        NA\n","rs4715208         NA       NA          NA         NA        NA        NA\n","rs571503          NA       NA          NA         NA        NA        NA\n","            pvalue_14 pvalue_15 pvalue_16 pvalue_17   pvalue_18   pvalue_19\n","rs2206271          NA        NA        NA        NA          NA          NA\n","rs35305915         NA        NA        NA        NA          NA          NA\n","rs72901027         NA        NA        NA        NA          NA          NA\n","rs9367397          NA        NA        NA        NA 0.000609782          NA\n","rs61660288         NA        NA        NA        NA          NA 0.000369504\n","rs141707622        NA        NA        NA        NA          NA          NA\n","rs2477488          NA        NA        NA        NA          NA          NA\n","rs13202872         NA        NA        NA        NA          NA          NA\n","rs62406441         NA        NA        NA        NA          NA          NA\n","rs17667036         NA        NA        NA        NA          NA          NA\n","rs75397901         NA        NA        NA        NA          NA          NA\n","rs10948573         NA        NA        NA        NA          NA          NA\n","rs4715208          NA        NA        NA        NA          NA          NA\n","rs571503           NA        NA        NA        NA          NA          NA\n","              pvalue_20 pvalue_21 pvalue_22   pvalue_23 pvalue_24 pvalue_25\n","rs2206271            NA        NA        NA          NA        NA        NA\n","rs35305915           NA        NA        NA          NA        NA        NA\n","rs72901027           NA        NA        NA          NA        NA        NA\n","rs9367397            NA        NA        NA          NA        NA        NA\n","rs61660288           NA        NA        NA          NA        NA        NA\n","rs141707622 0.000204908        NA        NA          NA        NA        NA\n","rs2477488            NA        NA        NA 6.03875e-10        NA        NA\n","rs13202872           NA        NA        NA          NA        NA        NA\n","rs62406441           NA        NA        NA          NA        NA        NA\n","rs17667036           NA        NA        NA          NA        NA        NA\n","rs75397901           NA        NA        NA          NA        NA        NA\n","rs10948573           NA        NA        NA          NA        NA        NA\n","rs4715208            NA        NA        NA          NA        NA        NA\n","rs571503             NA        NA        NA          NA        NA        NA\n","              pvalue_26 pvalue_27 pvalue_28 pvalue_29   pvalue_30 pvalue_31\n","rs2206271            NA        NA        NA        NA          NA        NA\n","rs35305915           NA        NA        NA        NA          NA        NA\n","rs72901027           NA        NA        NA        NA          NA        NA\n","rs9367397            NA        NA        NA        NA          NA        NA\n","rs61660288           NA        NA        NA        NA          NA        NA\n","rs141707622          NA        NA        NA        NA          NA        NA\n","rs2477488            NA        NA        NA        NA          NA        NA\n","rs13202872  1.27032e-19        NA        NA        NA          NA        NA\n","rs62406441           NA        NA        NA        NA 3.72902e-05        NA\n","rs17667036           NA        NA        NA        NA          NA        NA\n","rs75397901           NA        NA        NA        NA          NA        NA\n","rs10948573           NA        NA        NA        NA          NA        NA\n","rs4715208            NA        NA        NA        NA          NA        NA\n","rs571503             NA        NA        NA        NA          NA        NA\n","            pvalue_32   pvalue_33 pvalue_34  pvalue_35 pvalue_36 pvalue_37\n","rs2206271          NA          NA        NA 8.2487e-08        NA        NA\n","rs35305915         NA          NA        NA         NA        NA        NA\n","rs72901027         NA          NA        NA         NA        NA        NA\n","rs9367397          NA          NA        NA         NA        NA        NA\n","rs61660288         NA          NA        NA         NA        NA        NA\n","rs141707622        NA          NA        NA         NA        NA        NA\n","rs2477488          NA          NA        NA         NA        NA        NA\n","rs13202872         NA          NA        NA         NA        NA        NA\n","rs62406441         NA          NA        NA         NA        NA        NA\n","rs17667036         NA 1.35883e-05        NA         NA        NA        NA\n","rs75397901         NA          NA        NA         NA        NA        NA\n","rs10948573         NA          NA        NA         NA        NA        NA\n","rs4715208          NA          NA        NA         NA        NA        NA\n","rs571503           NA          NA        NA         NA        NA        NA\n","            pvalue_38   pvalue_39   pvalue_40   pvalue_41 pvalue_42 pvalue_43\n","rs2206271          NA          NA          NA          NA        NA        NA\n","rs35305915         NA          NA          NA          NA        NA        NA\n","rs72901027         NA          NA          NA          NA        NA        NA\n","rs9367397          NA          NA          NA          NA        NA        NA\n","rs61660288         NA          NA          NA          NA        NA        NA\n","rs141707622        NA          NA          NA          NA        NA        NA\n","rs2477488          NA          NA          NA          NA        NA        NA\n","rs13202872         NA          NA          NA          NA        NA        NA\n","rs62406441         NA          NA          NA          NA        NA        NA\n","rs17667036         NA          NA          NA          NA        NA        NA\n","rs75397901         NA 0.000625214          NA          NA        NA        NA\n","rs10948573         NA          NA 1.18765e-05          NA        NA        NA\n","rs4715208          NA          NA          NA 3.09081e-08        NA        NA\n","rs571503           NA          NA          NA          NA        NA        NA\n","            pvalue_44   pvalue_45 pvalue_46 pvalue_47 pvalue_48 pvalue_49\n","rs2206271          NA          NA        NA        NA        NA        NA\n","rs35305915         NA          NA        NA        NA        NA        NA\n","rs72901027         NA          NA        NA        NA        NA        NA\n","rs9367397          NA          NA        NA        NA        NA        NA\n","rs61660288         NA          NA        NA        NA        NA        NA\n","rs141707622        NA          NA        NA        NA        NA        NA\n","rs2477488          NA          NA        NA        NA        NA        NA\n","rs13202872         NA          NA        NA        NA        NA        NA\n","rs62406441         NA          NA        NA        NA        NA        NA\n","rs17667036         NA          NA        NA        NA        NA        NA\n","rs75397901         NA          NA        NA        NA        NA        NA\n","rs10948573         NA          NA        NA        NA        NA        NA\n","rs4715208          NA          NA        NA        NA        NA        NA\n","rs571503           NA 6.19974e-10        NA        NA        NA        NA\n","\n","$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n"]}]},{"cell_type":"code","source":["# Step 1: Load the required libraries:\n","library(base) # R package installed by default\n","library(utils) # R package installed by default\n","library(parallel) # R package installed by default\n","library(Matrix) # R package installed by default\n","library(methods) # R package installed by default\n","library(stats) # R package installed by default\n","library(lme4) # It needs to be installed separately\n","library(data.table) # It needs to be installed separately\n","\n","# Step 2: Main Mt_Robin function:\n","MR_MtRobin <- function(MR_MtRobin_input,pval_thresh=0.001){\n","  snpID <- MR_MtRobin_input$snpID\n","  gwas_betas <- MR_MtRobin_input$gwas_betas\n","  gwas_se <- MR_MtRobin_input$gwas_se\n","  eqtl_betas <- MR_MtRobin_input$eqtl_betas\n","  eqtl_se <- MR_MtRobin_input$eqtl_se\n","  eqtl_pvals <- MR_MtRobin_input$eqtl_pvals\n","  LD <- MR_MtRobin_input$LD\n","\n","  if(ncol(eqtl_betas) != ncol(eqtl_se) | ncol(eqtl_betas) != ncol(eqtl_pvals)){\n","    stop(\"eQTL betas, SE and p-values do not all have same number of columns\")\n","  }\n","\n","  ## determine number of tissues (or conditions/studies/etc.)\n","  nT <- ncol(eqtl_betas)\n","\n","  ## rename columns for standardized processing\n","  colnames(eqtl_betas) <- paste0(\"beta_\",1:nT)\n","  colnames(eqtl_se) <- paste0(\"SE_\",1:nT)\n","  colnames(eqtl_pvals) <- paste0(\"pvalue_\",1:nT)\n","\n","  ## combine eQTL results into data.table\n","  eqtl_res <- data.table::data.table(snpID,eqtl_betas,eqtl_se,eqtl_pvals)\n","\n","  # reshape to long format (each SNP-gene-tissue will be row)\n","  eQTL_res_melt <- data.table::melt(eqtl_res,id.vars=c(\"snpID\"),\n","                                    measure.vars=patterns(\"^beta_\",\"^SE_\",\"^pvalue_\"),value.name=c(\"beta\",\"se\",\"pvalue\"),\n","                                    variable.name=\"tissue\")\n","\n","  ## subset to strong instrumental variables (based on p-value threshold)\n","  eQTL_res_melt_PltThresh <- subset(eQTL_res_melt, pvalue < pval_thresh)\n","\n","  ## set up gwas data for merging\n","  gwas_res <- data.table::data.table(snpID=snpID,gwas_beta=gwas_betas,gwas_se=gwas_se)\n","\n","  ## merge eQTL and GWAS results\n","  data.table::setkey(eQTL_res_melt_PltThresh,snpID)\n","  data.table::setkey(gwas_res,snpID)\n","  merged_res <- data.table::merge.data.table(eQTL_res_melt_PltThresh,gwas_res)\n","\n","  ## set up coefficients for reverse regression\n","  beta_x <- matrix(merged_res$beta,ncol=1)\n","  beta_y <- matrix(merged_res$gwas_beta,ncol=1)\n","\n","  ## standard errors (for weights)\n","  se_x <- matrix(merged_res$se,ncol=1)\n","  se_y <- matrix(merged_res$gwas_se,ncol=1) ## not used by function; used in resampling (consider returning with results)\n","\n","  ## identifiers\n","  snpID <- merged_res$snpID\n","\n","  lme_res <- lme4::lmer(beta_x~(beta_y-1)+(beta_y-1|snpID),weights=1/se_x^2)\n","\n","  ## return results from weighted regression with random slopes (and correlated errors) along with GWAS SE and LD (needed for resampling)\n","  return(list(lme_res=lme_res,gwas_res=gwas_res,LD=LD))\n","}\n","\n","# Step 3: Read the previous file:\n","result_setup <- readRDS(\"/data/long_covid_setup_bs_4.rds\")\n","\n","# Step 4: Apply the main Mt_Robin function using parallel processing:\n","no_cores <- 36\n","\n","cl <- makeCluster(no_cores) # Create a cluster with the number of cores\n","clusterExport(cl, \"MR_MtRobin\") # Export the MR_MtRobin function to all the nodes of the cluster\n","result_list_MtRobin <- parLapply(cl, result_setup, function(MR_MtRobin_input) {\n","  tryCatch({\n","    MR_MtRobin(MR_MtRobin_input = MR_MtRobin_input)\n","  }, error = function(e) print(e))\n","})\n","stopCluster(cl) # Stop the cluster once the computation is done\n","\n","# Step 5: The names of the results will correspond to the gene ids:\n","names(result_list_MtRobin) <- names(result_setup)\n","\n","# Step 6: Save the previous result as a .rds file:\n","saveRDS(result_list_MtRobin, file = \"/data/long_covid_main_bs_4.rds\")"],"metadata":{"id":"bQ4WF6YSjf4C"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=36\n","#PBS -l mem=300GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/Mt_Robin/main_function.sif Rscript /data/main_function_bs_4.R\n","\n","# Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly\n","exit 0"],"metadata":{"id":"Dv0aj2Oqj4g2"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2023-08-30 10:32:18:\n","   Job Id:             93872803.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      2.25\n","   NCPUs Requested:    36                     NCPUs Used: 36\n","                                           CPU Time Used: 00:01:46\n","   Memory Requested:   300.0GB               Memory Used: 8.04GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:01:15\n","   JobFS requested:    200.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"LOubMP7Wj9q0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","# Open the output file:\n","result_main_bs_4 <- readRDS(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_4.rds\")\n","\n","# Filter out errors:\n","result_main_bs_4 <- result_main_bs_4[!sapply(result_main_bs_4, function(x) inherits(x, \"simpleError\"))]\n","\n","# Print the length and the first element:\n","print(length(result_main_bs_4))\n","print(length(unique(result_main_bs_4)))\n","print(head(result_main_bs_4))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ODc5uQuJj8zW","executionInfo":{"status":"ok","timestamp":1699488141678,"user_tz":-630,"elapsed":9282,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"8f5efd56-5785-499f-f4ab-d8eed2f2063b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 145\n","[1] 145\n","$ENSG00000008196\n","$ENSG00000008196$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 7.9038\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 0.000   \n"," Residual        5.069   \n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-2.161  \n","optimizer (nloptwrap) convergence code: 0 (OK) ; 0 optimizer warnings; 1 lme4 warnings \n","\n","$ENSG00000008196$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs10948573  0.00386748 0.0242689\n","2   rs13202872  0.01946290 0.0255439\n","3  rs141707622  0.12363800 0.0676588\n","4   rs17667036 -0.01505850 0.0241452\n","5    rs2206271  0.02738150 0.0236648\n","6    rs2477488  0.03828220 0.0224505\n","7   rs35305915 -0.03549410 0.0311378\n","8    rs4715208  0.02011210 0.0261452\n","9     rs571503 -0.00302458 0.0224935\n","10  rs61660288 -0.03065830 0.0354960\n","11  rs62406441  0.00770182 0.0275417\n","12  rs72901027 -0.01634200 0.0344939\n","13  rs75397901 -0.09779650 0.0923557\n","14   rs9367397  0.00006950 0.0471503\n","\n","$ENSG00000008196$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n","\n","$ENSG00000068781\n","$ENSG00000068781$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -28.1355\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 65.653  \n"," Residual         2.485  \n","Number of obs: 29, groups:  snpID, 2\n","Fixed Effects:\n","beta_y  \n"," 8.932  \n","\n","$ENSG00000068781$gwas_res\n","       snpID gwas_beta   gwas_se\n","1  rs2140551 0.0146944 0.0224589\n","2 rs62135453 0.0188507 0.0263231\n","\n","$ENSG00000068781$LD\n","             rs2140551  rs62135453\n","rs2140551   1.00000000 -0.01964898\n","rs62135453 -0.01964898  1.00000000\n","\n","\n","$ENSG00000112494\n","$ENSG00000112494$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 11.4942\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 23.710  \n"," Residual         3.311  \n","Number of obs: 9, groups:  snpID, 8\n","Fixed Effects:\n","beta_y  \n"," -6.11  \n","\n","$ENSG00000112494$gwas_res\n","        snpID   gwas_beta   gwas_se\n","1 rs117351628 -0.04908130 0.0815733\n","2 rs141514782  0.08384400 0.0598866\n","3 rs149028144 -0.02728360 0.1723020\n","4   rs1736600  0.01129700 0.0286659\n","5   rs3093009  0.00113459 0.0269883\n","6  rs55714481 -0.05503320 0.0621256\n","7  rs62438461  0.02744390 0.0245558\n","8   rs7766490 -0.03366910 0.0322543\n","\n","$ENSG00000112494$LD\n","            rs117351628  rs62438461    rs55714481 rs149028144   rs141514782\n","rs117351628  1.00000000  0.04888374  0.0720036892 -0.03624723  0.0452490319\n","rs62438461   0.04888374  1.00000000  0.0278140914  0.01209383 -0.0555439065\n","rs55714481   0.07200369  0.02781409  1.0000000000 -0.03467921 -0.0001274082\n","rs149028144 -0.03624723  0.01209383 -0.0346792109  1.00000000 -0.0138260032\n","rs141514782  0.04524903 -0.05554391 -0.0001274082 -0.01382600  1.0000000000\n","rs1736600   -0.05817461 -0.40929121  0.0077447039 -0.03015317 -0.0306809936\n","rs7766490   -0.01363783 -0.11982939 -0.0290568889 -0.01900376  0.0434365367\n","rs3093009   -0.00377826 -0.14430582 -0.0042937091 -0.03177324 -0.0506560837\n","               rs1736600   rs7766490    rs3093009\n","rs117351628 -0.058174609 -0.01363783 -0.003778260\n","rs62438461  -0.409291211 -0.11982939 -0.144305816\n","rs55714481   0.007744704 -0.02905689 -0.004293709\n","rs149028144 -0.030153172 -0.01900376 -0.031773236\n","rs141514782 -0.030680994  0.04343654 -0.050656084\n","rs1736600    1.000000000  0.12600742 -0.006354755\n","rs7766490    0.126007425  1.00000000 -0.218761737\n","rs3093009   -0.006354755 -0.21876174  1.000000000\n","\n","\n","$ENSG00000113905\n","$ENSG00000113905$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 25.9173\n","Random effects:\n"," Groups   Name   Std.Dev. \n"," snpID    beta_y 7.120e-07\n"," Residual        4.331e+00\n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-1.004  \n","optimizer (nloptwrap) convergence code: 0 (OK) ; 0 optimizer warnings; 1 lme4 warnings \n","\n","$ENSG00000113905$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1    rs1047115  0.01499370 0.0400648\n","2  rs112184129  0.11151000 0.0820665\n","3  rs114922365  0.00715984 0.0392488\n","4  rs140494728  0.00944623 0.0634965\n","5  rs144876286 -0.08669670 0.0554271\n","6  rs148601582  0.18900300 0.1288250\n","7    rs1868144  0.00296522 0.0299292\n","8    rs4564971 -0.03020450 0.0600564\n","9    rs4610218  0.00357956 0.0286432\n","10   rs6768529 -0.03159150 0.0282529\n","11  rs75211540 -0.33623900 0.1787070\n","12   rs7614563 -0.02256920 0.0283285\n","13  rs77264387 -0.05107820 0.0686270\n","14  rs78418844  0.03988440 0.0624887\n","\n","$ENSG00000113905$LD\n","              rs78418844  rs114922365   rs140494728  rs144876286   rs6768529\n","rs78418844   1.000000000 -0.049041226  0.0372760520 -0.015417463 -0.05840539\n","rs114922365 -0.049041226  1.000000000  0.0200185378 -0.016183532  0.01618302\n","rs140494728  0.037276052  0.020018538  1.0000000000 -0.038023416  0.01513508\n","rs144876286 -0.015417463 -0.016183532 -0.0380234158  1.000000000 -0.03912501\n","rs6768529   -0.058405392  0.016183019  0.0151350751 -0.039125011  1.00000000\n","rs112184129 -0.018220333  0.002736677  0.0272553987 -0.009747483 -0.04189426\n","rs1868144    0.011269378 -0.123567413 -0.0330853646  0.181246370 -0.01813303\n","rs7614563   -0.005237771 -0.039546627  0.0964756129  0.001626079 -0.08580770\n","rs4564971    0.043457967  0.028551916 -0.0003970324  0.054613924 -0.05460910\n","rs1047115   -0.002983096  0.171013423  0.4568884832 -0.017073426  0.01155008\n","rs77264387  -0.048140533 -0.014011149  0.0040859415  0.061341907 -0.01325065\n","rs4610218    0.041014094 -0.079111068 -0.2196378128 -0.052987992 -0.03957183\n","rs75211540   0.026292665 -0.037803432 -0.0483213675  0.006157921 -0.04051424\n","rs148601582  0.029135418  0.028638684 -0.0470038433 -0.046235467 -0.00840697\n","             rs112184129    rs1868144    rs7614563     rs4564971    rs1047115\n","rs78418844  -0.018220333  0.011269378 -0.005237771  0.0434579673 -0.002983096\n","rs114922365  0.002736677 -0.123567413 -0.039546627  0.0285519162  0.171013423\n","rs140494728  0.027255399 -0.033085365  0.096475613 -0.0003970324  0.456888483\n","rs144876286 -0.009747483  0.181246370  0.001626079  0.0546139244 -0.017073426\n","rs6768529   -0.041894258 -0.018133028 -0.085807703 -0.0546090978  0.011550077\n","rs112184129  1.000000000  0.036934859 -0.006472562  0.0837359010  0.042625994\n","rs1868144    0.036934859  1.000000000 -0.054654276  0.0761246422 -0.042221185\n","rs7614563   -0.006472562 -0.054654276  1.000000000 -0.0579202386  0.064351808\n","rs4564971    0.083735901  0.076124642 -0.057920239  1.0000000000  0.012769231\n","rs1047115    0.042625994 -0.042221185  0.064351808  0.0127692313  1.000000000\n","rs77264387   0.062536828 -0.004232278 -0.034626931 -0.0504431370  0.050633731\n","rs4610218    0.017925731 -0.035697786  0.033519213 -0.0439717467 -0.178405672\n","rs75211540   0.018979073  0.062158325 -0.002593466  0.0513491854 -0.035294427\n","rs148601582  0.068431974 -0.011475030 -0.048272949  0.0235988865  0.002483828\n","              rs77264387    rs4610218   rs75211540  rs148601582\n","rs78418844  -0.048140533  0.041014094  0.026292665  0.029135418\n","rs114922365 -0.014011149 -0.079111068 -0.037803432  0.028638684\n","rs140494728  0.004085941 -0.219637813 -0.048321368 -0.047003843\n","rs144876286  0.061341907 -0.052987992  0.006157921 -0.046235467\n","rs6768529   -0.013250651 -0.039571830 -0.040514241 -0.008406970\n","rs112184129  0.062536828  0.017925731  0.018979073  0.068431974\n","rs1868144   -0.004232278 -0.035697786  0.062158325 -0.011475030\n","rs7614563   -0.034626931  0.033519213 -0.002593466 -0.048272949\n","rs4564971   -0.050443137 -0.043971747  0.051349185  0.023598887\n","rs1047115    0.050633731 -0.178405672 -0.035294427  0.002483828\n","rs77264387   1.000000000  0.020129987 -0.007845105 -0.005914902\n","rs4610218    0.020129987  1.000000000  0.004554648  0.038764762\n","rs75211540  -0.007845105  0.004554648  1.000000000  0.088022541\n","rs148601582 -0.005914902  0.038764762  0.088022541  1.000000000\n","\n","\n","$ENSG00000117148\n","$ENSG00000117148$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -0.5075\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 8.100   \n"," Residual        4.359   \n","Number of obs: 5, groups:  snpID, 4\n","Fixed Effects:\n","beta_y  \n","-2.904  \n","\n","$ENSG00000117148$gwas_res\n","       snpID   gwas_beta   gwas_se\n","1 rs17362289 -0.00167287 0.0450904\n","2 rs34519427 -0.08333330 0.0530907\n","3  rs4141983  0.02863170 0.0240003\n","4   rs478698 -0.00256889 0.0321287\n","\n","$ENSG00000117148$LD\n","              rs478698   rs17362289  rs4141983   rs34519427\n","rs478698    1.00000000  0.038323461 0.09328150 -0.070593355\n","rs17362289  0.03832346  1.000000000 0.06398760 -0.006481147\n","rs4141983   0.09328150  0.063987598 1.00000000  0.059969317\n","rs34519427 -0.07059335 -0.006481147 0.05996932  1.000000000\n","\n","\n","$ENSG00000122965\n","$ENSG00000122965$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: -20.7081\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 8.687   \n"," Residual        4.152   \n","Number of obs: 49, groups:  snpID, 10\n","Fixed Effects:\n","beta_y  \n","-3.603  \n","\n","$ENSG00000122965$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs11066483 -0.02841120 0.0256192\n","2   rs11066771  0.01373880 0.0236739\n","3   rs11615428 -0.01659330 0.0310452\n","4  rs116976487  0.10003200 0.0907649\n","5   rs12369142 -0.01538810 0.0265626\n","6  rs147783556 -0.02014510 0.0542254\n","7    rs1870352  0.00221817 0.0256381\n","8    rs4767149  0.00304759 0.0268834\n","9    rs7977045 -0.02615790 0.0235846\n","10  rs80290311  0.02390540 0.0671365\n","\n","$ENSG00000122965$LD\n","               rs7977045  rs147783556    rs1870352   rs11066483 rs116976487\n","rs7977045    1.000000000 -0.206712295  0.081873465 -0.008454639 -0.05750897\n","rs147783556 -0.206712295  1.000000000 -0.007292528  0.035328443  0.05280153\n","rs1870352    0.081873465 -0.007292528  1.000000000  0.026063407 -0.04414327\n","rs11066483  -0.008454639  0.035328443  0.026063407  1.000000000  0.01662551\n","rs116976487 -0.057508966  0.052801529 -0.044143270  0.016625511  1.00000000\n","rs4767149    0.171418783 -0.138617529  0.020116315 -0.088660989 -0.01959316\n","rs12369142  -0.415600450  0.345253496 -0.076423313  0.007411732  0.06514148\n","rs11615428   0.505037949 -0.121173142  0.242169899 -0.020699028 -0.05879359\n","rs11066771  -0.616799801  0.336137832 -0.064410973  0.009494328  0.04829664\n","rs80290311   0.068027570 -0.084604531  0.208119182  0.063108318 -0.02321683\n","              rs4767149   rs12369142  rs11615428   rs11066771  rs80290311\n","rs7977045    0.17141878 -0.415600450  0.50503795 -0.616799801  0.06802757\n","rs147783556 -0.13861753  0.345253496 -0.12117314  0.336137832 -0.08460453\n","rs1870352    0.02011631 -0.076423313  0.24216990 -0.064410973  0.20811918\n","rs11066483  -0.08866099  0.007411732 -0.02069903  0.009494328  0.06310832\n","rs116976487 -0.01959316  0.065141483 -0.05879359  0.048296638 -0.02321683\n","rs4767149    1.00000000 -0.162747818  0.57376537 -0.288171436 -0.05362295\n","rs12369142  -0.16274782  1.000000000 -0.23656439  0.686835278 -0.08084646\n","rs11615428   0.57376537 -0.236564386  1.00000000 -0.332376400  0.23255788\n","rs11066771  -0.28817144  0.686835278 -0.33237640  1.000000000 -0.06798849\n","rs80290311  -0.05362295 -0.080846462  0.23255788 -0.067988491  1.00000000\n","\n","\n"]}]},{"cell_type":"code","source":["########################################################################\n","# GADI:\n","\n","# Obtain p-value from MR-MtRobin results:\n","# Uses a resampling procedure to estimate the p_value for a MR-MtRobin object:\n","# param MR_MtRobin_res a list object returned by \\code{MR_MtRobin}.\n","# param nsamp integer of the number of samples to use in estimating \\eqn{P}-value using a null distribution.\n","# param use_nonconverge logical of whether to use samples resulting in non-convergence of random slope\n","# return A list of two elements:\n","# \\tabular{ll}{\\code{pvalue} \\tab numeric of the estimated \\eqn{P}-value.\n","# \\code{nsamp_used} \\tab integer of the number of samples used in estimating the \\eqn{P}-value.\n","# \\code{nsamp_used} is returned because not all \\code{nsamp} may be used +\n","# If \\code{use_nonconverge=FALSE} (samples will be dropped if the model does not converge or results in a singular fit of the random slope).\n","# myRes <- MR_MtRobin(snpID,gwas_betas,gwas_se,eqtl_betas,eqtl_se,eqtl_pvals,LD)\n","# MR_MtRobin_resample(MR_MtRobin_res=myRes, nsamp=1e4, use_nonconverge=FALSE)\n","\n","# Load packages:\n","library(lme4)\n","library(mvtnorm)\n","library(data.table)\n","library(parallel)\n","\n","# Load main results:\n","results_1 <- readRDS(\"/scratch/sq95/sp6154/p_values/long_covid_main_bs_4.rds\")\n","# Filter out errors:\n","results_1 <- results_1[!sapply(results_1, function(x) inherits(x, \"simpleError\"))]\n","# Export:\n","saveRDS(results_1, \"/scratch/sq95/sp6154/p_values/long_covid_main_bs_4_no_errors.rds\")\n","\n","# Updated Function to calculate the p-value and FDR with parallel processing:\n","MR_MtRobin_resample <- function(results_1_path, nsamp=1e4, use_nonconverge=FALSE, output_file_path, num_cpus=1){\n","  # Check and load the results_1 dataset from the specified path:\n","  if(file.exists(results_1_path)){\n","    results_1 <- readRDS(results_1_path)\n","  } else {\n","    stop(\"The specified file path for results_1 does not exist.\")\n","  }\n","  # Initialize a list to store results for each gene:\n","  pvalues_results <- list()\n","  # Loop over each gene:\n","  for(gene_id in names(results_1)) {\n","    # Check if necessary components are present and not NULL:\n","    if(is.null(results_1[[gene_id]][[\"lme_res\"]]) || is.null(results_1[[gene_id]][[\"gwas_res\"]]) || is.null(results_1[[gene_id]][[\"LD\"]])) {\n","      next\n","    }\n","    # Create slopes for each information:\n","    lme_res <- results_1[[gene_id]][[\"lme_res\"]]\n","    gwas_res <- results_1[[gene_id]][[\"gwas_res\"]]\n","    LD <- results_1[[gene_id]][[\"LD\"]]\n","    # Extract data from MR-MtRobin results:\n","    eqtl_betas <- lme_res@frame$beta_x\n","    snpID <- lme_res@frame$snpID\n","    weights <- lme_res@frame$weights\n","    tstat_MR_MtRobin <- summary(lme_res)$coefficients[1,3]\n","    gwas_se <- gwas_res$gwas_se\n","    # Bootstrapped null distribution, accounting for LD correlations:\n","    beta_gwas_nullMat <- mvtnorm::rmvnorm(nsamp, mean=rep(0, length(gwas_se)), sigma=diag(gwas_se) %*% LD %*% diag(gwas_se))\n","    colnames(beta_gwas_nullMat) <- gwas_res$snpID\n","    # Return data types:\n","    tstat_nulls <- mclapply(1:nsamp, function(i) {\n","      beta_gwas_null <- beta_gwas_nullMat[i, snpID]\n","      tryCatch({\n","        # LMM equation:\n","        lme_null <- lme4::lmer(eqtl_betas~(beta_gwas_null-1)+(beta_gwas_null-1|snpID), weights=weights)\n","        return(summary(lme_null)$coefficients[1,3])\n","      }, error=function(e) {\n","        if(!use_nonconverge) return(NA)\n","      })\n","    }, mc.cores = num_cpus)\n","    tstat_nulls <- unlist(tstat_nulls)\n","    tstat_nulls <- tstat_nulls[!is.na(tstat_nulls)] # Remove NAs\n","    nsamp_used <- length(tstat_nulls)\n","    # Calculate the p-value and round it to 3 decimal places:\n","    pval <- round(mean(abs(tstat_nulls) >= abs(tstat_MR_MtRobin)), digits = 3)\n","    # Store the results for the current gene in the list:\n","    pvalues_results[[gene_id]] <- list(pvalue=pval, nsamp_used=nsamp_used)\n","  }\n","  # Calculate FDR:\n","  all_pvalues <- sapply(pvalues_results, function(x) x$pvalue)\n","  fdr_values <- p.adjust(all_pvalues, method = \"BH\")\n","  # Combine results with FDR values:\n","  for(i in seq_along(fdr_values)) {\n","    pvalues_results[[names(fdr_values)[i]]]$fdr <- fdr_values[i]\n","  }\n","  # Prepare results dataframe:\n","  results_df <- do.call(rbind, lapply(names(pvalues_results), function(gene_id) {\n","    data.frame(gene_id = gene_id,\n","               pvalue = sprintf(\"%.3f\", pvalues_results[[gene_id]]$pvalue), # Format pvalue with 3 decimal places\n","               nsamp_used = pvalues_results[[gene_id]]$nsamp_used,\n","               fdr = sprintf(\"%.15f\", pvalues_results[[gene_id]]$fdr), # Keep full precision for fdr\n","               stringsAsFactors = FALSE)\n","  }))\n","  # Write the results to the output file, ensuring the formatting is kept\n","  write.table(results_df, file = output_file_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = TRUE)\n","  return(pvalues_results)\n","}\n","\n","# Call the previous function:\n","MR_MtRobin_resample(\n","  results_1_path = \"/scratch/sq95/sp6154/p_values/long_covid_main_bs_4_no_errors.rds\",\n","  nsamp = 10000, # Number of bootstrap samples\n","  use_nonconverge = FALSE, # Whether to use non-converging results, typically FALSE\n","  output_file_path = \"/scratch/sq95/sp6154/p_values/MR_pvalue_FDR_results_bs4.txt\",\n","  num_cpus = 48 # Number of CPUs\n",")"],"metadata":{"id":"Ephz_Y59XXnk"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Specify the location of the program to be used:\n","#!/bin/bash\n","\n","# Specify the project code:\n","#PBS -P sq95\n","\n","# Specify the queue:\n","#PBS -q hugemem\n","\n","# Request resources:\n","#PBS -l ncpus=48\n","#PBS -l mem=500GB\n","#PBS -l jobfs=200GB\n","#PBS -l walltime=10:00:00\n","\n","# Request the system to enter and save the job in the working directory:\n","#PBS -l wd\n","\n","# Create different output files instead of overwriting the same file and specify the name:\n","date > test1_out_$PBS_JOBID.txt\n","sleep 60\n","date >> test1_out$PBS_JOBID.txt\n","\n","# Move the output to the directory where the job was submitted:\n","cd $PBS_O_WORKDIR\n","\n","# Receive alerts by email - aborted(a), begins(b), ends(e):\n","#PBS -M pinsy007@mymail.unisa.edu.au\n","#PBS -m abe\n","\n","###########################################################################################################\n","# Specify the work to be done:\n","\n","# Load the required modules:\n","module load R/4.2.2\n","module load gcc/12.2.0\n","module load intel-mkl/2021.2.0\n","module load singularity\n","\n","# Option 1: Run the R script using the Singularity container:\n","singularity exec --bind $PBS_O_WORKDIR:/data /scratch/sq95/sp6154/p_values/p_values.sif Rscript /data/p_values_bs_4.R\n","\n","# Option 2: Run the R script without any container:\n","#R --vanilla < parallelism_test.R > output_test\n","\n","###########################################################################################################\n","\n","# Exit cleanly:\n","exit 0"],"metadata":{"id":"xvA2o3YPXXr_"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["======================================================================================\n","                  Resource Usage on 2024-02-08 18:06:52:\n","   Job Id:             107837488.gadi-pbs\n","   Project:            sq95\n","   Exit Status:        0\n","   Service Units:      33.94\n","   NCPUs Requested:    24                     NCPUs Used: 24\n","                                           CPU Time Used: 10:08:22\n","   Memory Requested:   200.0GB               Memory Used: 19.28GB\n","   Walltime requested: 10:00:00            Walltime Used: 00:28:17\n","   JobFS requested:    100.0GB                JobFS used: 0B\n","======================================================================================"],"metadata":{"id":"Tw738ABkXX2a"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["%%R\n","\n","# Read the results with the p_value and FDR values:\n","results <- read.table(\"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_bs4.txt\", header=TRUE, sep=\"\\t\")\n","print(dim(results))\n","print(head(results))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"vr-60jGFXWzf","executionInfo":{"status":"ok","timestamp":1707376988560,"user_tz":-630,"elapsed":542,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"982af8ef-dc8b-40a4-f986-1a874236eb01"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] 145   4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000008196  0.258      10000 0.8382192\n","2 ENSG00000068781  0.878      10000 0.9583083\n","3 ENSG00000112494  0.655      10000 0.8952752\n","4 ENSG00000113905  0.776      10000 0.9132661\n","5 ENSG00000117148  0.745      10000 0.9132661\n","6 ENSG00000122965  0.374      10000 0.8382192\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# Call the analyse_gene_data() function to organize the Mt_Robin results for significant and non-significant genes:\n","\n","# Create the output paths:\n","save_paths <- list(\n","  all_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs4_all_genes_sig_non_sig.txt\",\n","  protein_coding_genes = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs4_protein_coding_genes_sig_not_sig.txt\"\n",")\n","\n","# Call the function:\n","analyze_gene_data(\n","  result_main_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_4.rds\",\n","  eqtl_long_covid_exposure_path = \"/content/drive/MyDrive/Colab/Long_COVID/eQTL/eqtl_all_tissues_110.txt\",\n","  save_files = save_paths\n",")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"9upr52RTXWzg","executionInfo":{"status":"ok","timestamp":1707377048997,"user_tz":-630,"elapsed":60439,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"f98d40d5-b793-4659-a9e1-7a6613de540b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved all_genes as all_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs4_all_genes_sig_non_sig.txt \n","Dimensions of all_genes : 145 10 \n","\n","          gene_id     gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000008196        TFAP2B protein_coding -2.161423  15    1\n","2 ENSG00000068781 STON1-GTF2A1L protein_coding  8.932349  29    2\n","3 ENSG00000112494        UNC93A protein_coding -6.110370   9    8\n","4 ENSG00000113905           HRG protein_coding -1.004191  15   14\n","5 ENSG00000117148         ACTL8 protein_coding -2.903997   5    4\n","6 ENSG00000122965         RBM19 protein_coding -3.602768  49   10\n","                                                                                                                                                   SNP_IDs\n","1     rs10948573,rs13202872,rs141707622,rs17667036,rs2206271,rs2477488,rs35305915,rs4715208,rs571503,rs61660288,rs62406441,rs72901027,rs75397901,rs9367397\n","2                                                                                                                                     rs2140551,rs62135453\n","3                                                                  rs117351628,rs141514782,rs149028144,rs1736600,rs3093009,rs55714481,rs62438461,rs7766490\n","4 rs1047115,rs112184129,rs114922365,rs140494728,rs144876286,rs148601582,rs1868144,rs4564971,rs4610218,rs6768529,rs75211540,rs7614563,rs77264387,rs78418844\n","5                                                                                                                 rs17362289,rs34519427,rs4141983,rs478698\n","6                                             rs11066483,rs11066771,rs11615428,rs116976487,rs12369142,rs147783556,rs1870352,rs4767149,rs7977045,rs80290311\n","         REML           SD Residual_SD\n","1   7.9037625 0.000000e+00    5.069014\n","2 -28.1355358 6.565327e+01    2.484851\n","3  11.4942247 2.371026e+01    3.310824\n","4  25.9172672 7.119931e-07    4.330643\n","5  -0.5075055 8.100030e+00    4.358663\n","6 -20.7081232 8.687471e+00    4.151621\n","\n","---\n","Saved protein_coding_genes as protein_coding_genes to /content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs4_protein_coding_genes_sig_not_sig.txt \n","Dimensions of protein_coding_genes : 28 10 \n","\n","          gene_id gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000186226     LCE1E protein_coding 250.61491   4    2\n","2 ENSG00000204345   CD300LD protein_coding -49.77334   4    2\n","3 ENSG00000242441   GTF2A1L protein_coding  37.00581  47    2\n","4 ENSG00000136696     IL36B protein_coding -32.82201   5    3\n","5 ENSG00000215009     ACSM4 protein_coding  30.38359   8    4\n","6 ENSG00000214513      NOTO protein_coding  25.83890   3    2\n","                                 SNP_IDs       REML        SD Residual_SD\n","1                  rs41268482,rs41268484 -10.089485 320.76952  1.44843557\n","2                    rs559743,rs59362027  -3.026900  95.04181  3.68822365\n","3                   rs2140551,rs35125065 -42.762211  42.76732  3.35812215\n","4        rs11123156,rs5833482,rs76336796  -1.092859  53.94958  3.23262992\n","5 rs10842265,rs12229849,rs2269354,rs8815 -10.265403  48.53811  0.96291899\n","6                 rs11894720,rs143097048 -13.649478  10.75561  0.02932065\n","\n","---\n"]}]},{"cell_type":"code","source":["%%R\n","\n","# BS4 final table:\n","# Call the function that retrieves more information from the Mt_Robin and filter the significant genes:\n","process_final_data(input_path_1 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/long_covid_main_bs_4_no_errors.rds\",\n","                   input_path_2 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs4_protein_coding_genes_sig_not_sig.txt\",\n","                   input_path_3 = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/MR_pvalue_FDR_results_bs4.txt\",\n","                   output_path = \"/content/drive/MyDrive/Colab/Long_COVID/MtRobin/bs4_protein_coding_genes_sig_full_table.txt\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"CnjIm5M9XWzg","executionInfo":{"status":"ok","timestamp":1707377054154,"user_tz":-630,"elapsed":5162,"user":{"displayName":"Sindy Pinero","userId":"07105275957931679855"}},"outputId":"f3283af4-2889-425e-b106-0b04285c630c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 145\n","$lme_res\n","Linear mixed model fit by REML ['lmerMod']\n","Formula: beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","Weights: 1/se_x^2\n","REML criterion at convergence: 7.9038\n","Random effects:\n"," Groups   Name   Std.Dev.\n"," snpID    beta_y 0.000   \n"," Residual        5.069   \n","Number of obs: 15, groups:  snpID, 14\n","Fixed Effects:\n","beta_y  \n","-2.161  \n","optimizer (nloptwrap) convergence code: 0 (OK) ; 0 optimizer warnings; 1 lme4 warnings \n","\n","$gwas_res\n","         snpID   gwas_beta   gwas_se\n","1   rs10948573  0.00386748 0.0242689\n","2   rs13202872  0.01946290 0.0255439\n","3  rs141707622  0.12363800 0.0676588\n","4   rs17667036 -0.01505850 0.0241452\n","5    rs2206271  0.02738150 0.0236648\n","6    rs2477488  0.03828220 0.0224505\n","7   rs35305915 -0.03549410 0.0311378\n","8    rs4715208  0.02011210 0.0261452\n","9     rs571503 -0.00302458 0.0224935\n","10  rs61660288 -0.03065830 0.0354960\n","11  rs62406441  0.00770182 0.0275417\n","12  rs72901027 -0.01634200 0.0344939\n","13  rs75397901 -0.09779650 0.0923557\n","14   rs9367397  0.00006950 0.0471503\n","\n","$LD\n","               rs2206271   rs35305915  rs72901027    rs9367397  rs61660288\n","rs2206271    1.000000000  0.181847333 -0.21528540 -0.071151068 -0.24230205\n","rs35305915   0.181847333  1.000000000 -0.06795493 -0.033229405  0.48066418\n","rs72901027  -0.215285397 -0.067954934  1.00000000 -0.026750052 -0.12654240\n","rs9367397   -0.071151068 -0.033229405 -0.02675005  1.000000000  0.04015302\n","rs61660288  -0.242302046  0.480664176 -0.12654240  0.040153019  1.00000000\n","rs141707622  0.119870789  0.009971749 -0.05598624  0.001138069 -0.02917965\n","rs2477488   -0.009077206 -0.228316423  0.09477027  0.007269130 -0.16061071\n","rs13202872  -0.136201309 -0.148212842 -0.07311959 -0.031303355 -0.14852754\n","rs62406441  -0.008002655 -0.111424559 -0.08150678 -0.090707064 -0.09612804\n","rs17667036   0.054885155  0.110191436  0.06151734  0.020875711  0.03804288\n","rs75397901   0.183026062  0.264094922  0.00593849 -0.055326238 -0.03668803\n","rs10948573  -0.364559036 -0.350527202 -0.24894668  0.084578073 -0.25732185\n","rs4715208   -0.323008364  0.240777328  0.56879634 -0.010220117  0.57630437\n","rs571503    -0.474874819 -0.327173864 -0.30756169  0.059012927 -0.33224409\n","             rs141707622    rs2477488  rs13202872   rs62406441   rs17667036\n","rs2206271    0.119870789 -0.009077206 -0.13620131 -0.008002655  0.054885155\n","rs35305915   0.009971749 -0.228316423 -0.14821284 -0.111424559  0.110191436\n","rs72901027  -0.055986244  0.094770273 -0.07311959 -0.081506783  0.061517343\n","rs9367397    0.001138069  0.007269130 -0.03130335 -0.090707064  0.020875711\n","rs61660288  -0.029179652 -0.160610711 -0.14852754 -0.096128035  0.038042883\n","rs141707622  1.000000000  0.132498616 -0.07725300  0.053178334 -0.128235768\n","rs2477488    0.132498616  1.000000000 -0.07366647 -0.074274924 -0.144630283\n","rs13202872  -0.077253005 -0.073666468  1.00000000  0.130442513 -0.105463006\n","rs62406441   0.053178334 -0.074274924  0.13044251  1.000000000 -0.005966075\n","rs17667036  -0.128235768 -0.144630283 -0.10546301 -0.005966075  1.000000000\n","rs75397901   0.054250285 -0.116781753 -0.03642830 -0.019579819 -0.005676509\n","rs10948573  -0.120390140  0.015492730  0.30662666  0.168116788 -0.046186252\n","rs4715208    0.019366695  0.019515913 -0.22105311 -0.122263117 -0.003934261\n","rs571503    -0.143230052  0.021697387  0.33163374  0.087491914 -0.038095251\n","              rs75397901  rs10948573    rs4715208    rs571503\n","rs2206271    0.183026062 -0.36455904 -0.323008364 -0.47487482\n","rs35305915   0.264094922 -0.35052720  0.240777328 -0.32717386\n","rs72901027   0.005938490 -0.24894668  0.568796336 -0.30756169\n","rs9367397   -0.055326238  0.08457807 -0.010220117  0.05901293\n","rs61660288  -0.036688029 -0.25732185  0.576304369 -0.33224409\n","rs141707622  0.054250285 -0.12039014  0.019366695 -0.14323005\n","rs2477488   -0.116781753  0.01549273  0.019515913  0.02169739\n","rs13202872  -0.036428295  0.30662666 -0.221053106  0.33163374\n","rs62406441  -0.019579819  0.16811679 -0.122263117  0.08749191\n","rs17667036  -0.005676509 -0.04618625 -0.003934261 -0.03809525\n","rs75397901   1.000000000 -0.10487956 -0.040319388 -0.09880661\n","rs10948573  -0.104879555  1.00000000 -0.441075032  0.70028947\n","rs4715208   -0.040319388 -0.44107503  1.000000000 -0.56173363\n","rs571503    -0.098806611  0.70028947 -0.561733625  1.00000000\n","\n","[1] \"Causal genes (protein coding genes): \"\n","[1] 28 10\n","          gene_id gene_name      gene_type    beta_y Obs SNPs\n","1 ENSG00000186226     LCE1E protein_coding 250.61491   4    2\n","2 ENSG00000204345   CD300LD protein_coding -49.77334   4    2\n","3 ENSG00000242441   GTF2A1L protein_coding  37.00581  47    2\n","4 ENSG00000136696     IL36B protein_coding -32.82201   5    3\n","5 ENSG00000215009     ACSM4 protein_coding  30.38359   8    4\n","6 ENSG00000214513      NOTO protein_coding  25.83890   3    2\n","                                 SNP_IDs       REML        SD Residual_SD\n","1                  rs41268482,rs41268484 -10.089485 320.76952  1.44843557\n","2                    rs559743,rs59362027  -3.026900  95.04181  3.68822365\n","3                   rs2140551,rs35125065 -42.762211  42.76732  3.35812215\n","4        rs11123156,rs5833482,rs76336796  -1.092859  53.94958  3.23262992\n","5 rs10842265,rs12229849,rs2269354,rs8815 -10.265403  48.53811  0.96291899\n","6                 rs11894720,rs143097048 -13.649478  10.75561  0.02932065\n","[1] \"Causal genes (noncoding genes and protein coding genes): \"\n","[1] 145   4\n","          gene_id pvalue nsamp_used       fdr\n","1 ENSG00000008196  0.258      10000 0.8382192\n","2 ENSG00000068781  0.878      10000 0.9583083\n","3 ENSG00000112494  0.655      10000 0.8952752\n","4 ENSG00000113905  0.776      10000 0.9132661\n","5 ENSG00000117148  0.745      10000 0.9132661\n","6 ENSG00000122965  0.374      10000 0.8382192\n","[1] \"Final table (protein coding genes): \"\n","[1]  0 22\n"," [1] gene_id    gene_name  SNPs       SNPs_IDs   Obs        beta_y    \n"," [7] Min        First_Q    Median     Third_Q    Max        SD_RE     \n","[13] Res_SD_RE  Var_RE     Res_Var_RE SE_FE      t_value_FE REML      \n","[19] p_value    FDR        CI_lower   CI_upper  \n","<0 rows> (or 0-length row.names)\n"]}]},{"cell_type":"markdown","source":["**Obs.:** No significant protein coding genes"],"metadata":{"id":"6WO5QaMXml4u"}},{"cell_type":"markdown","source":["## **Analysis**"],"metadata":{"id":"NkVt0jUESiwA"}},{"cell_type":"markdown","source":["**Results:**"],"metadata":{"id":"7foI8oC4g2SX"}},{"cell_type":"code","source":["$ ENSG00000102606:List of 3\n","  ..$ lme_res :Formal class 'lmerMod' [package \"lme4\"] with 13 slots\n","  .. .. ..@ resp   :Reference class 'lmerResp' [package \"lme4\"] with 9 fields\n","  .. .. .. ..$ Ptr    :<externalptr>\n","  .. .. .. ..$ mu     : num [1:37] 0.355 0.355 0.355 0.355 0.355 ...\n","  .. .. .. ..$ offset : num [1:37] 0 0 0 0 0 0 0 0 0 0 ...\n","  .. .. .. ..$ sqrtXwt: num [1:37] 42.3 31.1 25.2 28.3 18.4 ...\n","  .. .. .. ..$ sqrtrwt: num [1:37] 42.3 31.1 25.2 28.3 18.4 ...\n","  .. .. .. ..$ weights: num [1:37] 1786 966 634 803 338 ...\n","  .. .. .. ..$ wtres  : num [1:37] -3.0474 -0.9313 -0.0708 3.6392 0.5418 ...\n","  .. .. .. ..$ y      : num [1:37] 0.282 0.325 0.352 0.483 0.384 ...\n","  .. .. .. ..$ REML   : int 1\n","  .. .. .. ..and 28 methods, of which 14 are  possibly relevant:\n","  .. .. .. ..  allInfo, copy#envRefClass, initialize, initialize#lmResp,\n","  .. .. .. ..  initializePtr, initializePtr#lmResp, objective, ptr, ptr#lmResp,\n","  .. .. .. ..  setOffset, setResp, setWeights, updateMu, wrss\n","  .. .. ..@ Gp     : int [1:2] 0 7\n","  .. .. ..@ call   : language lme4::lmer(formula = beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID), weights = 1/se_x^2)\n","  .. .. ..@ frame  :'data.frame':\t37 obs. of  4 variables:\n","  .. .. .. ..$ beta_x   : num [1:37, 1] 0.282 0.325 0.352 0.483 0.384 ...\n","  .. .. .. ..$ beta_y   : num [1:37, 1] -0.056 -0.056 -0.056 -0.056 -0.056 ...\n","  .. .. .. ..$ snpID    : Factor w/ 7 levels \"rs2296353\",\"rs41275852\",..: 1 1 1 1 1 1 1 1 1 1 ...\n","  .. .. .. ..$ (weights): num [1:37, 1] 1786 966 634 803 338 ...\n","  .. .. .. ..- attr(*, \"terms\")=Classes 'terms', 'formula'  language beta_x ~ (beta_y - 1) + (beta_y - 1 + snpID)\n","  .. .. .. .. .. ..- attr(*, \"variables\")= language list(beta_x, beta_y, snpID)\n","  .. .. .. .. .. ..- attr(*, \"factors\")= int [1:3, 1:2] 0 1 0 0 0 1\n","  .. .. .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n","  .. .. .. .. .. .. .. ..$ : chr [1:3] \"beta_x\" \"beta_y\" \"snpID\"\n","  .. .. .. .. .. .. .. ..$ : chr [1:2] \"beta_y\" \"snpID\"\n","  .. .. .. .. .. ..- attr(*, \"term.labels\")= chr [1:2] \"beta_y\" \"snpID\"\n","  .. .. .. .. .. ..- attr(*, \"order\")= int [1:2] 1 1\n","  .. .. .. .. .. ..- attr(*, \"intercept\")= int 0\n","  .. .. .. .. .. ..- attr(*, \"response\")= int 1\n","  .. .. .. .. .. ..- attr(*, \".Environment\")=<environment: 0x0000013780e25478>\n","  .. .. .. .. .. ..- attr(*, \"predvars\")= language list(beta_x, beta_y, snpID)\n","  .. .. .. .. .. ..- attr(*, \"dataClasses\")= Named chr [1:4] \"nmatrix.1\" \"nmatrix.1\" \"character\" \"nmatrix.1\"\n","  .. .. .. .. .. .. ..- attr(*, \"names\")= chr [1:4] \"beta_x\" \"beta_y\" \"snpID\" \"(weights)\"\n","  .. .. .. .. .. ..- attr(*, \"predvars.fixed\")= language list(beta_x, beta_y)\n","  .. .. .. .. .. ..- attr(*, \"varnames.fixed\")= chr [1:3] \"beta_x\" \"beta_y\" \"(weights)\"\n","  .. .. .. .. .. ..- attr(*, \"predvars.random\")= language list(beta_x, beta_y, snpID)\n","  .. .. .. ..- attr(*, \"formula\")=Class 'formula'  language beta_x ~ (beta_y - 1) + (beta_y - 1 | snpID)\n","  .. .. .. .. .. ..- attr(*, \".Environment\")=<environment: 0x0000013780e25478>\n","  .. .. ..@ flist  :List of 1\n","  .. .. .. ..$ snpID: Factor w/ 7 levels \"rs2296353\",\"rs41275852\",..: 1 1 1 1 1 1 1 1 1 1 ...\n","  .. .. .. ..- attr(*, \"assign\")= int 1\n","  .. .. ..@ cnms   :List of 1\n","  .. .. .. ..$ snpID: chr \"beta_y\"\n","  .. .. ..@ lower  : num 0\n","  .. .. ..@ theta  : num 0.712\n","  .. .. ..@ beta   : num -4.76\n","  .. .. ..@ u      : num [1:7] -2.202 0.199 -0.271 -0.757 0.492 ...\n","  .. .. ..@ devcomp:List of 2\n","  .. .. .. ..$ cmp : Named num [1:10] 7.97 1.87 267.71 20.49 288.2 ...\n","  .. .. .. .. ..- attr(*, \"names\")= chr [1:10] \"ldL2\" \"ldRX2\" \"wrss\" \"ussq\" ...\n","  .. .. .. ..$ dims: Named int [1:12] 37 37 1 36 7 1 1 1 0 1 ...\n","  .. .. .. .. ..- attr(*, \"names\")= chr [1:12] \"N\" \"n\" \"p\" \"nmp\" ...\n","  .. .. ..@ pp     :Reference class 'merPredD' [package \"lme4\"] with 18 fields\n","  .. .. .. ..$ Lambdat:Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n","  .. .. .. .. .. ..@ i       : int [1:7] 0 1 2 3 4 5 6\n","  .. .. .. .. .. ..@ p       : int [1:8] 0 1 2 3 4 5 6 7\n","  .. .. .. .. .. ..@ Dim     : int [1:2] 7 7\n","  .. .. .. .. .. ..@ Dimnames:List of 2\n","  .. .. .. .. .. .. ..$ : NULL\n","  .. .. .. .. .. .. ..$ : NULL\n","  .. .. .. .. .. ..@ x       : num [1:7] 0.712 0.712 0.712 0.712 0.712 ...\n","  .. .. .. .. .. ..@ factors : list()\n","  .. .. .. ..$ LamtUt :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n","  .. .. .. .. .. ..@ i       : int [1:37] 0 0 0 0 0 0 0 0 0 0 ...\n","  .. .. .. .. .. ..@ p       : int [1:38] 0 1 2 3 4 5 6 7 8 9 ...\n","  .. .. .. .. .. ..@ Dim     : int [1:2] 7 37\n","  .. .. .. .. .. ..@ Dimnames:List of 2\n","  .. .. .. .. .. .. ..$ : NULL\n","  .. .. .. .. .. .. ..$ : chr [1:37] \"1\" \"2\" \"3\" \"4\" ...\n","  .. .. .. .. .. ..@ x       : num [1:37] -1.685 -1.239 -1.004 -1.13 -0.733 ...\n","  .. .. .. .. .. ..@ factors : list()\n","  .. .. .. ..$ Lind   : int [1:7] 1 1 1 1 1 1 1\n","  .. .. .. ..$ Ptr    :<externalptr>\n","  .. .. .. ..$ RZX    : num [1:7, 1] 9.3981 1.0636 1.424 0.4506 0.0181 ...\n","  .. .. .. ..$ Ut     :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n","  .. .. .. .. .. ..@ i       : int [1:37] 0 0 0 0 0 0 0 0 0 0 ...\n","  .. .. .. .. .. ..@ p       : int [1:38] 0 1 2 3 4 5 6 7 8 9 ...\n","  .. .. .. .. .. ..@ Dim     : int [1:2] 7 37\n","  .. .. .. .. .. ..@ Dimnames:List of 2\n","  .. .. .. .. .. .. ..$ : chr [1:7] \"rs2296353\" \"rs41275852\" \"rs9515381\" \"rs9522179\" ...\n","  .. .. .. .. .. .. ..$ : chr [1:37] \"1\" \"2\" \"3\" \"4\" ...\n","  .. .. .. .. .. ..@ x       : num [1:37] -2.37 -1.74 -1.41 -1.59 -1.03 ...\n","  .. .. .. .. .. ..@ factors : list()\n","  .. .. .. ..$ Utr    : num [1:7] -408.842 -6.915 -11.756 -3.559 0.411 ...\n","  .. .. .. ..$ V      : num [1:37, 1] -2.37 -1.74 -1.41 -1.59 -1.03 ...\n","  .. .. .. ..$ VtV    : num [1, 1] 110\n","  .. .. .. ..$ Vtr    : num -631\n","  .. .. .. ..$ X      : num [1:37, 1] -0.056 -0.056 -0.056 -0.056 -0.056 ...\n","  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n","  .. .. .. .. .. ..$ : chr [1:37] \"1\" \"2\" \"3\" \"4\" ...\n","  .. .. .. .. .. ..$ : chr \"beta_y\"\n","  .. .. .. .. ..- attr(*, \"assign\")= int 1\n","  .. .. .. .. ..- attr(*, \"msgScaleX\")= chr(0)\n","  .. .. .. ..$ Xwts   : num [1:37] 42.3 31.1 25.2 28.3 18.4 ...\n","  .. .. .. ..$ Zt     :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n","  .. .. .. .. .. ..@ i       : int [1:37] 0 0 0 0 0 0 0 0 0 0 ...\n","  .. .. .. .. .. ..@ p       : int [1:38] 0 1 2 3 4 5 6 7 8 9 ...\n","  .. .. .. .. .. ..@ Dim     : int [1:2] 7 37\n","  .. .. .. .. .. ..@ Dimnames:List of 2\n","  .. .. .. .. .. .. ..$ : chr [1:7] \"rs2296353\" \"rs41275852\" \"rs9515381\" \"rs9522179\" ...\n","  .. .. .. .. .. .. ..$ : chr [1:37] \"1\" \"2\" \"3\" \"4\" ...\n","  .. .. .. .. .. ..@ x       : num [1:37] -0.056 -0.056 -0.056 -0.056 -0.056 ...\n","  .. .. .. .. .. ..@ factors : list()\n","  .. .. .. ..$ beta0  : num 0\n","  .. .. .. ..$ delb   : num -4.76\n","  .. .. .. ..$ delu   : num [1:7] -2.202 0.199 -0.271 -0.757 0.492 ...\n","  .. .. .. ..$ theta  : num 0.712\n","  .. .. .. ..$ u0     : num [1:7] 0 0 0 0 0 0 0\n","  .. .. .. ..and 45 methods, of which 31 are  possibly relevant:\n","  .. .. .. ..  b, beta, CcNumer, copy#envRefClass, initialize, initializePtr,\n","  .. .. .. ..  installPars, L, ldL2, ldRX2, linPred, P, ptr, RX, RXdiag, RXi,\n","  .. .. .. ..  setBeta0, setDelb, setDelu, setTheta, setZt, solve, solveU, sqrL, u,\n","  .. .. .. ..  unsc, updateDecomp, updateL, updateLamtUt, updateRes, updateXwts\n","  .. .. ..@ optinfo:List of 8\n","  .. .. .. ..$ optimizer: chr \"nloptwrap\"\n","  .. .. .. ..$ control  :List of 1\n","  .. .. .. .. ..$ print_level: num 0\n","  .. .. .. ..$ derivs   :List of 2\n","  .. .. .. .. ..$ gradient: num 1.19e-06\n","  .. .. .. .. ..$ Hessian : num [1, 1] 17.1\n","  .. .. .. ..$ conv     :List of 2\n","  .. .. .. .. ..$ opt : num 0\n","  .. .. .. .. ..$ lme4: list()\n","  .. .. .. ..$ feval    : int 21\n","  .. .. .. ..$ message  : chr \"NLOPT_XTOL_REACHED: Optimization stopped because xtol_rel or xtol_abs (above) was reached.\"\n","  .. .. .. ..$ warnings : list()\n","  .. .. .. ..$ val      : num 0.712\n","  ..$ gwas_res:Classes ‘data.table’ and 'data.frame':\t7 obs. of  3 variables:\n","  .. ..$ snpID    : chr [1:7] \"rs2296353\" \"rs41275852\" \"rs9515381\" \"rs9522179\" ...\n","  .. ..$ gwas_beta: num [1:7] -0.05602 -0.04274 0.06575 0.07151 -0.00874 ...\n","  .. ..$ gwas_se  : num [1:7] 0.0321 0.0409 0.0312 0.0297 0.03 ...\n","  .. ..- attr(*, \".internal.selfref\")=<externalptr>\n","  .. ..- attr(*, \"sorted\")= chr \"snpID\"\n","  ..$ LD      : num [1:7, 1:7] 1 0.662 -0.41 0.415 -0.481 ...\n","  .. ..- attr(*, \"dimnames\")=List of 2\n","  .. .. ..$ : chr [1:7] \"rs2296353\" \"rs9560010\" \"rs9515381\" \"rs41275852\" ...\n","  .. .. ..$ : chr [1:7] \"rs2296353\" \"rs9560010\" \"rs9515381\" \"rs41275852\" ..."],"metadata":{"id":"kD5EJY9aNWjK"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**Interpretation of the Results:**\n","\n","**Linear Mixed Model (`lmerMod`) components:**\n","\n","1. `mu:` The vector of fitted values, or predicted means, of the response variable based on the fitted model.\n","  \n","2. `sqrtXwt:` Square root of the weights for the fixed effects. Useful for internal calculations.\n","\n","3. `sqrtrwt:` Square root of the \"residual weights.\" Useful for internal calculations.\n","\n","4. `weights:` Weights assigned to each observation, typically used to handle heteroscedasticity or to give certain observations more influence in the model fit.\n","\n","5. `wtres:` Weighted residuals, which are the differences between observed and fitted values, each multiplied by the square root of its corresponding weight.\n","\n","6. `y:` The response variable, (beta_x) in this case, which the model is trying to predict.\n","\n","7. `REML:` Indicates whether Restricted Maximum Likelihood (REML) was used. REML is a method for estimating the parameters in a linear mixed model.\n","- A lower (more negative) REML value indicates better fit but does not necessarily mean the gene is more biologically relevant to the outcome of interest. Biological plausibility and other external evidence should also be considered.\n","- It is the `Log-likelihood function value at convergence` that the method uses to estimate the parameters in a mixed model --> Used instead of the `Maximum Likelihood Estimation` because it produces lower unbiased estimates of variance components.\n","- Useful for comparing different models fitted to the same dataset. It is not useful for comparing models across different datasets.\n","- While a lower REML value indicates a better fit, it doesn't necessarily mean that the model is the best possible model. Overfitting is a concern if the model is too complex.\n","- It is generally used for comparing nested models, that is, models where one is a special case of the other (e.g., one model has additional predictors compared to the other).\n","- It is particularly useful to compare fixed and random effects in a model, as it provides a way to compare the goodness-of-fit taking both into account.\n","If you're using one dataset for exposures (e.g., gene expression levels for multiple genes) and another dataset for the outcome (e.g., a specific clinical trait), and you're interested in examining the relationship between each gene and the outcome, then you could potentially use REML as one of the metrics to assess how well each gene's model fits the data.\n","- Both input datasets should represent the same or very similar underlying populations for the REML criterion to be meaningful. If the datasets are drastically different, then the model's fit may not be valid for such a comparison.\n","- Ensure that the models for different genes are specified in the same way (i.e., using the same covariates, random effects structure, etc.) for the REML values to be comparable.\n","- REML is just one measure of model fit. Other metrics and validation methods should also be considered. For example, predictive accuracy on a held-out test set.\n","\n","**Data Frame (`frame`) components:**\n","\n","1. `beta_x:` The response variable in the model.\n","\n","2. `beta_y:` A predictor variable in the model.\n","\n","3. `lower:` Lower boundary for the parameter estimates.\n","\n","4. `theta:` The estimated variance components for the random effects.\n","\n","5. `beta:` The estimated fixed effects parameters, excluding the intercept.\n","\n","6. `u:` The random effects estimates.\n","\n","**Deviance Components (`devcomp`):**\n","\n","1. `cmp:` Various components needed to compute the model deviance, AIC, BIC, etc.\n","\n","2. `dims:` Various dimensions related to the model, like the number of observations, number of parameters, etc.\n","\n","**Prediction (`pp`) components:**\n","\n","1. `x:` Appears multiple times, these are various vectors/matrices used internally for prediction.\n","\n","2. `RZX:` Related to the rotated or \"whitened\" design matrix for the random effects.\n","\n","3. `Utr:` Related to a transformed version of the random effects.\n","\n","4. `V:` This may represent the \"pseudo-data\" in the penalized least squares formulation of the linear mixed model.\n","\n","5. `Xwts:` Weights for the predictor matrix (X).\n","\n","6. `delu:` Changes in the random effects estimates, typically during optimization.\n","\n","---\n","\n","**Downstream analysis:**\n","\n","**Basic Model Summary:**\n","1. `beta:` The relationship between the predictors and the response variable.\n","2. `theta:` The variability explained by random effects.\n","3. `u:` To make predictions for specific groups in the random effects.\n","\n","**Model Quality:**\n","1. `REML:` Whether REML or ML (Maximum Likelihood) was used can impact the comparability of models, especially if when comparing models with different fixed effects.\n","2. `cmp:` Components like AIC (Akaike Information Criterion) or BIC (Bayesian Information Criterion) can be computed from these, which are useful for model comparison.\n","3. `dims:` Dimensions (e.g., number of observations, number of parameters, etc.) to understand the scope of the model.\n","\n","**Residual Analysis:**\n","1. `wtres:` Weighted residuals for diagnostic plots to assess model fit.\n","2. `mu:` The fitted values can be compared to the actual values to assess how well the model is fitting the data.\n","\n","**Weights and Other Diagnostics:**\n","1. `weights:` Important to carry these weighted values forward into any downstream analyses.\n","2. `sqrtXwt, sqrtrwt:` Can be used in specific types of weighted or penalized analyses.\n","\n","**Specific Genetic Analysis:**\n","1. `gwas_res:` Crucial for any genetic-specific downstream analysis.\n","2. `LD:` Linkage disequilibrium information for genetic studies.\n","\n","**Enrichment Analysis:**\n","1. `beta:` The estimated fixed effects parameters can be crucial for enrichment analyses. These represent the relationship between the predictors and the response variable. A stronger effect (either positive or negative) can highlight genes more likely to be involved in the biological process.\n","2. `gwas_res:` Essencial to find enrichments specific to certain genetic loci.\n","3. `LD:` To understand whether multiple SNPs are independently contributing to a trait or if they are just proxies for one another one.\n","\n","**Prediction:**\n","1. `mu:` The fitted values can be used for making predictions on the training set. They can serve as a baseline for predicting the same outcome in a different cohort.\n","2. `X, Zt, etc.:` To do predictions for new data points based on the model.\n","3. `theta:` The estimated variance components for the random effects can help to understand how much of the variability in the outcome is accounted by the random effects, which might be important depending on the nature of the predictions.\n","4. `u:` Random effects estimates can also be useful for making predictions, especially for predicting outcomes for specific groups that are represented in the random effects.\n","\n","**Drug Repurposing:**\n","1. `beta:` The fixed effects estimates are crucial. Genes with strong effects are candidates for targeting by drugs.\n","2. `gwas_res:` The original GWAS results could provide additional context for why a gene might be a good target for drug repurposing.\n","\n","**General Important Variables:**\n","1. `weights:` Important when interpreting the results.\n","2. `REML:` Goodness of fit of the model.\n","\n","---\n","\n","**Most important variables for Drug Repurposing:**\n","\n","`gene_id:` ID of each gene.\n","\n","`gene_name:` Name of each gene.\n","\n","`beta_y:`\n","- Causal effect value of each gene.\n","- Higher absolute value (far from 0) --> stronger the causal relationship.\n","\n","`Obs:`\n","- Number of tissues in which the expression of a gene is associated or influenced by some specific SNPs (gene_expression-SNP association).\n","\n","`SNPs:`\n","- Number of SNPs that influence the expression of a specific gene (gene_expression-SNP association).\n","\n","`REML criterion at convergence:`\n","- Goodness of fit of the model.\n","- Lower values (more negative) --> Better fit.\n","\n","`SD:` Standard Deviation of the Random Effects.\n","\n","`Residual_SD:`\n","- The differences between the observed and predicted values for the dependent variable (gene_expression/SNPs).\n","- Smaller values --> The model is doing a better job of capturing the variation in the data.\n","\n","---\n","\n","**Most important variables for Prediction:**\n","\n","All the previous variables and:\n","\n","- ` theta:` Variance of the random effects.\n","- `u:` Random effect estimate.\n","- `mu:` The vector of fitted values, or predicted means, of the response variable based on the fitted model.\n","\n","---\n","\n","**Other important variables:**\n","- `weights:` Higher/Lower influence of the observations.\n","- `cmp:` Components of the model fit (AIC and BIC)\n","- `dims`: Number od observations/parameteres to understand the scope of the model.\n","\n","---\n","\n","**SD vs. Residual SD:**\n","- The Standard Deviation (SD) and the Residual Standard Deviation (Residual SD) refer to different aspects of variability in the data.\n","\n","**SD:**\n","- The SD of the random effects of `beta_y` across different levels of `snpID`.\n","- The SD for random effects helps us to understand how much the groups differ.\n","- The value that indicates the amount of variability in the random intercept `beta_y` across different `snpID` groups. `A higher value indicates that there is substantial variability across groups, suggesting that including a random effect is appropriate.`\n","\n","**Residual SD:**\n","- The differences between the observed and predicted values for the dependent variable (gene_expression/SNPs).\n","- The residual SD helps us to understand how well the model fits the data.\n","- The \"typical\" difference between the observed values of the dependent variable and the values predicted by the model. `Smaller values indicate that the model is doing a better job of capturing the variation in the data.`\n","\n","**SD and Variance vs. beta_y:**\n","\n","The Standard Deviation (SD) and Variance values are not directly comparable to the `beta_y` values for both fixed and random effects:\n","\n","**Different Units and Interpretations:**\n","\n","1. **Fixed Effects**: The SE for a fixed effect estimate `beta_y` shows how much that estimate would vary when taking many samples from the same population and estimating the model each time. It is not a measure of the \"size\" of the effect but rather of the precision of the estimate.\n","  \n","2. **Random Effects**: The variance and SD for a random effect are measures of how much that effect varies across different levels of the grouping variable (`snpID`) and cannot be compared to the fixed effect estimate.\n","\n","**Scale Sensitivity:**\n","\n","Both SD and variance are sensitive to the scale of the variables involved. If `beta_y` and `beta_x` are measured in different units or have different scales, then their SDs and variances will also be on different scales, making them incomparable.\n","\n","**Model Complexity and Sample Size:**\n","\n","The SD and variance can also be influenced by the complexity of the model and the sample size, which are independent of the value of the `beta_y`.\n","\n","**What I can Compare?:**\n","\n","1. **Coefficient Significance**: Assess the significance of the `beta_y` value by looking at its `t-value`, which takes the SE into account.\n","\n","2. **Confidence Intervals**: Compute confidence intervals for `beta_y` using its SD to get an idea of its possible range of values.\n","  \n","3. **Variance Components**: For random effects, look at the estimated variance and SD to understand the degree of variability in `beta_y` across different levels of `snpID`.\n","\n","4. **Model Comparison**: While the SD and variance aren't directly comparable to `beta_y`, they can be used for model comparison. For example, a model with lower SDs for its estimates is generally considered to be a better-fitting model, all else being equal."],"metadata":{"id":"q8ySou05KTv3"}}]}
